begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// REQUIRES: systemz-registered-target
end_comment

begin_comment
comment|// RUN: %clang_cc1 -target-cpu z13 -triple s390x-unknown-unknown \
end_comment

begin_comment
comment|// RUN: -Wall -Wno-unused -Werror -fsyntax-only -verify %s
end_comment

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|signed
name|char
name|vec_schar
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|signed
name|short
name|vec_sshort
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|signed
name|int
name|vec_sint
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|signed
name|long
name|long
name|vec_slong
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|unsigned
name|char
name|vec_uchar
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|unsigned
name|short
name|vec_ushort
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|unsigned
name|int
name|vec_uint
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|unsigned
name|long
name|long
name|vec_ulong
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|(vector_size(
literal|16
argument|))
argument_list|)
name|double
name|vec_double
typedef|;
end_typedef

begin_decl_stmt
specifier|volatile
name|vec_schar
name|vsc
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|vec_sshort
name|vss
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|vec_sint
name|vsi
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|vec_slong
name|vsl
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|vec_uchar
name|vuc
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|vec_ushort
name|vus
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|vec_uint
name|vui
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|vec_ulong
name|vul
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|vec_double
name|vd
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|volatile
name|unsigned
name|int
name|len
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|const
name|void
modifier|*
specifier|volatile
name|cptr
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|cc
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|test_core
parameter_list|(
name|void
parameter_list|)
block|{
name|__builtin_s390_lcbb
argument_list|(
name|cptr
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_lcbb
argument_list|(
name|cptr
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_lcbb
argument_list|(
name|cptr
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vlbb
argument_list|(
name|cptr
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vlbb
argument_list|(
name|cptr
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vlbb
argument_list|(
name|cptr
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vpdi
argument_list|(
name|vul
argument_list|,
name|vul
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vpdi
argument_list|(
name|vul
argument_list|,
name|vul
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vpdi
argument_list|(
name|vul
argument_list|,
name|vul
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
block|}
end_function

begin_function
name|void
name|test_integer
parameter_list|(
name|void
parameter_list|)
block|{
name|__builtin_s390_verimb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|__builtin_s390_verimb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|__builtin_s390_verimb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_verimh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|__builtin_s390_verimh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|__builtin_s390_verimh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_verimf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|__builtin_s390_verimf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|__builtin_s390_verimf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_verimg
argument_list|(
name|vul
argument_list|,
name|vul
argument_list|,
name|vul
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|__builtin_s390_verimg
argument_list|(
name|vul
argument_list|,
name|vul
argument_list|,
name|vul
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|__builtin_s390_verimg
argument_list|(
name|vul
argument_list|,
name|vul
argument_list|,
name|vul
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vsldb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vsldb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vsldb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
block|}
end_function

begin_function
name|void
name|test_string
parameter_list|(
name|void
parameter_list|)
block|{
name|__builtin_s390_vfaeb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaeb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaeb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaeh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaeh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaeh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaef
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaef
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaef
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaezb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaezh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaezf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrcb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrcb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrcb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrch
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrch
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrch
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrcf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrcf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrcf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrczb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczb
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrczh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczh
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrczf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczf
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaebs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaebs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaebs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaehs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaehs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaehs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaefs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaefs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaefs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaezbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaezhs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezhs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezhs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfaezfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfaezfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrcbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrcbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrcbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrchs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrchs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrchs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrcfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrcfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrcfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrczbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczbs
argument_list|(
name|vuc
argument_list|,
name|vuc
argument_list|,
name|vuc
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrczhs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczhs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczhs
argument_list|(
name|vus
argument_list|,
name|vus
argument_list|,
name|vus
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vstrczfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
literal|16
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vstrczfs
argument_list|(
name|vui
argument_list|,
name|vui
argument_list|,
name|vui
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
block|}
end_function

begin_function
name|void
name|test_float
parameter_list|(
name|void
parameter_list|)
block|{
name|__builtin_s390_vftcidb
argument_list|(
name|vd
argument_list|,
operator|-
literal|1
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 4095}}
name|__builtin_s390_vftcidb
argument_list|(
name|vd
argument_list|,
literal|4096
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 4095}}
name|__builtin_s390_vftcidb
argument_list|(
name|vd
argument_list|,
name|len
argument_list|,
operator|&
name|cc
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfidb
argument_list|(
name|vd
argument_list|,
operator|-
literal|1
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfidb
argument_list|(
name|vd
argument_list|,
literal|16
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfidb
argument_list|(
name|vd
argument_list|,
name|len
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
name|__builtin_s390_vfidb
argument_list|(
name|vd
argument_list|,
literal|0
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfidb
argument_list|(
name|vd
argument_list|,
literal|0
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|__builtin_s390_vfidb
argument_list|(
name|vd
argument_list|,
literal|0
argument_list|,
name|len
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a constant integer}}
block|}
end_function

end_unit

