begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// REQUIRES: mips-registered-target
end_comment

begin_comment
comment|// RUN: not %clang_cc1 -triple mips-unknown-linux-gnu -fsyntax-only %s \
end_comment

begin_comment
comment|// RUN:            -target-feature +msa -target-feature +fp64 \
end_comment

begin_comment
comment|// RUN:            -mfloat-abi hard -o - 2>&1 | FileCheck %s
end_comment

begin_include
include|#
directive|include
file|<msa.h>
end_include

begin_function
name|void
name|test
parameter_list|(
name|void
parameter_list|)
block|{
name|v16i8
name|v16i8_a
init|=
operator|(
name|v16i8
operator|)
block|{
literal|0
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|,
literal|6
block|,
literal|7
block|,
literal|8
block|,
literal|9
block|,
literal|10
block|,
literal|11
block|,
literal|12
block|,
literal|13
block|,
literal|14
block|,
literal|15
block|}
decl_stmt|;
name|v16i8
name|v16i8_r
decl_stmt|;
name|v8i16
name|v8i16_a
init|=
operator|(
name|v8i16
operator|)
block|{
literal|0
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|,
literal|6
block|,
literal|7
block|}
decl_stmt|;
name|v8i16
name|v8i16_r
decl_stmt|;
name|v4i32
name|v4i32_a
init|=
operator|(
name|v4i32
operator|)
block|{
literal|0
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|}
decl_stmt|;
name|v4i32
name|v4i32_r
decl_stmt|;
name|v2i64
name|v2i64_a
init|=
operator|(
name|v2i64
operator|)
block|{
literal|0
block|,
literal|1
block|}
decl_stmt|;
name|v2i64
name|v2i64_r
decl_stmt|;
name|v16u8
name|v16u8_a
init|=
operator|(
name|v16u8
operator|)
block|{
literal|0
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|,
literal|6
block|,
literal|7
block|,
literal|8
block|,
literal|9
block|,
literal|10
block|,
literal|11
block|,
literal|12
block|,
literal|13
block|,
literal|14
block|,
literal|15
block|}
decl_stmt|;
name|v16u8
name|v16u8_r
decl_stmt|;
name|v8u16
name|v8u16_a
init|=
operator|(
name|v8u16
operator|)
block|{
literal|0
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|,
literal|6
block|,
literal|7
block|}
decl_stmt|;
name|v8u16
name|v8u16_r
decl_stmt|;
name|v4u32
name|v4u32_a
init|=
operator|(
name|v4u32
operator|)
block|{
literal|0
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|}
decl_stmt|;
name|v4u32
name|v4u32_r
decl_stmt|;
name|v2u64
name|v2u64_a
init|=
operator|(
name|v2u64
operator|)
block|{
literal|0
block|,
literal|1
block|}
decl_stmt|;
name|v2u64
name|v2u64_r
decl_stmt|;
name|int
name|int_r
decl_stmt|;
name|long
name|long
name|ll_r
decl_stmt|;
name|v16u8_r
operator|=
name|__msa_addvi_b
argument_list|(
name|v16u8_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_addvi_h
argument_list|(
name|v8u16_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_addvi_w
argument_list|(
name|v4u32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_addvi_d
argument_list|(
name|v2u64_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_andi_b
argument_list|(
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v8i16_r
operator|=
name|__msa_andi_b
argument_list|(
name|v8i16_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v4i32_r
operator|=
name|__msa_andi_b
argument_list|(
name|v4i32_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v2i64_r
operator|=
name|__msa_andi_b
argument_list|(
name|v2i64_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_bclri_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_bclri_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_bclri_w
argument_list|(
name|v4i32_a
argument_list|,
literal|33
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_bclri_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_binsli_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_binsli_h
argument_list|(
name|v8i16_r
argument_list|,
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_binsli_w
argument_list|(
name|v4i32_r
argument_list|,
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_binsli_d
argument_list|(
name|v2i64_r
argument_list|,
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_binsri_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_binsri_h
argument_list|(
name|v8i16_r
argument_list|,
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_binsri_w
argument_list|(
name|v4i32_r
argument_list|,
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_binsri_d
argument_list|(
name|v2i64_r
argument_list|,
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_bmnzi_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_bmzi_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_bnegi_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_bnegi_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_bnegi_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_bnegi_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_bseli_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_bseti_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_bseti_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_bseti_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_bseti_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_ceqi_b
argument_list|(
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_ceqi_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_ceqi_w
argument_list|(
name|v4i32_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_ceqi_d
argument_list|(
name|v2i64_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16i8_r
operator|=
name|__msa_clei_s_b
argument_list|(
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_clei_s_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_clei_s_w
argument_list|(
name|v4i32_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_clei_s_d
argument_list|(
name|v2i64_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16u8_r
operator|=
name|__msa_clei_u_b
argument_list|(
name|v16u8_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_clei_u_h
argument_list|(
name|v8u16_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_clei_u_w
argument_list|(
name|v4u32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_clei_u_d
argument_list|(
name|v2u64_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_clti_s_b
argument_list|(
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_clti_s_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_clti_s_w
argument_list|(
name|v4i32_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_clti_s_d
argument_list|(
name|v2i64_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16u8_r
operator|=
name|__msa_clti_u_b
argument_list|(
name|v16u8_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_clti_u_h
argument_list|(
name|v8u16_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_clti_u_w
argument_list|(
name|v4u32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_clti_u_d
argument_list|(
name|v2u64_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|int_r
operator|=
name|__msa_copy_s_b
argument_list|(
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|int_r
operator|=
name|__msa_copy_s_h
argument_list|(
name|v8i16_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|int_r
operator|=
name|__msa_copy_s_w
argument_list|(
name|v4i32_a
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|ll_r
operator|=
name|__msa_copy_s_d
argument_list|(
name|v2i64_a
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|int_r
operator|=
name|__msa_copy_u_b
argument_list|(
name|v16u8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|int_r
operator|=
name|__msa_copy_u_h
argument_list|(
name|v8u16_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|int_r
operator|=
name|__msa_copy_u_w
argument_list|(
name|v4u32_a
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|ll_r
operator|=
name|__msa_copy_u_d
argument_list|(
name|v2i64_a
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_insve_b
argument_list|(
name|v16i8_r
argument_list|,
literal|16
argument_list|,
name|v16i8_a
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v8i16_r
operator|=
name|__msa_insve_h
argument_list|(
name|v8i16_r
argument_list|,
literal|8
argument_list|,
name|v8i16_a
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v4i32_r
operator|=
name|__msa_insve_w
argument_list|(
name|v4i32_r
argument_list|,
literal|4
argument_list|,
name|v4i32_a
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|v2i64_r
operator|=
name|__msa_insve_d
argument_list|(
name|v2i64_r
argument_list|,
literal|2
argument_list|,
name|v2i64_a
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_ld_b
argument_list|(
operator|&
name|v16i8_a
argument_list|,
literal|23
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a multiple of 16}}
name|v8i16_r
operator|=
name|__msa_ld_h
argument_list|(
operator|&
name|v8i16_a
argument_list|,
literal|77
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a multiple of 16}}
name|v4i32_r
operator|=
name|__msa_ld_w
argument_list|(
operator|&
name|v4i32_a
argument_list|,
literal|14
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a multiple of 16}}
name|v2i64_r
operator|=
name|__msa_ld_d
argument_list|(
operator|&
name|v2i64_a
argument_list|,
literal|23
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a multiple of 16}}
name|v16i8_r
operator|=
name|__msa_ld_b
argument_list|(
operator|&
name|v16i8_a
argument_list|,
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v8i16_r
operator|=
name|__msa_ld_h
argument_list|(
operator|&
name|v8i16_a
argument_list|,
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v4i32_r
operator|=
name|__msa_ld_w
argument_list|(
operator|&
name|v4i32_a
argument_list|,
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v2i64_r
operator|=
name|__msa_ld_d
argument_list|(
operator|&
name|v2i64_a
argument_list|,
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v16i8_r
operator|=
name|__msa_ldi_b
argument_list|(
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v8i16_r
operator|=
name|__msa_ldi_h
argument_list|(
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v4i32_r
operator|=
name|__msa_ldi_w
argument_list|(
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v2i64_r
operator|=
name|__msa_ldi_d
argument_list|(
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v16i8_r
operator|=
name|__msa_maxi_s_b
argument_list|(
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_maxi_s_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_maxi_s_w
argument_list|(
name|v4i32_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_maxi_s_d
argument_list|(
name|v2i64_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16u8_r
operator|=
name|__msa_maxi_u_b
argument_list|(
name|v16u8_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_maxi_u_h
argument_list|(
name|v8u16_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_maxi_u_w
argument_list|(
name|v4u32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_maxi_u_d
argument_list|(
name|v2u64_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_mini_s_b
argument_list|(
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_mini_s_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_mini_s_w
argument_list|(
name|v4i32_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_mini_s_d
argument_list|(
name|v2i64_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16u8_r
operator|=
name|__msa_mini_u_b
argument_list|(
name|v16u8_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_mini_u_h
argument_list|(
name|v8u16_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_mini_u_w
argument_list|(
name|v4u32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_mini_u_d
argument_list|(
name|v2u64_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_nori_b
argument_list|(
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_ori_b
argument_list|(
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_sat_s_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_sat_s_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_sat_s_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_sat_s_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_sat_u_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_sat_u_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_sat_u_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_sat_u_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_shf_b
argument_list|(
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v8i16_r
operator|=
name|__msa_shf_h
argument_list|(
name|v8i16_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v4i32_r
operator|=
name|__msa_shf_w
argument_list|(
name|v4i32_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_sld_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v8i16_r
operator|=
name|__msa_sld_h
argument_list|(
name|v8i16_r
argument_list|,
name|v8i16_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v4i32_r
operator|=
name|__msa_sld_w
argument_list|(
name|v4i32_r
argument_list|,
name|v4i32_a
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|v2i64_r
operator|=
name|__msa_sld_d
argument_list|(
name|v2i64_r
argument_list|,
name|v2i64_a
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_sldi_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v8i16_r
operator|=
name|__msa_sldi_h
argument_list|(
name|v8i16_r
argument_list|,
name|v8i16_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v4i32_r
operator|=
name|__msa_sldi_w
argument_list|(
name|v4i32_r
argument_list|,
name|v4i32_a
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|v2i64_r
operator|=
name|__msa_sldi_d
argument_list|(
name|v2i64_r
argument_list|,
name|v2i64_a
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_slli_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_slli_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_slli_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_slli_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_splati_b
argument_list|(
name|v16i8_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v8i16_r
operator|=
name|__msa_splati_h
argument_list|(
name|v8i16_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v4i32_r
operator|=
name|__msa_splati_w
argument_list|(
name|v4i32_a
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|v2i64_r
operator|=
name|__msa_splati_d
argument_list|(
name|v2i64_a
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_srai_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_srai_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_srai_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_srai_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_srari_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_srari_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_srari_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_srari_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_srli_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_srli_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_srli_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_srli_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_srlri_b
argument_list|(
name|v16i8_a
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_srlri_h
argument_list|(
name|v8i16_a
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_srlri_w
argument_list|(
name|v4i32_a
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_srlri_d
argument_list|(
name|v2i64_a
argument_list|,
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|__msa_st_b
argument_list|(
name|v16i8_b
argument_list|,
operator|&
name|v16i8_a
argument_list|,
literal|52
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a multiple of 16}}
name|__msa_st_h
argument_list|(
name|v8i16_b
argument_list|,
operator|&
name|v8i16_a
argument_list|,
literal|51
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a multiple of 16}}
name|__msa_st_w
argument_list|(
name|v4i32_b
argument_list|,
operator|&
name|v4i32_a
argument_list|,
literal|51
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a multiple of 16}}
name|__msa_st_d
argument_list|(
name|v2i64_b
argument_list|,
operator|&
name|v2i64_a
argument_list|,
literal|12
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a multiple of 16}}
name|__msa_st_b
argument_list|(
name|v16i8_b
argument_list|,
operator|&
name|v16i8_a
argument_list|,
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|__msa_st_h
argument_list|(
name|v8i16_b
argument_list|,
operator|&
name|v8i16_a
argument_list|,
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|__msa_st_w
argument_list|(
name|v4i32_b
argument_list|,
operator|&
name|v4i32_a
argument_list|,
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|__msa_st_d
argument_list|(
name|v2i64_b
argument_list|,
operator|&
name|v2i64_a
argument_list|,
literal|512
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v16i8_r
operator|=
name|__msa_subvi_b
argument_list|(
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8i16_r
operator|=
name|__msa_subvi_h
argument_list|(
name|v8i16_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4i32_r
operator|=
name|__msa_subvi_w
argument_list|(
name|v4i32_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_subvi_d
argument_list|(
name|v2i64_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_xori_b
argument_list|(
name|v16i8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v8i16_r
operator|=
name|__msa_xori_b
argument_list|(
name|v8i16_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v4i32_r
operator|=
name|__msa_xori_b
argument_list|(
name|v4i32_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v2i64_r
operator|=
name|__msa_xori_b
argument_list|(
name|v2i64_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16u8_r
operator|=
name|__msa_xori_b
argument_list|(
name|v16u8_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v8u16_r
operator|=
name|__msa_xori_b
argument_list|(
name|v8u16_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v4u32_r
operator|=
name|__msa_xori_b
argument_list|(
name|v4u32_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v2u64_r
operator|=
name|__msa_xori_b
argument_list|(
name|v2u64_a
argument_list|,
literal|256
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
comment|// Test the lower bounds
name|v16u8_r
operator|=
name|__msa_addvi_b
argument_list|(
name|v16u8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_addvi_h
argument_list|(
name|v8u16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_addvi_w
argument_list|(
name|v4u32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_addvi_d
argument_list|(
name|v2u64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_andi_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v8i16_r
operator|=
name|__msa_andi_b
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v4i32_r
operator|=
name|__msa_andi_b
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v2i64_r
operator|=
name|__msa_andi_b
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_bclri_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_bclri_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_bclri_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_bclri_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_binsli_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_binsli_h
argument_list|(
name|v8i16_r
argument_list|,
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_binsli_w
argument_list|(
name|v4i32_r
argument_list|,
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_binsli_d
argument_list|(
name|v2i64_r
argument_list|,
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_binsri_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_binsri_h
argument_list|(
name|v8i16_r
argument_list|,
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_binsri_w
argument_list|(
name|v4i32_r
argument_list|,
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_binsri_d
argument_list|(
name|v2i64_r
argument_list|,
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_bmnzi_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_bmzi_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_bnegi_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_bnegi_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_bnegi_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_bnegi_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_bseli_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_bseti_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_bseti_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_bseti_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_bseti_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_ceqi_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_ceqi_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_ceqi_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_ceqi_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16i8_r
operator|=
name|__msa_clei_s_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_clei_s_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_clei_s_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_clei_s_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16u8_r
operator|=
name|__msa_clei_u_b
argument_list|(
name|v16u8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_clei_u_h
argument_list|(
name|v8u16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_clei_u_w
argument_list|(
name|v4u32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_clei_u_d
argument_list|(
name|v2u64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_clti_s_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_clti_s_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_clti_s_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_clti_s_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16u8_r
operator|=
name|__msa_clti_u_b
argument_list|(
name|v16u8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_clti_u_h
argument_list|(
name|v8u16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_clti_u_w
argument_list|(
name|v4u32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_clti_u_d
argument_list|(
name|v2u64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|int_r
operator|=
name|__msa_copy_s_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|int_r
operator|=
name|__msa_copy_s_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|int_r
operator|=
name|__msa_copy_s_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|ll_r
operator|=
name|__msa_copy_s_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|int_r
operator|=
name|__msa_copy_u_b
argument_list|(
name|v16u8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|int_r
operator|=
name|__msa_copy_u_h
argument_list|(
name|v8u16_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|int_r
operator|=
name|__msa_copy_u_w
argument_list|(
name|v4u32_a
argument_list|,
operator|-
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|ll_r
operator|=
name|__msa_copy_u_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_insve_b
argument_list|(
name|v16i8_r
argument_list|,
literal|16
argument_list|,
name|v16i8_a
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v8i16_r
operator|=
name|__msa_insve_h
argument_list|(
name|v8i16_r
argument_list|,
literal|8
argument_list|,
name|v8i16_a
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v4i32_r
operator|=
name|__msa_insve_w
argument_list|(
name|v4i32_r
argument_list|,
literal|4
argument_list|,
name|v4i32_a
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|v2i64_r
operator|=
name|__msa_insve_d
argument_list|(
name|v2i64_r
argument_list|,
literal|2
argument_list|,
name|v2i64_a
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_ld_b
argument_list|(
operator|&
name|v16i8_a
argument_list|,
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v8i16_r
operator|=
name|__msa_ld_h
argument_list|(
operator|&
name|v8i16_a
argument_list|,
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v4i32_r
operator|=
name|__msa_ld_w
argument_list|(
operator|&
name|v4i32_a
argument_list|,
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v2i64_r
operator|=
name|__msa_ld_d
argument_list|(
operator|&
name|v2i64_a
argument_list|,
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v16i8_r
operator|=
name|__msa_ldi_b
argument_list|(
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v8i16_r
operator|=
name|__msa_ldi_h
argument_list|(
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v4i32_r
operator|=
name|__msa_ldi_w
argument_list|(
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v2i64_r
operator|=
name|__msa_ldi_d
argument_list|(
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v16i8_r
operator|=
name|__msa_maxi_s_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_maxi_s_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_maxi_s_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_maxi_s_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16u8_r
operator|=
name|__msa_maxi_u_b
argument_list|(
name|v16u8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_maxi_u_h
argument_list|(
name|v8u16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_maxi_u_w
argument_list|(
name|v4u32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_maxi_u_d
argument_list|(
name|v2u64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_mini_s_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v8i16_r
operator|=
name|__msa_mini_s_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v4i32_r
operator|=
name|__msa_mini_s_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v2i64_r
operator|=
name|__msa_mini_s_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -16 to 15}}
name|v16u8_r
operator|=
name|__msa_mini_u_b
argument_list|(
name|v16u8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8u16_r
operator|=
name|__msa_mini_u_h
argument_list|(
name|v8u16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4u32_r
operator|=
name|__msa_mini_u_w
argument_list|(
name|v4u32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2u64_r
operator|=
name|__msa_mini_u_d
argument_list|(
name|v2u64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_nori_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_ori_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_sat_s_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_sat_s_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_sat_s_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_sat_s_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_sat_u_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_sat_u_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_sat_u_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_sat_u_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_shf_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v8i16_r
operator|=
name|__msa_shf_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v4i32_r
operator|=
name|__msa_shf_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16i8_r
operator|=
name|__msa_sld_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v8i16_r
operator|=
name|__msa_sld_h
argument_list|(
name|v8i16_r
argument_list|,
name|v8i16_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v4i32_r
operator|=
name|__msa_sld_w
argument_list|(
name|v4i32_r
argument_list|,
name|v4i32_a
argument_list|,
operator|-
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|v2i64_r
operator|=
name|__msa_sld_d
argument_list|(
name|v2i64_r
argument_list|,
name|v2i64_a
argument_list|,
operator|-
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_sldi_b
argument_list|(
name|v16i8_r
argument_list|,
name|v16i8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v8i16_r
operator|=
name|__msa_sldi_h
argument_list|(
name|v8i16_r
argument_list|,
name|v8i16_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v4i32_r
operator|=
name|__msa_sldi_w
argument_list|(
name|v4i32_r
argument_list|,
name|v4i32_a
argument_list|,
operator|-
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|v2i64_r
operator|=
name|__msa_sldi_d
argument_list|(
name|v2i64_r
argument_list|,
name|v2i64_a
argument_list|,
operator|-
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_slli_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_slli_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_slli_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_slli_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_splati_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v8i16_r
operator|=
name|__msa_splati_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v4i32_r
operator|=
name|__msa_splati_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|v2i64_r
operator|=
name|__msa_splati_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|v16i8_r
operator|=
name|__msa_srai_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_srai_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_srai_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_srai_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_srari_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_srari_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_srari_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_srari_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_srli_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_srli_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_srli_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_srli_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|v16i8_r
operator|=
name|__msa_srlri_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|v8i16_r
operator|=
name|__msa_srlri_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|17
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|v4i32_r
operator|=
name|__msa_srlri_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_srlri_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|64
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 63}}
name|__msa_st_b
argument_list|(
name|v16i8_b
argument_list|,
operator|&
name|v16i8_a
argument_list|,
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|__msa_st_h
argument_list|(
name|v8i16_b
argument_list|,
operator|&
name|v8i16_a
argument_list|,
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|__msa_st_w
argument_list|(
name|v4i32_b
argument_list|,
operator|&
name|v4i32_a
argument_list|,
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|__msa_st_d
argument_list|(
name|v2i64_b
argument_list|,
operator|&
name|v2i64_a
argument_list|,
operator|-
literal|513
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from -512 to 511}}
name|v16i8_r
operator|=
name|__msa_subvi_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v8i16_r
operator|=
name|__msa_subvi_h
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v4i32_r
operator|=
name|__msa_subvi_w
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v2i64_r
operator|=
name|__msa_subvi_d
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
name|v16i8_r
operator|=
name|__msa_xori_b
argument_list|(
name|v16i8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v8i16_r
operator|=
name|__msa_xori_b
argument_list|(
name|v8i16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v4i32_r
operator|=
name|__msa_xori_b
argument_list|(
name|v4i32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v2i64_r
operator|=
name|__msa_xori_b
argument_list|(
name|v2i64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v16u8_r
operator|=
name|__msa_xori_b
argument_list|(
name|v16u8_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v8u16_r
operator|=
name|__msa_xori_b
argument_list|(
name|v8u16_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v4u32_r
operator|=
name|__msa_xori_b
argument_list|(
name|v4u32_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
name|v2u64_r
operator|=
name|__msa_xori_b
argument_list|(
name|v2u64_a
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// CHECK: warning: argument should be a value from 0 to 255}}
block|}
end_function

end_unit

