begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -analyze -analyzer-experimental-checks -analyzer-experimental-internal-checks -analyzer-check-objc-mem -analyzer-store=region -analyzer-constraints=basic -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -analyze -analyzer-experimental-checks -analyzer-experimental-internal-checks -analyzer-check-objc-mem -analyzer-store=region -analyzer-constraints=range -verify %s
end_comment

begin_function
name|int
name|string_literal_init
parameter_list|()
block|{
name|char
name|a
index|[]
init|=
literal|"abc"
decl_stmt|;
name|char
name|b
index|[
literal|2
index|]
init|=
literal|"abc"
decl_stmt|;
comment|// expected-warning{{too long}}
name|char
name|c
index|[
literal|5
index|]
init|=
literal|"abc"
decl_stmt|;
if|if
condition|(
name|a
index|[
literal|1
index|]
operator|!=
literal|'b'
condition|)
return|return
literal|0
return|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|b
index|[
literal|1
index|]
operator|!=
literal|'b'
condition|)
return|return
literal|0
return|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|c
index|[
literal|1
index|]
operator|!=
literal|'b'
condition|)
return|return
literal|0
return|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|a
index|[
literal|3
index|]
operator|!=
literal|0
condition|)
return|return
literal|0
return|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|c
index|[
literal|3
index|]
operator|!=
literal|0
condition|)
return|return
literal|0
return|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|c
index|[
literal|4
index|]
operator|!=
literal|0
condition|)
return|return
literal|0
return|;
comment|// expected-warning{{never executed}}
return|return
literal|42
return|;
block|}
end_function

begin_function
name|void
name|nested_compound_literals
parameter_list|(
name|int
name|rad
parameter_list|)
block|{
name|int
name|vec
index|[
literal|6
index|]
index|[
literal|2
index|]
init|=
block|{
block|{
literal|0.195
block|,
literal|0.02
block|}
block|,
block|{
literal|0.383
block|,
literal|0.067
block|}
block|,
block|{
literal|0.55
block|,
literal|0.169
block|}
block|,
block|{
literal|0.831
block|,
literal|0.45
block|}
block|,
block|{
literal|0.924
block|,
literal|0.617
block|}
block|,
block|{
literal|0.98
block|,
literal|0.805
block|}
block|}
decl_stmt|;
name|int
name|a
decl_stmt|;
for|for
control|(
name|a
operator|=
literal|0
init|;
name|a
operator|<
literal|6
condition|;
operator|++
name|a
control|)
block|{
name|vec
index|[
name|a
index|]
index|[
literal|0
index|]
operator|*=
name|rad
expr_stmt|;
comment|// no-warning
name|vec
index|[
name|a
index|]
index|[
literal|1
index|]
operator|*=
name|rad
expr_stmt|;
comment|// no-warning
block|}
block|}
end_function

begin_function
name|void
name|nested_compound_literals_float
parameter_list|(
name|float
name|rad
parameter_list|)
block|{
name|float
name|vec
index|[
literal|6
index|]
index|[
literal|2
index|]
init|=
block|{
block|{
literal|0.195
block|,
literal|0.02
block|}
block|,
block|{
literal|0.383
block|,
literal|0.067
block|}
block|,
block|{
literal|0.55
block|,
literal|0.169
block|}
block|,
block|{
literal|0.831
block|,
literal|0.45
block|}
block|,
block|{
literal|0.924
block|,
literal|0.617
block|}
block|,
block|{
literal|0.98
block|,
literal|0.805
block|}
block|}
decl_stmt|;
name|int
name|a
decl_stmt|;
for|for
control|(
name|a
operator|=
literal|0
init|;
name|a
operator|<
literal|6
condition|;
operator|++
name|a
control|)
block|{
name|vec
index|[
name|a
index|]
index|[
literal|0
index|]
operator|*=
name|rad
expr_stmt|;
comment|// no-warning
name|vec
index|[
name|a
index|]
index|[
literal|1
index|]
operator|*=
name|rad
expr_stmt|;
comment|// no-warning
block|}
block|}
end_function

end_unit

