begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -analyze -analyzer-experimental-internal-checks -analyzer-check-objc-mem -analyzer-experimental-checks -verify %s
end_comment

begin_comment
comment|// Trigger a warning if the analyzer reaches this point in the control flow.
end_comment

begin_define
define|#
directive|define
name|WARN
value|((void)*(char*)0)
end_define

begin_comment
comment|// There should be no warnings unless otherwise indicated.
end_comment

begin_function
name|void
name|testComparisons
parameter_list|(
name|int
name|a
parameter_list|)
block|{
comment|// Sema can already catch the simple comparison a==a,
comment|// since that's usually a logic error (and not path-dependent).
name|int
name|b
init|=
name|a
decl_stmt|;
if|if
condition|(
operator|!
operator|(
name|b
operator|==
name|a
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|!
operator|(
name|b
operator|>=
name|a
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|!
operator|(
name|b
operator|<=
name|a
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|b
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|b
operator|>
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|b
operator|<
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
block|}
end_function

begin_function
name|void
name|testSelfOperations
parameter_list|(
name|int
name|a
parameter_list|)
block|{
if|if
condition|(
operator|(
name|a
operator||
name|a
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|&
name|a
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|^
name|a
operator|)
operator|!=
literal|0
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|-
name|a
operator|)
operator|!=
literal|0
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
block|}
end_function

begin_function
name|void
name|testIdempotent
parameter_list|(
name|int
name|a
parameter_list|)
block|{
if|if
condition|(
operator|(
name|a
operator|*
literal|1
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|/
literal|1
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|+
literal|0
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|-
literal|0
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|<<
literal|0
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|>>
literal|0
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|^
literal|0
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|&
operator|(
operator|~
literal|0
operator|)
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator||
literal|0
operator|)
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
block|}
end_function

begin_function
name|void
name|testReductionToConstant
parameter_list|(
name|int
name|a
parameter_list|)
block|{
if|if
condition|(
operator|(
name|a
operator|*
literal|0
operator|)
operator|!=
literal|0
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator|&
literal|0
operator|)
operator|!=
literal|0
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
name|a
operator||
operator|(
operator|~
literal|0
operator|)
operator|)
operator|!=
operator|(
operator|~
literal|0
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
block|}
end_function

begin_function
name|void
name|testSymmetricIntSymOperations
parameter_list|(
name|int
name|a
parameter_list|)
block|{
if|if
condition|(
operator|(
literal|2
operator|+
name|a
operator|)
operator|!=
operator|(
name|a
operator|+
literal|2
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
literal|2
operator|*
name|a
operator|)
operator|!=
operator|(
name|a
operator|*
literal|2
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
literal|2
operator|&
name|a
operator|)
operator|!=
operator|(
name|a
operator|&
literal|2
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
literal|2
operator|^
name|a
operator|)
operator|!=
operator|(
name|a
operator|^
literal|2
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
literal|2
operator||
name|a
operator|)
operator|!=
operator|(
name|a
operator||
literal|2
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
block|}
end_function

begin_function
name|void
name|testAsymmetricIntSymOperations
parameter_list|(
name|int
name|a
parameter_list|)
block|{
if|if
condition|(
operator|(
operator|(
operator|~
literal|0
operator|)
operator|>>
name|a
operator|)
operator|!=
operator|(
operator|~
literal|0
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
literal|0
operator|>>
name|a
operator|)
operator|!=
literal|0
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|(
literal|0
operator|<<
name|a
operator|)
operator|!=
literal|0
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
comment|// Unsigned right shift shifts in zeroes.
if|if
condition|(
operator|(
operator|(
call|(
name|unsigned
call|)
argument_list|(
operator|~
literal|0
argument_list|)
operator|)
operator|>>
operator|(
operator|(
name|unsigned
operator|)
name|a
operator|)
operator|)
operator|!=
operator|(
call|(
name|unsigned
call|)
argument_list|(
operator|~
literal|0
argument_list|)
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{}}
block|}
end_function

begin_function
name|void
name|testLocations
parameter_list|(
name|char
modifier|*
name|a
parameter_list|)
block|{
name|char
modifier|*
name|b
init|=
name|a
decl_stmt|;
if|if
condition|(
operator|!
operator|(
name|b
operator|==
name|a
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|!
operator|(
name|b
operator|>=
name|a
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
operator|!
operator|(
name|b
operator|<=
name|a
operator|)
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|b
operator|!=
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|b
operator|>
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|b
operator|<
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
if|if
condition|(
name|b
operator|-
name|a
condition|)
name|WARN
expr_stmt|;
comment|// expected-warning{{never executed}}
block|}
end_function

end_unit

