begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_analyze_cc1 -triple x86_64-unknown-freebsd %s
end_comment

begin_include
include|#
directive|include
file|"Inputs/system-header-simulator.h"
end_include

begin_define
define|#
directive|define
name|M_ZERO
value|0x0100
end_define

begin_define
define|#
directive|define
name|NULL
value|((void *)0)
end_define

begin_function_decl
name|void
modifier|*
name|malloc
parameter_list|(
name|size_t
parameter_list|,
name|void
modifier|*
parameter_list|,
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_struct
struct|struct
name|test
block|{ }
struct|;
end_struct

begin_function_decl
name|void
name|foo
parameter_list|(
name|struct
name|test
modifier|*
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|test_zeroed
parameter_list|()
block|{
name|struct
name|test
modifier|*
modifier|*
name|list
decl_stmt|,
modifier|*
name|t
decl_stmt|;
name|int
name|i
decl_stmt|;
name|list
operator|=
name|malloc
argument_list|(
sizeof|sizeof
argument_list|(
operator|*
name|list
argument_list|)
operator|*
literal|10
argument_list|,
name|NULL
argument_list|,
name|M_ZERO
argument_list|)
expr_stmt|;
if|if
condition|(
name|list
operator|==
name|NULL
condition|)
return|return;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|10
condition|;
name|i
operator|++
control|)
block|{
name|t
operator|=
name|list
index|[
name|i
index|]
expr_stmt|;
name|foo
argument_list|(
name|t
argument_list|)
expr_stmt|;
block|}
name|free
argument_list|(
name|list
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_function
name|void
name|test_nonzero
parameter_list|()
block|{
name|struct
name|test
modifier|*
modifier|*
name|list
decl_stmt|,
modifier|*
name|t
decl_stmt|;
name|int
name|i
decl_stmt|;
name|list
operator|=
name|malloc
argument_list|(
sizeof|sizeof
argument_list|(
operator|*
name|list
argument_list|)
operator|*
literal|10
argument_list|,
name|NULL
argument_list|,
literal|0
argument_list|)
expr_stmt|;
if|if
condition|(
name|list
operator|==
name|NULL
condition|)
return|return;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|10
condition|;
name|i
operator|++
control|)
block|{
name|t
operator|=
name|list
index|[
name|i
index|]
expr_stmt|;
comment|// expected-warning{{undefined}}
name|foo
argument_list|(
name|t
argument_list|)
expr_stmt|;
block|}
name|free
argument_list|(
name|list
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test_indeterminate
parameter_list|(
name|int
name|flags
parameter_list|)
block|{
name|struct
name|test
modifier|*
modifier|*
name|list
decl_stmt|,
modifier|*
name|t
decl_stmt|;
name|int
name|i
decl_stmt|;
name|list
operator|=
name|malloc
argument_list|(
sizeof|sizeof
argument_list|(
operator|*
name|list
argument_list|)
operator|*
literal|10
argument_list|,
name|NULL
argument_list|,
name|flags
argument_list|)
expr_stmt|;
if|if
condition|(
name|list
operator|==
name|NULL
condition|)
return|return;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|10
condition|;
name|i
operator|++
control|)
block|{
name|t
operator|=
name|list
index|[
name|i
index|]
expr_stmt|;
comment|// expected-warning{{undefined}}
name|foo
argument_list|(
name|t
argument_list|)
expr_stmt|;
block|}
name|free
argument_list|(
name|list
argument_list|)
expr_stmt|;
block|}
end_function

end_unit

