begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -triple i386-apple-darwin10 -analyze -analyzer-experimental-internal-checks -std=gnu99 -checker-cfref -verify %s -analyzer-constraints=basic -analyzer-store=basic
end_comment

begin_comment
comment|// RUN: clang-cc -triple i386-apple-darwin10 -analyze -analyzer-experimental-internal-checks -std=gnu99 -checker-cfref -verify %s -analyzer-constraints=range -analyzer-store=basic
end_comment

begin_comment
comment|// RUN: clang-cc -triple i386-apple-darwin10 -analyze -analyzer-experimental-internal-checks -std=gnu99 -checker-cfref -analyzer-store=region -analyzer-constraints=range -analyzer-no-purge-dead -verify %s
end_comment

begin_comment
comment|// RUN: clang-cc -triple i386-apple-darwin10 -analyze -analyzer-experimental-internal-checks -std=gnu99 -checker-cfref -analyzer-store=region -analyzer-constraints=range -verify %s
end_comment

begin_typedef
typedef|typedef
name|unsigned
name|uintptr_t
typedef|;
end_typedef

begin_function_decl
specifier|extern
name|void
name|__assert_fail
parameter_list|(
name|__const
name|char
modifier|*
name|__assertion
parameter_list|,
name|__const
name|char
modifier|*
name|__file
parameter_list|,
name|unsigned
name|int
name|__line
parameter_list|,
name|__const
name|char
modifier|*
name|__function
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__noreturn__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_define
define|#
directive|define
name|assert
parameter_list|(
name|expr
parameter_list|)
define|\
value|((expr)  ? (void)(0)  : __assert_fail (#expr, __FILE__, __LINE__, __func__))
end_define

begin_function
name|void
name|f1
parameter_list|(
name|int
modifier|*
name|p
parameter_list|)
block|{
if|if
condition|(
name|p
condition|)
operator|*
name|p
operator|=
literal|1
expr_stmt|;
else|else
operator|*
name|p
operator|=
literal|0
expr_stmt|;
comment|// expected-warning{{ereference}}
block|}
end_function

begin_struct
struct|struct
name|foo_struct
block|{
name|int
name|x
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
name|f2
parameter_list|(
name|struct
name|foo_struct
modifier|*
name|p
parameter_list|)
block|{
if|if
condition|(
name|p
condition|)
name|p
operator|->
name|x
operator|=
literal|1
expr_stmt|;
return|return
name|p
operator|->
name|x
operator|++
return|;
comment|// expected-warning{{Dereference of null pointer}}
block|}
end_function

begin_function
name|int
name|f3
parameter_list|(
name|char
modifier|*
name|x
parameter_list|)
block|{
name|int
name|i
init|=
literal|2
decl_stmt|;
if|if
condition|(
name|x
condition|)
return|return
name|x
index|[
name|i
operator|-
literal|1
index|]
return|;
return|return
name|x
index|[
name|i
operator|+
literal|1
index|]
return|;
comment|// expected-warning{{Dereference of null pointer}}
block|}
end_function

begin_function
name|int
name|f3_b
parameter_list|(
name|char
modifier|*
name|x
parameter_list|)
block|{
name|int
name|i
init|=
literal|2
decl_stmt|;
if|if
condition|(
name|x
condition|)
return|return
name|x
index|[
name|i
operator|-
literal|1
index|]
return|;
return|return
name|x
index|[
name|i
operator|+
literal|1
index|]
operator|++
return|;
comment|// expected-warning{{Dereference of null pointer}}
block|}
end_function

begin_function
name|int
name|f4
parameter_list|(
name|int
modifier|*
name|p
parameter_list|)
block|{
name|uintptr_t
name|x
init|=
operator|(
name|uintptr_t
operator|)
name|p
decl_stmt|;
if|if
condition|(
name|x
condition|)
return|return
literal|1
return|;
name|int
modifier|*
name|q
init|=
operator|(
name|int
operator|*
operator|)
name|x
decl_stmt|;
return|return
operator|*
name|q
return|;
comment|// expected-warning{{Dereference of null pointer loaded from variable 'q'}}
block|}
end_function

begin_function
name|int
name|f4_b
parameter_list|()
block|{
name|short
name|array
index|[
literal|2
index|]
decl_stmt|;
name|uintptr_t
name|x
init|=
name|array
decl_stmt|;
comment|// expected-warning{{incompatible pointer to integer conversion initializing}}
name|short
modifier|*
name|p
init|=
name|x
decl_stmt|;
comment|// expected-warning{{incompatible integer to pointer conversion initializing}}
comment|// The following branch should be infeasible.
if|if
condition|(
operator|!
operator|(
name|p
operator|=
operator|&
name|array
index|[
literal|0
index|]
operator|)
condition|)
block|{
name|p
operator|=
literal|0
expr_stmt|;
operator|*
name|p
operator|=
literal|1
expr_stmt|;
comment|// no-warning
block|}
if|if
condition|(
name|p
condition|)
block|{
operator|*
name|p
operator|=
literal|5
expr_stmt|;
comment|// no-warning
name|p
operator|=
literal|0
expr_stmt|;
block|}
else|else
return|return;
comment|// expected-warning {{non-void function 'f4_b' should return a value}}
operator|*
name|p
operator|+=
literal|10
expr_stmt|;
comment|// expected-warning{{Dereference of null pointer}}
return|return
literal|0
return|;
block|}
end_function

begin_function
name|int
name|f5
parameter_list|()
block|{
name|char
modifier|*
name|s
init|=
literal|"hello world"
decl_stmt|;
return|return
name|s
index|[
literal|0
index|]
return|;
comment|// no-warning
block|}
end_function

begin_function_decl
name|int
name|bar
parameter_list|(
name|int
modifier|*
name|p
parameter_list|,
name|int
name|q
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function
name|int
name|f6
parameter_list|(
name|int
modifier|*
name|p
parameter_list|)
block|{
return|return
operator|!
name|p
condition|?
name|bar
argument_list|(
name|p
argument_list|,
literal|1
argument_list|)
comment|// expected-warning {{Null pointer passed as an argument to a 'nonnull' parameter}}
else|:
name|bar
argument_list|(
name|p
argument_list|,
literal|0
argument_list|)
return|;
comment|// no-warning
block|}
end_function

begin_function_decl
name|int
name|bar2
parameter_list|(
name|int
modifier|*
name|p
parameter_list|,
name|int
name|q
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
parameter_list|(
function_decl|1
end_function_decl

begin_empty_stmt
unit|)))
empty_stmt|;
end_empty_stmt

begin_function
name|int
name|f6b
parameter_list|(
name|int
modifier|*
name|p
parameter_list|)
block|{
return|return
operator|!
name|p
condition|?
name|bar2
argument_list|(
name|p
argument_list|,
literal|1
argument_list|)
comment|// expected-warning {{Null pointer passed as an argument to a 'nonnull' parameter}}
else|:
name|bar2
argument_list|(
name|p
argument_list|,
literal|0
argument_list|)
return|;
comment|// no-warning
block|}
end_function

begin_function_decl
name|int
name|bar3
parameter_list|(
name|int
modifier|*
name|p
parameter_list|,
name|int
name|q
parameter_list|,
name|int
modifier|*
name|r
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
parameter_list|(
function_decl|1
operator|,
function_decl|3
end_function_decl

begin_empty_stmt
unit|)))
empty_stmt|;
end_empty_stmt

begin_function
name|int
name|f6c
parameter_list|(
name|int
modifier|*
name|p
parameter_list|,
name|int
modifier|*
name|q
parameter_list|)
block|{
return|return
operator|!
name|p
condition|?
name|bar3
argument_list|(
name|q
argument_list|,
literal|2
argument_list|,
name|p
argument_list|)
comment|// expected-warning {{Null pointer passed as an argument to a 'nonnull' parameter}}
else|:
name|bar3
argument_list|(
name|p
argument_list|,
literal|2
argument_list|,
name|q
argument_list|)
return|;
comment|// no-warning
block|}
end_function

begin_function
name|void
name|f6d
parameter_list|(
name|int
modifier|*
name|p
parameter_list|)
block|{
name|bar
argument_list|(
name|p
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// At this point, 'p' cannot be null.
if|if
condition|(
operator|!
name|p
condition|)
block|{
name|int
modifier|*
name|q
init|=
literal|0
decl_stmt|;
operator|*
name|q
operator|=
literal|0xDEADBEEF
expr_stmt|;
comment|// no-warning
block|}
block|}
end_function

begin_function_decl
name|int
modifier|*
name|qux
parameter_list|()
function_decl|;
end_function_decl

begin_function
name|int
name|f7
parameter_list|(
name|int
name|x
parameter_list|)
block|{
name|int
modifier|*
name|p
init|=
literal|0
decl_stmt|;
if|if
condition|(
literal|0
operator|==
name|x
condition|)
name|p
operator|=
name|qux
argument_list|()
expr_stmt|;
if|if
condition|(
literal|0
operator|==
name|x
condition|)
operator|*
name|p
operator|=
literal|1
expr_stmt|;
comment|// no-warning
return|return
name|x
return|;
block|}
end_function

begin_function
name|int
modifier|*
name|f7b
parameter_list|(
name|int
modifier|*
name|x
parameter_list|)
block|{
name|int
modifier|*
name|p
init|=
literal|0
decl_stmt|;
if|if
condition|(
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|==
name|x
condition|)
name|p
operator|=
name|qux
argument_list|()
expr_stmt|;
if|if
condition|(
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|==
name|x
condition|)
operator|*
name|p
operator|=
literal|1
expr_stmt|;
comment|// no-warning
return|return
name|x
return|;
block|}
end_function

begin_function
name|int
modifier|*
name|f7c
parameter_list|(
name|int
modifier|*
name|x
parameter_list|)
block|{
name|int
modifier|*
name|p
init|=
literal|0
decl_stmt|;
if|if
condition|(
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|==
name|x
condition|)
name|p
operator|=
name|qux
argument_list|()
expr_stmt|;
if|if
condition|(
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|!=
name|x
condition|)
return|return
name|x
return|;
comment|// If we reach here then 'p' is not null.
operator|*
name|p
operator|=
literal|1
expr_stmt|;
comment|// no-warning
return|return
name|x
return|;
block|}
end_function

begin_function
name|int
modifier|*
name|f7c2
parameter_list|(
name|int
modifier|*
name|x
parameter_list|)
block|{
name|int
modifier|*
name|p
init|=
literal|0
decl_stmt|;
if|if
condition|(
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|==
name|x
condition|)
name|p
operator|=
name|qux
argument_list|()
expr_stmt|;
if|if
condition|(
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|==
name|x
condition|)
return|return
name|x
return|;
operator|*
name|p
operator|=
literal|1
expr_stmt|;
comment|// expected-warning{{null}}
return|return
name|x
return|;
block|}
end_function

begin_function
name|void
name|f8
parameter_list|(
name|int
modifier|*
name|p
parameter_list|,
name|int
modifier|*
name|q
parameter_list|)
block|{
if|if
condition|(
operator|!
name|p
condition|)
if|if
condition|(
name|p
condition|)
operator|*
name|p
operator|=
literal|1
expr_stmt|;
comment|// no-warning
if|if
condition|(
name|q
condition|)
if|if
condition|(
operator|!
name|q
condition|)
operator|*
name|q
operator|=
literal|1
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_function_decl
name|int
modifier|*
name|qux
parameter_list|()
function_decl|;
end_function_decl

begin_function
name|int
name|f9
parameter_list|(
name|unsigned
name|len
parameter_list|)
block|{
name|assert
argument_list|(
name|len
operator|!=
literal|0
argument_list|)
expr_stmt|;
name|int
modifier|*
name|p
init|=
literal|0
decl_stmt|;
name|unsigned
name|i
decl_stmt|;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
name|len
condition|;
operator|++
name|i
control|)
name|p
operator|=
name|qux
argument_list|(
name|i
argument_list|)
expr_stmt|;
return|return
operator|*
name|p
operator|++
return|;
comment|// no-warning
block|}
end_function

begin_function
name|int
name|f9b
parameter_list|(
name|unsigned
name|len
parameter_list|)
block|{
name|assert
argument_list|(
name|len
operator|>
literal|0
argument_list|)
expr_stmt|;
comment|// note use of '>'
name|int
modifier|*
name|p
init|=
literal|0
decl_stmt|;
name|unsigned
name|i
decl_stmt|;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
name|len
condition|;
operator|++
name|i
control|)
name|p
operator|=
name|qux
argument_list|(
name|i
argument_list|)
expr_stmt|;
return|return
operator|*
name|p
operator|++
return|;
comment|// no-warning
block|}
end_function

begin_function
name|int
modifier|*
name|f10
parameter_list|(
name|int
modifier|*
name|p
parameter_list|,
name|signed
name|char
name|x
parameter_list|,
name|int
name|y
parameter_list|)
block|{
comment|// This line tests symbolication with compound assignments where the
comment|// LHS and RHS have different bitwidths.  The new symbolic value
comment|// for 'x' should have a bitwidth of 8.
name|x
operator|&=
name|y
expr_stmt|;
comment|// This tests that our symbolication worked, and that we correctly test
comment|// x against 0 (with the same bitwidth).
if|if
condition|(
operator|!
name|x
condition|)
block|{
if|if
condition|(
operator|!
name|p
condition|)
return|return;
comment|// expected-warning {{non-void function 'f10' should return a value}}
operator|*
name|p
operator|=
literal|10
expr_stmt|;
block|}
else|else
name|p
operator|=
literal|0
expr_stmt|;
if|if
condition|(
operator|!
name|x
condition|)
operator|*
name|p
operator|=
literal|5
expr_stmt|;
comment|// no-warning
return|return
name|p
return|;
block|}
end_function

begin_comment
comment|// Test case from<rdar://problem/6407949>
end_comment

begin_function
name|void
name|f11
parameter_list|(
name|unsigned
name|i
parameter_list|)
block|{
name|int
modifier|*
name|x
init|=
literal|0
decl_stmt|;
if|if
condition|(
name|i
operator|>=
literal|0
condition|)
block|{
comment|// always true
block|}
else|else
block|{
operator|*
name|x
operator|=
literal|42
expr_stmt|;
comment|// no-warning
block|}
block|}
end_function

begin_function
name|void
name|f11b
parameter_list|(
name|unsigned
name|i
parameter_list|)
block|{
name|int
modifier|*
name|x
init|=
literal|0
decl_stmt|;
if|if
condition|(
name|i
operator|<=
operator|~
operator|(
name|unsigned
operator|)
literal|0
condition|)
block|{
comment|// always true
block|}
else|else
block|{
operator|*
name|x
operator|=
literal|42
expr_stmt|;
comment|// no-warning
block|}
block|}
end_function

begin_comment
comment|// Test case for switch statements with weird case arms.
end_comment

begin_typedef
typedef|typedef
name|int
name|BOOL
typedef|,
modifier|*
name|PBOOL
typedef|,
modifier|*
name|LPBOOL
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|long
name|LONG_PTR
typedef|,
modifier|*
name|PLONG_PTR
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|long
name|ULONG_PTR
typedef|,
modifier|*
name|PULONG_PTR
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|ULONG_PTR
name|DWORD_PTR
typedef|,
modifier|*
name|PDWORD_PTR
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|LONG_PTR
name|LRESULT
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|struct
name|_F12ITEM
modifier|*
name|HF12ITEM
typedef|;
end_typedef

begin_function
name|void
name|f12
parameter_list|(
name|HF12ITEM
name|i
parameter_list|,
name|char
modifier|*
name|q
parameter_list|)
block|{
name|char
modifier|*
name|p
init|=
literal|0
decl_stmt|;
switch|switch
condition|(
operator|(
name|DWORD_PTR
operator|)
name|i
condition|)
block|{
case|case
literal|0
operator|...
literal|10
case|:
name|p
operator|=
name|q
expr_stmt|;
break|break;
case|case
call|(
name|DWORD_PTR
call|)
argument_list|(
operator|(
name|HF12ITEM
operator|)
operator|-
literal|65535
argument_list|)
case|:
return|return;
default|default:
return|return;
block|}
operator|*
name|p
operator|=
literal|1
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_comment
comment|// Test handling of translating between integer "pointers" and back.
end_comment

begin_function
name|void
name|f13
parameter_list|()
block|{
name|int
modifier|*
name|x
init|=
literal|0
decl_stmt|;
if|if
condition|(
operator|(
operator|(
operator|(
operator|(
name|int
operator|)
name|x
operator|)
operator|<<
literal|2
operator|)
operator|+
literal|1
operator|)
operator|>>
literal|1
condition|)
operator|*
name|x
operator|=
literal|1
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_comment
comment|// PR 4759 - Attribute non-null checking by the analyzer was not correctly
end_comment

begin_comment
comment|// handling pointer values that were undefined.
end_comment

begin_function_decl
name|void
name|pr4759_aux
parameter_list|(
name|int
modifier|*
name|p
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function
name|void
name|pr4759
parameter_list|()
block|{
name|int
modifier|*
name|p
decl_stmt|;
name|pr4759_aux
argument_list|(
name|p
argument_list|)
expr_stmt|;
comment|// expected-warning{{undefined}}
block|}
end_function

end_unit

