begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -analyze -checker-cfref -analyzer-store=basic -verify %s&&
end_comment

begin_comment
comment|// RUN: clang-cc -analyze -checker-cfref -analyzer-store=region -verify %s
end_comment

begin_include
include|#
directive|include
file|<stdlib.h>
end_include

begin_function
name|int
modifier|*
name|f1
parameter_list|()
block|{
name|int
name|x
init|=
literal|0
decl_stmt|;
return|return
operator|&
name|x
return|;
comment|// expected-warning{{Address of stack memory associated with local variable 'x' returned.}} expected-warning{{address of stack memory associated with local variable 'x' returned}}
block|}
end_function

begin_function
name|int
modifier|*
name|f2
parameter_list|(
name|int
name|y
parameter_list|)
block|{
return|return
operator|&
name|y
return|;
comment|// expected-warning{{Address of stack memory associated with local variable 'y' returned.}} expected-warning{{address of stack memory associated with local variable 'y' returned}}
block|}
end_function

begin_function
name|int
modifier|*
name|f3
parameter_list|(
name|int
name|x
parameter_list|,
name|int
modifier|*
name|y
parameter_list|)
block|{
name|int
name|w
init|=
literal|0
decl_stmt|;
if|if
condition|(
name|x
condition|)
name|y
operator|=
operator|&
name|w
expr_stmt|;
return|return
name|y
return|;
comment|// expected-warning{{Address of stack memory associated with local variable 'w' returned.}}
block|}
end_function

begin_function
name|void
modifier|*
name|compound_literal
parameter_list|(
name|int
name|x
parameter_list|,
name|int
name|y
parameter_list|)
block|{
if|if
condition|(
name|x
condition|)
return|return
operator|&
operator|(
name|unsigned
name|short
operator|)
block|{
operator|(
operator|(
name|unsigned
name|short
operator|)
literal|0x22EF
operator|)
block|}
return|;
comment|// expected-warning{{Address of stack memory}}
name|int
modifier|*
name|array
index|[]
init|=
block|{}
decl_stmt|;
struct|struct
name|s
block|{
name|int
name|z
decl_stmt|;
name|double
name|y
decl_stmt|;
name|int
name|w
decl_stmt|;
block|}
struct|;
if|if
condition|(
name|y
condition|)
return|return
operator|&
operator|(
operator|(
expr|struct
name|s
operator|)
block|{
literal|2
block|,
literal|0.4
block|,
literal|5
operator|*
literal|8
block|}
operator|)
return|;
comment|// expected-warning{{Address of stack memory}}
name|void
modifier|*
name|p
init|=
operator|&
operator|(
operator|(
expr|struct
name|s
operator|)
block|{
literal|42
block|,
literal|0.4
block|,
name|x
condition|?
literal|42
else|:
literal|0
block|}
operator|)
decl_stmt|;
return|return
name|p
return|;
comment|// expected-warning{{Address of stack memory}}
block|}
end_function

begin_function
name|void
modifier|*
name|alloca_test
parameter_list|()
block|{
name|void
modifier|*
name|p
init|=
name|__builtin_alloca
argument_list|(
literal|10
argument_list|)
decl_stmt|;
return|return
name|p
return|;
comment|// expected-warning{{Address of stack memory}}
block|}
end_function

end_unit

