begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -analyze -checker-cfref -analyzer-store=basic -fblocks -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -analyze -checker-cfref -analyzer-store=region -fblocks -verify %s
end_comment

begin_function
name|int
modifier|*
name|f1
parameter_list|()
block|{
name|int
name|x
init|=
literal|0
decl_stmt|;
return|return
operator|&
name|x
return|;
comment|// expected-warning{{Address of stack memory associated with local variable 'x' returned.}} expected-warning{{address of stack memory associated with local variable 'x' returned}}
block|}
end_function

begin_function
name|int
modifier|*
name|f2
parameter_list|(
name|int
name|y
parameter_list|)
block|{
return|return
operator|&
name|y
return|;
comment|// expected-warning{{Address of stack memory associated with local variable 'y' returned.}} expected-warning{{address of stack memory associated with local variable 'y' returned}}
block|}
end_function

begin_function
name|int
modifier|*
name|f3
parameter_list|(
name|int
name|x
parameter_list|,
name|int
modifier|*
name|y
parameter_list|)
block|{
name|int
name|w
init|=
literal|0
decl_stmt|;
if|if
condition|(
name|x
condition|)
name|y
operator|=
operator|&
name|w
expr_stmt|;
return|return
name|y
return|;
comment|// expected-warning{{Address of stack memory associated with local variable 'w' returned.}}
block|}
end_function

begin_function
name|void
modifier|*
name|compound_literal
parameter_list|(
name|int
name|x
parameter_list|,
name|int
name|y
parameter_list|)
block|{
if|if
condition|(
name|x
condition|)
return|return
operator|&
operator|(
name|unsigned
name|short
operator|)
block|{
operator|(
operator|(
name|unsigned
name|short
operator|)
literal|0x22EF
operator|)
block|}
return|;
comment|// expected-warning{{Address of stack memory}}
name|int
modifier|*
name|array
index|[]
init|=
block|{}
decl_stmt|;
struct|struct
name|s
block|{
name|int
name|z
decl_stmt|;
name|double
name|y
decl_stmt|;
name|int
name|w
decl_stmt|;
block|}
struct|;
if|if
condition|(
name|y
condition|)
return|return
operator|&
operator|(
operator|(
expr|struct
name|s
operator|)
block|{
literal|2
block|,
literal|0.4
block|,
literal|5
operator|*
literal|8
block|}
operator|)
return|;
comment|// expected-warning{{Address of stack memory}}
name|void
modifier|*
name|p
init|=
operator|&
operator|(
operator|(
expr|struct
name|s
operator|)
block|{
literal|42
block|,
literal|0.4
block|,
name|x
condition|?
literal|42
else|:
literal|0
block|}
operator|)
decl_stmt|;
return|return
name|p
return|;
comment|// expected-warning{{Address of stack memory}}
block|}
end_function

begin_function
name|void
modifier|*
name|alloca_test
parameter_list|()
block|{
name|void
modifier|*
name|p
init|=
name|__builtin_alloca
argument_list|(
literal|10
argument_list|)
decl_stmt|;
return|return
name|p
return|;
comment|// expected-warning{{Address of stack memory}}
block|}
end_function

begin_function
name|int
name|array_test
parameter_list|(
name|int
name|x
index|[
literal|2
index|]
parameter_list|)
block|{
return|return
name|x
index|[
literal|0
index|]
return|;
comment|// no-warning
block|}
end_function

begin_struct
struct|struct
name|baz
block|{
name|int
name|x
decl_stmt|;
name|int
name|y
index|[
literal|2
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
name|struct_test
parameter_list|(
name|struct
name|baz
name|byVal
parameter_list|,
name|int
name|flag
parameter_list|)
block|{
if|if
condition|(
name|flag
condition|)
return|return
name|byVal
operator|.
name|x
return|;
comment|// no-warning
else|else
block|{
return|return
name|byVal
operator|.
name|y
index|[
literal|0
index|]
return|;
comment|// no-warning
block|}
block|}
end_function

begin_typedef
typedef|typedef
name|int
function_decl|(
modifier|^
name|ComparatorBlock
function_decl|)
parameter_list|(
name|int
name|a
parameter_list|,
name|int
name|b
parameter_list|)
function_decl|;
end_typedef

begin_function
name|ComparatorBlock
name|test_return_block
parameter_list|(
name|void
parameter_list|)
block|{
name|ComparatorBlock
name|b
init|=
lambda|^
name|int
parameter_list|(
name|int
name|a
parameter_list|,
name|int
name|b
parameter_list|)
block|{
return|return
name|a
operator|>
name|b
return|;
block|}
decl_stmt|;
return|return
name|b
return|;
comment|// expected-warning{{Address of stack-allocated block declared on line 61 returned to caller}}
block|}
end_function

begin_function_decl
name|ComparatorBlock
name|test_return_block_neg_aux
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|ComparatorBlock
name|test_return_block_neg
parameter_list|(
name|void
parameter_list|)
block|{
name|ComparatorBlock
name|b
init|=
name|test_return_block_neg_aux
argument_list|()
decl_stmt|;
return|return
name|b
return|;
comment|// no-warning
block|}
end_function

end_unit

