begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -analyze -analyzer-check-objc-mem -analyzer-experimental-checks -analyzer-store region -verify %s
end_comment

begin_typedef
typedef|typedef
name|__typeof__
argument_list|(
argument|sizeof(int)
argument_list|)
name|size_t
expr_stmt|;
end_typedef

begin_typedef
typedef|typedef
name|struct
name|_IO_FILE
name|FILE
typedef|;
end_typedef

begin_define
define|#
directive|define
name|SEEK_SET
value|0
end_define

begin_comment
comment|/* Seek from beginning of file.  */
end_comment

begin_define
define|#
directive|define
name|SEEK_CUR
value|1
end_define

begin_comment
comment|/* Seek from current position.  */
end_comment

begin_define
define|#
directive|define
name|SEEK_END
value|2
end_define

begin_comment
comment|/* Seek from end of file.  */
end_comment

begin_function_decl
specifier|extern
name|FILE
modifier|*
name|fopen
parameter_list|(
specifier|const
name|char
modifier|*
name|path
parameter_list|,
specifier|const
name|char
modifier|*
name|mode
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
specifier|extern
name|size_t
name|fread
parameter_list|(
name|void
modifier|*
name|ptr
parameter_list|,
name|size_t
name|size
parameter_list|,
name|size_t
name|nmemb
parameter_list|,
name|FILE
modifier|*
name|stream
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
specifier|extern
name|int
name|fseek
parameter_list|(
name|FILE
modifier|*
name|__stream
parameter_list|,
name|long
name|int
name|__off
parameter_list|,
name|int
name|__whence
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
specifier|extern
name|long
name|int
name|ftell
parameter_list|(
name|FILE
modifier|*
name|__stream
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
specifier|extern
name|void
name|rewind
parameter_list|(
name|FILE
modifier|*
name|__stream
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|f1
parameter_list|(
name|void
parameter_list|)
block|{
name|FILE
modifier|*
name|p
init|=
name|fopen
argument_list|(
literal|"foo"
argument_list|,
literal|"r"
argument_list|)
decl_stmt|;
name|char
name|buf
index|[
literal|1024
index|]
decl_stmt|;
name|fread
argument_list|(
name|buf
argument_list|,
literal|1
argument_list|,
literal|1
argument_list|,
name|p
argument_list|)
expr_stmt|;
comment|// expected-warning {{Stream pointer might be NULL.}}
block|}
end_function

begin_function
name|void
name|f2
parameter_list|(
name|void
parameter_list|)
block|{
name|FILE
modifier|*
name|p
init|=
name|fopen
argument_list|(
literal|"foo"
argument_list|,
literal|"r"
argument_list|)
decl_stmt|;
name|fseek
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|SEEK_SET
argument_list|)
expr_stmt|;
comment|// expected-warning {{Stream pointer might be NULL.}}
block|}
end_function

begin_function
name|void
name|f3
parameter_list|(
name|void
parameter_list|)
block|{
name|FILE
modifier|*
name|p
init|=
name|fopen
argument_list|(
literal|"foo"
argument_list|,
literal|"r"
argument_list|)
decl_stmt|;
name|ftell
argument_list|(
name|p
argument_list|)
expr_stmt|;
comment|// expected-warning {{Stream pointer might be NULL.}}
block|}
end_function

begin_function
name|void
name|f4
parameter_list|(
name|void
parameter_list|)
block|{
name|FILE
modifier|*
name|p
init|=
name|fopen
argument_list|(
literal|"foo"
argument_list|,
literal|"r"
argument_list|)
decl_stmt|;
name|rewind
argument_list|(
name|p
argument_list|)
expr_stmt|;
comment|// expected-warning {{Stream pointer might be NULL.}}
block|}
end_function

begin_function
name|void
name|f5
parameter_list|(
name|void
parameter_list|)
block|{
name|FILE
modifier|*
name|p
init|=
name|fopen
argument_list|(
literal|"foo"
argument_list|,
literal|"r"
argument_list|)
decl_stmt|;
if|if
condition|(
operator|!
name|p
condition|)
return|return;
name|fseek
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|SEEK_SET
argument_list|)
expr_stmt|;
comment|// no-warning
name|fseek
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
literal|3
argument_list|)
expr_stmt|;
comment|// expected-warning {{The whence argument to fseek() should be SEEK_SET, SEEK_END, or SEEK_CUR.}}
block|}
end_function

end_unit

