begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -analyze -checker-cfref -verify %s&&
end_comment

begin_comment
comment|// RUN: clang-cc -analyze -checker-cfref -analyzer-store=region -verify %s
end_comment

begin_struct
struct|struct
name|FPRec
block|{
name|void
function_decl|(
modifier|*
name|my_func
function_decl|)
parameter_list|(
name|int
modifier|*
name|x
parameter_list|)
function_decl|;
block|}
struct|;
end_struct

begin_function_decl
name|int
name|bar
parameter_list|(
name|int
name|x
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|int
name|f1_a
parameter_list|(
name|struct
name|FPRec
modifier|*
name|foo
parameter_list|)
block|{
name|int
name|x
decl_stmt|;
call|(
modifier|*
name|foo
operator|->
name|my_func
call|)
argument_list|(
operator|&
name|x
argument_list|)
expr_stmt|;
return|return
name|bar
argument_list|(
name|x
argument_list|)
operator|+
literal|1
return|;
comment|// no-warning
block|}
end_function

begin_function
name|int
name|f1_b
parameter_list|()
block|{
name|int
name|x
decl_stmt|;
return|return
name|bar
argument_list|(
name|x
argument_list|)
operator|+
literal|1
return|;
comment|// expected-warning{{Pass-by-value argument in function call is undefined.}}
block|}
end_function

begin_function
name|int
name|f2
parameter_list|()
block|{
name|int
name|x
decl_stmt|;
if|if
condition|(
name|x
operator|+
literal|1
condition|)
comment|// expected-warning{{Branch}}
return|return
literal|1
return|;
return|return
literal|2
return|;
block|}
end_function

begin_function
name|int
name|f2_b
parameter_list|()
block|{
name|int
name|x
decl_stmt|;
return|return
operator|(
operator|(
name|x
operator|+
literal|1
operator|)
operator|+
literal|2
operator|+
operator|(
operator|(
name|x
operator|)
operator|)
operator|)
operator|+
literal|1
condition|?
literal|1
else|:
literal|2
return|;
comment|// expected-warning{{Branch}}
block|}
end_function

begin_function
name|int
name|f3
parameter_list|(
name|void
parameter_list|)
block|{
name|int
name|i
decl_stmt|;
name|int
modifier|*
name|p
init|=
operator|&
name|i
decl_stmt|;
if|if
condition|(
operator|*
name|p
operator|>
literal|0
condition|)
comment|// expected-warning{{Branch condition evaluates to an uninitialized value}}
return|return
literal|0
return|;
else|else
return|return
literal|1
return|;
block|}
end_function

begin_function_decl
name|void
name|f4_aux
parameter_list|(
name|float
modifier|*
name|x
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|float
name|f4
parameter_list|(
name|void
parameter_list|)
block|{
name|float
name|x
decl_stmt|;
name|f4_aux
argument_list|(
operator|&
name|x
argument_list|)
expr_stmt|;
return|return
name|x
return|;
comment|// no-warning
block|}
end_function

begin_struct
struct|struct
name|f5_struct
block|{
name|int
name|x
decl_stmt|;
block|}
struct|;
end_struct

begin_function_decl
name|void
name|f5_aux
parameter_list|(
name|struct
name|f5_struct
modifier|*
name|s
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|int
name|f5
parameter_list|(
name|void
parameter_list|)
block|{
name|struct
name|f5_struct
name|s
decl_stmt|;
name|f5_aux
argument_list|(
operator|&
name|s
argument_list|)
expr_stmt|;
return|return
name|s
operator|.
name|x
return|;
comment|// no-warning
block|}
end_function

begin_function
name|int
name|ret_uninit
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
name|int
modifier|*
name|p
init|=
operator|&
name|i
decl_stmt|;
return|return
operator|*
name|p
return|;
comment|// expected-warning{{Uninitialized or undefined value returned to caller.}}
block|}
end_function

begin_comment
comment|//<rdar://problem/6451816>
end_comment

begin_typedef
typedef|typedef
name|unsigned
name|char
name|Boolean
typedef|;
end_typedef

begin_typedef
typedef|typedef
specifier|const
name|struct
name|__CFNumber
modifier|*
name|CFNumberRef
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|signed
name|long
name|CFIndex
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|CFIndex
name|CFNumberType
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|long
name|UInt32
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|UInt32
name|CFStringEncoding
typedef|;
end_typedef

begin_typedef
typedef|typedef
specifier|const
name|struct
name|__CFString
modifier|*
name|CFStringRef
typedef|;
end_typedef

begin_function_decl
specifier|extern
name|Boolean
name|CFNumberGetValue
parameter_list|(
name|CFNumberRef
name|number
parameter_list|,
name|CFNumberType
name|theType
parameter_list|,
name|void
modifier|*
name|valuePtr
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
specifier|extern
name|CFStringRef
name|CFStringConvertEncodingToIANACharSetName
parameter_list|(
name|CFStringEncoding
name|encoding
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|CFStringRef
name|rdar_6451816
parameter_list|(
name|CFNumberRef
name|nr
parameter_list|)
block|{
name|CFStringEncoding
name|encoding
decl_stmt|;
comment|//&encoding is casted to void*.  This test case tests whether or not
comment|// we properly invalidate the value of 'encoding'.
name|CFNumberGetValue
argument_list|(
name|nr
argument_list|,
literal|9
argument_list|,
operator|&
name|encoding
argument_list|)
expr_stmt|;
return|return
name|CFStringConvertEncodingToIANACharSetName
argument_list|(
name|encoding
argument_list|)
return|;
comment|// no-warning
block|}
end_function

end_unit

