begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -analyze -checker-cfref -analyzer-store=region -verify %s
end_comment

begin_struct
struct|struct
name|s
block|{
name|int
name|data
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|s
name|global
decl_stmt|;
end_decl_stmt

begin_function_decl
name|void
name|g
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|f4
parameter_list|()
block|{
name|int
name|a
decl_stmt|;
if|if
condition|(
name|global
operator|.
name|data
operator|==
literal|0
condition|)
name|a
operator|=
literal|3
expr_stmt|;
if|if
condition|(
name|global
operator|.
name|data
operator|==
literal|0
condition|)
comment|// When the true branch is feasible 'a = 3'.
name|g
argument_list|(
name|a
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_comment
comment|// Test uninitialized value due to part of the structure being uninitialized.
end_comment

begin_struct
struct|struct
name|TestUninit
block|{
name|int
name|x
decl_stmt|;
name|int
name|y
decl_stmt|;
block|}
struct|;
end_struct

begin_function_decl
name|struct
name|TestUninit
name|test_uninit_aux
parameter_list|()
function_decl|;
end_function_decl

begin_function_decl
name|void
name|test_unit_aux2
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|test_uninit_pos
parameter_list|()
block|{
name|struct
name|TestUninit
name|v1
init|=
block|{
literal|0
block|,
literal|0
block|}
decl_stmt|;
name|struct
name|TestUninit
name|v2
init|=
name|test_uninit_aux
argument_list|()
decl_stmt|;
name|int
name|z
decl_stmt|;
name|v1
operator|.
name|y
operator|=
name|z
expr_stmt|;
comment|// expected-warning{{Assigned value is garbage or undefined}}
name|test_unit_aux2
argument_list|(
name|v2
operator|.
name|x
operator|+
name|v1
operator|.
name|y
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test_uninit_pos_2
parameter_list|()
block|{
name|struct
name|TestUninit
name|v1
init|=
block|{
literal|0
block|,
literal|0
block|}
decl_stmt|;
name|struct
name|TestUninit
name|v2
decl_stmt|;
name|test_unit_aux2
argument_list|(
name|v2
operator|.
name|x
operator|+
name|v1
operator|.
name|y
argument_list|)
expr_stmt|;
comment|// expected-warning{{The left operand of '+' is a garbage value}}
block|}
end_function

begin_function
name|void
name|test_uninit_pos_3
parameter_list|()
block|{
name|struct
name|TestUninit
name|v1
init|=
block|{
literal|0
block|,
literal|0
block|}
decl_stmt|;
name|struct
name|TestUninit
name|v2
decl_stmt|;
name|test_unit_aux2
argument_list|(
name|v1
operator|.
name|y
operator|+
name|v2
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning{{The right operand of '+' is a garbage value}}
block|}
end_function

begin_function
name|void
name|test_uninit_neg
parameter_list|()
block|{
name|struct
name|TestUninit
name|v1
init|=
block|{
literal|0
block|,
literal|0
block|}
decl_stmt|;
name|struct
name|TestUninit
name|v2
init|=
name|test_uninit_aux
argument_list|()
decl_stmt|;
name|test_unit_aux2
argument_list|(
name|v2
operator|.
name|x
operator|+
name|v1
operator|.
name|y
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
end_function

end_unit

