begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang -cc1 -analyze -warn-uninit-values -verify %s
end_comment

begin_function
name|int
name|f1
parameter_list|()
block|{
name|int
name|x
decl_stmt|;
return|return
name|x
return|;
comment|// expected-warning {{use of uninitialized variable}}
block|}
end_function

begin_function
name|int
name|f2
parameter_list|(
name|int
name|x
parameter_list|)
block|{
name|int
name|y
decl_stmt|;
name|int
name|z
init|=
name|x
operator|+
name|y
decl_stmt|;
comment|// expected-warning {{use of uninitialized variable}}
return|return
name|z
return|;
block|}
end_function

begin_function
name|int
name|f3
parameter_list|(
name|int
name|x
parameter_list|)
block|{
name|int
name|y
decl_stmt|;
return|return
name|x
condition|?
literal|1
else|:
name|y
return|;
comment|// expected-warning {{use of uninitialized variable}}
block|}
end_function

begin_function
name|int
name|f4
parameter_list|(
name|int
name|x
parameter_list|)
block|{
name|int
name|y
decl_stmt|;
if|if
condition|(
name|x
condition|)
name|y
operator|=
literal|1
expr_stmt|;
return|return
name|y
return|;
comment|// expected-warning {{use of uninitialized variable}}
block|}
end_function

begin_function
name|void
name|f5
parameter_list|()
block|{
name|int
name|a
decl_stmt|;
name|a
operator|=
literal|30
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_function
name|void
name|f6
parameter_list|(
name|int
name|i
parameter_list|)
block|{
name|int
name|x
decl_stmt|;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|10
condition|;
name|i
operator|++
control|)
name|printf
argument_list|(
literal|"%d"
argument_list|,
name|x
operator|++
argument_list|)
expr_stmt|;
comment|// expected-warning {{use of uninitialized variable}} \
comment|// expected-warning{{implicitly declaring C library function 'printf' with type 'int (char const *, ...)'}} \
comment|// expected-note{{please include the header<stdio.h> or explicitly provide a declaration for 'printf'}}
block|}
end_function

begin_function
name|void
name|f7
parameter_list|(
name|int
name|i
parameter_list|)
block|{
name|int
name|x
init|=
name|i
decl_stmt|;
name|int
name|y
decl_stmt|;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|10
condition|;
name|i
operator|++
control|)
block|{
name|printf
argument_list|(
literal|"%d"
argument_list|,
name|x
operator|++
argument_list|)
expr_stmt|;
comment|// no-warning
name|x
operator|+=
name|y
expr_stmt|;
comment|// expected-warning {{use of uninitialized variable}}
block|}
block|}
end_function

begin_function
name|int
name|f8
parameter_list|(
name|int
name|j
parameter_list|)
block|{
name|int
name|x
init|=
literal|1
decl_stmt|,
name|y
init|=
name|x
operator|+
literal|1
decl_stmt|;
if|if
condition|(
name|y
condition|)
comment|// no-warning
return|return
name|x
return|;
return|return
name|y
return|;
block|}
end_function

end_unit

