begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -fsyntax-only -verify %s
end_comment

begin_function_decl
specifier|extern
name|int
name|funcInt
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
specifier|extern
name|float
name|funcFloat
parameter_list|(
name|float
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
specifier|extern
name|double
name|funcDouble
parameter_list|(
name|double
parameter_list|)
function_decl|;
end_function_decl

begin_comment
comment|// figure out why "char *" doesn't work (with gcc, nothing to do with clang)
end_comment

begin_comment
comment|//extern void funcCharPtr(char *);
end_comment

begin_define
define|#
directive|define
name|func
parameter_list|(
name|expr
parameter_list|)
define|\
value|do { \     typeof(expr) tmp; \     if (__builtin_types_compatible_p(typeof(expr), int)) funcInt(tmp); \     else if (__builtin_types_compatible_p(typeof(expr), float)) funcFloat(tmp); \     else if (__builtin_types_compatible_p(typeof(expr), double)) funcDouble(tmp); \   } while (0)
end_define

begin_define
define|#
directive|define
name|func_choose
parameter_list|(
name|expr
parameter_list|)
define|\
value|__builtin_choose_expr(__builtin_types_compatible_p(typeof(expr), int), funcInt(expr), \     __builtin_choose_expr(__builtin_types_compatible_p(typeof(expr), float), funcFloat(expr), \       __builtin_choose_expr(__builtin_types_compatible_p(typeof(expr), double), funcDouble(expr), (void)0)))
end_define

begin_function
specifier|static
name|void
name|test
parameter_list|()
block|{
name|int
name|a
decl_stmt|;
name|float
name|b
decl_stmt|;
name|double
name|d
decl_stmt|;
name|func
argument_list|(
name|a
argument_list|)
expr_stmt|;
name|func
argument_list|(
name|b
argument_list|)
expr_stmt|;
name|func
argument_list|(
name|d
argument_list|)
expr_stmt|;
name|a
operator|=
name|func_choose
argument_list|(
name|a
argument_list|)
expr_stmt|;
name|b
operator|=
name|func_choose
argument_list|(
name|b
argument_list|)
expr_stmt|;
name|d
operator|=
name|func_choose
argument_list|(
name|d
argument_list|)
expr_stmt|;
name|int
name|c
decl_stmt|;
struct|struct
name|xx
block|{
name|int
name|a
decl_stmt|;
block|}
name|x
struct|,
name|y
struct|;
name|c
operator|=
name|__builtin_choose_expr
argument_list|(
name|a
operator|+
literal|3
operator|-
literal|7
argument_list|,
name|b
argument_list|,
name|x
argument_list|)
expr_stmt|;
comment|// expected-error{{'__builtin_choose_expr' requires a constant expression}}
name|c
operator|=
name|__builtin_choose_expr
argument_list|(
literal|0
argument_list|,
name|b
argument_list|,
name|x
argument_list|)
expr_stmt|;
comment|// expected-error{{incompatible type assigning 'struct xx', expected 'int'}}
name|c
operator|=
name|__builtin_choose_expr
argument_list|(
literal|5
operator|+
literal|3
operator|-
literal|7
argument_list|,
name|b
argument_list|,
name|x
argument_list|)
expr_stmt|;
name|y
operator|=
name|__builtin_choose_expr
argument_list|(
literal|4
operator|+
literal|3
operator|-
literal|7
argument_list|,
name|b
argument_list|,
name|x
argument_list|)
expr_stmt|;
block|}
end_function

end_unit

