begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s
end_comment

begin_struct
struct|struct
name|foo
block|{
name|int
name|a
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
name|main
parameter_list|()
block|{
name|int
name|a
decl_stmt|;
name|float
name|b
decl_stmt|;
name|double
name|d
decl_stmt|;
name|struct
name|foo
name|s
decl_stmt|;
specifier|static
name|int
name|ary
index|[
name|__builtin_classify_type
argument_list|(
name|a
argument_list|)
index|]
decl_stmt|;
specifier|static
name|int
name|ary2
index|[
call|(
name|__builtin_classify_type
call|)
argument_list|(
name|a
argument_list|)
index|]
decl_stmt|;
comment|// expected-error{{variable length array declaration can not have 'static' storage duration}}
specifier|static
name|int
name|ary3
index|[
call|(
modifier|*
name|__builtin_classify_type
call|)
argument_list|(
name|a
argument_list|)
index|]
decl_stmt|;
comment|// expected-error{{variable length array declaration can not have 'static' storage duration}}
name|int
name|result
decl_stmt|;
name|result
operator|=
name|__builtin_classify_type
argument_list|(
name|a
argument_list|)
expr_stmt|;
name|result
operator|=
name|__builtin_classify_type
argument_list|(
name|b
argument_list|)
expr_stmt|;
name|result
operator|=
name|__builtin_classify_type
argument_list|(
name|d
argument_list|)
expr_stmt|;
name|result
operator|=
name|__builtin_classify_type
argument_list|(
name|s
argument_list|)
expr_stmt|;
block|}
end_function

end_unit

