begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s -Wno-unreachable-code
end_comment

begin_function
name|void
name|test1
parameter_list|()
block|{
block|{
empty_stmt|;
block|{
empty_stmt|;
empty_stmt|;
block|}
block|}
empty_stmt|;
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test2
parameter_list|()
block|{
if|if
condition|(
literal|0
condition|)
block|{
if|if
condition|(
literal|1
condition|)
block|{}
block|}
else|else
block|{ }
do|do
block|{ }
do|while
condition|(
literal|0
condition|)
do|;
while|while
condition|(
literal|0
condition|)
while|while
condition|(
literal|0
condition|)
do|do
empty_stmt|;
do|while
condition|(
literal|0
condition|)
do|;
for|for
control|(
operator|(
name|void
operator|)
literal|0
init|;
literal|0
condition|;
operator|(
name|void
operator|)
literal|0
control|)
for|for
control|(
init|;
condition|;
control|)
for|for
control|(
operator|(
name|void
operator|)
literal|9
init|;
literal|0
condition|;
operator|(
name|void
operator|)
literal|2
control|)
empty_stmt|;
for|for
control|(
name|int
name|X
init|=
literal|0
init|;
literal|0
condition|;
operator|(
name|void
operator|)
literal|0
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test3
parameter_list|()
block|{
switch|switch
condition|(
literal|0
condition|)
block|{
case|case
literal|4
case|:
if|if
condition|(
literal|0
condition|)
block|{
case|case
literal|6
case|:
empty_stmt|;
block|}
default|default:
empty_stmt|;
block|}
block|}
end_function

begin_function
name|void
name|test4
parameter_list|()
block|{
if|if
condition|(
literal|0
condition|)
empty_stmt|;
comment|// expected-warning {{if statement has empty body}} expected-note {{put the semicolon on a separate line to silence this warning}}
name|int
name|X
decl_stmt|;
comment|// declaration in a block.
name|foo
label|:
if|if
condition|(
literal|0
condition|)
empty_stmt|;
comment|// expected-warning {{if statement has empty body}} expected-note {{put the semicolon on a separate line to silence this warning}}
block|}
end_function

begin_typedef
typedef|typedef
name|int
name|t
typedef|;
end_typedef

begin_function
name|void
name|test5
parameter_list|()
block|{
if|if
condition|(
literal|0
condition|)
empty_stmt|;
comment|// expected-warning {{if statement has empty body}} expected-note {{put the semicolon on a separate line to silence this warning}}
name|t
name|x
init|=
literal|0
decl_stmt|;
if|if
condition|(
literal|0
condition|)
empty_stmt|;
comment|// expected-warning {{if statement has empty body}} expected-note {{put the semicolon on a separate line to silence this warning}}
block|}
end_function

begin_function
name|void
name|test6
parameter_list|(
name|void
parameter_list|)
block|{
do|do
operator|.
comment|// expected-error {{expected expression}}
while|while
condition|(
literal|0
condition|)
empty_stmt|;
do|}
name|int
name|test7
parameter_list|()
block|{
return|return
literal|4
comment|// expected-error {{expected ';' after return statement}}
block|}
name|void
name|test8
parameter_list|()
block|{
comment|// Should not skip '}' and produce a "expected '}'" error.
name|undecl
comment|// expected-error {{use of undeclared identifier 'undecl'}}
block|}
end_function

end_unit

