begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -fopenmp -verify %s
end_comment

begin_comment
comment|// expected-error@+1 {{unexpected OpenMP directive '#pragma omp distribute simd'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
end_pragma

begin_comment
comment|// expected-error@+1 {{unexpected OpenMP directive '#pragma omp distribute simd'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|foo
end_pragma

begin_comment
comment|// expected-error@+1 {{unexpected OpenMP directive '#pragma omp distribute simd'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
name|)
end_pragma

begin_function
name|void
name|test_no_clause
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{statement after '#pragma omp distribute simd' must be a for loop}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
operator|++
name|i
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test_branch_protected_scope
parameter_list|()
block|{
name|int
name|i
init|=
literal|0
decl_stmt|;
name|L1
label|:
operator|++
name|i
expr_stmt|;
name|int
name|x
index|[
literal|24
index|]
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
block|{
if|if
condition|(
name|i
operator|==
literal|5
condition|)
goto|goto
name|L1
goto|;
comment|// expected-error {{use of undeclared label 'L1'}}
elseif|else
if|if
condition|(
name|i
operator|==
literal|6
condition|)
return|return;
comment|// expected-error {{cannot return from OpenMP region}}
elseif|else
if|if
condition|(
name|i
operator|==
literal|7
condition|)
goto|goto
name|L2
goto|;
elseif|else
if|if
condition|(
name|i
operator|==
literal|8
condition|)
block|{
name|L2
label|:
name|x
index|[
name|i
index|]
operator|++
expr_stmt|;
block|}
block|}
if|if
condition|(
name|x
index|[
literal|0
index|]
operator|==
literal|0
condition|)
goto|goto
name|L2
goto|;
comment|// expected-error {{use of undeclared label 'L2'}}
elseif|else
if|if
condition|(
name|x
index|[
literal|1
index|]
operator|==
literal|1
condition|)
goto|goto
name|L1
goto|;
block|}
end_function

begin_function
name|void
name|test_invalid_clause
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp distribute simd' are ignored}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|foo
name|bar
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_non_identifiers
parameter_list|()
block|{
name|int
name|i
decl_stmt|,
name|x
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp distribute simd' are ignored}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp distribute simd' are ignored}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|x
name|)
name|;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp distribute simd' are ignored}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|,
name|private
name|(
name|x
name|)
name|;
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function_decl
specifier|extern
name|int
name|foo
parameter_list|()
function_decl|;
end_function_decl

begin_function
name|void
name|test_safelen
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}  expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-warning@+2 {{extra tokens at the end of '#pragma omp distribute simd' are ignored}}
comment|// expected-error@+1 {{expected '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// xxpected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
name|,
name|,
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|4
name|,
name|8
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expression is not an integer constant expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|2.5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expression is not an integer constant expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|foo
name|(
name|)
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'safelen' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|-
name|5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'safelen' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|0
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'safelen' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|5
name|-
name|5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_simdlen
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}  expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-warning@+2 {{extra tokens at the end of '#pragma omp distribute simd' are ignored}}
comment|// expected-error@+1 {{expected '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|4
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|4
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|4
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|4
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|4
name|,
name|,
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|4
name|,
name|8
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expression is not an integer constant expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|2.5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expression is not an integer constant expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|foo
name|(
name|)
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'simdlen' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|-
name|5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'simdlen' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|0
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'simdlen' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|5
name|-
name|5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_safelen_simdlen
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{the value of 'simdlen' parameter must be less than or equal to the value of the 'safelen' parameter}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|simdlen
name|(
name|6
name|)
name|safelen
name|(
name|5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{the value of 'simdlen' parameter must be less than or equal to the value of the 'safelen' parameter}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|safelen
name|(
name|5
name|)
name|simdlen
name|(
name|6
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_collapse
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}  expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-warning@+2 {{extra tokens at the end of '#pragma omp distribute simd' are ignored}}
comment|// expected-error@+1 {{expected '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}} expected-note@+1 {{as specified in 'collapse' clause}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|4
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
comment|// expected-error {{expected 4 for loops after '#pragma omp distribute simd', but found only 1}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}} expected-note@+1 {{as specified in 'collapse' clause}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|4
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
comment|// expected-error {{expected 4 for loops after '#pragma omp distribute simd', but found only 1}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}} expected-note@+1 {{as specified in 'collapse' clause}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|4
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
comment|// expected-error {{expected 4 for loops after '#pragma omp distribute simd', but found only 1}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// xxpected-error@+1 {{expected expression}} expected-note@+1 {{as specified in 'collapse' clause}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
comment|// expected-error {{expected 4 for loops after '#pragma omp distribute simd', but found only 1}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}} expected-note@+1 {{as specified in 'collapse' clause}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|4
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
comment|// expected-error {{expected 4 for loops after '#pragma omp distribute simd', but found only 1}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}} expected-note@+1 {{as specified in 'collapse' clause}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|4
name|,
name|,
name|4
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
comment|// expected-error {{expected 4 for loops after '#pragma omp distribute simd', but found only 1}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|4
name|)
for|for
control|(
name|int
name|i1
init|=
literal|0
init|;
name|i1
operator|<
literal|16
condition|;
operator|++
name|i1
control|)
for|for
control|(
name|int
name|i2
init|=
literal|0
init|;
name|i2
operator|<
literal|16
condition|;
operator|++
name|i2
control|)
for|for
control|(
name|int
name|i3
init|=
literal|0
init|;
name|i3
operator|<
literal|16
condition|;
operator|++
name|i3
control|)
for|for
control|(
name|int
name|i4
init|=
literal|0
init|;
name|i4
operator|<
literal|16
condition|;
operator|++
name|i4
control|)
name|foo
argument_list|()
expr_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}}
comment|// expected-note@+1 {{to match this '('}} expected-note@+1 {{as specified in 'collapse' clause}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|4
name|,
name|8
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
comment|// expected-error {{expected 4 for loops after '#pragma omp distribute simd', but found only 1}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expression is not an integer constant expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|2.5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expression is not an integer constant expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|foo
name|(
name|)
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'collapse' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|-
name|5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'collapse' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|0
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument to 'collapse' clause must be a strictly positive integer value}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|5
name|-
name|5
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
comment|// expected-note@+3 {{defined as reduction}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|collapse
name|(
name|2
name|)
name|reduction
name|(
name|+
name|:
name|i
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
comment|// expected-note@+1 {{variable with automatic storage duration is predetermined as private; perhaps you forget to enclose 'omp for' directive into a parallel or another task region?}}
for|for
control|(
name|int
name|j
init|=
literal|0
init|;
name|j
operator|<
literal|16
condition|;
operator|++
name|j
control|)
comment|// expected-error@+2 2 {{reduction variable must be shared}}
comment|// expected-error@+1 {{OpenMP constructs may not be nested inside a simd region}}
pragma|#
directive|pragma
name|omp
name|for
name|reduction
name|(
name|+
name|:
name|i
name|,
name|j
name|)
for|for
control|(
name|int
name|k
init|=
literal|0
init|;
name|k
operator|<
literal|16
condition|;
operator|++
name|k
control|)
name|i
operator|+=
name|j
expr_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
for|for
control|(
name|int
name|j
init|=
literal|0
init|;
name|j
operator|<
literal|16
condition|;
operator|++
name|j
control|)
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
name|:
name|i
name|,
name|j
name|)
for|for
control|(
name|int
name|k
init|=
literal|0
init|;
name|k
operator|<
literal|16
condition|;
operator|++
name|k
control|)
name|i
operator|+=
name|j
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test_linear
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected expression}}
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected expression}}
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|int
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|0
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{use of undeclared identifier 'x'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{use of undeclared identifier 'x'}}
comment|// expected-error@+1 {{use of undeclared identifier 'y'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|,
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+3 {{use of undeclared identifier 'x'}}
comment|// expected-error@+2 {{use of undeclared identifier 'y'}}
comment|// expected-error@+1 {{use of undeclared identifier 'z'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|,
name|y
name|,
name|z
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
name|int
name|x
decl_stmt|,
name|y
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|:
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|:
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|:
name|1
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|:
name|2
name|*
name|2
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|:
name|1
name|,
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|:
name|1
name|,
name|y
name|,
name|z
name|:
name|1
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-note@+2 {{defined as linear}}
comment|// expected-error@+1 {{linear variable cannot be linear}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|)
name|linear
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-note@+2 {{defined as private}}
comment|// expected-error@+1 {{private variable cannot be linear}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|x
name|)
name|linear
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-note@+2 {{defined as linear}}
comment|// expected-error@+1 {{linear variable cannot be private}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|)
name|private
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-warning@+1 {{zero linear step (x and other variables in clause should probably be const)}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|,
name|y
name|:
name|0
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-note@+2 {{defined as linear}}
comment|// expected-error@+1 {{linear variable cannot be lastprivate}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|x
name|)
name|lastprivate
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-note@+2 {{defined as lastprivate}}
comment|// expected-error@+1 {{lastprivate variable cannot be linear}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|x
name|)
name|linear
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_aligned
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected expression}}
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected expression}}
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|int
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|0
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{use of undeclared identifier 'x'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{use of undeclared identifier 'x'}}
comment|// expected-error@+1 {{use of undeclared identifier 'y'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|,
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+3 {{use of undeclared identifier 'x'}}
comment|// expected-error@+2 {{use of undeclared identifier 'y'}}
comment|// expected-error@+1 {{use of undeclared identifier 'z'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|,
name|y
name|,
name|z
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
name|int
modifier|*
name|x
decl_stmt|,
name|y
decl_stmt|,
name|z
index|[
literal|25
index|]
decl_stmt|;
comment|// expected-note 4 {{'y' defined here}}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|z
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|:
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}} expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|:
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|:
name|1
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|:
name|2
name|*
name|2
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|:
name|1
name|,
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|:
name|1
name|,
name|y
name|,
name|z
name|:
name|1
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument of aligned clause should be array or pointer, not 'int'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|,
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{argument of aligned clause should be array or pointer, not 'int'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|,
name|y
name|,
name|z
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-note@+2 {{defined as aligned}}
comment|// expected-error@+1 {{a variable cannot appear in more than one aligned clause}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|)
name|aligned
name|(
name|z
name|,
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-note@+3 {{defined as aligned}}
comment|// expected-error@+2 {{a variable cannot appear in more than one aligned clause}}
comment|// expected-error@+1 2 {{argument of aligned clause should be array or pointer, not 'int'}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|aligned
name|(
name|x
name|,
name|y
name|,
name|z
name|)
name|aligned
name|(
name|y
name|,
name|z
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_private
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected expression}}
comment|// expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|int
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|0
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
name|int
name|x
decl_stmt|,
name|y
decl_stmt|,
name|z
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|x
name|,
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|private
name|(
name|x
name|,
name|y
name|,
name|z
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
block|{
name|x
operator|=
name|y
operator|*
name|i
operator|+
name|z
expr_stmt|;
block|}
block|}
end_function

begin_function
name|void
name|test_firstprivate
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|firstprivate
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_lastprivate
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|,
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|int
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|0
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
name|int
name|x
decl_stmt|,
name|y
decl_stmt|,
name|z
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|x
name|,
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|lastprivate
name|(
name|x
name|,
name|y
name|,
name|z
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_reduction
parameter_list|()
block|{
name|int
name|i
decl_stmt|,
name|x
decl_stmt|,
name|y
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+3 {{expected ')'}} expected-note@+3 {{to match this '('}}
comment|// expected-error@+2 {{expected identifier}}
comment|// expected-warning@+1 {{missing ':' after reduction identifier - ignoring}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected identifier}}
comment|// expected-warning@+1 {{missing ':' after reduction identifier - ignoring}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{expected expression}}
comment|// expected-warning@+1 {{missing ':' after reduction identifier - ignoring}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected identifier}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+3 {{expected ')'}} expected-note@+3 {{to match this '('}}
comment|// expected-error@+2 {{expected identifier}}
comment|// expected-warning@+1 {{missing ':' after reduction identifier - ignoring}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|,
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+3 {{expected ')'}} expected-note@+3 {{to match this '('}}
comment|// expected-error@+2 {{expected expression}}
comment|// expected-warning@+1 {{missing ':' after reduction identifier - ignoring}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+3 {{expected ')'}} expected-note@+3 {{to match this '('}}
comment|//
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
name|:
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
name|:
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
name|:
name|,
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
name|:
name|x
name|,
name|+
name|:
name|y
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected identifier}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|%
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|*
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|-
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|&
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name||
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|^
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|&&
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|||
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|max
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|min
name|:
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
struct|struct
name|X
block|{
name|int
name|x
decl_stmt|;
block|}
struct|;
name|struct
name|X
name|X
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
name|:
name|X
name|.
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|reduction
name|(
name|+
name|:
name|x
name|+
name|x
name|)
for|for
control|(
name|i
operator|=
literal|0
init|;
name|i
operator|<
literal|16
condition|;
operator|++
name|i
control|)
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test_loop_messages
parameter_list|()
block|{
name|float
name|a
index|[
literal|100
index|]
decl_stmt|,
name|b
index|[
literal|100
index|]
decl_stmt|,
name|c
index|[
literal|100
index|]
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{variable must be of integer or pointer type}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
for|for
control|(
name|float
name|fi
init|=
literal|0
init|;
name|fi
operator|<
literal|10.0
condition|;
name|fi
operator|++
control|)
block|{
name|c
index|[
operator|(
name|int
operator|)
name|fi
index|]
operator|=
name|a
index|[
operator|(
name|int
operator|)
name|fi
index|]
operator|+
name|b
index|[
operator|(
name|int
operator|)
name|fi
index|]
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
comment|// expected-error@+2 {{variable must be of integer or pointer type}}
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
for|for
control|(
name|double
name|fi
init|=
literal|0
init|;
name|fi
operator|<
literal|10.0
condition|;
name|fi
operator|++
control|)
block|{
name|c
index|[
operator|(
name|int
operator|)
name|fi
index|]
operator|=
name|a
index|[
operator|(
name|int
operator|)
name|fi
index|]
operator|+
name|b
index|[
operator|(
name|int
operator|)
name|fi
index|]
expr_stmt|;
block|}
block|}
end_function

begin_function
name|void
name|linear_modifiers
parameter_list|(
name|int
name|argc
parameter_list|)
block|{
name|int
name|f
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|f
name|)
for|for
control|(
name|int
name|k
init|=
literal|0
init|;
name|k
operator|<
name|argc
condition|;
operator|++
name|k
control|)
operator|++
name|k
expr_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|val
name|(
name|f
name|)
name|)
for|for
control|(
name|int
name|k
init|=
literal|0
init|;
name|k
operator|<
name|argc
condition|;
operator|++
name|k
control|)
operator|++
name|k
expr_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|uval
name|(
name|f
name|)
name|)
comment|// expected-error {{expected 'val' modifier}}
for|for
control|(
name|int
name|k
init|=
literal|0
init|;
name|k
operator|<
name|argc
condition|;
operator|++
name|k
control|)
operator|++
name|k
expr_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|ref
name|(
name|f
name|)
name|)
comment|// expected-error {{expected 'val' modifier}}
for|for
control|(
name|int
name|k
init|=
literal|0
init|;
name|k
operator|<
name|argc
condition|;
operator|++
name|k
control|)
operator|++
name|k
expr_stmt|;
pragma|#
directive|pragma
name|omp
name|target
pragma|#
directive|pragma
name|omp
name|teams
pragma|#
directive|pragma
name|omp
name|distribute
name|simd
name|linear
name|(
name|foo
name|(
name|f
name|)
name|)
comment|// expected-error {{expected 'val' modifier}}
for|for
control|(
name|int
name|k
init|=
literal|0
init|;
name|k
operator|<
name|argc
condition|;
operator|++
name|k
control|)
operator|++
name|k
expr_stmt|;
block|}
end_function

end_unit

