begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -fopenmp=libiomp5 -verify %s
end_comment

begin_function_decl
name|void
name|foo
parameter_list|()
function_decl|;
end_function_decl

begin_comment
comment|// expected-error@+1 {{unexpected OpenMP directive '#pragma omp sections'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|omp
name|sections
end_pragma

begin_comment
comment|// expected-error@+1 {{unexpected OpenMP directive '#pragma omp sections'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|omp
name|sections
name|foo
end_pragma

begin_function
name|void
name|test_no_clause
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|sections
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
comment|// expected-error@+2 {{the statement for '#pragma omp sections' must be a compound statement}}
pragma|#
directive|pragma
name|omp
name|sections
operator|++
name|i
expr_stmt|;
pragma|#
directive|pragma
name|omp
name|sections
block|{
name|foo
argument_list|()
expr_stmt|;
name|foo
argument_list|()
expr_stmt|;
comment|// expected-error {{statement in 'omp sections' directive must be enclosed into a section region}}
block|}
block|}
end_function

begin_function
name|void
name|test_branch_protected_scope
parameter_list|()
block|{
name|int
name|i
init|=
literal|0
decl_stmt|;
name|L1
label|:
operator|++
name|i
expr_stmt|;
name|int
name|x
index|[
literal|24
index|]
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
block|{
if|if
condition|(
name|i
operator|==
literal|5
condition|)
goto|goto
name|L1
goto|;
comment|// expected-error {{use of undeclared label 'L1'}}
elseif|else
if|if
condition|(
name|i
operator|==
literal|6
condition|)
return|return;
comment|// expected-error {{cannot return from OpenMP region}}
elseif|else
if|if
condition|(
name|i
operator|==
literal|7
condition|)
goto|goto
name|L2
goto|;
elseif|else
if|if
condition|(
name|i
operator|==
literal|8
condition|)
block|{
name|L2
label|:
name|x
index|[
name|i
index|]
operator|++
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|section
if|if
condition|(
name|i
operator|==
literal|5
condition|)
goto|goto
name|L1
goto|;
comment|// expected-error {{use of undeclared label 'L1'}}
elseif|else
if|if
condition|(
name|i
operator|==
literal|6
condition|)
return|return;
comment|// expected-error {{cannot return from OpenMP region}}
elseif|else
if|if
condition|(
name|i
operator|==
literal|7
condition|)
goto|goto
name|L3
goto|;
elseif|else
if|if
condition|(
name|i
operator|==
literal|8
condition|)
block|{
name|L3
label|:
name|x
index|[
name|i
index|]
operator|++
expr_stmt|;
block|}
block|}
if|if
condition|(
name|x
index|[
literal|0
index|]
operator|==
literal|0
condition|)
goto|goto
name|L2
goto|;
comment|// expected-error {{use of undeclared label 'L2'}}
elseif|else
if|if
condition|(
name|x
index|[
literal|1
index|]
operator|==
literal|1
condition|)
goto|goto
name|L1
goto|;
goto|goto
name|L3
goto|;
comment|// expected-error {{use of undeclared label 'L3'}}
block|}
end_function

begin_function
name|void
name|test_invalid_clause
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp sections' are ignored}}
pragma|#
directive|pragma
name|omp
name|sections
name|foo
name|bar
block|{
name|foo
argument_list|()
expr_stmt|;
comment|// expected-error@+1 {{unexpected OpenMP clause 'nowait' in directive '#pragma omp section'}}
pragma|#
directive|pragma
name|omp
name|section
name|nowait
empty_stmt|;
block|}
block|}
end_function

begin_function
name|void
name|test_non_identifiers
parameter_list|()
block|{
name|int
name|i
decl_stmt|,
name|x
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp sections' are ignored}}
pragma|#
directive|pragma
name|omp
name|sections
name|;
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+2 {{unexpected OpenMP clause 'linear' in directive '#pragma omp sections'}}
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp sections' are ignored}}
pragma|#
directive|pragma
name|omp
name|sections
name|linear
name|(
name|x
name|)
name|;
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp sections' are ignored}}
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|x
name|)
name|;
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-warning@+1 {{extra tokens at the end of '#pragma omp sections' are ignored}}
pragma|#
directive|pragma
name|omp
name|sections
name|,
name|private
name|(
name|x
name|)
name|;
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
block|}
end_function

begin_function
name|void
name|test_private
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+2 {{expected expression}}
comment|// expected-error@+1 {{expected ')'}} expected-note@+1 {{to match this '('}}
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|,
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|,
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|int
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|0
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
name|int
name|x
decl_stmt|,
name|y
decl_stmt|,
name|z
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|x
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|x
name|,
name|y
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|private
name|(
name|x
name|,
name|y
name|,
name|z
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
block|}
end_function

begin_function
name|void
name|test_lastprivate
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|,
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|,
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|int
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|0
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
name|int
name|x
decl_stmt|,
name|y
decl_stmt|,
name|z
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|x
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|x
name|,
name|y
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|x
name|,
name|y
name|,
name|z
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
block|}
end_function

begin_function
name|void
name|test_firstprivate
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|firstprivate
name|(
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+2 {{expected ')'}} expected-note@+2 {{to match this '('}}
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|firstprivate
name|(
name|,
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 2 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|firstprivate
name|(
name|,
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|firstprivate
name|(
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected expression}}
pragma|#
directive|pragma
name|omp
name|sections
name|firstprivate
name|(
name|int
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
comment|// expected-error@+1 {{expected variable name}}
pragma|#
directive|pragma
name|omp
name|sections
name|firstprivate
name|(
name|0
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
name|int
name|x
decl_stmt|,
name|y
decl_stmt|,
name|z
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|x
name|)
name|firstprivate
name|(
name|x
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|x
name|,
name|y
name|)
name|firstprivate
name|(
name|x
name|,
name|y
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|lastprivate
name|(
name|x
name|,
name|y
name|,
name|z
name|)
name|firstprivate
name|(
name|x
name|,
name|y
name|,
name|z
name|)
block|{
name|foo
argument_list|()
expr_stmt|;
block|}
block|}
end_function

begin_function
name|void
name|test_nowait
parameter_list|()
block|{
pragma|#
directive|pragma
name|omp
name|parallel
pragma|#
directive|pragma
name|omp
name|sections
name|nowait
name|nowait
comment|// expected-error {{directive '#pragma omp sections' cannot contain more than one 'nowait' clause}}
block|{
empty_stmt|;
block|}
block|}
end_function

end_unit

