begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-apple-macos10.7.0 -verify -fopenmp -ferror-limit 100 -o - %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-apple-macos10.7.0 -verify -fopenmp -ferror-limit 100 -o - -x c++ %s
end_comment

begin_function
name|int
name|main
parameter_list|(
name|int
name|argc
parameter_list|,
name|char
modifier|*
modifier|*
name|argv
parameter_list|)
block|{
name|int
name|r
decl_stmt|;
pragma|#
directive|pragma
name|omp
name|target
name|exit
name|data
comment|// expected-error {{expected at least one 'map' clause for '#pragma omp target exit data'}}
pragma|#
directive|pragma
name|omp
name|target
name|exit
name|data
name|map
name|(
name|r
name|)
comment|// expected-error {{map type must be specified for '#pragma omp target exit data'}}
pragma|#
directive|pragma
name|omp
name|target
name|exit
name|data
name|map
name|(
name|tofrom
name|:
name|r
name|)
comment|// expected-error {{map type 'tofrom' is not allowed for '#pragma omp target exit data'}}
pragma|#
directive|pragma
name|omp
name|target
name|exit
name|data
name|map
name|(
name|always
name|,
name|from
name|:
name|r
name|)
pragma|#
directive|pragma
name|omp
name|target
name|exit
name|data
name|map
name|(
name|delete
name|:
name|r
name|)
pragma|#
directive|pragma
name|omp
name|target
name|exit
name|data
name|map
name|(
name|release
name|:
name|r
name|)
pragma|#
directive|pragma
name|omp
name|target
name|exit
name|data
name|map
name|(
name|always
name|,
name|alloc
name|:
name|r
name|)
comment|// expected-error {{map type 'alloc' is not allowed for '#pragma omp target exit data'}}
pragma|#
directive|pragma
name|omp
name|target
name|exit
name|data
name|map
name|(
name|to
name|:
name|r
name|)
comment|// expected-error {{map type 'to' is not allowed for '#pragma omp target exit data'}}
return|return
literal|0
return|;
block|}
end_function

end_unit

