begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_typedef
typedef|typedef
name|struct
name|objc_selector
modifier|*
name|SEL
typedef|;
end_typedef

begin_expr_stmt
unit|@
name|interface
name|Foo
operator|-
operator|(
name|void
operator|)
name|NotOK
expr_stmt|;
end_expr_stmt

begin_expr_stmt
unit|@
name|end
expr|@
name|implementation
name|Foo
operator|-
operator|(
name|void
operator|)
name|foo
block|{
name|SEL
name|a
operator|=
expr|@
name|selector
argument_list|(
name|b1ar
argument_list|)
block|;
name|a
operator|=
expr|@
name|selector
argument_list|(
name|b1ar
argument_list|)
block|;
name|a
operator|=
expr|@
name|selector
argument_list|(
name|bar
argument_list|)
block|;
name|a
operator|=
expr|@
name|selector
argument_list|(
name|ok
argument_list|)
block|;
comment|// expected-warning {{unimplemented selector 'ok'}}
name|a
operator|=
expr|@
name|selector
argument_list|(
name|ok
argument_list|)
block|;
name|a
operator|=
expr|@
name|selector
argument_list|(
name|NotOK
argument_list|)
block|;
comment|// expected-warning {{unimplemented selector 'NotOK'}}
name|a
operator|=
expr|@
name|selector
argument_list|(
name|NotOK
argument_list|)
block|;
name|a
operator|=
expr|@
name|selector
argument_list|(
name|clNotOk
argument_list|)
block|;
comment|// expected-warning {{unimplemented selector 'clNotOk'}}
name|a
operator|=
expr|@
name|selector
argument_list|(
name|cl1
argument_list|)
block|;
name|a
operator|=
expr|@
name|selector
argument_list|(
name|cl2
argument_list|)
block|;
name|a
operator|=
expr|@
name|selector
argument_list|(
name|instNotOk
argument_list|)
block|;
comment|// expected-warning {{unimplemented selector 'instNotOk'}}
block|}
expr|@
name|end
end_expr_stmt

end_unit

