begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple aarch64-linux-gnu -target-feature +neon -ffreestanding -fsyntax-only -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -triple arm64-linux-gnu -target-feature +neon -ffreestanding -fsyntax-only -verify %s
end_comment

begin_include
include|#
directive|include
file|<arm_neon.h>
end_include

begin_function
name|void
name|test_vext_8bit
parameter_list|(
name|int8x8_t
name|small
parameter_list|,
name|int8x16_t
name|big
parameter_list|)
block|{
name|vext_s8
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vext_u8
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vext_p8
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vextq_s8
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vextq_u8
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vextq_p8
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vext_s8
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vext_u8
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vext_p8
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vextq_s8
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vextq_u8
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vextq_p8
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
block|}
end_function

begin_function
name|void
name|test_mul_lane_f64
parameter_list|(
name|float64x1_t
name|small
parameter_list|,
name|float64x2_t
name|big
parameter_list|,
name|float64x2_t
name|rhs
parameter_list|)
block|{
name|vmul_lane_f64
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vmul_laneq_f64
argument_list|(
name|small
argument_list|,
name|big
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vmulq_lane_f64
argument_list|(
name|big
argument_list|,
name|small
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vmulq_laneq_f64
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vfma_lane_f64
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
name|small
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vfma_laneq_f64
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
name|big
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vfmaq_lane_f64
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
name|small
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vfmaq_laneq_f64
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
name|big
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vmul_lane_f64
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vmul_laneq_f64
argument_list|(
name|small
argument_list|,
name|big
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vfma_lane_f64
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
name|small
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vfma_laneq_f64
argument_list|(
name|small
argument_list|,
name|small
argument_list|,
name|big
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vfmaq_laneq_f64
argument_list|(
name|big
argument_list|,
name|big
argument_list|,
name|big
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
block|}
end_function

begin_function
name|void
name|test_ld1st1
parameter_list|(
name|int8x8_t
name|small
parameter_list|,
name|int8x16_t
name|big
parameter_list|,
name|void
modifier|*
name|addr
parameter_list|)
block|{
name|vld1_lane_s8
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vld1_lane_s16
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vld1_lane_s32
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vld1_lane_s64
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vld1q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vld1q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vld1q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vld1q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vld1_lane_s8
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vld1_lane_s16
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vld1_lane_s32
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vld1_lane_s64
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vld1q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vld1q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vld1q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vld1q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vst1_lane_s8
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vst1_lane_s16
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vst1_lane_s32
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vst1_lane_s64
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vst1q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vst1q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vst1q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vst1q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vst1_lane_s8
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vst1_lane_s16
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vst1_lane_s32
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vst1_lane_s64
argument_list|(
name|addr
argument_list|,
name|small
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vst1q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vst1q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vst1q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vst1q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
block|}
end_function

begin_function
name|void
name|test_ld2st2
parameter_list|(
name|int8x8x2_t
name|small8
parameter_list|,
name|int8x16x2_t
name|big8
parameter_list|,
name|int16x4x2_t
name|small16
parameter_list|,
name|int16x8x2_t
name|big16
parameter_list|,
name|int32x2x2_t
name|small32
parameter_list|,
name|int32x4x2_t
name|big32
parameter_list|,
name|int64x1x2_t
name|small64
parameter_list|,
name|int64x2x2_t
name|big64
parameter_list|,
name|void
modifier|*
name|addr
parameter_list|)
block|{
name|vld2_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vld2_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vld2_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vld2_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vld2q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vld2q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vld2q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vld2q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vld2_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vld2_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vld2_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vld2_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vld2q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vld2q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vld2q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vld2q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vst2_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vst2_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vst2_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vst2_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vst2q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vst2q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vst2q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vst2q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vst2_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vst2_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vst2_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vst2_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vst2q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vst2q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vst2q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vst2q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
block|}
end_function

begin_function
name|void
name|test_ld3st3
parameter_list|(
name|int8x8x3_t
name|small8
parameter_list|,
name|int8x16x3_t
name|big8
parameter_list|,
name|int16x4x3_t
name|small16
parameter_list|,
name|int16x8x3_t
name|big16
parameter_list|,
name|int32x2x3_t
name|small32
parameter_list|,
name|int32x4x3_t
name|big32
parameter_list|,
name|int64x1x3_t
name|small64
parameter_list|,
name|int64x2x3_t
name|big64
parameter_list|,
name|void
modifier|*
name|addr
parameter_list|)
block|{
name|vld3_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vld3_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vld3_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vld3_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vld3q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vld3q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vld3q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vld3q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vld3_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vld3_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vld3_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vld3_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vld3q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vld3q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vld3q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vld3q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vst3_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vst3_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vst3_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vst3_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vst3q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vst3q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vst3q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vst3q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vst3_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vst3_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vst3_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vst3_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vst3q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vst3q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vst3q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vst3q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
block|}
end_function

begin_function
name|void
name|test_ld4st4
parameter_list|(
name|int8x8x4_t
name|small8
parameter_list|,
name|int8x16x4_t
name|big8
parameter_list|,
name|int16x4x4_t
name|small16
parameter_list|,
name|int16x8x4_t
name|big16
parameter_list|,
name|int32x2x4_t
name|small32
parameter_list|,
name|int32x4x4_t
name|big32
parameter_list|,
name|int64x1x4_t
name|small64
parameter_list|,
name|int64x2x4_t
name|big64
parameter_list|,
name|void
modifier|*
name|addr
parameter_list|)
block|{
name|vld4_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vld4_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vld4_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vld4_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vld4q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vld4q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vld4q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vld4q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vld4_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vld4_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vld4_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vld4_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vld4q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vld4q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vld4q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vld4q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vst4_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vst4_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vst4_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vst4_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|vst4q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|15
argument_list|)
expr_stmt|;
name|vst4q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|7
argument_list|)
expr_stmt|;
name|vst4q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|3
argument_list|)
expr_stmt|;
name|vst4q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|vst4_lane_s8
argument_list|(
name|addr
argument_list|,
name|small8
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vst4_lane_s16
argument_list|(
name|addr
argument_list|,
name|small16
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vst4_lane_s32
argument_list|(
name|addr
argument_list|,
name|small32
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
name|vst4_lane_s64
argument_list|(
name|addr
argument_list|,
name|small64
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 0}}
name|vst4q_lane_s8
argument_list|(
name|addr
argument_list|,
name|big8
argument_list|,
literal|16
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 15}}
name|vst4q_lane_s16
argument_list|(
name|addr
argument_list|,
name|big16
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
name|vst4q_lane_s32
argument_list|(
name|addr
argument_list|,
name|big32
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
name|vst4q_lane_s64
argument_list|(
name|addr
argument_list|,
name|big64
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 1}}
block|}
end_function

end_unit

