begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s
end_comment

begin_include
include|#
directive|include
file|<stdint.h>
end_include

begin_function_decl
specifier|extern
name|void
name|f1
parameter_list|(
name|int
modifier|*
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
specifier|extern
name|void
name|f2
parameter_list|(
name|char
modifier|*
parameter_list|)
function_decl|;
end_function_decl

begin_struct
struct|struct
name|Ok
block|{
name|char
name|c
decl_stmt|;
name|int
name|x
decl_stmt|;
block|}
struct|;
end_struct

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(packed)
argument_list|)
name|Arguable
block|{
name|char
name|c0
decl_stmt|;
name|int
name|x
decl_stmt|;
name|char
name|c1
decl_stmt|;
block|}
struct|;
end_struct

begin_union
union|union
name|__attribute__
argument_list|(
argument|(packed)
argument_list|)
name|UnionArguable
block|{
name|char
name|c
decl_stmt|;
name|int
name|x
decl_stmt|;
block|}
union|;
end_union

begin_typedef
typedef|typedef
name|struct
name|Arguable
name|ArguableT
typedef|;
end_typedef

begin_function_decl
name|struct
name|Arguable
modifier|*
name|get_arguable
parameter_list|()
function_decl|;
end_function_decl

begin_function_decl
name|void
name|to_void
parameter_list|(
name|void
modifier|*
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|to_intptr
parameter_list|(
name|intptr_t
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|g0
parameter_list|(
name|void
parameter_list|)
block|{
block|{
name|struct
name|Ok
name|ok
decl_stmt|;
name|f1
argument_list|(
operator|&
name|ok
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|f2
argument_list|(
operator|&
name|ok
operator|.
name|c
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
block|{
name|struct
name|Arguable
name|arguable
decl_stmt|;
name|f2
argument_list|(
operator|&
name|arguable
operator|.
name|c0
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{packed member 'x' of class or structure 'Arguable'}}
name|f2
argument_list|(
operator|&
name|arguable
operator|.
name|c1
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|(
name|int
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_void
argument_list|(
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|void
modifier|*
name|p
init|=
operator|&
name|arguable
operator|.
name|x
decl_stmt|;
comment|// no-warning
name|to_void
argument_list|(
name|p
argument_list|)
expr_stmt|;
name|to_intptr
argument_list|(
operator|(
name|intptr_t
operator|)
name|p
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
block|{
name|union
name|UnionArguable
name|arguable
decl_stmt|;
name|f2
argument_list|(
operator|&
name|arguable
operator|.
name|c
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{packed member 'x' of class or structure 'UnionArguable'}}
name|f1
argument_list|(
operator|(
name|int
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_void
argument_list|(
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_intptr
argument_list|(
operator|(
name|intptr_t
operator|)
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
block|{
name|ArguableT
name|arguable
decl_stmt|;
name|f2
argument_list|(
operator|&
name|arguable
operator|.
name|c0
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{packed member 'x' of class or structure 'Arguable'}}
name|f2
argument_list|(
operator|&
name|arguable
operator|.
name|c1
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|(
name|int
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_void
argument_list|(
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_intptr
argument_list|(
operator|(
name|intptr_t
operator|)
operator|&
name|arguable
operator|.
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
block|{
name|struct
name|Arguable
modifier|*
name|arguable
init|=
name|get_arguable
argument_list|()
decl_stmt|;
name|f2
argument_list|(
operator|&
name|arguable
operator|->
name|c0
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|&
name|arguable
operator|->
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{packed member 'x' of class or structure 'Arguable'}}
name|f2
argument_list|(
operator|&
name|arguable
operator|->
name|c1
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|(
name|int
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
operator|&
name|arguable
operator|->
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_void
argument_list|(
operator|&
name|arguable
operator|->
name|c1
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_intptr
argument_list|(
operator|(
name|intptr_t
operator|)
operator|&
name|arguable
operator|->
name|c1
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
block|{
name|ArguableT
modifier|*
name|arguable
init|=
name|get_arguable
argument_list|()
decl_stmt|;
name|f2
argument_list|(
operator|&
operator|(
name|arguable
operator|->
name|c0
operator|)
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|&
operator|(
name|arguable
operator|->
name|x
operator|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{packed member 'x' of class or structure 'Arguable'}}
name|f2
argument_list|(
operator|&
operator|(
name|arguable
operator|->
name|c1
operator|)
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|(
name|int
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
operator|&
operator|(
name|arguable
operator|->
name|x
operator|)
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_void
argument_list|(
operator|&
operator|(
name|arguable
operator|->
name|c1
operator|)
argument_list|)
expr_stmt|;
comment|// no-warning
name|to_intptr
argument_list|(
operator|(
name|intptr_t
operator|)
operator|&
operator|(
name|arguable
operator|->
name|c1
operator|)
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
block|}
end_function

begin_struct
struct|struct
name|S1
block|{
name|char
name|c
decl_stmt|;
name|int
name|i
name|__attribute__
argument_list|(
operator|(
name|packed
operator|)
argument_list|)
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
modifier|*
name|g1
parameter_list|(
name|struct
name|S1
modifier|*
name|s1
parameter_list|)
block|{
return|return
operator|&
name|s1
operator|->
name|i
return|;
comment|// expected-warning {{packed member 'i' of class or structure 'S1'}}
block|}
end_function

begin_struct
struct|struct
name|S2_i
block|{
name|int
name|i
decl_stmt|;
block|}
struct|;
end_struct

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(packed)
argument_list|)
name|S2
block|{
name|char
name|c
decl_stmt|;
name|struct
name|S2_i
name|inner
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
modifier|*
name|g2
parameter_list|(
name|struct
name|S2
modifier|*
name|s2
parameter_list|)
block|{
return|return
operator|&
name|s2
operator|->
name|inner
operator|.
name|i
return|;
comment|// expected-warning {{packed member 'inner' of class or structure 'S2'}}
block|}
end_function

begin_struct
struct|struct
name|S2_a
block|{
name|char
name|c
decl_stmt|;
name|struct
name|S2_i
name|inner
name|__attribute__
argument_list|(
operator|(
name|packed
operator|)
argument_list|)
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
modifier|*
name|g2_a
parameter_list|(
name|struct
name|S2_a
modifier|*
name|s2_a
parameter_list|)
block|{
return|return
operator|&
name|s2_a
operator|->
name|inner
operator|.
name|i
return|;
comment|// expected-warning {{packed member 'inner' of class or structure 'S2_a'}}
block|}
end_function

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(packed)
argument_list|)
name|S3
block|{
name|char
name|c
decl_stmt|;
struct|struct
block|{
name|int
name|i
decl_stmt|;
block|}
name|inner
struct|;
block|}
struct|;
end_struct

begin_function
name|int
modifier|*
name|g3
parameter_list|(
name|struct
name|S3
modifier|*
name|s3
parameter_list|)
block|{
return|return
operator|&
name|s3
operator|->
name|inner
operator|.
name|i
return|;
comment|// expected-warning {{packed member 'inner' of class or structure 'S3'}}
block|}
end_function

begin_struct
struct|struct
name|S4
block|{
name|char
name|c
decl_stmt|;
name|struct
name|__attribute__
argument_list|(
operator|(
name|packed
operator|)
argument_list|)
block|{
name|int
name|i
decl_stmt|;
block|}
name|inner
expr_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
modifier|*
name|g4
parameter_list|(
name|struct
name|S4
modifier|*
name|s4
parameter_list|)
block|{
return|return
operator|&
name|s4
operator|->
name|inner
operator|.
name|i
return|;
comment|// expected-warning {{packed member 'i' of class or structure 'S4::(anonymous)'}}
block|}
end_function

begin_struct
struct|struct
name|S5
block|{
name|char
name|c
decl_stmt|;
struct|struct
block|{
name|char
name|c1
decl_stmt|;
name|int
name|i
name|__attribute__
argument_list|(
operator|(
name|packed
operator|)
argument_list|)
decl_stmt|;
block|}
name|inner
struct|;
block|}
struct|;
end_struct

begin_function
name|int
modifier|*
name|g5
parameter_list|(
name|struct
name|S5
modifier|*
name|s5
parameter_list|)
block|{
return|return
operator|&
name|s5
operator|->
name|inner
operator|.
name|i
return|;
comment|// expected-warning {{packed member 'i' of class or structure 'S5::(anonymous)'}}
block|}
end_function

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(packed, aligned(
literal|2
argument|))
argument_list|)
name|AlignedTo2
block|{
name|int
name|x
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|char
modifier|*
name|g6
parameter_list|(
name|struct
name|AlignedTo2
modifier|*
name|s
parameter_list|)
block|{
return|return
operator|(
name|char
operator|*
operator|)
operator|&
name|s
operator|->
name|x
return|;
comment|// no-warning
block|}
end_function

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(packed, aligned(
literal|2
argument|))
argument_list|)
name|AlignedTo2Bis
block|{
name|int
name|x
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|struct
name|AlignedTo2Bis
modifier|*
name|g7
parameter_list|(
name|struct
name|AlignedTo2
modifier|*
name|s
parameter_list|)
block|{
return|return
operator|(
expr|struct
name|AlignedTo2Bis
operator|*
operator|)
operator|&
name|s
operator|->
name|x
return|;
comment|// no-warning
block|}
end_function

begin_typedef
typedef|typedef
struct|struct
block|{
name|char
name|c
decl_stmt|;
name|int
name|x
decl_stmt|;
block|}
name|__attribute__
typedef|((
name|packed
typedef|))
name|TypedefStructArguable
typedef|;
end_typedef

begin_typedef
typedef|typedef
union|union
block|{
name|char
name|c
decl_stmt|;
name|int
name|x
decl_stmt|;
block|}
name|__attribute
typedef|((
name|packed
typedef|))
name|TypedefUnionArguable
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|TypedefStructArguable
name|TypedefStructArguableTheSecond
typedef|;
end_typedef

begin_function
name|int
modifier|*
name|typedef1
parameter_list|(
name|TypedefStructArguable
modifier|*
name|s
parameter_list|)
block|{
return|return
operator|&
name|s
operator|->
name|x
return|;
comment|// expected-warning {{packed member 'x' of class or structure 'TypedefStructArguable'}}
block|}
end_function

begin_function
name|int
modifier|*
name|typedef2
parameter_list|(
name|TypedefStructArguableTheSecond
modifier|*
name|s
parameter_list|)
block|{
return|return
operator|&
name|s
operator|->
name|x
return|;
comment|// expected-warning {{packed member 'x' of class or structure 'TypedefStructArguable'}}
block|}
end_function

begin_function
name|int
modifier|*
name|typedef3
parameter_list|(
name|TypedefUnionArguable
modifier|*
name|s
parameter_list|)
block|{
return|return
operator|&
name|s
operator|->
name|x
return|;
comment|// expected-warning {{packed member 'x' of class or structure 'TypedefUnionArguable'}}
block|}
end_function

begin_struct
struct|struct
name|S6
block|{
union|union
block|{
name|char
name|c
decl_stmt|;
name|int
name|x
decl_stmt|;
block|}
name|__attribute__
argument_list|(
operator|(
name|packed
operator|)
argument_list|)
union|;
block|}
struct|;
end_struct

begin_function
name|int
modifier|*
name|anonymousInnerUnion
parameter_list|(
name|struct
name|S6
modifier|*
name|s
parameter_list|)
block|{
return|return
operator|&
name|s
operator|->
name|x
return|;
comment|// expected-warning {{packed member 'x' of class or structure 'S6::(anonymous)'}}
block|}
end_function

begin_struct
struct|struct
name|S6a
block|{
name|int
name|a
decl_stmt|;
name|int
name|_
decl_stmt|;
name|int
name|c
decl_stmt|;
name|char
name|__
decl_stmt|;
name|int
name|d
decl_stmt|;
block|}
name|__attribute__
argument_list|(
operator|(
name|packed
operator|,
name|aligned
argument_list|(
literal|16
argument_list|)
operator|)
argument_list|)
name|s6
struct|;
end_struct

begin_function
name|void
name|g8
parameter_list|()
block|{
name|f1
argument_list|(
operator|&
name|s6
operator|.
name|a
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|&
name|s6
operator|.
name|c
argument_list|)
expr_stmt|;
comment|// no-warning
name|f1
argument_list|(
operator|&
name|s6
operator|.
name|d
argument_list|)
expr_stmt|;
comment|// expected-warning {{packed member 'd' of class or structure 'S6a'}}
block|}
end_function

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(packed, aligned(
literal|1
argument|))
argument_list|)
name|MisalignedContainee
block|{
name|double
name|d
decl_stmt|;
block|}
struct|;
end_struct

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(aligned(
literal|8
argument|))
argument_list|)
name|AlignedContainer
block|{
name|struct
name|MisalignedContainee
name|b
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|AlignedContainer
modifier|*
name|p
decl_stmt|;
end_decl_stmt

begin_function
name|double
modifier|*
name|g9
parameter_list|()
block|{
return|return
operator|&
name|p
operator|->
name|b
operator|.
name|d
return|;
comment|// no-warning
block|}
end_function

begin_union
union|union
name|OneUnion
block|{
name|uint32_t
name|a
decl_stmt|;
name|uint32_t
name|b
range|:
literal|1
decl_stmt|;
block|}
union|;
end_union

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(packed)
argument_list|)
name|S7
block|{
name|uint8_t
name|length
decl_stmt|;
name|uint8_t
name|stuff
decl_stmt|;
name|uint8_t
name|padding
index|[
literal|2
index|]
decl_stmt|;
name|union
name|OneUnion
name|one_union
decl_stmt|;
block|}
struct|;
end_struct

begin_union
union|union
name|AnotherUnion
block|{
name|long
name|data
decl_stmt|;
name|struct
name|S7
name|s
decl_stmt|;
block|}
modifier|*
name|au
union|;
end_union

begin_function
name|union
name|OneUnion
modifier|*
name|get_OneUnion
parameter_list|(
name|void
parameter_list|)
block|{
return|return
operator|&
name|au
operator|->
name|s
operator|.
name|one_union
return|;
comment|// no-warning
block|}
end_function

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(packed)
argument_list|)
name|S8
block|{
name|uint8_t
name|data1
decl_stmt|;
name|uint8_t
name|data2
decl_stmt|;
name|uint16_t
name|wider_data
decl_stmt|;
block|}
struct|;
end_struct

begin_define
define|#
directive|define
name|LE_READ_2
parameter_list|(
name|p
parameter_list|)
define|\
value|((uint16_t)					\ 	 ((((const uint8_t *)(p))[0]      ) |		\ 	  (((const uint8_t *)(p))[1]<<  8)))
end_define

begin_function
name|uint32_t
name|get_wider_data
parameter_list|(
name|struct
name|S8
modifier|*
name|s
parameter_list|)
block|{
return|return
name|LE_READ_2
argument_list|(
operator|&
name|s
operator|->
name|wider_data
argument_list|)
return|;
comment|// no-warning
block|}
end_function

begin_struct
struct|struct
name|S9
block|{
name|uint32_t
name|x
decl_stmt|;
name|uint8_t
name|y
index|[
literal|2
index|]
decl_stmt|;
name|uint16_t
name|z
decl_stmt|;
block|}
name|__attribute__
argument_list|(
operator|(
name|__packed__
operator|)
argument_list|)
struct|;
end_struct

begin_typedef
typedef|typedef
name|struct
name|S9
name|__attribute__
argument_list|(
operator|(
name|__aligned__
argument_list|(
literal|16
argument_list|)
operator|)
argument_list|)
name|aligned_S9
typedef|;
end_typedef

begin_function
name|void
name|g10
parameter_list|()
block|{
name|struct
name|S9
name|x
decl_stmt|;
name|struct
name|S9
name|__attribute__
argument_list|(
operator|(
name|__aligned__
argument_list|(
literal|8
argument_list|)
operator|)
argument_list|)
name|y
decl_stmt|;
name|aligned_S9
name|z
decl_stmt|;
name|uint32_t
modifier|*
name|p32
decl_stmt|;
name|p32
operator|=
operator|&
name|x
operator|.
name|x
expr_stmt|;
comment|// expected-warning {{packed member 'x' of class or structure 'S9'}}
name|p32
operator|=
operator|&
name|y
operator|.
name|x
expr_stmt|;
comment|// no-warning
name|p32
operator|=
operator|&
name|z
operator|.
name|x
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_typedef
typedef|typedef
struct|struct
block|{
name|uint32_t
name|msgh_bits
decl_stmt|;
name|uint32_t
name|msgh_size
decl_stmt|;
name|int32_t
name|msgh_voucher_port
decl_stmt|;
name|int32_t
name|msgh_id
decl_stmt|;
block|}
name|S10Header
typedef|;
end_typedef

begin_typedef
typedef|typedef
struct|struct
block|{
name|uint32_t
name|t
decl_stmt|;
name|uint64_t
name|m
decl_stmt|;
name|uint32_t
name|p
decl_stmt|;
union|union
block|{
struct|struct
block|{
name|uint32_t
name|a
decl_stmt|;
name|double
name|z
decl_stmt|;
block|}
name|__attribute__
argument_list|(
operator|(
name|aligned
argument_list|(
literal|8
argument_list|)
operator|,
name|packed
operator|)
argument_list|)
name|a
struct|;
struct|struct
block|{
name|uint32_t
name|b
decl_stmt|;
name|double
name|z
decl_stmt|;
name|uint32_t
name|a
decl_stmt|;
block|}
name|__attribute__
argument_list|(
operator|(
name|aligned
argument_list|(
literal|8
argument_list|)
operator|,
name|packed
operator|)
argument_list|)
name|b
struct|;
block|}
union|;
block|}
name|__attribute__
typedef|((
name|aligned
typedef|(8)
operator|,
name|packed
typedef|))
name|S10Data
typedef|;
end_typedef

begin_typedef
typedef|typedef
struct|struct
block|{
name|S10Header
name|hdr
decl_stmt|;
name|uint32_t
name|size
decl_stmt|;
name|uint8_t
name|count
decl_stmt|;
name|S10Data
name|data
index|[]
name|__attribute__
decl|((
name|aligned
decl|(8
decl_stmt|)
block|)
typedef|);
end_typedef

begin_expr_stmt
unit|} __attribute__
operator|(
operator|(
name|aligned
argument_list|(
literal|8
argument_list|)
operator|,
name|packed
operator|)
operator|)
name|S10
expr_stmt|;
end_expr_stmt

begin_function_decl
name|void
name|g11
parameter_list|(
name|S10Header
modifier|*
name|hdr
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|g12
parameter_list|(
name|S10
modifier|*
name|s
parameter_list|)
block|{
name|g11
argument_list|(
operator|&
name|s
operator|->
name|hdr
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_struct
struct|struct
name|S11
block|{
name|uint32_t
name|x
decl_stmt|;
block|}
name|__attribute__
argument_list|(
operator|(
name|__packed__
operator|)
argument_list|)
struct|;
end_struct

begin_function
name|void
name|g13
parameter_list|(
name|void
parameter_list|)
block|{
name|struct
name|S11
name|__attribute__
argument_list|(
operator|(
name|__aligned__
argument_list|(
literal|4
argument_list|)
operator|)
argument_list|)
name|a
index|[
literal|4
index|]
decl_stmt|;
name|uint32_t
modifier|*
name|p32
decl_stmt|;
name|p32
operator|=
operator|&
name|a
index|[
literal|0
index|]
operator|.
name|x
expr_stmt|;
comment|// no-warning
block|}
end_function

end_unit

