begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -faltivec -verify -pedantic -fsyntax-only
end_comment

begin_typedef
typedef|typedef
name|int
name|v4
name|__attribute
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|short
name|v8
name|__attribute
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_function
name|v8
name|foo
parameter_list|(
name|void
parameter_list|)
block|{
name|v8
name|a
decl_stmt|;
name|v4
name|b
decl_stmt|;
name|a
operator|=
operator|(
name|v8
operator|)
block|{
literal|4
block|,
literal|2
block|}
expr_stmt|;
comment|// expected-error {{too few elements in vector initialization (expected 8 elements, have 2)}}
name|b
operator|=
call|(
name|v4
call|)
argument_list|(
literal|5
argument_list|,
literal|6
argument_list|,
literal|7
argument_list|,
literal|8
argument_list|,
literal|9
argument_list|)
expr_stmt|;
comment|// expected-warning {{excess elements in vector initializer}}
name|b
operator|=
call|(
name|v4
call|)
argument_list|(
literal|5
argument_list|,
literal|6
argument_list|,
literal|8
argument_list|,
literal|8.0f
argument_list|)
expr_stmt|;
return|return
operator|(
name|v8
operator|)
block|{
literal|0
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|}
return|;
comment|// FIXME: test that (type)(fn)(args) still works with -faltivec
comment|// FIXME: test that c++ overloaded commas still work -faltivec
block|}
end_function

end_unit

