begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -triple=powerpc-apple-darwin8 -target-feature +altivec -verify -pedantic -fsyntax-only
end_comment

begin_typedef
typedef|typedef
name|int
name|v4
name|__attribute
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|short
name|v8
name|__attribute
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_function
name|v8
name|foo
parameter_list|(
name|void
parameter_list|)
block|{
name|v8
name|a
decl_stmt|;
name|v4
name|b
decl_stmt|;
name|a
operator|=
operator|(
name|v8
operator|)
block|{
literal|4
block|,
literal|2
block|}
expr_stmt|;
name|b
operator|=
call|(
name|v4
call|)
argument_list|(
literal|5
argument_list|,
literal|6
argument_list|,
literal|7
argument_list|,
literal|8
argument_list|,
literal|9
argument_list|)
expr_stmt|;
comment|// expected-warning {{excess elements in vector initializer}}
name|b
operator|=
call|(
name|v4
call|)
argument_list|(
literal|5
argument_list|,
literal|6
argument_list|,
literal|8
argument_list|,
literal|8.0f
argument_list|)
expr_stmt|;
name|vector
name|int
name|vi
decl_stmt|;
name|vi
operator|=
call|(
name|vector
name|int
call|)
argument_list|(
literal|1
argument_list|)
expr_stmt|;
name|vi
operator|=
call|(
name|vector
name|int
call|)
argument_list|(
literal|1
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{number of elements must be either one or match the size of the vector}}
name|vi
operator|=
call|(
name|vector
name|int
call|)
argument_list|(
literal|1
argument_list|,
literal|2
argument_list|,
literal|3
argument_list|,
literal|4
argument_list|)
expr_stmt|;
name|vi
operator|=
call|(
name|vector
name|int
call|)
argument_list|(
literal|1
argument_list|,
literal|2
argument_list|,
literal|3
argument_list|,
literal|4
argument_list|,
literal|5
argument_list|)
expr_stmt|;
comment|// expected-warning {{excess elements in vector initializer}}
name|vi
operator|=
operator|(
name|vector
name|int
operator|)
block|{
literal|1
block|}
expr_stmt|;
name|vi
operator|=
operator|(
name|vector
name|int
operator|)
block|{
literal|1
block|,
literal|2
block|}
expr_stmt|;
name|vi
operator|=
operator|(
name|vector
name|int
operator|)
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|}
expr_stmt|;
comment|// expected-warning {{excess elements in vector initializer}}
name|vector
name|float
name|vf
decl_stmt|;
name|vf
operator|=
call|(
name|vector
name|float
call|)
argument_list|(
literal|1.0
argument_list|)
expr_stmt|;
return|return
operator|(
name|v8
operator|)
block|{
literal|0
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|}
return|;
comment|// FIXME: test that (type)(fn)(args) still works with -maltivec
comment|// FIXME: test that c++ overloaded commas still work -maltivec
block|}
end_function

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|__overloadable__
operator|)
argument_list|)
name|f
argument_list|(
name|v4
name|a
argument_list|)
block|{ }
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|__overloadable__
operator|)
argument_list|)
name|f
argument_list|(
name|int
name|a
argument_list|)
block|{ }
end_decl_stmt

begin_function
name|void
name|test
parameter_list|()
block|{
name|v4
name|vGCC
decl_stmt|;
name|vector
name|int
name|vAltiVec
decl_stmt|;
name|f
argument_list|(
name|vAltiVec
argument_list|)
expr_stmt|;
name|vGCC
operator|=
name|vAltiVec
expr_stmt|;
name|int
name|res
init|=
name|vGCC
operator|>
name|vAltiVec
decl_stmt|;
name|vAltiVec
operator|=
literal|0
condition|?
name|vGCC
else|:
name|vGCC
expr_stmt|;
block|}
end_function

end_unit

