begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -fsyntax-only -verify
end_comment

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|annotate
argument_list|(
literal|"foo"
argument_list|)
operator|)
argument_list|)
name|foo
argument_list|(
name|float
operator|*
name|a
argument_list|)
block|{
name|__attribute__
argument_list|(
argument|(annotate(
literal|"bar"
argument|))
argument_list|)
name|int
name|x
decl_stmt|;
name|__attribute__
argument_list|(
argument|(annotate(
literal|1
argument|))
argument_list|)
name|int
name|y
decl_stmt|;
comment|// expected-error {{argument to annotate attribute was not a string literal}}
name|__attribute__
argument_list|(
argument|(annotate(
literal|"bar"
argument|,
literal|1
argument|))
argument_list|)
name|int
name|z
decl_stmt|;
comment|// expected-error {{attribute takes one argument}}
name|int
name|u
init|=
name|__builtin_annotation
argument_list|(
name|z
argument_list|,
operator|(
name|char
operator|*
operator|)
literal|0
argument_list|)
decl_stmt|;
comment|// expected-error {{__builtin_annotation requires a non wide string constant}}
name|int
name|v
init|=
name|__builtin_annotation
argument_list|(
name|z
argument_list|,
operator|(
name|char
operator|*
operator|)
literal|L"bar"
argument_list|)
decl_stmt|;
comment|// expected-error {{__builtin_annotation requires a non wide string constant}}
name|int
name|w
init|=
name|__builtin_annotation
argument_list|(
name|z
argument_list|,
literal|"foo"
argument_list|)
decl_stmt|;
block|}
end_decl_stmt

end_unit

