begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -pedantic -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -Wgnu -Wc11-extensions -verify %s
end_comment

begin_comment
comment|// REQUIRES: LP64
end_comment

begin_function_decl
specifier|extern
name|int
name|foof
parameter_list|()
init|=
literal|1
function_decl|;
end_function_decl

begin_comment
comment|// expected-error{{illegal initializer (only variables can be initialized)}}
end_comment

begin_decl_stmt
specifier|static
name|int
name|x
decl_stmt|,
name|y
decl_stmt|,
name|z
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|static
name|int
name|ary
index|[]
init|=
block|{
name|x
block|,
name|y
block|,
name|z
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error{{initializer element is not a compile-time constant}}
end_comment

begin_decl_stmt
name|int
name|ary2
index|[]
init|=
block|{
name|x
block|,
name|y
block|,
name|z
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error{{initializer element is not a compile-time constant}}
end_comment

begin_decl_stmt
specifier|extern
name|int
name|fileScopeExtern
index|[
literal|3
index|]
init|=
block|{
literal|1
block|,
literal|3
block|,
literal|5
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{'extern' variable has an initializer}}
end_comment

begin_decl_stmt
specifier|static
name|long
name|ary3
index|[]
init|=
block|{
literal|1
block|,
literal|"abc"
block|,
literal|3
block|,
literal|4
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{incompatible pointer to integer conversion initializing 'long' with an expression of type 'char [4]'}}
end_comment

begin_function
name|void
name|func
parameter_list|()
block|{
name|int
name|x
init|=
literal|1
decl_stmt|;
typedef|typedef
name|int
name|TInt
init|=
literal|1
typedef|;
comment|// expected-error{{illegal initializer (only variables can be initialized)}}
name|int
name|xComputeSize
index|[]
init|=
block|{
literal|1
block|,
literal|3
block|,
literal|5
block|}
decl_stmt|;
name|int
name|x3
index|[
name|x
index|]
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
comment|// expected-error{{variable-sized object may not be initialized}}
name|int
name|x4
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
comment|// expected-warning{{excess elements in scalar initializer}}
name|int
name|y
index|[
literal|4
index|]
index|[
literal|3
index|]
init|=
block|{
block|{
literal|1
block|,
literal|3
block|,
literal|5
block|}
block|,
block|{
literal|2
block|,
literal|4
block|,
literal|6
block|}
block|,
block|{
literal|3
block|,
literal|5
block|,
literal|7
block|}
block|,   }
decl_stmt|;
name|int
name|y2
index|[
literal|4
index|]
index|[
literal|3
index|]
init|=
block|{
literal|1
block|,
literal|3
block|,
literal|5
block|,
literal|2
block|,
literal|4
block|,
literal|6
block|,
literal|3
block|,
literal|5
block|,
literal|7
block|}
decl_stmt|;
name|int
name|y3
index|[
literal|4
index|]
index|[
literal|3
index|]
init|=
block|{
block|{
literal|1
block|,
literal|3
block|,
literal|5
block|}
block|,
block|{
literal|2
block|,
literal|4
block|,
literal|6
block|}
block|,
block|{
literal|3
block|,
literal|5
block|,
literal|7
block|}
block|,
block|{
literal|4
block|,
literal|6
block|,
literal|8
block|}
block|,
block|{
literal|5
block|}
block|,
comment|// expected-warning{{excess elements in array initializer}}
block|}
decl_stmt|;
struct|struct
name|threeElements
block|{
name|int
name|a
decl_stmt|,
name|b
decl_stmt|,
name|c
decl_stmt|;
block|}
name|z
init|=
block|{
literal|1
block|}
struct|;
name|struct
name|threeElements
modifier|*
name|p
init|=
literal|7
decl_stmt|;
comment|// expected-warning{{incompatible integer to pointer conversion initializing 'struct threeElements *' with an expression of type 'int'}}
specifier|extern
name|int
name|blockScopeExtern
index|[
literal|3
index|]
init|=
block|{
literal|1
block|,
literal|3
block|,
literal|5
block|}
decl_stmt|;
comment|// expected-error{{'extern' variable cannot have an initializer}}
specifier|static
name|long
name|x2
index|[
literal|3
index|]
init|=
block|{
literal|1.0
block|,
literal|"abc"
block|,
comment|// expected-warning{{incompatible pointer to integer conversion initializing 'long' with an expression of type 'char [4]'}}
literal|5.8
block|}
decl_stmt|;
comment|// expected-warning {{implicit conversion from 'double' to 'long' changes value from 5.8 to 5}}
block|}
end_function

begin_function
name|void
name|test
parameter_list|()
block|{
name|int
name|y1
index|[
literal|3
index|]
init|=
block|{
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|}
comment|// expected-warning{{excess elements in scalar initializer}}
block|}
decl_stmt|;
name|int
name|y3
index|[
literal|4
index|]
index|[
literal|3
index|]
init|=
block|{
block|{
literal|1
block|,
literal|3
block|,
literal|5
block|}
block|,
block|{
literal|2
block|,
literal|4
block|,
literal|6
block|}
block|,
block|{
literal|3
block|,
literal|5
block|,
literal|7
block|}
block|,
block|{
literal|4
block|,
literal|6
block|,
literal|8
block|}
block|,
block|{  }
block|,
comment|// expected-warning{{use of GNU empty initializer extension}} expected-warning{{excess elements in array initializer}}
block|}
decl_stmt|;
name|int
name|y4
index|[
literal|4
index|]
index|[
literal|3
index|]
init|=
block|{
block|{
literal|1
block|,
literal|3
block|,
literal|5
block|,
literal|2
block|}
block|,
comment|// expected-warning{{excess elements in array initializer}}
block|{
literal|4
block|,
literal|6
block|}
block|,
block|{
literal|3
block|,
literal|5
block|,
literal|7
block|}
block|,
block|{
literal|4
block|,
literal|6
block|,
literal|8
block|}
block|,   }
decl_stmt|;
block|}
end_function

begin_function
name|void
name|allLegalAndSynonymous
parameter_list|()
block|{
name|short
name|q
index|[
literal|4
index|]
index|[
literal|3
index|]
index|[
literal|2
index|]
init|=
block|{
block|{
literal|1
block|}
block|,
block|{
literal|2
block|,
literal|3
block|}
block|,
block|{
literal|4
block|,
literal|5
block|,
literal|6
block|}
block|}
decl_stmt|;
name|short
name|q2
index|[
literal|4
index|]
index|[
literal|3
index|]
index|[
literal|2
index|]
init|=
block|{
block|{
literal|1
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|}
block|,
block|{
literal|2
block|,
literal|3
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|}
block|,
block|{
literal|4
block|,
literal|5
block|,
literal|6
block|}
block|}
decl_stmt|;
name|short
name|q3
index|[
literal|4
index|]
index|[
literal|3
index|]
index|[
literal|2
index|]
init|=
block|{
block|{
block|{
literal|1
block|}
block|,     }
block|,
block|{
block|{
literal|2
block|,
literal|3
block|}
block|,     }
block|,
block|{
block|{
literal|4
block|,
literal|5
block|}
block|,
block|{
literal|6
block|}
block|,     }
block|,   }
decl_stmt|;
block|}
end_function

begin_function
name|void
name|legal
parameter_list|()
block|{
name|short
name|q
index|[]
index|[
literal|3
index|]
index|[
literal|2
index|]
init|=
block|{
block|{
literal|1
block|}
block|,
block|{
literal|2
block|,
literal|3
block|}
block|,
block|{
literal|4
block|,
literal|5
block|,
literal|6
block|}
block|}
decl_stmt|;
name|int
name|q_sizecheck
index|[
operator|(
sizeof|sizeof
argument_list|(
name|q
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
name|short
index|[
literal|3
index|]
index|[
literal|2
index|]
argument_list|)
operator|)
operator|==
literal|3
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
block|}
end_function

begin_decl_stmt
name|unsigned
name|char
name|asso_values
index|[]
init|=
block|{
literal|34
block|}
decl_stmt|;
end_decl_stmt

begin_function
name|int
name|legal2
parameter_list|()
block|{
return|return
name|asso_values
index|[
literal|0
index|]
return|;
block|}
end_function

begin_function
name|void
name|illegal
parameter_list|()
block|{
name|short
name|q2
index|[
literal|4
index|]
index|[]
index|[
literal|2
index|]
init|=
block|{
comment|// expected-error{{array has incomplete element type 'short [][2]'}}
block|{
literal|1
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|}
block|,
block|{
literal|2
block|,
literal|3
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|,
literal|0
block|}
block|,
block|{
literal|4
block|,
literal|5
block|,
literal|6
block|}
block|}
decl_stmt|;
name|short
name|q3
index|[
literal|4
index|]
index|[
literal|3
index|]
index|[]
init|=
block|{
comment|// expected-error{{array has incomplete element type 'short []'}}
block|{
block|{
literal|1
block|}
block|,     }
block|,
block|{
block|{
literal|2
block|,
literal|3
block|}
block|,     }
block|,
block|{
block|{
literal|4
block|,
literal|5
block|}
block|,
block|{
literal|6
block|}
block|,     }
block|,   }
decl_stmt|;
name|int
name|a
index|[]
index|[]
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
comment|// expected-error{{array has incomplete element type 'int []'}}
block|}
end_function

begin_typedef
typedef|typedef
name|int
name|AryT
index|[]
typedef|;
end_typedef

begin_function
name|void
name|testTypedef
parameter_list|()
block|{
name|AryT
name|a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|,
name|b
init|=
block|{
literal|3
block|,
literal|4
block|,
literal|5
block|}
decl_stmt|;
name|int
name|a_sizecheck
index|[
operator|(
sizeof|sizeof
argument_list|(
name|a
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
name|int
argument_list|)
operator|)
operator|==
literal|2
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
name|int
name|b_sizecheck
index|[
operator|(
sizeof|sizeof
argument_list|(
name|b
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
name|int
argument_list|)
operator|)
operator|==
literal|3
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
block|}
end_function

begin_decl_stmt
specifier|static
name|char
specifier|const
name|xx
index|[]
init|=
literal|"test"
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|xx_sizecheck
index|[
operator|(
sizeof|sizeof
argument_list|(
name|xx
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
name|char
argument_list|)
operator|)
operator|==
literal|5
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|static
name|char
specifier|const
name|yy
index|[
literal|5
index|]
init|=
literal|"test"
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|static
name|char
specifier|const
name|zz
index|[
literal|3
index|]
init|=
literal|"test"
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{initializer-string for char array is too long}}
end_comment

begin_function
name|void
name|charArrays
parameter_list|()
block|{
specifier|static
name|char
specifier|const
name|test
index|[]
init|=
literal|"test"
decl_stmt|;
name|int
name|test_sizecheck
index|[
operator|(
sizeof|sizeof
argument_list|(
name|test
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
name|char
argument_list|)
operator|)
operator|==
literal|5
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
specifier|static
name|char
specifier|const
name|test2
index|[]
init|=
block|{
literal|"weird stuff"
block|}
decl_stmt|;
specifier|static
name|char
specifier|const
name|test3
index|[]
init|=
block|{
literal|"test"
block|,
literal|"excess stuff"
block|}
decl_stmt|;
comment|// expected-warning{{excess elements in char array initializer}}
name|char
modifier|*
name|cp
index|[]
init|=
block|{
literal|"Hello"
block|}
decl_stmt|;
name|char
name|c
index|[]
init|=
block|{
literal|"Hello"
block|}
decl_stmt|;
name|int
name|l
index|[
sizeof|sizeof
argument_list|(
name|c
argument_list|)
operator|==
literal|6
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
name|int
name|i
index|[]
init|=
block|{
literal|"Hello "
block|}
decl_stmt|;
comment|// expected-warning{{incompatible pointer to integer conversion initializing 'int' with an expression of type 'char [7]'}}
name|char
name|c2
index|[]
init|=
block|{
literal|"Hello"
block|,
literal|"Good bye"
block|}
decl_stmt|;
comment|//expected-warning{{excess elements in char array initializer}}
name|int
name|i2
index|[
literal|1
index|]
init|=
block|{
literal|"Hello"
block|}
decl_stmt|;
comment|//expected-warning{{incompatible pointer to integer conversion initializing 'int' with an expression of type 'char [6]'}}
name|char
name|c3
index|[
literal|5
index|]
init|=
block|{
literal|"Hello"
block|}
decl_stmt|;
name|char
name|c4
index|[
literal|4
index|]
init|=
block|{
literal|"Hello"
block|}
decl_stmt|;
comment|//expected-warning{{initializer-string for char array is too long}}
name|int
name|i3
index|[]
init|=
block|{}
decl_stmt|;
comment|//expected-warning{{zero size arrays are an extension}} expected-warning{{use of GNU empty initializer extension}}
block|}
end_function

begin_function
name|void
name|variableArrayInit
parameter_list|()
block|{
name|int
name|a
init|=
literal|4
decl_stmt|;
name|char
name|strlit
index|[
name|a
index|]
init|=
literal|"foo"
decl_stmt|;
comment|//expected-error{{variable-sized object may not be initialized}}
name|int
name|b
index|[
name|a
index|]
init|=
block|{
literal|1
block|,
literal|2
block|,
literal|4
block|}
decl_stmt|;
comment|//expected-error{{variable-sized object may not be initialized}}
block|}
end_function

begin_comment
comment|// Pure array tests
end_comment

begin_decl_stmt
name|float
name|r1
index|[
literal|10
index|]
init|=
block|{
block|{
literal|7
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|//expected-warning{{braces around scalar initializer}}
end_comment

begin_decl_stmt
name|float
name|r2
index|[]
init|=
block|{
block|{
literal|8
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|//expected-warning{{braces around scalar initializer}}
end_comment

begin_decl_stmt
name|char
name|r3
index|[]
index|[
literal|5
index|]
init|=
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|,
literal|6
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|r3_sizecheck
index|[
operator|(
sizeof|sizeof
argument_list|(
name|r3
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
name|char
index|[
literal|5
index|]
argument_list|)
operator|)
operator|==
literal|2
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|r3_2
index|[
sizeof|sizeof
name|r3
operator|==
literal|10
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|float
name|r4
index|[
literal|1
index|]
index|[
literal|2
index|]
init|=
block|{
literal|1
block|,
block|{
literal|2
block|}
block|,
literal|3
block|,
literal|4
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|//expected-warning{{braces around scalar initializer}} expected-warning{{excess elements in array initializer}}
end_comment

begin_decl_stmt
name|char
name|r5
index|[]
index|[
literal|5
index|]
init|=
block|{
literal|"aa"
block|,
literal|"bbb"
block|,
literal|"ccccc"
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|r6
index|[
sizeof|sizeof
name|r5
operator|==
literal|15
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|const
name|char
name|r7
index|[]
init|=
literal|"zxcv"
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|r8
index|[
literal|5
index|]
init|=
literal|"5char"
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|r9
index|[
literal|5
index|]
init|=
literal|"6chars"
decl_stmt|;
end_decl_stmt

begin_comment
comment|//expected-warning{{initializer-string for char array is too long}}
end_comment

begin_decl_stmt
name|unsigned
name|char
name|r10
index|[]
init|=
name|__extension__
argument_list|(
generic_selection|_Generic(
literal|0
generic_selection|,
name|int
association|:
operator|(
name|__extension__
literal|"foo"
operator|)
generic_selection|)
argument_list|)
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|r11
index|[
literal|0
index|]
init|=
block|{}
decl_stmt|;
end_decl_stmt

begin_comment
comment|//expected-warning{{zero size arrays are an extension}} expected-warning{{use of GNU empty initializer extension}}
end_comment

begin_comment
comment|// Some struct tests
end_comment

begin_function
name|void
name|autoStructTest
parameter_list|()
block|{
struct|struct
name|s1
block|{
name|char
name|a
decl_stmt|;
name|char
name|b
decl_stmt|;
block|}
name|t1
struct|;
struct|struct
name|s2
block|{
name|struct
name|s1
name|c
decl_stmt|;
block|}
name|t2
init|=
block|{
name|t1
block|}
struct|;
comment|// The following is a less than great diagnostic (though it's on par with EDG).
name|struct
name|s1
name|t3
index|[]
init|=
block|{
name|t1
block|,
name|t1
block|,
literal|"abc"
block|,
literal|0
block|}
decl_stmt|;
comment|//expected-warning{{incompatible pointer to integer conversion initializing 'char' with an expression of type 'char [4]'}}
name|int
name|t4
index|[
sizeof|sizeof
name|t3
operator|==
literal|6
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
block|}
end_function

begin_struct
struct|struct
name|foo
block|{
name|int
name|z
decl_stmt|;
block|}
name|w
struct|;
end_struct

begin_function
name|int
name|bar
parameter_list|(
name|void
parameter_list|)
block|{
name|struct
name|foo
name|z
init|=
block|{
name|w
block|}
decl_stmt|;
comment|//expected-error{{initializing 'int' with an expression of incompatible type 'struct foo'}}
return|return
name|z
operator|.
name|z
return|;
block|}
end_function

begin_struct
struct|struct
name|s3
block|{
name|void
function_decl|(
modifier|*
name|a
function_decl|)
parameter_list|(
name|void
parameter_list|)
function_decl|;
block|}
name|t5
init|=
block|{
name|autoStructTest
block|}
struct|;
end_struct

begin_struct
struct|struct
block|{
name|int
name|a
decl_stmt|;
name|int
name|b
index|[]
decl_stmt|;
block|}
name|t6
init|=
block|{
literal|1
block|,
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|}
block|}
struct|;
end_struct

begin_comment
comment|// expected-warning{{flexible array initialization is a GNU extension}} \
end_comment

begin_comment
comment|// expected-note{{initialized flexible array member 'b' is here}}
end_comment

begin_union
union|union
block|{
name|char
name|a
decl_stmt|;
name|int
name|b
decl_stmt|;
block|}
name|t7
index|[]
init|=
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|}
union|;
end_union

begin_decl_stmt
name|int
name|t8
index|[
sizeof|sizeof
name|t7
operator|==
operator|(
literal|3
operator|*
sizeof|sizeof
argument_list|(
name|int
argument_list|)
operator|)
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|bittest
block|{
name|int
label|:
literal|31
operator|,
name|a
operator|,
operator|:
literal|21
operator|,
operator|:
literal|12
operator|,
name|b
expr_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|bittest
name|bittestvar
init|=
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|//expected-warning{{excess elements in struct initializer}}
end_comment

begin_comment
comment|// Not completely sure what should happen here...
end_comment

begin_decl_stmt
name|int
name|u1
init|=
block|{}
decl_stmt|;
end_decl_stmt

begin_comment
comment|//expected-warning{{use of GNU empty initializer extension}} expected-error{{scalar initializer cannot be empty}}
end_comment

begin_decl_stmt
name|int
name|u2
init|=
block|{
block|{
literal|3
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|//expected-warning{{too many braces around scalar initializer}}
end_comment

begin_comment
comment|// PR2362
end_comment

begin_function
name|void
name|varArray
parameter_list|()
block|{
name|int
name|c
index|[]
index|[
name|x
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
comment|//expected-error{{variable-sized object may not be initialized}}
block|}
end_function

begin_comment
comment|// PR2151
end_comment

begin_function
name|void
name|emptyInit
parameter_list|()
block|{
struct|struct
block|{}
name|x
index|[]
init|=
block|{
literal|6
block|}
struct|;
block|}
end_function

begin_comment
comment|//expected-warning{{empty struct is a GNU extension}} \
end_comment

begin_comment
comment|// expected-error{{initializer for aggregate with no elements}}
end_comment

begin_function
name|void
name|noNamedInit
parameter_list|()
block|{
struct|struct
block|{
name|int
label|:
literal|5
expr_stmt|;
block|}
name|x
index|[]
init|=
block|{
literal|6
block|}
struct|;
comment|//expected-error{{initializer for aggregate with no elements}} \
comment|// expected-warning {{struct without named members is a GNU extension}}
block|}
end_function

begin_struct
struct|struct
block|{
name|int
name|a
decl_stmt|;
name|int
label|:
literal|5
expr_stmt|;
block|}
name|noNamedImplicit
index|[]
init|=
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|}
struct|;
end_struct

begin_decl_stmt
name|int
name|noNamedImplicitCheck
index|[
sizeof|sizeof
argument_list|(
name|noNamedImplicit
argument_list|)
operator|==
literal|3
operator|*
sizeof|sizeof
argument_list|(
operator|*
name|noNamedImplicit
argument_list|)
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_comment
comment|// ptrs are constant
end_comment

begin_struct
struct|struct
name|soft_segment_descriptor
block|{
name|long
name|ssd_base
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
specifier|static
name|int
name|dblfault_tss
decl_stmt|;
end_decl_stmt

begin_union
union|union
name|uniao
block|{
name|int
name|ola
decl_stmt|;
block|}
name|xpto
index|[
literal|1
index|]
union|;
end_union

begin_decl_stmt
name|struct
name|soft_segment_descriptor
name|gdt_segs
index|[]
init|=
block|{
block|{
operator|(
name|long
operator|)
operator|&
name|dblfault_tss
block|}
block|,
block|{
operator|(
name|long
operator|)
name|xpto
block|}
block|, }
decl_stmt|;
end_decl_stmt

begin_function_decl
specifier|static
name|void
name|sppp_ipv6cp_up
parameter_list|()
function_decl|;
end_function_decl

begin_struct
specifier|const
struct|struct
block|{}
name|ipcp
init|=
block|{
name|sppp_ipv6cp_up
block|}
struct|;
end_struct

begin_comment
comment|//expected-warning{{empty struct is a GNU extension}} \
end_comment

begin_comment
comment|// expected-warning{{excess elements in struct initializer}}
end_comment

begin_struct
struct|struct
name|_Matrix
block|{
union|union
block|{
name|float
name|m
index|[
literal|4
index|]
index|[
literal|4
index|]
decl_stmt|;
block|}
union|;
block|}
struct|;
end_struct

begin_comment
comment|//expected-warning{{anonymous unions are a C11 extension}}
end_comment

begin_typedef
typedef|typedef
name|struct
name|_Matrix
name|Matrix
typedef|;
end_typedef

begin_function
name|void
name|test_matrix
parameter_list|()
block|{
specifier|const
name|Matrix
name|mat1
init|=
block|{
block|{
block|{
literal|1.0f
block|,
literal|2.0f
block|,
literal|3.0f
block|,
literal|4.0f
block|,
literal|5.0f
block|,
literal|6.0f
block|,
literal|7.0f
block|,
literal|8.0f
block|,
literal|9.0f
block|,
literal|10.0f
block|,
literal|11.0f
block|,
literal|12.0f
block|,
literal|13.0f
block|,
literal|14.0f
block|,
literal|15.0f
block|,
literal|16.0f
block|}
block|}
block|}
decl_stmt|;
specifier|const
name|Matrix
name|mat2
init|=
block|{
literal|1.0f
block|,
literal|2.0f
block|,
literal|3.0f
block|,
literal|4.0f
block|,
literal|5.0f
block|,
literal|6.0f
block|,
literal|7.0f
block|,
literal|8.0f
block|,
literal|9.0f
block|,
literal|10.0f
block|,
literal|11.0f
block|,
literal|12.0f
block|,
literal|13.0f
block|,
literal|14.0f
block|,
literal|15.0f
block|,
literal|16.0f
block|}
decl_stmt|;
block|}
end_function

begin_decl_stmt
name|char
name|badchararray
index|[
literal|1
index|]
init|=
block|{
name|badchararray
index|[
literal|0
index|]
block|,
literal|"asdf"
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning {{excess elements in array initializer}} expected-error {{initializer element is not a compile-time constant}}
end_comment

begin_comment
comment|// Test the GNU extension for initializing an array from an array
end_comment

begin_comment
comment|// compound literal. PR9261.
end_comment

begin_typedef
typedef|typedef
name|int
name|int5
index|[
literal|5
index|]
typedef|;
end_typedef

begin_decl_stmt
name|int
name|a1
index|[
literal|5
index|]
init|=
operator|(
name|int
index|[]
operator|)
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{initialization of an array of type 'int [5]' from a compound literal of type 'int [5]' is a GNU extension}}
end_comment

begin_decl_stmt
name|int
name|a2
index|[
literal|5
index|]
init|=
operator|(
name|int
index|[
literal|5
index|]
operator|)
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{initialization of an array of type 'int [5]' from a compound literal of type 'int [5]' is a GNU extension}}
end_comment

begin_decl_stmt
name|int
name|a3
index|[]
init|=
operator|(
operator|(
name|int
index|[]
operator|)
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|}
operator|)
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{initialization of an array of type 'int []' from a compound literal of type 'int [5]' is a GNU extension}}
end_comment

begin_decl_stmt
name|int
name|a4
index|[]
init|=
operator|(
name|int
index|[
literal|5
index|]
operator|)
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{initialization of an array of type 'int []' from a compound literal of type 'int [5]' is a GNU extension}}
end_comment

begin_decl_stmt
name|int
name|a5
index|[]
init|=
operator|(
name|int5
operator|)
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
literal|5
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{initialization of an array of type 'int []' from a compound literal of type 'int5' (aka 'int [5]') is a GNU extension}}
end_comment

begin_decl_stmt
name|int
name|a6
index|[
literal|5
index|]
init|=
operator|(
name|int
index|[]
operator|)
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error{{cannot initialize array of type 'int [5]' with array of type 'int [3]'}}
end_comment

begin_function_decl
name|int
name|nonconst_value
parameter_list|()
function_decl|;
end_function_decl

begin_decl_stmt
name|int
name|a7
index|[
literal|5
index|]
init|=
operator|(
name|int
index|[
literal|5
index|]
operator|)
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|,
literal|4
block|,
name|nonconst_value
argument_list|()
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error{{initializer element is not a compile-time constant}}
end_comment

begin_comment
comment|//<rdar://problem/10636946>
end_comment

begin_macro
name|__attribute__
argument_list|(
argument|(weak)
argument_list|)
end_macro

begin_decl_stmt
specifier|const
name|unsigned
name|int
name|test10_bound
init|=
literal|10
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|test10_global
index|[
name|test10_bound
index|]
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error {{variable length array declaration not allowed at file scope}}
end_comment

begin_function
name|void
name|test10
parameter_list|()
block|{
name|char
name|test10_local
index|[
name|test10_bound
index|]
init|=
literal|"help"
decl_stmt|;
comment|// expected-error {{variable-sized object may not be initialized}}
block|}
end_function

end_unit

