begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -triple=i686-apple-darwin9 -verify
end_comment

begin_comment
comment|// rdar://13973577
end_comment

begin_struct
struct|struct
name|foo
block|{
name|int
name|big
index|[
literal|128
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_struct
struct|struct
name|bar
block|{
name|char
name|c
index|[
literal|3
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|bar
name|smallThing
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|foo
name|bigThing
decl_stmt|;
end_decl_stmt

begin_decl_stmt
atomic|_Atomic
argument_list|(
expr|struct
name|foo
argument_list|)
name|bigAtomic
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|structAtomicStore
parameter_list|()
block|{
name|struct
name|foo
name|f
init|=
block|{
literal|0
block|}
decl_stmt|;
name|__c11_atomic_store
argument_list|(
operator|&
name|bigAtomic
argument_list|,
name|f
argument_list|,
literal|5
argument_list|)
expr_stmt|;
comment|// expected-error {{atomic store requires runtime support that is not available for this target}}
name|struct
name|bar
name|b
init|=
block|{
literal|0
block|}
decl_stmt|;
name|__atomic_store
argument_list|(
operator|&
name|smallThing
argument_list|,
operator|&
name|b
argument_list|,
literal|5
argument_list|)
expr_stmt|;
name|__atomic_store
argument_list|(
operator|&
name|bigThing
argument_list|,
operator|&
name|f
argument_list|,
literal|5
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|structAtomicLoad
parameter_list|()
block|{
name|struct
name|foo
name|f
init|=
name|__c11_atomic_load
argument_list|(
operator|&
name|bigAtomic
argument_list|,
literal|5
argument_list|)
decl_stmt|;
comment|// expected-error {{atomic load requires runtime support that is not available for this target}}
name|struct
name|bar
name|b
decl_stmt|;
name|__atomic_load
argument_list|(
operator|&
name|smallThing
argument_list|,
operator|&
name|b
argument_list|,
literal|5
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
operator|&
name|bigThing
argument_list|,
operator|&
name|f
argument_list|,
literal|5
argument_list|)
expr_stmt|;
block|}
end_function

end_unit

