begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -verify -ffreestanding -fsyntax-only -triple=i686-linux-gnu -std=c11
end_comment

begin_comment
comment|// Basic parsing/Sema tests for __c11_atomic_*
end_comment

begin_include
include|#
directive|include
file|<stdatomic.h>
end_include

begin_struct
struct|struct
name|S
block|{
name|char
name|c
index|[
literal|3
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_BOOL_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_BOOL_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_BOOL_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_CHAR_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR16_T_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR16_T_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_CHAR16_T_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR32_T_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR32_T_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_CHAR32_T_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_WCHAR_T_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_WCHAR_T_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_WCHAR_T_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_SHORT_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_SHORT_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_SHORT_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_INT_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_INT_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_INT_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_LONG_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_LONG_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_LONG_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_ifdef
ifdef|#
directive|ifdef
name|__i386__
end_ifdef

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_LLONG_LOCK_FREE
operator|==
literal|1
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_else
else|#
directive|else
end_else

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_LLONG_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_endif
endif|#
directive|endif
end_endif

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_LLONG_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_LLONG_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_POINTER_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_POINTER_LOCK_FREE
operator|==
name|__CLANG_ATOMIC_POINTER_LOCK_FREE
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|1
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|2
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|3
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|4
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|17
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|1
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|2
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|3
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|4
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|8
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|16
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|17
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|atomic_is_lock_free
argument_list|(
operator|(
name|atomic_char
operator|*
operator|)
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|atomic_is_lock_free
argument_list|(
operator|(
name|atomic_short
operator|*
operator|)
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|atomic_is_lock_free
argument_list|(
operator|(
name|atomic_int
operator|*
operator|)
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|atomic_is_lock_free
argument_list|(
operator|(
name|atomic_long
operator|*
operator|)
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error@+1 {{__int128 is not supported on this target}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|atomic_is_lock_free
argument_list|(
operator|(
expr|_Atomic
operator|(
name|__int128
operator|)
operator|*
operator|)
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|atomic_is_lock_free
argument_list|(
literal|0
operator|+
operator|(
name|atomic_char
operator|*
operator|)
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_decl_stmt
name|char
name|i8
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|short
name|i16
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|i32
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|__attribute__
argument_list|(
operator|(
name|vector_size
argument_list|(
literal|8
argument_list|)
operator|)
argument_list|)
name|i64
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|Incomplete
modifier|*
name|incomplete
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-note {{forward declaration of 'struct Incomplete'}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|1
argument_list|,
operator|&
name|i8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|1
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i32
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|8
argument_list|,
operator|&
name|i32
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|8
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|1
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|3
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|8
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|16
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|17
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|1
argument_list|,
name|incomplete
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
name|incomplete
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
name|incomplete
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|1
argument_list|,
operator|&
name|i8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|1
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i32
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|8
argument_list|,
operator|&
name|i32
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|8
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_define
define|#
directive|define
name|_AS1
value|__attribute__((address_space(1)))
end_define

begin_define
define|#
directive|define
name|_AS2
value|__attribute__((address_space(2)))
end_define

begin_function
name|void
name|f
parameter_list|(
atomic|_Atomic
argument_list|(
name|int
argument_list|)
modifier|*
name|i
parameter_list|,
specifier|const
atomic|_Atomic
argument_list|(
name|int
argument_list|)
modifier|*
name|ci
parameter_list|,
atomic|_Atomic
argument_list|(
name|int
operator|*
argument_list|)
modifier|*
name|p
parameter_list|,
atomic|_Atomic
argument_list|(
name|float
argument_list|)
modifier|*
name|d
parameter_list|,
name|int
modifier|*
name|I
parameter_list|,
specifier|const
name|int
modifier|*
name|CI
parameter_list|,
name|int
modifier|*
modifier|*
name|P
parameter_list|,
name|float
modifier|*
name|D
parameter_list|,
name|struct
name|S
modifier|*
name|s1
parameter_list|,
name|struct
name|S
modifier|*
name|s2
parameter_list|)
block|{
name|__c11_atomic_init
argument_list|(
name|I
argument_list|,
literal|5
argument_list|)
expr_stmt|;
comment|// expected-error {{pointer to _Atomic}}
name|__c11_atomic_init
argument_list|(
name|ci
argument_list|,
literal|5
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
name|__c11_atomic_load
argument_list|(
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{too few arguments to function}}
name|__c11_atomic_load
argument_list|(
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to function}}
name|__c11_atomic_store
argument_list|(
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic builtin must be a pointer}}
name|__c11_atomic_store
argument_list|(
operator|(
name|int
operator|*
operator|)
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to _Atomic}}
name|__c11_atomic_store
argument_list|(
name|i
argument_list|,
literal|0
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
name|__c11_atomic_store
argument_list|(
name|ci
argument_list|,
literal|0
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
name|__c11_atomic_load
argument_list|(
name|i
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_load
argument_list|(
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_load
argument_list|(
name|d
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_load
argument_list|(
name|ci
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
name|int
name|load_n_1
init|=
name|__atomic_load_n
argument_list|(
name|I
argument_list|,
name|memory_order_relaxed
argument_list|)
decl_stmt|;
name|int
modifier|*
name|load_n_2
init|=
name|__atomic_load_n
argument_list|(
name|P
argument_list|,
name|memory_order_relaxed
argument_list|)
decl_stmt|;
name|float
name|load_n_3
init|=
name|__atomic_load_n
argument_list|(
name|D
argument_list|,
name|memory_order_relaxed
argument_list|)
decl_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_load_n
argument_list|(
name|s1
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|load_n_1
operator|=
name|__atomic_load_n
argument_list|(
name|CI
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
name|i
argument_list|,
name|I
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to a trivially-copyable type}}
name|__atomic_load
argument_list|(
name|CI
argument_list|,
name|I
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
name|I
argument_list|,
name|i
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing '_Atomic(int) *' to parameter of type 'int *'}}
name|__atomic_load
argument_list|(
name|I
argument_list|,
operator|*
name|P
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
name|I
argument_list|,
operator|*
name|P
argument_list|,
name|memory_order_relaxed
argument_list|,
literal|42
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments}}
operator|(
name|int
operator|)
name|__atomic_load
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
name|__atomic_load
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
name|CI
argument_list|,
name|I
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
name|I
argument_list|,
name|CI
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const int *' to parameter of type 'int *' discards qualifiers}}
name|__atomic_load
argument_list|(
name|CI
argument_list|,
name|CI
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const int *' to parameter of type 'int *' discards qualifiers}}
name|__c11_atomic_store
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_store
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible integer to pointer conversion}}
operator|(
name|int
operator|)
name|__c11_atomic_store
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
name|__atomic_store_n
argument_list|(
name|I
argument_list|,
literal|4
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|__atomic_store_n
argument_list|(
name|I
argument_list|,
literal|4.0
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|__atomic_store_n
argument_list|(
name|CI
argument_list|,
literal|4
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const type ('const int *' invalid)}}
name|__atomic_store_n
argument_list|(
name|I
argument_list|,
name|P
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{parameter of type 'int'}}
name|__atomic_store_n
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_store_n
argument_list|(
name|s1
argument_list|,
operator|*
name|s2
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_store_n
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible pointer to integer conversion passing 'int *' to parameter of type 'int'; dereference with *}}
name|__atomic_store
argument_list|(
name|I
argument_list|,
operator|*
name|P
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|__atomic_store
argument_list|(
name|CI
argument_list|,
name|I
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const type ('const int *' invalid)}}
name|__atomic_store
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|__atomic_store
argument_list|(
name|i
argument_list|,
name|I
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{trivially-copyable}}
name|int
name|exchange_1
init|=
name|__c11_atomic_exchange
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|int
name|exchange_2
init|=
name|__c11_atomic_exchange
argument_list|(
name|I
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-error {{must be a pointer to _Atomic}}
name|int
name|exchange_3
init|=
name|__atomic_exchange_n
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|int
name|exchange_4
init|=
name|__atomic_exchange_n
argument_list|(
name|I
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|__atomic_exchange
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
name|s2
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_exchange
argument_list|(
name|s1
argument_list|,
name|I
argument_list|,
name|P
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning 2{{parameter of type 'struct S *'}}
operator|(
name|int
operator|)
name|__atomic_exchange
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
name|s2
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
name|__atomic_exchange
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|I
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_exchange
argument_list|(
name|CI
argument_list|,
name|I
argument_list|,
name|I
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const type ('const int *' invalid)}}
name|__atomic_exchange
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|CI
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const int *' to parameter of type 'int *' discards qualifiers}}
name|__c11_atomic_fetch_add
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_fetch_add
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_fetch_add
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer or pointer}}
name|__atomic_fetch_add
argument_list|(
name|i
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{pointer to integer or pointer}}
name|__atomic_fetch_sub
argument_list|(
name|I
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_sub
argument_list|(
name|P
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_sub
argument_list|(
name|D
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_fetch_sub
argument_list|(
name|s1
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__c11_atomic_fetch_and
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_fetch_and
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer}}
name|__c11_atomic_fetch_and
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer}}
name|__atomic_fetch_and
argument_list|(
name|i
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{pointer to integer}}
name|__atomic_fetch_or
argument_list|(
name|I
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_xor
argument_list|(
name|P
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer}}
name|__atomic_fetch_or
argument_list|(
name|D
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer}}
name|__atomic_fetch_and
argument_list|(
name|s1
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer}}
name|_Bool
name|cmpexch_1
init|=
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|i
argument_list|,
name|I
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|_Bool
name|cmpexch_2
init|=
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|p
argument_list|,
name|P
argument_list|,
operator|(
name|int
operator|*
operator|)
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|_Bool
name|cmpexch_3
init|=
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|d
argument_list|,
name|I
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{incompatible pointer types}}
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|i
argument_list|,
name|CI
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const int *' to parameter of type 'int *' discards qualifiers}}
name|_Bool
name|cmpexchw_1
init|=
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|i
argument_list|,
name|I
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|_Bool
name|cmpexchw_2
init|=
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|p
argument_list|,
name|P
argument_list|,
operator|(
name|int
operator|*
operator|)
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|_Bool
name|cmpexchw_3
init|=
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|d
argument_list|,
name|I
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{incompatible pointer types}}
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|i
argument_list|,
name|CI
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const int *' to parameter of type 'int *' discards qualifiers}}
name|_Bool
name|cmpexch_4
init|=
name|__atomic_compare_exchange_n
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
literal|5
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|_Bool
name|cmpexch_5
init|=
name|__atomic_compare_exchange_n
argument_list|(
name|I
argument_list|,
name|P
argument_list|,
literal|5
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{; dereference with *}}
name|_Bool
name|cmpexch_6
init|=
name|__atomic_compare_exchange_n
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|P
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{passing 'int **' to parameter of type 'int'}}
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|CI
argument_list|,
name|I
argument_list|,
literal|5
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const type ('const int *' invalid)}}
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|I
argument_list|,
name|CI
argument_list|,
literal|5
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const int *' to parameter of type 'int *' discards qualifiers}}
name|_Bool
name|cmpexch_7
init|=
name|__atomic_compare_exchange
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
literal|5
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{passing 'int' to parameter of type 'int *'}}
name|_Bool
name|cmpexch_8
init|=
name|__atomic_compare_exchange
argument_list|(
name|I
argument_list|,
name|P
argument_list|,
name|I
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{; dereference with *}}
name|_Bool
name|cmpexch_9
init|=
name|__atomic_compare_exchange
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|I
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|CI
argument_list|,
name|I
argument_list|,
name|I
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const type ('const int *' invalid)}}
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|I
argument_list|,
name|CI
argument_list|,
name|I
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const int *' to parameter of type 'int *' discards qualifiers}}
comment|// Pointers to different address spaces are allowed.
name|_Bool
name|cmpexch_10
init|=
name|__c11_atomic_compare_exchange_strong
argument_list|(
operator|(
expr|_Atomic
name|int
name|_AS1
operator|*
operator|)
literal|0x308
argument_list|,
operator|(
name|int
name|_AS2
operator|*
operator|)
literal|0x309
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
specifier|const
specifier|volatile
name|int
name|flag_k
init|=
literal|0
decl_stmt|;
specifier|volatile
name|int
name|flag
init|=
literal|0
decl_stmt|;
operator|(
name|void
operator|)
operator|(
name|int
operator|)
name|__atomic_test_and_set
argument_list|(
operator|&
name|flag_k
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const volatile int *' to parameter of type 'volatile void *'}}
operator|(
name|void
operator|)
operator|(
name|int
operator|)
name|__atomic_test_and_set
argument_list|(
operator|&
name|flag
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_clear
argument_list|(
operator|&
name|flag_k
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const volatile int *' to parameter of type 'volatile void *'}}
name|__atomic_clear
argument_list|(
operator|&
name|flag
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|int
operator|)
name|__atomic_clear
argument_list|(
operator|&
name|flag
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
name|__c11_atomic_init
argument_list|(
name|ci
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
name|__c11_atomic_store
argument_list|(
name|ci
argument_list|,
literal|0
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
name|__c11_atomic_load
argument_list|(
name|ci
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
comment|// Ensure the<stdatomic.h> macros behave appropriately.
name|atomic_int
name|n
init|=
name|ATOMIC_VAR_INIT
argument_list|(
literal|123
argument_list|)
decl_stmt|;
name|atomic_init
argument_list|(
operator|&
name|n
argument_list|,
literal|456
argument_list|)
expr_stmt|;
name|atomic_init
argument_list|(
operator|&
name|n
argument_list|,
operator|(
name|void
operator|*
operator|)
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'void *' to parameter of type 'int'}}
specifier|const
name|atomic_wchar_t
name|cawt
decl_stmt|;
name|atomic_init
argument_list|(
operator|&
name|cawt
argument_list|,
literal|L'
expr|x'
argument_list|)
expr_stmt|;
comment|// expected-error {{non-const}}
name|atomic_wchar_t
name|awt
decl_stmt|;
name|atomic_init
argument_list|(
operator|&
name|awt
argument_list|,
literal|L'
expr|x'
argument_list|)
expr_stmt|;
name|int
name|x
init|=
name|kill_dependency
argument_list|(
literal|12
argument_list|)
decl_stmt|;
name|atomic_thread_fence
argument_list|()
expr_stmt|;
comment|// expected-error {{too few arguments to function call}}
name|atomic_thread_fence
argument_list|(
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|atomic_signal_fence
argument_list|(
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|void
function_decl|(
modifier|*
name|pfn
function_decl|)
parameter_list|(
name|memory_order
parameter_list|)
init|=
operator|&
name|atomic_thread_fence
function_decl|;
name|pfn
operator|=
operator|&
name|atomic_signal_fence
expr_stmt|;
name|int
name|k
init|=
name|atomic_load_explicit
argument_list|(
operator|&
name|n
argument_list|,
name|memory_order_relaxed
argument_list|)
decl_stmt|;
name|atomic_store_explicit
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
name|atomic_store
argument_list|(
operator|&
name|n
argument_list|,
name|atomic_load
argument_list|(
operator|&
name|n
argument_list|)
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_exchange
argument_list|(
operator|&
name|n
argument_list|,
literal|72
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_exchange_explicit
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|atomic_compare_exchange_strong
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|,
name|k
argument_list|)
expr_stmt|;
comment|// expected-warning {{take the address with&}}
name|atomic_compare_exchange_weak
argument_list|(
operator|&
name|n
argument_list|,
operator|&
name|k
argument_list|,
name|k
argument_list|)
expr_stmt|;
name|atomic_compare_exchange_strong_explicit
argument_list|(
operator|&
name|n
argument_list|,
operator|&
name|k
argument_list|,
name|k
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{too few arguments}}
name|atomic_compare_exchange_weak_explicit
argument_list|(
operator|&
name|n
argument_list|,
operator|&
name|k
argument_list|,
name|k
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
name|atomic_fetch_add
argument_list|(
operator|&
name|k
argument_list|,
name|n
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to _Atomic}}
name|k
operator|=
name|atomic_fetch_add
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_sub
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_and
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_or
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_xor
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_add_explicit
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_sub_explicit
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_and_explicit
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_or_explicit
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
name|k
operator|=
name|atomic_fetch_xor_explicit
argument_list|(
operator|&
name|n
argument_list|,
name|k
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// C11 7.17.1/4: atomic_flag is a structure type.
name|struct
name|atomic_flag
name|must_be_struct
init|=
name|ATOMIC_FLAG_INIT
decl_stmt|;
comment|// C11 7.17.8/5 implies that it is also a typedef type.
name|atomic_flag
name|guard
init|=
name|ATOMIC_FLAG_INIT
decl_stmt|;
name|_Bool
name|old_val
init|=
name|atomic_flag_test_and_set
argument_list|(
operator|&
name|guard
argument_list|)
decl_stmt|;
if|if
condition|(
name|old_val
condition|)
name|atomic_flag_clear
argument_list|(
operator|&
name|guard
argument_list|)
expr_stmt|;
name|old_val
operator|=
call|(
name|atomic_flag_test_and_set
call|)
argument_list|(
operator|&
name|guard
argument_list|)
expr_stmt|;
if|if
condition|(
name|old_val
condition|)
call|(
name|atomic_flag_clear
call|)
argument_list|(
operator|&
name|guard
argument_list|)
expr_stmt|;
specifier|const
name|atomic_flag
name|const_guard
decl_stmt|;
name|atomic_flag_test_and_set
argument_list|(
operator|&
name|const_guard
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const atomic_bool *' (aka 'const _Atomic(_Bool) *') invalid)}}
name|atomic_flag_clear
argument_list|(
operator|&
name|const_guard
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const atomic_bool *' (aka 'const _Atomic(_Bool) *') invalid)}}
block|}
end_function

begin_decl_stmt
atomic|_Atomic
argument_list|(
name|int
operator|*
argument_list|)
name|PR12527_a
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|PR12527
parameter_list|()
block|{
name|int
modifier|*
name|b
init|=
name|PR12527_a
decl_stmt|;
block|}
end_function

begin_function
name|void
name|PR16931
parameter_list|(
name|int
modifier|*
name|x
parameter_list|)
block|{
comment|// expected-note {{passing argument to parameter 'x' here}}
typedef|typedef
struct|struct
block|{
atomic|_Atomic
argument_list|(
name|_Bool
argument_list|)
name|flag
decl_stmt|;
block|}
name|flag
typedef|;
name|flag
name|flagvar
init|=
block|{
literal|0
block|}
decl_stmt|;
name|PR16931
argument_list|(
operator|&
name|flagvar
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible pointer types}}
block|}
end_function

begin_function
name|void
name|memory_checks
parameter_list|(
atomic|_Atomic
argument_list|(
name|int
argument_list|)
modifier|*
name|Ap
parameter_list|,
name|int
modifier|*
name|p
parameter_list|,
name|int
name|val
parameter_list|)
block|{
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
literal|42
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
operator|(
expr|struct
name|Incomplete
operator|*
expr|_Atomic
operator|*
operator|)
literal|0
argument_list|,
comment|// expected-error {{incomplete type 'struct Incomplete'}}
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_acquire
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_consume
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_release
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_acq_rel
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_acquire
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_consume
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_release
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_acq_rel
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|nullPointerWarning
parameter_list|(
atomic|_Atomic
argument_list|(
name|int
argument_list|)
modifier|*
name|Ap
parameter_list|,
name|int
modifier|*
name|p
parameter_list|,
name|int
name|val
parameter_list|)
block|{
comment|// The 'expected' pointer shouldn't be NULL.
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|NULL
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee that requires a non-null argument}}
operator|(
name|void
operator|)
name|atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
argument_list|,
name|val
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee that requires a non-null argument}}
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|NULL
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee that requires a non-null argument}}
block|}
end_function

end_unit

