begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -verify -fsyntax-only -triple=i686-linux-gnu -std=c11
end_comment

begin_comment
comment|// Basic parsing/Sema tests for __c11_atomic_*
end_comment

begin_typedef
typedef|typedef
enum|enum
name|memory_order
block|{
name|memory_order_relaxed
block|,
name|memory_order_consume
block|,
name|memory_order_acquire
block|,
name|memory_order_release
block|,
name|memory_order_acq_rel
block|,
name|memory_order_seq_cst
block|}
name|memory_order
typedef|;
end_typedef

begin_struct
struct|struct
name|S
block|{
name|char
name|c
index|[
literal|3
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_BOOL_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR16_T_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_CHAR32_T_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_WCHAR_T_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_SHORT_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_INT_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_LONG_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_ifdef
ifdef|#
directive|ifdef
name|__i386__
end_ifdef

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_LLONG_LOCK_FREE
operator|==
literal|1
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_else
else|#
directive|else
end_else

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_LLONG_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_endif
endif|#
directive|endif
end_endif

begin_assert
assert|_Static_assert
argument_list|(
name|__GCC_ATOMIC_POINTER_LOCK_FREE
operator|==
literal|2
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|1
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|2
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|3
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|4
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__c11_atomic_is_lock_free
argument_list|(
literal|17
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|1
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|2
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|3
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|4
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|8
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|16
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|17
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_decl_stmt
name|char
name|i8
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|short
name|i16
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|i32
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|__attribute__
argument_list|(
operator|(
name|vector_size
argument_list|(
literal|8
argument_list|)
operator|)
argument_list|)
name|i64
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|Incomplete
modifier|*
name|incomplete
decl_stmt|;
end_decl_stmt

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|1
argument_list|,
operator|&
name|i8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|1
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i32
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|8
argument_list|,
operator|&
name|i32
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_comment
comment|// expected-error {{not an integral constant expression}}
end_comment

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_is_lock_free
argument_list|(
literal|8
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|1
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|3
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|8
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|16
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|17
argument_list|,
literal|0
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|1
argument_list|,
name|incomplete
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
name|incomplete
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
name|incomplete
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|1
argument_list|,
operator|&
name|i8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|1
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i8
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|2
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i16
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i32
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|4
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
operator|!
name|__atomic_always_lock_free
argument_list|(
literal|8
argument_list|,
operator|&
name|i32
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_assert
assert|_Static_assert
argument_list|(
name|__atomic_always_lock_free
argument_list|(
literal|8
argument_list|,
operator|&
name|i64
argument_list|)
argument_list|,
literal|""
argument_list|)
assert|;
end_assert

begin_function
name|void
name|f
parameter_list|(
atomic|_Atomic
argument_list|(
name|int
argument_list|)
modifier|*
name|i
parameter_list|,
atomic|_Atomic
argument_list|(
name|int
operator|*
argument_list|)
modifier|*
name|p
parameter_list|,
atomic|_Atomic
argument_list|(
name|float
argument_list|)
modifier|*
name|d
parameter_list|,
name|int
modifier|*
name|I
parameter_list|,
name|int
modifier|*
modifier|*
name|P
parameter_list|,
name|float
modifier|*
name|D
parameter_list|,
name|struct
name|S
modifier|*
name|s1
parameter_list|,
name|struct
name|S
modifier|*
name|s2
parameter_list|)
block|{
name|__c11_atomic_init
argument_list|(
name|I
argument_list|,
literal|5
argument_list|)
expr_stmt|;
comment|// expected-error {{pointer to _Atomic}}
name|__c11_atomic_load
argument_list|(
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{too few arguments to function}}
name|__c11_atomic_load
argument_list|(
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to function}}
name|__c11_atomic_store
argument_list|(
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic builtin must be a pointer}}
name|__c11_atomic_store
argument_list|(
operator|(
name|int
operator|*
operator|)
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to _Atomic}}
name|__c11_atomic_load
argument_list|(
name|i
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_load
argument_list|(
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_load
argument_list|(
name|d
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|int
name|load_n_1
init|=
name|__atomic_load_n
argument_list|(
name|I
argument_list|,
name|memory_order_relaxed
argument_list|)
decl_stmt|;
name|int
modifier|*
name|load_n_2
init|=
name|__atomic_load_n
argument_list|(
name|P
argument_list|,
name|memory_order_relaxed
argument_list|)
decl_stmt|;
name|float
name|load_n_3
init|=
name|__atomic_load_n
argument_list|(
name|D
argument_list|,
name|memory_order_relaxed
argument_list|)
decl_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_load_n
argument_list|(
name|s1
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_load
argument_list|(
name|i
argument_list|,
name|I
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to a trivially-copyable type}}
name|__atomic_load
argument_list|(
name|I
argument_list|,
name|i
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing '_Atomic(int) *' to parameter of type 'int *'}}
name|__atomic_load
argument_list|(
name|I
argument_list|,
operator|*
name|P
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
name|I
argument_list|,
operator|*
name|P
argument_list|,
name|memory_order_relaxed
argument_list|,
literal|42
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments}}
operator|(
name|int
operator|)
name|__atomic_load
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
name|__atomic_load
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
name|__c11_atomic_store
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_store
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible integer to pointer conversion}}
operator|(
name|int
operator|)
name|__c11_atomic_store
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
name|__atomic_store_n
argument_list|(
name|I
argument_list|,
literal|4
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|__atomic_store_n
argument_list|(
name|I
argument_list|,
literal|4.0
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|__atomic_store_n
argument_list|(
name|I
argument_list|,
name|P
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{parameter of type 'int'}}
name|__atomic_store_n
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_store_n
argument_list|(
name|s1
argument_list|,
operator|*
name|s2
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_store
argument_list|(
name|I
argument_list|,
operator|*
name|P
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|__atomic_store
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
name|__atomic_store
argument_list|(
name|i
argument_list|,
name|I
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{trivially-copyable}}
name|int
name|exchange_1
init|=
name|__c11_atomic_exchange
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|int
name|exchange_2
init|=
name|__c11_atomic_exchange
argument_list|(
name|I
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-error {{must be a pointer to _Atomic}}
name|int
name|exchange_3
init|=
name|__atomic_exchange_n
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|int
name|exchange_4
init|=
name|__atomic_exchange_n
argument_list|(
name|I
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|__atomic_exchange
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
name|s2
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_exchange
argument_list|(
name|s1
argument_list|,
name|I
argument_list|,
name|P
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning 2{{parameter of type 'struct S *'}}
operator|(
name|int
operator|)
name|__atomic_exchange
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
name|s2
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
name|__c11_atomic_fetch_add
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_fetch_add
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_fetch_add
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer or pointer}}
name|__atomic_fetch_add
argument_list|(
name|i
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{pointer to integer or pointer}}
name|__atomic_fetch_sub
argument_list|(
name|I
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_sub
argument_list|(
name|P
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_sub
argument_list|(
name|D
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__atomic_fetch_sub
argument_list|(
name|s1
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer or pointer}}
name|__c11_atomic_fetch_and
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__c11_atomic_fetch_and
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer}}
name|__c11_atomic_fetch_and
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer}}
name|__atomic_fetch_and
argument_list|(
name|i
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{pointer to integer}}
name|__atomic_fetch_or
argument_list|(
name|I
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_xor
argument_list|(
name|P
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer}}
name|__atomic_fetch_or
argument_list|(
name|D
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer}}
name|__atomic_fetch_and
argument_list|(
name|s1
argument_list|,
literal|3
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to integer}}
name|_Bool
name|cmpexch_1
init|=
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|i
argument_list|,
literal|0
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|_Bool
name|cmpexch_2
init|=
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|p
argument_list|,
literal|0
argument_list|,
operator|(
name|int
operator|*
operator|)
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|_Bool
name|cmpexch_3
init|=
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|d
argument_list|,
operator|(
name|int
operator|*
operator|)
literal|0
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{incompatible pointer types}}
name|_Bool
name|cmpexch_4
init|=
name|__atomic_compare_exchange_n
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
literal|5
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
name|_Bool
name|cmpexch_5
init|=
name|__atomic_compare_exchange_n
argument_list|(
name|I
argument_list|,
name|P
argument_list|,
literal|5
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{; dereference with *}}
name|_Bool
name|cmpexch_6
init|=
name|__atomic_compare_exchange_n
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|P
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{passing 'int **' to parameter of type 'int'}}
name|_Bool
name|cmpexch_7
init|=
name|__atomic_compare_exchange
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
literal|5
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{passing 'int' to parameter of type 'int *'}}
name|_Bool
name|cmpexch_8
init|=
name|__atomic_compare_exchange
argument_list|(
name|I
argument_list|,
name|P
argument_list|,
name|I
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
comment|// expected-warning {{; dereference with *}}
name|_Bool
name|cmpexch_9
init|=
name|__atomic_compare_exchange
argument_list|(
name|I
argument_list|,
name|I
argument_list|,
name|I
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
decl_stmt|;
specifier|const
specifier|volatile
name|int
name|flag_k
init|=
literal|0
decl_stmt|;
specifier|volatile
name|int
name|flag
init|=
literal|0
decl_stmt|;
operator|(
name|void
operator|)
operator|(
name|int
operator|)
name|__atomic_test_and_set
argument_list|(
operator|&
name|flag_k
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const volatile int *' to parameter of type 'volatile void *'}}
operator|(
name|void
operator|)
operator|(
name|int
operator|)
name|__atomic_test_and_set
argument_list|(
operator|&
name|flag
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_clear
argument_list|(
operator|&
name|flag_k
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const volatile int *' to parameter of type 'volatile void *'}}
name|__atomic_clear
argument_list|(
operator|&
name|flag
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|int
operator|)
name|__atomic_clear
argument_list|(
operator|&
name|flag
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
specifier|const
atomic|_Atomic
argument_list|(
name|int
argument_list|)
name|const_atomic
decl_stmt|;
name|__c11_atomic_init
argument_list|(
operator|&
name|const_atomic
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
name|__c11_atomic_store
argument_list|(
operator|&
name|const_atomic
argument_list|,
literal|0
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
name|__c11_atomic_load
argument_list|(
operator|&
name|const_atomic
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(int) *' invalid)}}
block|}
end_function

begin_decl_stmt
atomic|_Atomic
argument_list|(
name|int
operator|*
argument_list|)
name|PR12527_a
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|PR12527
parameter_list|()
block|{
name|int
modifier|*
name|b
init|=
name|PR12527_a
decl_stmt|;
block|}
end_function

begin_function
name|void
name|PR16931
parameter_list|(
name|int
modifier|*
name|x
parameter_list|)
block|{
comment|// expected-note {{passing argument to parameter 'x' here}}
typedef|typedef
struct|struct
block|{
atomic|_Atomic
argument_list|(
name|_Bool
argument_list|)
name|flag
decl_stmt|;
block|}
name|flag
typedef|;
name|flag
name|flagvar
init|=
block|{
literal|0
block|}
decl_stmt|;
name|PR16931
argument_list|(
operator|&
name|flagvar
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible pointer types}}
block|}
end_function

begin_function
name|void
name|memory_checks
parameter_list|(
atomic|_Atomic
argument_list|(
name|int
argument_list|)
modifier|*
name|Ap
parameter_list|,
name|int
modifier|*
name|p
parameter_list|,
name|int
name|val
parameter_list|)
block|{
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
operator|-
literal|1
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_load
argument_list|(
name|Ap
argument_list|,
literal|42
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__c11_atomic_store
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_add
argument_list|(
name|Ap
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_init
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_sub
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_and
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_or
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_fetch_xor
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_exchange
argument_list|(
name|Ap
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_strong
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__c11_atomic_compare_exchange_weak
argument_list|(
name|Ap
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_load_n
argument_list|(
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
comment|// expected-warning {{memory order argument to atomic operation is invalid}}
operator|(
name|void
operator|)
name|__atomic_store_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_sub
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_add_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_sub_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_or
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_xor
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_fetch_nand
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_and_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_or_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_xor_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_nand_fetch
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange_n
argument_list|(
name|p
argument_list|,
name|val
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acquire
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_consume
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_release
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_acq_rel
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_acquire
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_consume
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_release
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_acq_rel
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|p
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_relaxed
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_acquire
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_consume
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_release
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_acq_rel
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__atomic_compare_exchange_n
argument_list|(
name|p
argument_list|,
name|p
argument_list|,
name|val
argument_list|,
literal|0
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_relaxed
argument_list|)
expr_stmt|;
block|}
end_function

end_unit

