begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -verify -fsyntax-only
end_comment

begin_comment
comment|// Basic parsing/Sema tests for __atomic_*
end_comment

begin_comment
comment|// FIXME: Need to implement __atomic_is_lock_free
end_comment

begin_typedef
typedef|typedef
enum|enum
name|memory_order
block|{
name|memory_order_relaxed
block|,
name|memory_order_consume
block|,
name|memory_order_acquire
block|,
name|memory_order_release
block|,
name|memory_order_acq_rel
block|,
name|memory_order_seq_cst
block|}
name|memory_order
typedef|;
end_typedef

begin_function
name|void
name|f
parameter_list|(
atomic|_Atomic
argument_list|(
name|int
argument_list|)
modifier|*
name|i
parameter_list|,
atomic|_Atomic
argument_list|(
name|int
operator|*
argument_list|)
modifier|*
name|p
parameter_list|,
atomic|_Atomic
argument_list|(
name|float
argument_list|)
modifier|*
name|d
parameter_list|)
block|{
name|__atomic_load
argument_list|(
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{too few arguments to function}}
name|__atomic_load
argument_list|(
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to function}}
name|__atomic_store
argument_list|(
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{first argument to atomic operation}}
name|__atomic_store
argument_list|(
operator|(
name|int
operator|*
operator|)
literal|0
argument_list|,
literal|0
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{first argument to atomic operation}}
name|__atomic_load
argument_list|(
name|i
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
name|p
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_load
argument_list|(
name|d
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_store
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_store
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible integer to pointer conversion}}
operator|(
name|int
operator|)
name|__atomic_store
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{operand of type 'void'}}
name|__atomic_fetch_add
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_add
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_add
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer or pointer}}
name|__atomic_fetch_and
argument_list|(
name|i
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_fetch_and
argument_list|(
name|p
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer}}
name|__atomic_fetch_and
argument_list|(
name|d
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-error {{must be a pointer to atomic integer}}
name|__atomic_compare_exchange_strong
argument_list|(
name|i
argument_list|,
literal|0
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_compare_exchange_strong
argument_list|(
name|p
argument_list|,
literal|0
argument_list|,
operator|(
name|int
operator|*
operator|)
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
name|__atomic_compare_exchange_strong
argument_list|(
name|d
argument_list|,
operator|(
name|int
operator|*
operator|)
literal|0
argument_list|,
literal|1
argument_list|,
name|memory_order_seq_cst
argument_list|,
name|memory_order_seq_cst
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible pointer types}}
block|}
end_function

end_unit

