begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple i386-apple-darwin9 -fsyntax-only -verify %s
end_comment

begin_decl_stmt
name|int
name|x
name|__attribute__
argument_list|(
operator|(
name|aligned
argument_list|(
literal|3
argument_list|)
operator|)
argument_list|)
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error {{requested alignment is not a power of 2}}
end_comment

begin_decl_stmt
name|int
name|y
name|__attribute__
argument_list|(
operator|(
name|aligned
argument_list|(
literal|1
operator|<<
literal|29
argument_list|)
operator|)
argument_list|)
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error {{requested alignment must be 268435456 bytes or smaller}}
end_comment

begin_comment
comment|// PR3254
end_comment

begin_decl_stmt
name|short
name|g0
index|[
literal|3
index|]
name|__attribute__
argument_list|(
operator|(
name|aligned
operator|)
argument_list|)
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|short
name|g0_chk
index|[
name|__alignof__
argument_list|(
name|g0
argument_list|)
operator|==
literal|16
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_comment
comment|//<rdar://problem/6840045>
end_comment

begin_typedef
typedef|typedef
name|char
name|ueber_aligned_char
name|__attribute__
typedef|((
name|aligned
typedef|(8)));
end_typedef

begin_struct
struct|struct
name|struct_with_ueber_char
block|{
name|ueber_aligned_char
name|c
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|char
name|a
init|=
literal|0
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|a0
index|[
name|__alignof__
argument_list|(
name|ueber_aligned_char
argument_list|)
operator|==
literal|8
condition|?
literal|1
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|a1
index|[
name|__alignof__
argument_list|(
expr|struct
name|struct_with_ueber_char
argument_list|)
operator|==
literal|8
condition|?
literal|1
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|a2
index|[
name|__alignof__
argument_list|(
name|a
argument_list|)
operator|==
literal|1
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|a3
index|[
sizeof|sizeof
argument_list|(
name|a
argument_list|)
operator|==
literal|1
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_typedef
typedef|typedef
name|long
name|long
name|__attribute__
typedef|((
name|aligned
typedef|(1)))
name|underaligned_longlong
typedef|;
end_typedef

begin_decl_stmt
name|char
name|a4
index|[
name|__alignof__
argument_list|(
name|underaligned_longlong
argument_list|)
operator|==
literal|1
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_typedef
typedef|typedef
name|long
name|long
name|__attribute__
typedef|((
name|aligned
typedef|(1)))
name|underaligned_complex_longlong
typedef|;
end_typedef

begin_decl_stmt
name|char
name|a5
index|[
name|__alignof__
argument_list|(
name|underaligned_complex_longlong
argument_list|)
operator|==
literal|1
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// rdar://problem/8335865
end_comment

begin_decl_stmt
name|int
name|b
name|__attribute__
argument_list|(
operator|(
name|aligned
argument_list|(
literal|2
argument_list|)
operator|)
argument_list|)
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|b1
index|[
name|__alignof__
argument_list|(
name|b
argument_list|)
operator|==
literal|2
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|C
block|{
name|int
name|member
name|__attribute__
argument_list|(
operator|(
name|aligned
argument_list|(
literal|2
argument_list|)
operator|)
argument_list|)
decl_stmt|;
block|}
name|c
struct|;
end_struct

begin_decl_stmt
name|char
name|c1
index|[
name|__alignof__
argument_list|(
name|c
argument_list|)
operator|==
literal|4
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|c2
index|[
name|__alignof__
argument_list|(
name|c
operator|.
name|member
argument_list|)
operator|==
literal|4
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|D
block|{
name|int
name|member
name|__attribute__
argument_list|(
operator|(
name|aligned
argument_list|(
literal|2
argument_list|)
operator|)
argument_list|)
name|__attribute__
argument_list|(
operator|(
name|packed
operator|)
argument_list|)
decl_stmt|;
block|}
name|d
struct|;
end_struct

begin_decl_stmt
name|char
name|d1
index|[
name|__alignof__
argument_list|(
name|d
argument_list|)
operator|==
literal|2
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|d2
index|[
name|__alignof__
argument_list|(
name|d
operator|.
name|member
argument_list|)
operator|==
literal|2
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|E
block|{
name|int
name|member
name|__attribute__
argument_list|(
operator|(
name|aligned
argument_list|(
literal|2
argument_list|)
operator|)
argument_list|)
decl_stmt|;
block|}
name|__attribute__
argument_list|(
operator|(
name|packed
operator|)
argument_list|)
struct|;
end_struct

begin_decl_stmt
name|struct
name|E
name|e
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|e1
index|[
name|__alignof__
argument_list|(
name|e
argument_list|)
operator|==
literal|2
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|e2
index|[
name|__alignof__
argument_list|(
name|e
operator|.
name|member
argument_list|)
operator|==
literal|2
condition|?
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_typedef
typedef|typedef
name|char
name|overaligned_char
name|__attribute__
typedef|((
name|aligned
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|overaligned_char
name|array_with_overaligned_char
index|[
literal|11
index|]
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|char
name|array_with_align_attr
index|[
literal|11
index|]
name|__attribute__
typedef|((
name|aligned
typedef|(16)));
end_typedef

begin_decl_stmt
name|char
name|f0
index|[
name|__alignof__
argument_list|(
name|array_with_overaligned_char
argument_list|)
operator|==
literal|16
condition|?
literal|1
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|f1
index|[
name|__alignof__
argument_list|(
name|array_with_align_attr
argument_list|)
operator|==
literal|16
condition|?
literal|1
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|array_with_overaligned_char
name|F2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|f2
index|[
name|__alignof__
argument_list|(
name|F2
argument_list|)
operator|==
literal|16
condition|?
literal|1
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|array_with_align_attr
name|F3
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|f3
index|[
name|__alignof__
argument_list|(
name|F3
argument_list|)
operator|==
literal|16
condition|?
literal|1
else|:
operator|-
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
end_decl_stmt

end_unit

