begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-apple-darwin9 -fsyntax-only -fblocks -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -D WARN_PARTIAL -Wpartial-availability -triple x86_64-apple-darwin9 -fsyntax-only -fblocks -verify %s
end_comment

begin_comment
comment|//
end_comment

begin_function_decl
name|void
name|f0
parameter_list|()
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|10.4
parameter_list|,
name|deprecated
init|=
literal|10.2
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-warning{{feature cannot be deprecated in macOS version 10.2 before it was introduced in version 10.4; attribute ignored}}
end_comment

begin_function_decl
name|void
name|f1
parameter_list|()
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|obsoleted
init|=
literal|2.1
parameter_list|,
name|deprecated
init|=
literal|3.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-warning{{feature cannot be obsoleted in iOS version 2.1 before it was deprecated in version 3.0; attribute ignored}}
end_comment

begin_function_decl
name|void
name|f2
parameter_list|()
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|introduced
init|=
literal|2.1
parameter_list|,
name|deprecated
init|=
literal|2.1
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|void
name|f3
parameter_list|()
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|otheros
parameter_list|,
name|introduced
init|=
literal|2.2
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-warning{{unknown platform 'otheros' in availability macro}}
end_comment

begin_comment
comment|// rdar://10095131
end_comment

begin_function_decl
specifier|extern
name|void
name|ATSFontGetName
parameter_list|(
specifier|const
name|char
modifier|*
name|oName
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|8.0
parameter_list|,
name|deprecated
init|=
literal|9.0
parameter_list|,
name|message
init|=
literal|"use CTFontCopyFullName"
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note {{'ATSFontGetName' has been explicitly marked deprecated here}}
end_comment

begin_function_decl
specifier|extern
name|void
name|ATSFontGetPostScriptName
parameter_list|(
name|int
name|flags
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|8.0
parameter_list|,
name|obsoleted
init|=
literal|9.0
parameter_list|,
name|message
init|=
literal|"use ATSFontGetFullPostScriptName"
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note {{'ATSFontGetPostScriptName' has been explicitly marked unavailable here}}
end_comment

begin_if
if|#
directive|if
name|defined
argument_list|(
name|WARN_PARTIAL
argument_list|)
end_if

begin_comment
comment|// expected-note@+3 {{has been explicitly marked partial here}}
end_comment

begin_endif
endif|#
directive|endif
end_endif

begin_function_decl
specifier|extern
name|void
name|PartiallyAvailable
parameter_list|()
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|10.8
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|introduced
operator|=
literal|10.8
argument_list|)
operator|)
argument_list|)
name|PartialEnum
block|{
name|kPartialEnumConstant
operator|,
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_function
name|void
name|test_10095131
parameter_list|()
block|{
name|ATSFontGetName
argument_list|(
literal|"Hello"
argument_list|)
expr_stmt|;
comment|// expected-warning {{'ATSFontGetName' is deprecated: first deprecated in macOS 9.0 - use CTFontCopyFullName}}
name|ATSFontGetPostScriptName
argument_list|(
literal|100
argument_list|)
expr_stmt|;
comment|// expected-error {{'ATSFontGetPostScriptName' is unavailable: obsoleted in macOS 9.0 - use ATSFontGetFullPostScriptName}}
if|#
directive|if
name|defined
argument_list|(
name|WARN_PARTIAL
argument_list|)
comment|// expected-warning@+2 {{is only available on macOS 10.8 or newer}} expected-note@+2 {{enclose 'PartiallyAvailable' in a __builtin_available check to silence this warning}}
endif|#
directive|endif
name|PartiallyAvailable
argument_list|()
expr_stmt|;
block|}
end_function

begin_function_decl
specifier|extern
name|void
name|PartiallyAvailable
parameter_list|()
function_decl|;
end_function_decl

begin_function
name|void
name|with_redeclaration
parameter_list|()
block|{
name|PartiallyAvailable
argument_list|()
expr_stmt|;
comment|// Don't warn.
comment|// enums should never warn.
name|enum
name|PartialEnum
name|p
init|=
name|kPartialEnumConstant
decl_stmt|;
block|}
end_function

begin_comment
comment|// rdar://10711037
end_comment

begin_expr_stmt
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macos
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
comment|// expected-warning {{attribute 'availability' is ignored}}
expr|enum
block|{
name|NSDataWritingFileProtectionWriteOnly
operator|=
literal|0x30000000
block|,
name|NSDataWritingFileProtectionCompleteUntilUserAuthentication
operator|=
literal|0x40000000
block|, }
expr_stmt|;
end_expr_stmt

begin_function_decl
name|void
name|f4
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|deprecated
init|=
literal|3.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|void
name|f4
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|introduced
init|=
literal|4.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-warning {{feature cannot be deprecated in iOS version 3.0 before it was introduced in version 4.0; attribute ignored}}
end_comment

begin_function_decl
name|void
name|f5
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|deprecated
init|=
literal|3.0
parameter_list|)
operator|,
function_decl|availability
parameter_list|(
name|ios
parameter_list|,
name|introduced
init|=
literal|4.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-warning {{feature cannot be deprecated in iOS version 3.0 before it was introduced in version 4.0; attribute ignored}}
end_comment

begin_function_decl
name|void
name|f6
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|deprecated
init|=
literal|3.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note {{previous attribute is here}}
end_comment

begin_function_decl
name|void
name|f6
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|deprecated
init|=
literal|4.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-warning {{availability does not match previous declaration}}
end_comment

begin_function_decl
name|void
name|f7
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|introduced
init|=
literal|2.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|void
name|f7
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|deprecated
init|=
literal|3.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note {{previous attribute is here}}
end_comment

begin_function_decl
name|void
name|f7
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|deprecated
init|=
literal|4.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-warning {{availability does not match previous declaration}}
end_comment

begin_comment
comment|//<rdar://problem/11886458>
end_comment

begin_if
if|#
directive|if
operator|!
name|__has_feature
argument_list|(
name|attribute_availability_with_message
argument_list|)
end_if

begin_error
error|#
directive|error
literal|"Missing __has_feature"
end_error

begin_endif
endif|#
directive|endif
end_endif

begin_decl_stmt
specifier|extern
name|int
name|x
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|introduced
operator|=
literal|10.5
argument_list|)
operator|)
argument_list|)
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|extern
name|int
name|x
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|f8
parameter_list|()
block|{
name|int
function_decl|(
modifier|^
name|b
function_decl|)
parameter_list|(
name|int
parameter_list|)
function_decl|;
name|b
operator|=
operator|^
operator|(
name|int
name|i
operator|)
name|__attribute__
argument_list|(
argument|(availability(macosx,introduced=
literal|10.2
argument|))
argument_list|)
block|{
return|return
literal|1
return|;
block|}
empty_stmt|;
comment|// expected-warning {{'availability' attribute only applies to named declarations}}
block|}
end_function

begin_decl_stmt
specifier|extern
name|int
name|x2
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|introduced
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-note {{previous attribute is here}}
end_comment

begin_decl_stmt
specifier|extern
name|int
name|x2
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|introduced
operator|=
literal|10.5
argument_list|)
operator|)
argument_list|)
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning {{availability does not match previous declaration}}
end_comment

begin_enum
enum|enum
name|Original
block|{
name|OriginalDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
block|,
comment|// expected-note + {{'OriginalDeprecated' has been explicitly marked deprecated here}}
name|OriginalUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
comment|// expected-note + {{'OriginalUnavailable' has been explicitly marked unavailable here}}
block|}
enum|;
end_enum

begin_enum
enum|enum
name|AllDeprecated
block|{
name|AllDeprecatedCase
block|,
comment|// expected-note + {{'AllDeprecatedCase' has been explicitly marked deprecated here}}
name|AllDeprecatedUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
comment|// expected-note + {{'AllDeprecatedUnavailable' has been explicitly marked unavailable here}}
block|}
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
enum|;
end_enum

begin_enum
enum|enum
name|AllUnavailable
block|{
name|AllUnavailableCase
block|,
comment|// expected-note + {{'AllUnavailableCase' has been explicitly marked unavailable here}}
block|}
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
enum|;
end_enum

begin_enum
enum|enum
name|User
block|{
name|UserOD
init|=
name|OriginalDeprecated
block|,
comment|// expected-warning {{deprecated}}
name|UserODDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalDeprecated
block|,
name|UserODUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalDeprecated
block|,
name|UserOU
init|=
name|OriginalUnavailable
block|,
comment|// expected-error {{unavailable}}
name|UserOUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalUnavailable
block|,
comment|// expected-error {{unavailable}}
name|UserOUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalUnavailable
block|,
name|UserAD
init|=
name|AllDeprecatedCase
block|,
comment|// expected-warning {{deprecated}}
name|UserADDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedCase
block|,
name|UserADUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedCase
block|,
name|UserADU
init|=
name|AllDeprecatedUnavailable
block|,
comment|// expected-error {{unavailable}}
name|UserADUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedUnavailable
block|,
comment|// expected-error {{unavailable}}
name|UserADUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedUnavailable
block|,
name|UserAU
init|=
name|AllUnavailableCase
block|,
comment|// expected-error {{unavailable}}
name|UserAUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllUnavailableCase
block|,
comment|// expected-error {{unavailable}}
name|UserAUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllUnavailableCase
block|, }
enum|;
end_enum

begin_enum
enum|enum
name|UserDeprecated
block|{
name|UserDeprecatedOD
init|=
name|OriginalDeprecated
block|,
name|UserDeprecatedODDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalDeprecated
block|,
name|UserDeprecatedODUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalDeprecated
block|,
name|UserDeprecatedOU
init|=
name|OriginalUnavailable
block|,
comment|// expected-error {{unavailable}}
name|UserDeprecatedOUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalUnavailable
block|,
comment|// expected-error {{unavailable}}
name|UserDeprecatedOUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalUnavailable
block|,
name|UserDeprecatedAD
init|=
name|AllDeprecatedCase
block|,
name|UserDeprecatedADDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedCase
block|,
name|UserDeprecatedADUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedCase
block|,
name|UserDeprecatedADU
init|=
name|AllDeprecatedUnavailable
block|,
comment|// expected-error {{unavailable}}
name|UserDeprecatedADUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedUnavailable
block|,
comment|// expected-error {{unavailable}}
name|UserDeprecatedADUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedUnavailable
block|,
name|UserDeprecatedAU
init|=
name|AllUnavailableCase
block|,
comment|// expected-error {{unavailable}}
name|UserDeprecatedAUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllUnavailableCase
block|,
comment|// expected-error {{unavailable}}
name|UserDeprecatedAUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllUnavailableCase
block|, }
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
enum|;
end_enum

begin_enum
enum|enum
name|UserUnavailable
block|{
name|UserUnavailableOD
init|=
name|OriginalDeprecated
block|,
name|UserUnavailableODDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalDeprecated
block|,
name|UserUnavailableODUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalDeprecated
block|,
name|UserUnavailableOU
init|=
name|OriginalUnavailable
block|,
name|UserUnavailableOUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalUnavailable
block|,
name|UserUnavailableOUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|OriginalUnavailable
block|,
name|UserUnavailableAD
init|=
name|AllDeprecatedCase
block|,
name|UserUnavailableADDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedCase
block|,
name|UserUnavailableADUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedCase
block|,
name|UserUnavailableADU
init|=
name|AllDeprecatedUnavailable
block|,
name|UserUnavailableADUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedUnavailable
block|,
name|UserUnavailableADUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllDeprecatedUnavailable
block|,
name|UserUnavailableAU
init|=
name|AllUnavailableCase
block|,
name|UserUnavailableAUDeprecated
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|deprecated
operator|=
literal|10.2
argument_list|)
operator|)
argument_list|)
init|=
name|AllUnavailableCase
block|,
name|UserUnavailableAUUnavailable
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
init|=
name|AllUnavailableCase
block|, }
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|unavailable
argument_list|)
operator|)
argument_list|)
enum|;
end_enum

end_unit

