begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 "-triple" "x86_64-apple-darwin9.0.0" -fsyntax-only -verify %s
end_comment

begin_function_decl
name|void
name|f0
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|10.4
parameter_list|,
name|deprecated
init|=
literal|10.6
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|void
name|f1
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|10.5
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|void
name|f2
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|10.4
parameter_list|,
name|deprecated
init|=
literal|10.5
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note {{'f2' declared here}}
end_comment

begin_function_decl
name|void
name|f3
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|10.6
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|void
name|f4
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|macosx
parameter_list|,
name|introduced
init|=
literal|10.1
parameter_list|,
name|deprecated
init|=
literal|10.3
parameter_list|,
name|obsoleted
init|=
literal|10.5
parameter_list|)
operator|,
function_decl|availability
parameter_list|(
name|ios
parameter_list|,
name|introduced
init|=
literal|2.0
parameter_list|,
name|deprecated
init|=
literal|3.0
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note{{explicitly marked unavailable}}
end_comment

begin_function_decl
name|void
name|f5
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(availability
parameter_list|(
name|ios
parameter_list|,
name|introduced
init|=
literal|3.2
parameter_list|)
operator|,
function_decl|availability
parameter_list|(
name|macosx
parameter_list|,
name|unavailable
parameter_list|)
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note{{function has been explicitly marked unavailable here}}
end_comment

begin_function
name|void
name|test
parameter_list|()
block|{
name|f0
argument_list|(
literal|0
argument_list|)
expr_stmt|;
name|f1
argument_list|(
literal|0
argument_list|)
expr_stmt|;
name|f2
argument_list|(
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning{{'f2' is deprecated: first deprecated in OS X 10.5}}
name|f3
argument_list|(
literal|0
argument_list|)
expr_stmt|;
name|f4
argument_list|(
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error{{f4' is unavailable: obsoleted in OS X 10.5}}
name|f5
argument_list|(
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error{{'f5' is unavailable: not available on OS X}}
block|}
end_function

begin_comment
comment|// rdar://10535640
end_comment

begin_enum
enum|enum
block|{
name|foo
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|introduced
operator|=
literal|8.0
argument_list|,
name|deprecated
operator|=
literal|9.0
argument_list|)
operator|)
argument_list|)
block|}
enum|;
end_enum

begin_enum
enum|enum
block|{
name|bar
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|introduced
operator|=
literal|8.0
argument_list|,
name|deprecated
operator|=
literal|9.0
argument_list|)
operator|)
argument_list|)
init|=
name|foo
block|}
enum|;
end_enum

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|availability
argument_list|(
name|macosx
argument_list|,
name|introduced
operator|=
literal|8.0
argument_list|,
name|deprecated
operator|=
literal|9.0
argument_list|)
operator|)
argument_list|)
block|{
name|bar1
operator|=
name|foo
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

end_unit

