begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only %s -verify -fblocks
end_comment

begin_function_decl
name|void
name|I
parameter_list|(
name|void
function_decl|(
modifier|^
function_decl|)
parameter_list|(
name|void
parameter_list|)
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
function_decl|(
modifier|^
name|noop
function_decl|)
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|nothing
parameter_list|()
function_decl|;
end_function_decl

begin_function_decl
name|int
name|printf
parameter_list|(
specifier|const
name|char
modifier|*
parameter_list|,
modifier|...
parameter_list|)
function_decl|;
end_function_decl

begin_typedef
typedef|typedef
name|void
function_decl|(
modifier|^
name|T
function_decl|)
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_typedef

begin_function_decl
name|void
name|takeblock
parameter_list|(
name|T
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|int
name|takeintint
parameter_list|(
name|int
function_decl|(
modifier|^
name|C
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|)
block|{
return|return
name|C
argument_list|(
literal|4
argument_list|)
return|;
block|}
end_function

begin_function
name|T
name|somefunction
parameter_list|()
block|{
if|if
condition|(
lambda|^
block|{ }
condition|)
name|nothing
argument_list|()
expr_stmt|;
name|noop
operator|=
lambda|^
block|{}
expr_stmt|;
name|noop
operator|=
lambda|^
block|{
name|printf
argument_list|(
literal|"\nClosure\n"
argument_list|)
expr_stmt|;
block|}
expr_stmt|;
name|I
argument_list|(
lambda|^
block|{ }
argument_list|)
expr_stmt|;
return|return
lambda|^
block|{
name|printf
argument_list|(
literal|"\nClosure\n"
argument_list|)
expr_stmt|;
block|}
return|;
block|}
end_function

begin_function
name|void
name|test2
parameter_list|()
block|{
name|int
name|x
init|=
literal|4
decl_stmt|;
name|takeblock
argument_list|(
lambda|^
block|{
name|printf
argument_list|(
literal|"%d\n"
argument_list|,
name|x
argument_list|)
expr_stmt|;
block|}
argument_list|)
expr_stmt|;
while|while
condition|(
literal|1
condition|)
block|{
name|takeblock
argument_list|(
lambda|^
block|{
break|break;
comment|// expected-error {{'break' statement not in loop or switch statement}}
continue|continue;
comment|// expected-error {{'continue' statement not in loop statement}}
while|while
condition|(
literal|1
condition|)
break|break;
comment|// ok
goto|goto
name|foo
goto|;
comment|// expected-error {{use of undeclared label 'foo'}}
name|a
label|:
goto|goto
name|a
goto|;
comment|// ok
block|}
argument_list|)
expr_stmt|;
break|break;
block|}
name|foo
label|:
name|takeblock
argument_list|(
lambda|^
block|{
name|x
operator|=
literal|4
expr_stmt|;
block|}
argument_list|)
expr_stmt|;
comment|// expected-error {{variable is not assignable (missing __block type specifier)}}
specifier|__block
name|y
operator|=
literal|7
expr_stmt|;
comment|// expected-warning {{type specifier missing, defaults to 'int'}}
name|takeblock
argument_list|(
lambda|^
block|{
name|y
operator|=
literal|8
expr_stmt|;
block|}
argument_list|)
expr_stmt|;
block|}
end_function

begin_macro
name|void
argument_list|(
argument|^test3()
argument_list|)
end_macro

begin_expr_stmt
operator|(
name|void
operator|)
block|{
return|return
lambda|^
block|{
block|}
expr_stmt|;
end_expr_stmt

begin_macro
unit|}  void
name|test4
argument_list|()
end_macro

begin_block
block|{
name|void
function_decl|(
modifier|^
name|noop
function_decl|)
parameter_list|(
name|void
parameter_list|)
init|=
lambda|^
block|{}
function_decl|;
name|void
function_decl|(
modifier|*
name|noop2
function_decl|)
parameter_list|()
init|=
literal|0
function_decl|;
block|}
end_block

begin_function
name|void
name|myfunc
parameter_list|(
name|int
function_decl|(
modifier|^
name|block
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|)
block|{}
end_function

begin_function_decl
name|void
name|myfunc3
parameter_list|(
specifier|const
name|int
modifier|*
name|x
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|test5
parameter_list|()
block|{
name|int
name|a
decl_stmt|;
name|myfunc
argument_list|(
lambda|^
parameter_list|(
name|int
name|abcd
parameter_list|)
block|{
name|myfunc3
argument_list|(
operator|&
name|a
argument_list|)
expr_stmt|;
return|return
literal|1
return|;
block|}
argument_list|)
expr_stmt|;
block|}
end_function

begin_decl_stmt
name|void
modifier|*
name|X
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|test_arguments
parameter_list|()
block|{
name|int
name|y
decl_stmt|;
name|int
function_decl|(
modifier|^
name|c
function_decl|)
parameter_list|(
name|char
parameter_list|)
function_decl|;
operator|(
literal|1
condition|?
name|c
else|:
literal|0
operator|)
operator|(
literal|'x'
operator|)
expr_stmt|;
operator|(
literal|1
condition|?
literal|0
else|:
name|c
operator|)
operator|(
literal|'x'
operator|)
expr_stmt|;
operator|(
literal|1
condition|?
name|c
else|:
name|c
operator|)
operator|(
literal|'x'
operator|)
expr_stmt|;
block|}
end_function

begin_decl_stmt
specifier|static
name|int
name|global_x
init|=
literal|10
decl_stmt|;
end_decl_stmt

begin_function_decl
name|void
function_decl|(
modifier|^
name|global_block
function_decl|)
parameter_list|(
name|void
parameter_list|)
init|=
lambda|^
block|{
name|printf
argument_list|(
literal|"global x is %d\n"
argument_list|,
name|global_x
argument_list|)
function_decl|;
end_function_decl

begin_typedef
unit|};
typedef|typedef
name|void
function_decl|(
modifier|^
name|void_block_t
function_decl|)
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_typedef

begin_decl_stmt
specifier|static
specifier|const
name|void_block_t
name|myBlock
init|=
lambda|^
block|{ }
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|static
specifier|const
name|void_block_t
name|myBlock2
init|=
lambda|^
name|void
parameter_list|(
name|void
parameter_list|)
block|{ }
decl_stmt|;
end_decl_stmt

end_unit

