begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s -fblocks
end_comment

begin_function_decl
name|int
function_decl|(
modifier|*
name|FP
function_decl|)
parameter_list|()
function_decl|;
end_function_decl

begin_function_decl
name|int
function_decl|(
modifier|^
name|IFP
function_decl|)
parameter_list|()
function_decl|;
end_function_decl

begin_function_decl
name|int
function_decl|(
modifier|^
name|II
function_decl|)
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|int
name|main
parameter_list|()
block|{
name|int
function_decl|(
modifier|*
name|FPL
function_decl|)
parameter_list|(
name|int
parameter_list|)
init|=
name|FP
function_decl|;
comment|// C doesn't consider this an error.
comment|// For Blocks, the ASTContext::typesAreBlockCompatible() makes sure this is an error.
name|int
function_decl|(
modifier|^
name|PFR
function_decl|)
parameter_list|(
name|int
parameter_list|)
init|=
name|IFP
function_decl|;
comment|// OK
name|PFR
operator|=
name|II
expr_stmt|;
comment|// OK
name|int
function_decl|(
modifier|^
name|IFP
function_decl|)
parameter_list|()
init|=
name|PFR
function_decl|;
comment|// OK
specifier|const
name|int
function_decl|(
modifier|^
name|CIC
function_decl|)
parameter_list|()
init|=
name|IFP
function_decl|;
comment|// expected-error {{incompatible block pointer types initializing 'int const (^)()' with an expression of type 'int (^)()'}} \
comment|// expected-warning{{type qualifier on return type has no effect}}
specifier|const
name|int
function_decl|(
modifier|^
name|CICC
function_decl|)
parameter_list|()
init|=
name|CIC
function_decl|;
comment|// expected-warning{{type qualifier on return type has no effect}}
name|int
modifier|*
specifier|const
function_decl|(
modifier|^
name|IPCC
function_decl|)
parameter_list|()
init|=
literal|0
function_decl|;
name|int
modifier|*
specifier|const
function_decl|(
modifier|^
name|IPCC1
function_decl|)
parameter_list|()
init|=
name|IPCC
function_decl|;
name|int
modifier|*
function_decl|(
modifier|^
name|IPCC2
function_decl|)
parameter_list|()
init|=
name|IPCC
function_decl|;
comment|// expected-error {{incompatible block pointer types initializing 'int *(^)()' with an expression of type 'int *const (^)()'}}
name|int
function_decl|(
modifier|^
name|IPCC3
function_decl|)
parameter_list|(
specifier|const
name|int
parameter_list|)
init|=
name|PFR
function_decl|;
name|int
function_decl|(
modifier|^
name|IPCC4
function_decl|)
parameter_list|(
name|int
parameter_list|,
name|char
function_decl|(
modifier|^
name|CArg
function_decl|)
parameter_list|(
name|double
parameter_list|)
parameter_list|)
function_decl|;
name|int
function_decl|(
modifier|^
name|IPCC5
function_decl|)
parameter_list|(
name|int
parameter_list|,
name|char
function_decl|(
modifier|^
name|CArg
function_decl|)
parameter_list|(
name|double
parameter_list|)
parameter_list|)
init|=
name|IPCC4
function_decl|;
name|int
function_decl|(
modifier|^
name|IPCC6
function_decl|)
parameter_list|(
name|int
parameter_list|,
name|char
function_decl|(
modifier|^
name|CArg
function_decl|)
parameter_list|(
name|float
parameter_list|)
parameter_list|)
init|=
name|IPCC4
function_decl|;
comment|// expected-error {{incompatible block pointer types initializing 'int (^)(int, char (^)(float))' with an expression of type 'int (^)(int, char (^)(double))'}}
name|IPCC2
operator|=
literal|0
expr_stmt|;
name|IPCC2
operator|=
literal|1
expr_stmt|;
comment|// expected-error {{invalid block pointer conversion assigning to 'int *(^)()' from 'int'}}
name|int
function_decl|(
modifier|^
name|x
function_decl|)
parameter_list|()
init|=
literal|0
function_decl|;
name|int
function_decl|(
modifier|^
name|y
function_decl|)
parameter_list|()
init|=
literal|3
function_decl|;
comment|// expected-error {{invalid block pointer conversion initializing 'int (^)()' with an expression of type 'int'}}
name|int
name|a
init|=
literal|1
decl_stmt|;
name|int
function_decl|(
modifier|^
name|z
function_decl|)
parameter_list|()
init|=
name|a
operator|+
literal|4
function_decl|;
comment|// expected-error {{invalid block pointer conversion initializing 'int (^)()' with an expression of type 'int'}}
block|}
end_function

begin_function
name|int
name|blah
parameter_list|()
block|{
name|int
function_decl|(
modifier|^
name|IFP
function_decl|)
parameter_list|(
name|float
parameter_list|)
function_decl|;
name|char
function_decl|(
modifier|^
name|PCP
function_decl|)
parameter_list|(
name|double
parameter_list|,
name|double
parameter_list|,
name|char
parameter_list|)
function_decl|;
name|IFP
argument_list|(
literal|1.0
argument_list|)
expr_stmt|;
name|IFP
argument_list|(
literal|1.0
argument_list|,
literal|2.0
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to block call}}
name|char
name|ch
init|=
name|PCP
argument_list|(
literal|1.0
argument_list|,
literal|2.0
argument_list|,
literal|'a'
argument_list|)
decl_stmt|;
return|return
name|PCP
argument_list|(
literal|1.0
argument_list|,
literal|2.0
argument_list|)
return|;
comment|// expected-error {{too few arguments to block}}
block|}
end_function

end_unit

