begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -ObjC -fsyntax-only %s -verify -fblocks
end_comment

begin_comment
comment|// FIXME: should compile
end_comment

begin_comment
comment|// Test for blocks with explicit return type specified.
end_comment

begin_typedef
typedef|typedef
name|float
modifier|*
name|PF
typedef|;
end_typedef

begin_decl_stmt
name|float
name|gf
decl_stmt|;
end_decl_stmt

begin_expr_stmt
unit|@
name|interface
name|NSView
operator|-
operator|(
name|id
operator|)
name|some_method_that_returns_id
expr_stmt|;
end_expr_stmt

begin_decl_stmt
unit|@
name|end
name|NSView
modifier|*
name|some_object
decl_stmt|;
end_decl_stmt

begin_function_decl
name|void
name|some_func
parameter_list|(
name|NSView
modifier|*
function_decl|(
modifier|^
function_decl|)
parameter_list|(
name|id
parameter_list|)
parameter_list|)
function_decl|;
end_function_decl

begin_typedef
typedef|typedef
name|struct
name|dispatch_item_s
modifier|*
name|dispatch_item_t
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|void
function_decl|(
modifier|^
name|completion_block_t
function_decl|)
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_typedef

begin_typedef
typedef|typedef
name|double
function_decl|(
modifier|^
name|myblock
function_decl|)
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_typedef

begin_function_decl
name|double
name|test
parameter_list|(
name|myblock
name|I
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|int
name|main
parameter_list|()
block|{
specifier|__block
name|int
name|x
init|=
literal|1
decl_stmt|;
specifier|__block
name|int
name|y
init|=
literal|2
decl_stmt|;
operator|(
name|void
operator|)
lambda|^
name|void
modifier|*
block|{
return|return
literal|0
return|;
block|}
expr_stmt|;
operator|(
name|void
operator|)
lambda|^
name|float
parameter_list|(
name|float
name|y
parameter_list|)
block|{
return|return
name|y
return|;
block|}
expr_stmt|;
operator|(
name|void
operator|)
lambda|^
name|double
parameter_list|(
name|float
name|y
parameter_list|,
name|double
name|d
parameter_list|)
block|{
if|if
condition|(
name|y
condition|)
return|return
name|d
return|;
else|else
return|return
name|y
return|;
block|}
expr_stmt|;
specifier|const
name|char
modifier|*
function_decl|(
modifier|^
name|chb
function_decl|)
parameter_list|(
name|int
name|flag
parameter_list|,
specifier|const
name|char
modifier|*
name|arg
parameter_list|,
name|char
modifier|*
name|arg1
parameter_list|)
init|=
lambda|^
specifier|const
name|char
modifier|*
parameter_list|(
name|int
name|flag
parameter_list|,
specifier|const
name|char
modifier|*
name|arg
parameter_list|,
name|char
modifier|*
name|arg1
parameter_list|)
block|{
if|if
condition|(
name|flag
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|flag
operator|==
literal|1
condition|)
return|return
name|arg
return|;
elseif|else
if|if
condition|(
name|flag
operator|==
literal|2
condition|)
return|return
literal|""
return|;
return|return
name|arg1
return|;
block|}
end_function

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_expr_stmt
operator|(
name|void
operator|)
lambda|^
name|PF
block|{
return|return
operator|&
name|gf
return|;
block|}
expr_stmt|;
end_expr_stmt

begin_expr_stmt
name|some_func
argument_list|(
lambda|^
name|NSView
modifier|*
parameter_list|(
name|id
name|whatever
parameter_list|)
block|{
return|return
index|[
name|some_object
name|some_method_that_returns_id
index|]
return|;
block|}
argument_list|)
expr_stmt|;
end_expr_stmt

begin_decl_stmt
name|double
name|res
init|=
name|test
argument_list|(
lambda|^
parameter_list|(
name|int
name|z
parameter_list|)
block|{
name|x
operator|=
name|y
operator|+
name|z
expr_stmt|;
return|return
operator|(
name|double
operator|)
name|z
return|;
block|}
argument_list|)
decl_stmt|;
end_decl_stmt

begin_macro
unit|}  void
name|func
argument_list|()
end_macro

begin_block
block|{
name|completion_block_t
name|X
decl_stmt|;
name|completion_block_t
function_decl|(
modifier|^
name|blockx
function_decl|)
parameter_list|(
name|dispatch_item_t
parameter_list|)
init|=
lambda|^
name|completion_block_t
parameter_list|(
name|dispatch_item_t
name|item
parameter_list|)
block|{
return|return
name|X
return|;
block|}
end_block

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_function_decl
name|completion_block_t
function_decl|(
modifier|^
name|blocky
function_decl|)
parameter_list|(
name|dispatch_item_t
parameter_list|)
init|=
lambda|^
parameter_list|(
name|dispatch_item_t
name|item
parameter_list|)
block|{
end_function_decl

begin_return
return|return
name|X
return|;
end_return

begin_expr_stmt
unit|};
name|blockx
operator|=
name|blocky
expr_stmt|;
end_expr_stmt

begin_comment
unit|}
comment|// intent: block taking int returning block that takes char,int and returns int
end_comment

begin_expr_stmt
unit|int
operator|(
operator|^
operator|(
operator|^
name|block
operator|)
operator|(
name|double
name|x
operator|)
operator|)
operator|(
name|char
operator|,
name|short
operator|)
expr_stmt|;
end_expr_stmt

begin_function
name|void
name|foo
parameter_list|()
block|{
name|int
name|one
init|=
literal|1
decl_stmt|;
name|block
operator|=
lambda|^
parameter_list|(
name|double
name|x
parameter_list|)
block|{
return|return
lambda|^
parameter_list|(
name|char
name|c
parameter_list|,
name|short
name|y
parameter_list|)
block|{
return|return
name|one
operator|+
name|c
operator|+
name|y
return|;
block|}
return|;
block|}
expr_stmt|;
comment|// expected-error {{returning block that lives on the local stack}}
comment|// or:
name|block
operator|=
lambda|^
parameter_list|(
name|double
name|x
parameter_list|)
block|{
return|return
lambda|^
parameter_list|(
name|char
name|c
parameter_list|,
name|short
name|y
parameter_list|)
block|{
return|return
name|one
operator|+
operator|(
name|int
operator|)
name|c
operator|+
name|y
return|;
block|}
return|;
block|}
expr_stmt|;
comment|// expected-error {{returning block that lives on the local stack}}
block|}
end_function

end_unit

