begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s
end_comment

begin_function
name|void
name|f
parameter_list|(
name|int
name|x
parameter_list|,
name|int
name|y
parameter_list|,
name|int
name|z
parameter_list|)
block|{
name|int
name|a
decl_stmt|,
name|b
decl_stmt|;
if|if
condition|(
operator|(
name|a
operator|>
literal|2
operator|)
operator|>
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant 1 with boolean expression is always false}}
if|if
condition|(
name|a
operator|>
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|<
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>=
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|<=
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|==
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|!=
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>
literal|2
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>=
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>=
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>=
literal|2
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>=
operator|-
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|<=
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|<=
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|<=
literal|2
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|<=
operator|-
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|>
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|>
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant 1 with boolean expression is always false}}
if|if
condition|(
operator|!
name|a
operator|>
literal|2
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always false}}
if|if
condition|(
operator|!
name|a
operator|>
name|y
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|>
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|>
operator|-
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always true}}
if|if
condition|(
operator|!
name|a
operator|<
literal|0
condition|)
block|{}
comment|// expected-warning {{comparison of constant 0 with boolean expression is always false}}
if|if
condition|(
operator|!
name|a
operator|<
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|<
literal|2
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always true}}
if|if
condition|(
operator|!
name|a
operator|<
name|y
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|<
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|<
operator|-
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always false}}
if|if
condition|(
operator|!
name|a
operator|>=
literal|0
condition|)
block|{}
comment|// expected-warning {{comparison of constant 0 with boolean expression is always true}}
if|if
condition|(
operator|!
name|a
operator|>=
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|>=
literal|2
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always false}}
if|if
condition|(
operator|!
name|a
operator|>=
name|y
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|>=
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|>=
operator|-
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always true}}
if|if
condition|(
operator|!
name|a
operator|<=
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|<=
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant 1 with boolean expression is always true}}
if|if
condition|(
operator|!
name|a
operator|<=
literal|2
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always true}}
if|if
condition|(
operator|!
name|a
operator|<=
name|y
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|<=
name|b
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|!
name|a
operator|<=
operator|-
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|||
name|b
operator|)
operator|>
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|||
name|b
operator|)
operator|>
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant 1 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|||
name|b
operator|)
operator|>
literal|4
condition|)
block|{}
comment|// expected-warning {{comparison of constant 4 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|||
name|b
operator|)
operator|>
operator|-
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always true}}
if|if
condition|(
operator|(
name|a
operator|&&
name|b
operator|)
operator|>
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|&&
name|b
operator|)
operator|>
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant 1 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|&&
name|b
operator|)
operator|>
literal|4
condition|)
block|{}
comment|// expected-warning {{comparison of constant 4 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|>
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|>
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant 1 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|>
literal|4
condition|)
block|{}
comment|// expected-warning {{comparison of constant 4 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|>
name|z
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|>
operator|-
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always true}}
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|==
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|==
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|==
literal|2
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|==
name|z
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|==
operator|-
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always false}}
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|!=
literal|0
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|!=
literal|1
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|!=
literal|2
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always true}}
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|!=
name|z
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|!=
operator|-
literal|1
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always true}}
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|==
name|z
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|a
operator|>
name|y
operator|<
name|z
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|>
name|z
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|>
operator|(
name|z
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|==
operator|(
name|z
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|!=
operator|(
name|z
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|z
operator|==
name|x
operator|)
operator|<
operator|(
name|y
operator|==
name|z
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|a
operator|<
name|y
operator|)
operator|!=
operator|(
operator|(
name|z
operator|==
name|x
operator|)
operator|<
operator|(
name|y
operator|==
name|z
operator|)
operator|)
condition|)
block|{}
comment|//no warning
if|if
condition|(
literal|0
operator|>
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant 0 with boolean expression is always false}}
if|if
condition|(
literal|1
operator|>
operator|!
name|a
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|2
operator|>
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always true}}
if|if
condition|(
name|y
operator|>
operator|!
name|a
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|-
literal|1
operator|>
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always false}}
if|if
condition|(
literal|0
operator|<
operator|!
name|a
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|1
operator|<
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant 1 with boolean expression is always false}}
if|if
condition|(
literal|2
operator|<
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always false}}
if|if
condition|(
name|y
operator|<
operator|!
name|a
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|-
literal|1
operator|<
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always true}}
if|if
condition|(
literal|0
operator|>=
operator|!
name|a
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|1
operator|>=
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant 1 with boolean expression is always true}}
if|if
condition|(
literal|2
operator|>=
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always true}}
if|if
condition|(
name|y
operator|>=
operator|!
name|a
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|-
literal|1
operator|>=
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always false}}
if|if
condition|(
literal|0
operator|<=
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant 0 with boolean expression is always true}}
if|if
condition|(
literal|1
operator|<=
operator|!
name|a
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|2
operator|<=
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always false}}
if|if
condition|(
name|y
operator|<=
operator|!
name|a
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|-
literal|1
operator|<=
operator|!
name|a
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always true}}
if|if
condition|(
literal|0
operator|>
operator|(
name|a
operator|||
name|b
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 0 with boolean expression is always false}}
if|if
condition|(
literal|1
operator|>
operator|(
name|a
operator|||
name|b
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|4
operator|>
operator|(
name|a
operator|||
name|b
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 4 with boolean expression is always true}}
if|if
condition|(
literal|0
operator|>
operator|(
name|a
operator|&&
name|b
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 0 with boolean expression is always false}}
if|if
condition|(
literal|1
operator|>
operator|(
name|a
operator|&&
name|b
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|4
operator|>
operator|(
name|a
operator|&&
name|b
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 4 with boolean expression is always true}}
if|if
condition|(
literal|0
operator|>
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 0 with boolean expression is always false}}
if|if
condition|(
literal|1
operator|>
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|4
operator|>
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 4 with boolean expression is always true}}
if|if
condition|(
name|z
operator|>
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|-
literal|1
operator|>
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always false}}
if|if
condition|(
literal|0
operator|==
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|1
operator|==
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|2
operator|==
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always false}}
if|if
condition|(
name|z
operator|==
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|-
literal|1
operator|==
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always false}}
if|if
condition|(
literal|0
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|1
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
literal|2
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always true}}
if|if
condition|(
name|z
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|-
literal|1
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always true}}
if|if
condition|(
name|z
operator|==
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|z
operator|<
name|a
operator|>
name|y
condition|)
block|{}
comment|// no warning
if|if
condition|(
name|z
operator|>
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|z
operator|<
name|y
operator|)
operator|>
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|z
operator|<
name|y
operator|)
operator|==
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|z
operator|<
name|y
operator|)
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
name|y
operator|==
name|z
operator|)
operator|<
operator|(
name|z
operator|==
name|x
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
operator|(
name|z
operator|==
name|x
operator|)
operator|<
operator|(
name|y
operator|==
name|z
operator|)
operator|)
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
operator|(
name|z
operator|==
name|x
operator|)
operator|<
operator|(
operator|-
literal|1
operator|==
name|z
operator|)
operator|)
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
operator|(
name|z
operator|==
name|x
operator|)
operator|<
operator|(
name|z
operator|==
operator|-
literal|1
operator|)
operator|)
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
if|if
condition|(
operator|(
operator|(
name|z
operator|==
name|x
operator|)
operator|<
operator|-
literal|1
operator|)
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant -1 with boolean expression is always false}}
if|if
condition|(
operator|(
operator|(
name|z
operator|==
name|x
operator|)
operator|<
literal|2
operator|)
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// expected-warning {{comparison of constant 2 with boolean expression is always true}}
if|if
condition|(
operator|(
operator|(
name|z
operator|==
name|x
operator|)
operator|<
operator|(
name|z
operator|>
literal|2
operator|)
operator|)
operator|!=
operator|(
name|a
operator|<
name|y
operator|)
condition|)
block|{}
comment|// no warning
block|}
end_function

end_unit

