begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple i386-unknown-unknown -fsyntax-only -fno-spell-checking -verify %s
end_comment

begin_typedef
typedef|typedef
name|int
name|v4si
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|float
name|v4f
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|double
name|v2d
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|long
name|long
name|v2ll
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|long
name|long
name|v4ll
name|__attribute__
typedef|((
name|vector_size
typedef|(32)));
end_typedef

begin_typedef
typedef|typedef
name|long
name|long
name|v8ll
name|__attribute__
typedef|((
name|vector_size
typedef|(64)));
end_typedef

begin_function
name|void
name|call_x86_64_builtins
parameter_list|(
name|void
parameter_list|)
block|{
name|unsigned
name|long
name|long
modifier|*
name|ullp
decl_stmt|;
name|void
modifier|*
name|vp
decl_stmt|;
name|v4f
name|vec4floats
decl_stmt|;
name|v2d
name|vec2doubles
decl_stmt|;
name|v2ll
name|vec2longlongs
decl_stmt|;
name|v4ll
name|vec4longlongs
decl_stmt|;
name|v8ll
name|vec8longlongs
decl_stmt|;
operator|(
name|void
operator|)
name|__builtin_ia32_readeflags_u64
argument_list|()
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_writeeflags_u64
argument_list|(
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_cvtss2si64
argument_list|(
name|vec4floats
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_cvttss2si64
argument_list|(
name|vec4floats
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_cvtsd2si64
argument_list|(
name|vec2doubles
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_cvttsd2si64
argument_list|(
name|vec2doubles
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_crc32di
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_rdfsbase64
argument_list|()
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_rdgsbase64
argument_list|()
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_wrfsbase64
argument_list|(
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_wrgsbase64
argument_list|(
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_fxrstor64
argument_list|(
name|vp
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_fxsave64
argument_list|(
name|vp
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_xsave64
argument_list|(
name|vp
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_xrstor64
argument_list|(
name|vp
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_xsaveopt64
argument_list|(
name|vp
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_xrstors64
argument_list|(
name|vp
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_xsavec64
argument_list|(
name|vp
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_xsaves64
argument_list|(
name|vp
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_addcarryx_u64
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|,
literal|4
argument_list|,
name|ullp
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_addcarry_u64
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|,
literal|4
argument_list|,
name|ullp
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_subborrow_u64
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|,
literal|4
argument_list|,
name|ullp
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_rdseed64_step
argument_list|(
name|ullp
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_bextr_u64
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_bzhi_di
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_pdep_di
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_pext_di
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_bextri_u64
argument_list|(
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_pbroadcastq512_gpr_mask
argument_list|(
literal|4
argument_list|,
name|vec8longlongs
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_pbroadcastq128_gpr_mask
argument_list|(
literal|4
argument_list|,
name|vec2longlongs
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_pbroadcastq256_gpr_mask
argument_list|(
literal|4
argument_list|,
name|vec4longlongs
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_vcvtsd2si64
argument_list|(
name|vec2doubles
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_vcvtsd2usi64
argument_list|(
name|vec2doubles
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_vcvtss2si64
argument_list|(
name|vec4floats
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_vcvtss2usi64
argument_list|(
name|vec4floats
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_vcvttsd2si64
argument_list|(
name|vec2doubles
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_vcvttsd2usi64
argument_list|(
name|vec2doubles
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_vcvttss2si64
argument_list|(
name|vec4floats
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_vcvttss2usi64
argument_list|(
name|vec4floats
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_cvtsi2sd64
argument_list|(
name|vec2doubles
argument_list|,
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_cvtsi2ss64
argument_list|(
name|vec4floats
argument_list|,
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_cvtusi2sd64
argument_list|(
name|vec2doubles
argument_list|,
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
operator|(
name|void
operator|)
name|__builtin_ia32_cvtusi2ss64
argument_list|(
name|vec4floats
argument_list|,
literal|4
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{use of unknown builtin}}
block|}
end_function

end_unit

