begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s
end_comment

begin_function
name|void
name|foo
parameter_list|()
block|{
name|int
name|a
decl_stmt|;
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|)
expr_stmt|;
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|,
literal|1
argument_list|)
expr_stmt|;
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|,
literal|1
argument_list|,
literal|2
argument_list|)
expr_stmt|;
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|,
literal|1
argument_list|,
literal|9
argument_list|,
literal|3
argument_list|)
expr_stmt|;
comment|// expected-error{{too many arguments to function}}
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|,
literal|"hello"
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error{{argument to '__builtin_prefetch' must be a constant integer}}
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|,
name|a
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error{{argument to '__builtin_prefetch' must be a constant integer}}
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error{{argument should be a value from 0 to 1}}
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|,
literal|0
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{argument should be a value from 0 to 3}}
name|__builtin_prefetch
argument_list|(
operator|&
name|a
argument_list|,
operator|-
literal|1
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error{{argument should be a value from 0 to 1}}
block|}
end_function

end_unit

