begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple armv7 -fsyntax-only -verify %s
end_comment

begin_struct
struct|struct
name|Simple
block|{
name|char
name|a
decl_stmt|,
name|b
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
name|test_ldrex
parameter_list|(
name|char
modifier|*
name|addr
parameter_list|)
block|{
name|int
name|sum
init|=
literal|0
decl_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
name|addr
argument_list|)
expr_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
name|short
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
name|int
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
name|long
name|long
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
name|float
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
name|double
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|sum
operator|+=
operator|*
name|__builtin_arm_ldrex
argument_list|(
operator|(
name|int
operator|*
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
expr|struct
name|Simple
operator|*
operator|*
operator|)
name|addr
argument_list|)
operator|->
name|a
expr_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
specifier|volatile
name|char
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
specifier|const
specifier|volatile
name|char
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
comment|// In principle this might be valid, but stick to ints and floats for scalar
comment|// types at the moment.
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
expr|struct
name|Simple
operator|*
operator|)
name|addr
argument_list|)
operator|.
name|a
expr_stmt|;
comment|// expected-error {{address argument to atomic builtin must be a pointer to}}
name|sum
operator|+=
name|__builtin_arm_ldrex
argument_list|(
operator|(
name|__int128
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
comment|// expected-error {{__int128 is not supported on this target}} expected-error {{address argument to load or store exclusive builtin must be a pointer to 1,2,4 or 8 byte type}}
name|__builtin_arm_ldrex
argument_list|()
expr_stmt|;
comment|// expected-error {{too few arguments to function call}}
name|__builtin_arm_ldrex
argument_list|(
literal|1
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to function call}}
return|return
name|sum
return|;
block|}
end_function

begin_function
name|int
name|test_strex
parameter_list|(
name|char
modifier|*
name|addr
parameter_list|)
block|{
name|int
name|res
init|=
literal|0
decl_stmt|;
name|struct
name|Simple
name|var
init|=
block|{
literal|0
block|}
decl_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|4
argument_list|,
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|42
argument_list|,
operator|(
name|short
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|42
argument_list|,
operator|(
name|int
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|42
argument_list|,
operator|(
name|long
name|long
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|2.71828f
argument_list|,
operator|(
name|float
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|3.14159
argument_list|,
operator|(
name|double
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
operator|&
name|var
argument_list|,
operator|(
expr|struct
name|Simple
operator|*
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|42
argument_list|,
operator|(
specifier|volatile
name|char
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|42
argument_list|,
operator|(
name|char
operator|*
specifier|const
operator|)
name|addr
argument_list|)
expr_stmt|;
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|42
argument_list|,
operator|(
specifier|const
name|char
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing 'const char *' to parameter of type 'volatile char *' discards qualifiers}}
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
name|var
argument_list|,
operator|(
expr|struct
name|Simple
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
comment|// expected-error {{address argument to atomic builtin must be a pointer to}}
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
name|var
argument_list|,
operator|(
expr|struct
name|Simple
operator|*
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
comment|// expected-error {{passing 'struct Simple' to parameter of incompatible type 'struct Simple *'}}
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
operator|&
name|var
argument_list|,
operator|(
expr|struct
name|Simple
operator|*
operator|*
operator|)
name|addr
argument_list|)
operator|.
name|a
expr_stmt|;
comment|// expected-error {{is not a structure or union}}
name|res
operator||=
name|__builtin_arm_strex
argument_list|(
literal|1
argument_list|,
operator|(
name|__int128
operator|*
operator|)
name|addr
argument_list|)
expr_stmt|;
comment|// expected-error {{__int128 is not supported on this target}} expected-error {{address argument to load or store exclusive builtin must be a pointer to 1,2,4 or 8 byte type}}
name|__builtin_arm_strex
argument_list|(
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{too few arguments to function call}}
name|__builtin_arm_strex
argument_list|(
literal|1
argument_list|,
literal|2
argument_list|,
literal|3
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to function call}}
return|return
name|res
return|;
block|}
end_function

begin_function
name|void
name|test_clrex
parameter_list|()
block|{
name|__builtin_arm_clrex
argument_list|()
expr_stmt|;
name|__builtin_arm_clrex
argument_list|(
literal|1
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to function call}}
block|}
end_function

end_unit

