begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple=x86_64-apple-darwin -fsyntax-only -verify %s
end_comment

begin_typedef
typedef|typedef
name|long
name|long
name|__m128i
name|__attribute__
typedef|((
name|__vector_size__
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|float
name|__m128
name|__attribute__
typedef|((
name|__vector_size__
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|double
name|__m128d
name|__attribute__
typedef|((
name|__vector_size__
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|long
name|long
name|__m512i
name|__attribute__
typedef|((
name|__vector_size__
typedef|(64)));
end_typedef

begin_typedef
typedef|typedef
name|float
name|__m512
name|__attribute__
typedef|((
name|__vector_size__
typedef|(64)));
end_typedef

begin_typedef
typedef|typedef
name|double
name|__m512d
name|__attribute__
typedef|((
name|__vector_size__
typedef|(64)));
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|char
name|__mmask8
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|short
name|__mmask16
typedef|;
end_typedef

begin_function
name|__m128
name|test__builtin_ia32_cmpps
parameter_list|(
name|__m128
name|__a
parameter_list|,
name|__m128
name|__b
parameter_list|)
block|{
name|__builtin_ia32_cmpps
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
block|}
end_function

begin_function
name|__m128d
name|test__builtin_ia32_cmppd
parameter_list|(
name|__m128d
name|__a
parameter_list|,
name|__m128d
name|__b
parameter_list|)
block|{
name|__builtin_ia32_cmppd
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
block|}
end_function

begin_function
name|__m128
name|test__builtin_ia32_cmpss
parameter_list|(
name|__m128
name|__a
parameter_list|,
name|__m128
name|__b
parameter_list|)
block|{
name|__builtin_ia32_cmpss
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
block|}
end_function

begin_function
name|__m128d
name|test__builtin_ia32_cmpsd
parameter_list|(
name|__m128d
name|__a
parameter_list|,
name|__m128d
name|__b
parameter_list|)
block|{
name|__builtin_ia32_cmpsd
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|32
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
block|}
end_function

begin_function
name|__mmask16
name|test__builtin_ia32_cmpps512_mask
parameter_list|(
name|__m512d
name|__a
parameter_list|,
name|__m512d
name|__b
parameter_list|)
block|{
name|__builtin_ia32_cmpps512_mask
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|32
argument_list|,
operator|-
literal|1
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
block|}
end_function

begin_function
name|__mmask8
name|test__builtin_ia32_cmppd512_mask
parameter_list|(
name|__m512d
name|__a
parameter_list|,
name|__m512d
name|__b
parameter_list|)
block|{
name|__builtin_ia32_cmppd512_mask
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|32
argument_list|,
operator|-
literal|1
argument_list|,
literal|4
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 31}}
block|}
end_function

begin_function
name|__m128i
name|test__builtin_ia32_vpcomub
parameter_list|(
name|__m128i
name|__a
parameter_list|,
name|__m128i
name|__b
parameter_list|)
block|{
name|__builtin_ia32_vpcomub
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
block|}
end_function

begin_function
name|__m128i
name|test__builtin_ia32_vpcomuw
parameter_list|(
name|__m128i
name|__a
parameter_list|,
name|__m128i
name|__b
parameter_list|)
block|{
name|__builtin_ia32_vpcomuw
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
block|}
end_function

begin_function
name|__m128i
name|test__builtin_ia32_vpcomud
parameter_list|(
name|__m128i
name|__a
parameter_list|,
name|__m128i
name|__b
parameter_list|)
block|{
name|__builtin_ia32_vpcomud
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
block|}
end_function

begin_function
name|__m128i
name|test__builtin_ia32_vpcomuq
parameter_list|(
name|__m128i
name|__a
parameter_list|,
name|__m128i
name|__b
parameter_list|)
block|{
name|__builtin_ia32_vpcomuq
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
block|}
end_function

begin_function
name|__m128i
name|test__builtin_ia32_vpcomb
parameter_list|(
name|__m128i
name|__a
parameter_list|,
name|__m128i
name|__b
parameter_list|)
block|{
name|__builtin_ia32_vpcomub
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
block|}
end_function

begin_function
name|__m128i
name|test__builtin_ia32_vpcomw
parameter_list|(
name|__m128i
name|__a
parameter_list|,
name|__m128i
name|__b
parameter_list|)
block|{
name|__builtin_ia32_vpcomuw
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
block|}
end_function

begin_function
name|__m128i
name|test__builtin_ia32_vpcomd
parameter_list|(
name|__m128i
name|__a
parameter_list|,
name|__m128i
name|__b
parameter_list|)
block|{
name|__builtin_ia32_vpcomud
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
block|}
end_function

begin_function
name|__m128i
name|test__builtin_ia32_vpcomq
parameter_list|(
name|__m128i
name|__a
parameter_list|,
name|__m128i
name|__b
parameter_list|)
block|{
name|__builtin_ia32_vpcomuq
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|8
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 7}}
block|}
end_function

begin_function
name|__mmask16
name|test__builtin_ia32_cmpps512_mask_rounding
parameter_list|(
name|__m512
name|__a
parameter_list|,
name|__m512
name|__b
parameter_list|,
name|__mmask16
name|__u
parameter_list|)
block|{
name|__builtin_ia32_cmpps512_mask
argument_list|(
name|__a
argument_list|,
name|__b
argument_list|,
literal|0
argument_list|,
name|__u
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error {{invalid rounding argument}}
block|}
end_function

begin_function
name|__m128i
name|test_mm_mask_i32gather_epi32
parameter_list|(
name|__m128i
name|a
parameter_list|,
name|int
specifier|const
modifier|*
name|b
parameter_list|,
name|__m128i
name|c
parameter_list|,
name|__m128i
name|mask
parameter_list|)
block|{
return|return
name|__builtin_ia32_gatherd_d
argument_list|(
name|a
argument_list|,
name|b
argument_list|,
name|c
argument_list|,
name|mask
argument_list|,
literal|5
argument_list|)
return|;
comment|// expected-error {{scale argument must be 1, 2, 4, or 8}}
block|}
end_function

begin_function
name|__m512i
name|_mm512_mask_prefetch_i32gather_ps
parameter_list|(
name|__m512i
name|index
parameter_list|,
name|__mmask16
name|mask
parameter_list|,
name|int
specifier|const
modifier|*
name|addr
parameter_list|)
block|{
return|return
name|__builtin_ia32_gatherpfdps
argument_list|(
name|mask
argument_list|,
name|index
argument_list|,
name|addr
argument_list|,
literal|5
argument_list|,
literal|1
argument_list|)
return|;
comment|// expected-error {{scale argument must be 1, 2, 4, or 8}}
block|}
end_function

begin_function
name|__m512
name|_mm512_mask_prefetch_i32gather_ps_2
parameter_list|(
name|__m512i
name|index
parameter_list|,
name|__mmask16
name|mask
parameter_list|,
name|int
specifier|const
modifier|*
name|addr
parameter_list|)
block|{
return|return
name|__builtin_ia32_gatherpfdps
argument_list|(
name|mask
argument_list|,
name|index
argument_list|,
name|addr
argument_list|,
literal|1
argument_list|,
literal|1
argument_list|)
return|;
comment|// expected-error {{argument should be a value from 2 to 3}}
block|}
end_function

end_unit

