begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple i386-mingw32 -fms-extensions -fsyntax-only -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-unknown-unknown -fsyntax-only -verify %s
end_comment

begin_function_decl
name|int
name|nonconst
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|isconst
parameter_list|(
name|void
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(const
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|int
name|ispure
parameter_list|(
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(pure
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function
name|int
name|foo
parameter_list|(
name|int
modifier|*
name|a
parameter_list|,
name|int
name|i
parameter_list|)
block|{
ifdef|#
directive|ifdef
name|_MSC_VER
name|__assume
argument_list|(
name|i
operator|!=
literal|4
argument_list|)
expr_stmt|;
name|__assume
argument_list|(
operator|++
name|i
operator|>
literal|2
argument_list|)
expr_stmt|;
comment|//expected-warning {{the argument to '__assume' has side effects that will be discarded}}
name|__assume
argument_list|(
name|nonconst
argument_list|()
operator|>
literal|2
argument_list|)
expr_stmt|;
comment|//expected-warning {{the argument to '__assume' has side effects that will be discarded}}
name|__assume
argument_list|(
name|isconst
argument_list|()
operator|>
literal|2
argument_list|)
expr_stmt|;
name|__assume
argument_list|(
name|ispure
argument_list|(
name|i
argument_list|)
operator|>
literal|2
argument_list|)
expr_stmt|;
name|__assume
argument_list|(
name|ispure
argument_list|(
operator|++
name|i
argument_list|)
operator|>
literal|2
argument_list|)
expr_stmt|;
comment|//expected-warning {{the argument to '__assume' has side effects that will be discarded}}
name|int
name|test
init|=
expr|sizeof
operator|(
expr|struct
block|{
name|char
name|qq
index|[
operator|(
name|__assume
argument_list|(
name|i
operator|!=
literal|5
argument_list|)
expr|,
literal|7
operator|)
index|]
block|;}
operator|)
decl_stmt|;
else|#
directive|else
name|__builtin_assume
argument_list|(
name|i
operator|!=
literal|4
argument_list|)
expr_stmt|;
name|__builtin_assume
argument_list|(
operator|++
name|i
operator|>
literal|2
argument_list|)
expr_stmt|;
comment|//expected-warning {{the argument to '__builtin_assume' has side effects that will be discarded}}
name|__builtin_assume
argument_list|(
name|nonconst
argument_list|()
operator|>
literal|2
argument_list|)
expr_stmt|;
comment|//expected-warning {{the argument to '__builtin_assume' has side effects that will be discarded}}
name|__builtin_assume
argument_list|(
name|isconst
argument_list|()
operator|>
literal|2
argument_list|)
expr_stmt|;
name|__builtin_assume
argument_list|(
name|ispure
argument_list|(
name|i
argument_list|)
operator|>
literal|2
argument_list|)
expr_stmt|;
name|__builtin_assume
argument_list|(
name|ispure
argument_list|(
operator|++
name|i
argument_list|)
operator|>
literal|2
argument_list|)
expr_stmt|;
comment|//expected-warning {{the argument to '__builtin_assume' has side effects that will be discarded}}
name|int
name|test
init|=
expr|sizeof
operator|(
expr|struct
block|{
name|char
name|qq
index|[
operator|(
name|__builtin_assume
argument_list|(
name|i
operator|!=
literal|5
argument_list|)
expr|,
literal|7
operator|)
index|]
block|;}
operator|)
decl_stmt|;
endif|#
directive|endif
return|return
name|a
index|[
name|i
index|]
return|;
block|}
end_function

end_unit

