begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -triple x86_64-apple-darwin9 -verify %s
end_comment

begin_decl_stmt
name|int
name|a
index|[
literal|10
index|]
decl_stmt|;
end_decl_stmt

begin_function
name|int
name|f0
parameter_list|()
block|{
return|return
name|__builtin_object_size
argument_list|(
operator|&
name|a
argument_list|)
return|;
comment|// expected-error {{too few arguments to function}}
block|}
end_function

begin_function
name|int
name|f1
parameter_list|()
block|{
return|return
operator|(
name|__builtin_object_size
argument_list|(
operator|&
name|a
argument_list|,
literal|0
argument_list|)
operator|+
name|__builtin_object_size
argument_list|(
operator|&
name|a
argument_list|,
literal|1
argument_list|)
operator|+
name|__builtin_object_size
argument_list|(
operator|&
name|a
argument_list|,
literal|2
argument_list|)
operator|+
name|__builtin_object_size
argument_list|(
operator|&
name|a
argument_list|,
literal|3
argument_list|)
operator|)
return|;
block|}
end_function

begin_function
name|int
name|f2
parameter_list|()
block|{
return|return
name|__builtin_object_size
argument_list|(
operator|&
name|a
argument_list|,
operator|-
literal|1
argument_list|)
return|;
comment|// expected-error {{argument should be a value from 0 to 3}}
block|}
end_function

begin_function
name|int
name|f3
parameter_list|()
block|{
return|return
name|__builtin_object_size
argument_list|(
operator|&
name|a
argument_list|,
literal|4
argument_list|)
return|;
comment|// expected-error {{argument should be a value from 0 to 3}}
block|}
end_function

begin_comment
comment|// rdar://6252231 - cannot call vsnprintf with va_list on x86_64
end_comment

begin_function
name|void
name|f4
parameter_list|(
specifier|const
name|char
modifier|*
name|fmt
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_va_list
name|args
decl_stmt|;
name|__builtin___vsnprintf_chk
argument_list|(
literal|0
argument_list|,
literal|42
argument_list|,
literal|0
argument_list|,
literal|11
argument_list|,
name|fmt
argument_list|,
name|args
argument_list|)
expr_stmt|;
comment|// expected-warning {{'__builtin___vsnprintf_chk' will always overflow destination buffer}}
block|}
end_function

begin_comment
comment|// rdar://18334276
end_comment

begin_typedef
typedef|typedef
name|__typeof__
argument_list|(
argument|sizeof(int)
argument_list|)
name|size_t
expr_stmt|;
end_typedef

begin_function_decl
name|void
modifier|*
name|memcset
parameter_list|(
name|void
modifier|*
specifier|restrict
name|dst
parameter_list|,
name|int
name|src
parameter_list|,
name|size_t
name|n
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
modifier|*
name|memcpy
parameter_list|(
name|void
modifier|*
specifier|restrict
name|dst
parameter_list|,
specifier|const
name|void
modifier|*
specifier|restrict
name|src
parameter_list|,
name|size_t
name|n
parameter_list|)
function_decl|;
end_function_decl

begin_define
define|#
directive|define
name|memset
parameter_list|(
name|dest
parameter_list|,
name|src
parameter_list|,
name|len
parameter_list|)
value|__builtin___memset_chk(dest, src, len, __builtin_object_size(dest, 0))
end_define

begin_define
define|#
directive|define
name|memcpy
parameter_list|(
name|dest
parameter_list|,
name|src
parameter_list|,
name|len
parameter_list|)
value|__builtin___memcpy_chk(dest, src, len, __builtin_object_size(dest, 0))
end_define

begin_define
define|#
directive|define
name|memcpy1
parameter_list|(
name|dest
parameter_list|,
name|src
parameter_list|,
name|len
parameter_list|)
value|__builtin___memcpy_chk(dest, src, len, __builtin_object_size(dest, 4))
end_define

begin_define
define|#
directive|define
name|NULL
value|((void *)0)
end_define

begin_function
name|void
name|f5
parameter_list|(
name|void
parameter_list|)
block|{
name|char
name|buf
index|[
literal|10
index|]
decl_stmt|;
name|memset
argument_list|(
operator|(
name|void
operator|*
operator|)
literal|0x100000000ULL
argument_list|,
literal|0
argument_list|,
literal|0x1000
argument_list|)
expr_stmt|;
name|memcpy
argument_list|(
operator|(
name|char
operator|*
operator|)
name|NULL
operator|+
literal|0x10000
argument_list|,
name|buf
argument_list|,
literal|0x10
argument_list|)
expr_stmt|;
name|memcpy1
argument_list|(
operator|(
name|char
operator|*
operator|)
name|NULL
operator|+
literal|0x10000
argument_list|,
name|buf
argument_list|,
literal|0x10
argument_list|)
expr_stmt|;
comment|// expected-error {{argument should be a value from 0 to 3}}
block|}
end_function

begin_comment
comment|// rdar://18431336
end_comment

begin_function
name|void
name|f6
parameter_list|(
name|void
parameter_list|)
block|{
name|char
name|b
index|[
literal|5
index|]
decl_stmt|;
name|char
name|buf
index|[
literal|10
index|]
decl_stmt|;
name|__builtin___memccpy_chk
argument_list|(
name|buf
argument_list|,
name|b
argument_list|,
literal|'\0'
argument_list|,
sizeof|sizeof
argument_list|(
name|b
argument_list|)
argument_list|,
name|__builtin_object_size
argument_list|(
name|buf
argument_list|,
literal|0
argument_list|)
argument_list|)
expr_stmt|;
name|__builtin___memccpy_chk
argument_list|(
name|b
argument_list|,
name|buf
argument_list|,
literal|'\0'
argument_list|,
sizeof|sizeof
argument_list|(
name|buf
argument_list|)
argument_list|,
name|__builtin_object_size
argument_list|(
name|b
argument_list|,
literal|0
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{'__builtin___memccpy_chk' will always overflow destination buffer}}
block|}
end_function

begin_function
name|int
name|pr28314
parameter_list|(
name|void
parameter_list|)
block|{
struct|struct
block|{
name|struct
name|InvalidField
name|a
decl_stmt|;
comment|// expected-error{{has incomplete type}} expected-note 3{{forward declaration of 'struct InvalidField'}}
name|char
name|b
index|[
literal|0
index|]
decl_stmt|;
block|}
modifier|*
name|p
struct|;
struct|struct
block|{
name|struct
name|InvalidField
name|a
decl_stmt|;
comment|// expected-error{{has incomplete type}}
name|char
name|b
index|[
literal|1
index|]
decl_stmt|;
block|}
modifier|*
name|p2
struct|;
struct|struct
block|{
name|struct
name|InvalidField
name|a
decl_stmt|;
comment|// expected-error{{has incomplete type}}
name|char
name|b
index|[
literal|2
index|]
decl_stmt|;
block|}
modifier|*
name|p3
struct|;
name|int
name|a
init|=
literal|0
decl_stmt|;
name|a
operator|+=
name|__builtin_object_size
argument_list|(
operator|&
name|p
operator|->
name|a
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|a
operator|+=
name|__builtin_object_size
argument_list|(
name|p
operator|->
name|b
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|a
operator|+=
name|__builtin_object_size
argument_list|(
name|p2
operator|->
name|b
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|a
operator|+=
name|__builtin_object_size
argument_list|(
name|p3
operator|->
name|b
argument_list|,
literal|0
argument_list|)
expr_stmt|;
return|return
name|a
return|;
block|}
end_function

end_unit

