begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -fsyntax-only -verify -pedantic
end_comment

begin_function_decl
name|void
name|check
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|a
parameter_list|()
block|{
name|check
argument_list|(
name|__builtin_isfinite
argument_list|(
literal|1.0f
argument_list|)
argument_list|)
expr_stmt|;
name|check
argument_list|(
name|__builtin_isinf
argument_list|(
literal|1.0
argument_list|)
argument_list|)
expr_stmt|;
name|check
argument_list|(
name|__builtin_isinf_sign
argument_list|(
literal|1.0L
argument_list|)
argument_list|)
expr_stmt|;
name|check
argument_list|(
name|__builtin_isnan
argument_list|(
literal|1.0f
argument_list|)
argument_list|)
expr_stmt|;
name|check
argument_list|(
name|__builtin_isnormal
argument_list|(
literal|1.0f
argument_list|)
argument_list|)
expr_stmt|;
name|check
argument_list|(
name|__builtin_isfinite
argument_list|(
literal|1
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-error{{requires argument of floating point type}}
name|check
argument_list|(
name|__builtin_isinf
argument_list|()
argument_list|)
expr_stmt|;
comment|// expected-error{{too few arguments}}
name|check
argument_list|(
name|__builtin_isnan
argument_list|(
literal|1
argument_list|,
literal|2
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-error{{too many arguments}}
block|}
end_function

end_unit

