begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -triple x86_64-unknown-unknown %s -verify
end_comment

begin_typedef
typedef|typedef
struct|struct
block|{
name|unsigned
name|long
name|bits
index|[
operator|(
operator|(
operator|(
literal|1
operator|)
operator|+
operator|(
literal|64
operator|)
operator|-
literal|1
operator|)
operator|/
operator|(
literal|64
operator|)
operator|)
index|]
decl_stmt|;
block|}
name|cpumask_t
typedef|;
end_typedef

begin_decl_stmt
name|cpumask_t
name|x
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|foo
parameter_list|()
block|{
operator|(
name|void
operator|)
name|x
expr_stmt|;
block|}
end_function

begin_function
name|void
name|bar
parameter_list|()
block|{
name|char
modifier|*
name|a
decl_stmt|;
name|double
name|b
decl_stmt|;
name|b
operator|=
operator|(
name|double
operator|)
name|a
expr_stmt|;
comment|// expected-error {{pointer cannot be cast to type}}
name|a
operator|=
operator|(
name|char
operator|*
operator|)
name|b
expr_stmt|;
comment|// expected-error {{cannot be cast to a pointer type}}
block|}
end_function

begin_function
name|long
name|bar1
parameter_list|(
name|long
modifier|*
name|next
parameter_list|)
block|{
return|return
call|(
name|long
call|)
argument_list|(
operator|*
name|next
argument_list|)
operator|++
return|;
block|}
end_function

begin_typedef
typedef|typedef
name|_Bool
name|Bool
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|int
name|Int
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|long
name|Long
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|float
name|Float
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|double
name|Double
typedef|;
end_typedef

begin_typedef
typedef|typedef
specifier|_Complex
name|int
name|CInt
typedef|;
end_typedef

begin_typedef
typedef|typedef
specifier|_Complex
name|long
name|CLong
typedef|;
end_typedef

begin_typedef
typedef|typedef
specifier|_Complex
name|float
name|CFloat
typedef|;
end_typedef

begin_typedef
typedef|typedef
specifier|_Complex
name|double
name|CDouble
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|void
modifier|*
name|VoidPtr
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|char
modifier|*
name|CharPtr
typedef|;
end_typedef

begin_function
name|void
name|testBool
parameter_list|(
name|Bool
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|VoidPtr
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CharPtr
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testInt
parameter_list|(
name|Int
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|VoidPtr
operator|)
name|v
expr_stmt|;
comment|// expected-warning{{cast to 'VoidPtr' (aka 'void *') from smaller integer type 'Int' (aka 'int')}}
operator|(
name|void
operator|)
operator|(
name|CharPtr
operator|)
name|v
expr_stmt|;
comment|// expected-warning{{cast to 'CharPtr' (aka 'char *') from smaller integer type 'Int' (aka 'int')}}
comment|// Test that casts to void* can be controlled separately
comment|// from other -Wint-to-pointer-cast warnings.
pragma|#
directive|pragma
name|clang
name|diagnostic
name|push
pragma|#
directive|pragma
name|clang
name|diagnostic
name|ignored
literal|"-Wint-to-void-pointer-cast"
operator|(
name|void
operator|)
operator|(
name|VoidPtr
operator|)
name|v
expr_stmt|;
comment|// no-warning
operator|(
name|void
operator|)
operator|(
name|CharPtr
operator|)
name|v
expr_stmt|;
comment|// expected-warning{{cast to 'CharPtr' (aka 'char *') from smaller integer type 'Int' (aka 'int')}}
pragma|#
directive|pragma
name|clang
name|diagnostic
name|pop
block|}
end_function

begin_function
name|void
name|testLong
parameter_list|(
name|Long
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|VoidPtr
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CharPtr
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testFloat
parameter_list|(
name|Float
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testDouble
parameter_list|(
name|Double
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testCI
parameter_list|(
name|CInt
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testCLong
parameter_list|(
name|CLong
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testCFloat
parameter_list|(
name|CFloat
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testCDouble
parameter_list|(
name|CDouble
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Float
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Double
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CInt
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CLong
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CFloat
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CDouble
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testVoidPtr
parameter_list|(
name|VoidPtr
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|VoidPtr
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CharPtr
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_function
name|void
name|testCharPtr
parameter_list|(
name|CharPtr
name|v
parameter_list|)
block|{
operator|(
name|void
operator|)
operator|(
name|Bool
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Int
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|Long
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|VoidPtr
operator|)
name|v
expr_stmt|;
operator|(
name|void
operator|)
operator|(
name|CharPtr
operator|)
name|v
expr_stmt|;
block|}
end_function

begin_typedef
typedef|typedef
enum|enum
block|{
name|x_a
block|,
name|x_b
block|}
name|X
typedef|;
end_typedef

begin_function
name|void
modifier|*
name|intToPointerCast2
parameter_list|(
name|X
name|x
parameter_list|)
block|{
return|return
operator|(
name|void
operator|*
operator|)
name|x
return|;
block|}
end_function

begin_function
name|void
modifier|*
name|intToPointerCast3
parameter_list|()
block|{
return|return
operator|(
name|void
operator|*
operator|)
operator|(
literal|1
operator|+
literal|3
operator|)
return|;
block|}
end_function

end_unit

