begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only %s -verify
end_comment

begin_typedef
typedef|typedef
struct|struct
block|{
name|unsigned
name|long
name|bits
index|[
operator|(
operator|(
operator|(
literal|1
operator|)
operator|+
operator|(
literal|64
operator|)
operator|-
literal|1
operator|)
operator|/
operator|(
literal|64
operator|)
operator|)
index|]
decl_stmt|;
block|}
name|cpumask_t
typedef|;
end_typedef

begin_decl_stmt
name|cpumask_t
name|x
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|foo
parameter_list|()
block|{
operator|(
name|void
operator|)
name|x
expr_stmt|;
block|}
end_function

begin_function
name|void
name|bar
parameter_list|()
block|{
name|char
modifier|*
name|a
decl_stmt|;
name|double
name|b
decl_stmt|;
name|b
operator|=
operator|(
name|double
operator|)
name|a
expr_stmt|;
comment|// expected-error {{pointer cannot be cast to type}}
name|a
operator|=
operator|(
name|char
operator|*
operator|)
name|b
expr_stmt|;
comment|// expected-error {{cannot be cast to a pointer type}}
block|}
end_function

begin_function
name|long
name|bar1
parameter_list|(
name|long
modifier|*
name|next
parameter_list|)
block|{
return|return
call|(
name|long
call|)
argument_list|(
operator|*
name|next
argument_list|)
operator|++
return|;
block|}
end_function

end_unit

