begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -triple x86_64-apple-darwin -fsyntax-only -pedantic -verify -Wsign-compare %s
end_comment

begin_function
name|int
name|test
parameter_list|(
name|char
modifier|*
name|C
parameter_list|)
block|{
comment|// nothing here should warn.
return|return
name|C
operator|!=
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
return|;
return|return
name|C
operator|!=
operator|(
name|void
operator|*
operator|)
literal|0
return|;
return|return
name|C
operator|!=
literal|0
return|;
return|return
name|C
operator|!=
literal|1
return|;
comment|// expected-warning {{comparison between pointer and integer ('char *' and 'int')}}
block|}
end_function

begin_function
name|int
name|ints
parameter_list|(
name|long
name|a
parameter_list|,
name|unsigned
name|long
name|b
parameter_list|)
block|{
enum|enum
name|EnumA
block|{
name|A
block|}
enum|;
enum|enum
name|EnumB
block|{
name|B
block|}
enum|;
enum|enum
name|EnumC
block|{
name|C
init|=
literal|0x10000
block|}
enum|;
return|return
comment|// (a,b)
operator|(
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|==
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|==
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|long
operator|)
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|long
operator|)
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
comment|// (A,b)
operator|(
name|A
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|A
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|A
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|A
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|A
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|A
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|A
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|A
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|A
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|A
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|A
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|A
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|A
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|A
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|A
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|A
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|A
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// (a,B)
operator|(
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
name|B
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
name|B
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
name|B
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
name|B
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|B
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
name|B
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|<
name|B
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|<
name|B
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
name|B
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|B
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|B
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|B
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|B
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
comment|// (C,b)
operator|(
name|C
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|C
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|C
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|C
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|C
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|C
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|C
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|C
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|C
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|C
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|C
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|C
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|C
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|C
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|C
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|C
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|C
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// (a,C)
operator|(
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
name|C
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
name|C
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
name|C
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
name|C
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
name|C
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|<
name|C
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|<
name|C
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
name|C
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
comment|// (0x80000,b)
operator|(
literal|0x80000
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
literal|0x80000
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
literal|0x80000
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
literal|0x80000
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
literal|0x80000
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
literal|0x80000
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
literal|0x80000
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
literal|0x80000
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
literal|0x80000
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
literal|0x80000
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
literal|0x80000
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
literal|0x80000
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
literal|0x80000
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
literal|0x80000
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
literal|0x80000
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
literal|0x80000
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
literal|0x80000
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// (a,0x80000)
operator|(
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|<
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|<
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
literal|10
return|;
block|}
end_function

begin_function
name|int
name|equal
parameter_list|(
name|char
modifier|*
name|a
parameter_list|,
specifier|const
name|char
modifier|*
name|b
parameter_list|)
block|{
return|return
name|a
operator|==
name|b
return|;
block|}
end_function

begin_decl_stmt
name|int
name|arrays
argument_list|(
name|char
argument_list|(
operator|*
name|a
argument_list|)
index|[
literal|5
index|]
argument_list|,
name|char
argument_list|(
operator|*
name|b
argument_list|)
index|[
literal|10
index|]
argument_list|,
name|char
argument_list|(
operator|*
name|c
argument_list|)
index|[
literal|5
index|]
argument_list|)
block|{
name|int
name|d
init|=
operator|(
name|a
operator|==
name|c
operator|)
decl_stmt|;
return|return
name|a
operator|==
name|b
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
block|}
end_decl_stmt

begin_function
name|int
name|pointers
parameter_list|(
name|int
modifier|*
name|a
parameter_list|)
block|{
return|return
name|a
operator|>
literal|0
return|;
comment|// expected-warning {{ordered comparison between pointer and zero ('int *' and 'int') is an extension}}
return|return
name|a
operator|>
literal|42
return|;
comment|// expected-warning {{ordered comparison between pointer and integer ('int *' and 'int')}}
return|return
name|a
operator|>
operator|(
name|void
operator|*
operator|)
literal|0
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
block|}
end_function

begin_function
name|int
name|function_pointers
parameter_list|(
name|int
function_decl|(
modifier|*
name|a
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|,
name|int
function_decl|(
modifier|*
name|b
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|,
name|void
function_decl|(
modifier|*
name|c
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|)
block|{
return|return
name|a
operator|>
name|b
return|;
comment|// expected-warning {{ordered comparison of function pointers}}
return|return
name|function_pointers
operator|>
name|function_pointers
return|;
comment|// expected-warning {{ordered comparison of function pointers}}
return|return
name|a
operator|>
name|c
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
return|return
name|a
operator|==
operator|(
name|void
operator|*
operator|)
literal|0
return|;
return|return
name|a
operator|==
operator|(
name|void
operator|*
operator|)
literal|1
return|;
comment|// expected-warning {{equality comparison between function pointer and void pointer}}
block|}
end_function

begin_function
name|int
name|void_pointers
parameter_list|(
name|void
modifier|*
name|foo
parameter_list|)
block|{
return|return
name|foo
operator|==
operator|(
name|void
operator|*
operator|)
literal|0
return|;
return|return
name|foo
operator|==
operator|(
name|void
operator|*
operator|)
literal|1
return|;
block|}
end_function

begin_function
name|int
name|test1
parameter_list|(
name|int
name|i
parameter_list|)
block|{
enum|enum
name|en
block|{
name|zero
block|}
enum|;
return|return
name|i
operator|>
name|zero
return|;
block|}
end_function

end_unit

