begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -verify %s
end_comment

begin_function
name|void
name|f1
parameter_list|()
block|{
name|int
name|a
init|=
literal|1
decl_stmt|;
name|int
name|b
init|=
name|__imag
name|a
decl_stmt|;
name|int
modifier|*
name|c
init|=
operator|&
name|__real
name|a
decl_stmt|;
name|int
modifier|*
name|d
init|=
operator|&
name|__imag
name|a
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue of type 'int'}}
block|}
end_function

begin_function
name|void
name|f2
parameter_list|()
block|{
specifier|_Complex
name|int
name|a
init|=
literal|1
decl_stmt|;
name|int
name|b
init|=
name|__imag
name|a
decl_stmt|;
name|int
modifier|*
name|c
init|=
operator|&
name|__real
name|a
decl_stmt|;
name|int
modifier|*
name|d
init|=
operator|&
name|__imag
name|a
decl_stmt|;
block|}
end_function

begin_function
name|void
name|f3
parameter_list|()
block|{
name|double
name|a
init|=
literal|1
decl_stmt|;
name|double
name|b
init|=
name|__imag
name|a
decl_stmt|;
name|double
modifier|*
name|c
init|=
operator|&
name|__real
name|a
decl_stmt|;
name|double
modifier|*
name|d
init|=
operator|&
name|__imag
name|a
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue of type 'double'}}
block|}
end_function

begin_function
name|void
name|f4
parameter_list|()
block|{
specifier|_Complex
name|double
name|a
init|=
literal|1
decl_stmt|;
name|double
name|b
init|=
name|__imag
name|a
decl_stmt|;
name|double
modifier|*
name|c
init|=
operator|&
name|__real
name|a
decl_stmt|;
name|double
modifier|*
name|d
init|=
operator|&
name|__imag
name|a
decl_stmt|;
block|}
end_function

end_unit

