begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc %s -verify -fsyntax-only
end_comment

begin_function
name|void
name|a
parameter_list|()
block|{
name|__complex__
name|int
name|arr
decl_stmt|;
name|__complex__
name|short
name|brr
decl_stmt|;
name|__complex__
name|unsigned
name|xx
decl_stmt|;
name|__complex__
name|signed
name|yy
decl_stmt|;
name|__complex__
name|int
name|result
decl_stmt|;
name|int
name|ii
decl_stmt|;
name|int
name|aa
init|=
literal|1 + 1.0iF
decl_stmt|;
name|result
operator|=
name|arr
operator|*
name|ii
expr_stmt|;
name|result
operator|=
name|ii
operator|*
name|brr
expr_stmt|;
name|result
operator|=
name|arr
operator|*
name|brr
expr_stmt|;
name|result
operator|=
name|xx
operator|*
name|yy
expr_stmt|;
switch|switch
condition|(
name|arr
condition|)
block|{
comment|// expected-error{{statement requires expression of integer type ('_Complex int' invalid)}}
case|case
name|brr
case|:
empty_stmt|;
comment|// expected-error{{expression is not an integer constant expression}}
case|case
name|xx
case|:
empty_stmt|;
comment|// expected-error{{expression is not an integer constant expression}}
block|}
block|}
end_function

begin_function
name|void
name|Tester
parameter_list|()
block|{
name|__complex
name|short
name|a1
decl_stmt|;
name|__complex
name|int
name|a2
decl_stmt|;
name|__complex
name|float
name|a3
decl_stmt|;
name|__complex
name|double
name|a4
decl_stmt|;
name|short
name|a5
decl_stmt|;
name|int
name|a6
decl_stmt|;
name|float
name|a7
decl_stmt|;
name|double
name|a8
decl_stmt|;
define|#
directive|define
name|TestPair
parameter_list|(
name|m
parameter_list|,
name|n
parameter_list|)
value|int x##m##n = a##m+a##n;
define|#
directive|define
name|TestPairs
parameter_list|(
name|m
parameter_list|)
value|TestPair(m,1) TestPair(m,2) \                     TestPair(m,3) TestPair(m,4) \                     TestPair(m,5) TestPair(m,6) \                     TestPair(m,7) TestPair(m,8)
name|TestPairs
argument_list|(
literal|1
argument_list|)
expr_stmt|;
name|TestPairs
argument_list|(
literal|2
argument_list|)
expr_stmt|;
name|TestPairs
argument_list|(
literal|3
argument_list|)
expr_stmt|;
name|TestPairs
argument_list|(
literal|4
argument_list|)
expr_stmt|;
name|TestPairs
argument_list|(
literal|5
argument_list|)
expr_stmt|;
name|TestPairs
argument_list|(
literal|6
argument_list|)
expr_stmt|;
name|TestPairs
argument_list|(
literal|7
argument_list|)
expr_stmt|;
name|TestPairs
argument_list|(
literal|8
argument_list|)
expr_stmt|;
block|}
end_function

begin_comment
comment|// rdar://6097730
end_comment

begin_function
name|void
name|test3
parameter_list|(
specifier|_Complex
name|int
modifier|*
name|x
parameter_list|)
block|{
operator|*
name|x
operator|=
operator|~
operator|*
name|x
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test4
parameter_list|(
specifier|_Complex
name|float
modifier|*
name|x
parameter_list|)
block|{
operator|*
name|x
operator|=
operator|~
operator|*
name|x
expr_stmt|;
block|}
end_function

end_unit

