begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-apple-darwin -fsyntax-only -pedantic -verify -Wsign-compare %s -Wno-unreachable-code
end_comment

begin_function
name|int
name|test
parameter_list|(
name|char
modifier|*
name|C
parameter_list|)
block|{
comment|// nothing here should warn.
return|return
name|C
operator|!=
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
return|;
return|return
name|C
operator|!=
operator|(
name|void
operator|*
operator|)
literal|0
return|;
return|return
name|C
operator|!=
literal|0
return|;
return|return
name|C
operator|!=
literal|1
return|;
comment|// expected-warning {{comparison between pointer and integer ('char *' and 'int')}}
block|}
end_function

begin_function
name|int
name|ints
parameter_list|(
name|long
name|a
parameter_list|,
name|unsigned
name|long
name|b
parameter_list|)
block|{
enum|enum
name|EnumA
block|{
name|A
block|}
enum|;
enum|enum
name|EnumB
block|{
name|B
block|}
enum|;
enum|enum
name|EnumC
block|{
name|C
init|=
literal|0x10000
block|}
enum|;
return|return
comment|// (a,b)
operator|(
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|==
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|==
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|long
operator|)
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|long
operator|)
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// (A,b)
operator|(
name|A
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|A
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|A
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|A
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|A
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|A
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|A
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|A
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|A
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
name|A
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|A
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|A
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|A
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|A
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|A
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|A
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|A
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|A
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// (a,B)
operator|(
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
name|B
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
name|B
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
name|B
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
name|B
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|B
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|B
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
name|B
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|<
name|B
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|<
name|B
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
name|B
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|B
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|B
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|B
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|B
operator|)
operator|+
comment|// (C,b)
operator|(
name|C
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of constant 'C' (65536) with expression of type 'unsigned short' is always false}}
operator|(
name|C
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of constant 'C' (65536) with expression of type 'unsigned char' is always false}}
operator|(
operator|(
name|long
operator|)
name|C
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|C
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|C
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|C
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|C
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|C
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|C
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|C
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
name|C
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of constant 'C' (65536) with expression of type 'unsigned short' is always false}}
operator|(
name|C
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of constant 'C' (65536) with expression of type 'unsigned char' is always false}}
operator|(
operator|(
name|long
operator|)
name|C
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|C
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|C
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|C
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|C
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|C
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|C
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|C
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// (a,C)
operator|(
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
name|C
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
name|C
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of constant 'C' (65536) with expression of type 'short' is always false}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of constant 'C' (65536) with expression of type 'signed char' is always false}}
operator|(
operator|(
name|long
operator|)
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|C
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
name|C
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|<
name|C
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|<
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of constant 'C' (65536) with expression of type 'short' is always true}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of constant 'C' (65536) with expression of type 'signed char' is always true}}
operator|(
operator|(
name|long
operator|)
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
name|C
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
name|C
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
name|C
operator|)
operator|+
comment|// (0x80000,b)
operator|(
literal|0x80000
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of constant 524288 with expression of type 'unsigned short' is always false}}
operator|(
literal|0x80000
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of constant 524288 with expression of type 'unsigned char' is always false}}
operator|(
operator|(
name|long
operator|)
literal|0x80000
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
literal|0x80000
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
literal|0x80000
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
literal|0x80000
operator|==
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
literal|0x80000
operator|==
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
literal|0x80000
operator|==
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
literal|0x80000
operator|==
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
literal|0x80000
operator|==
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
literal|0x80000
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of constant 524288 with expression of type 'unsigned short' is always false}}
operator|(
literal|0x80000
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// expected-warning {{comparison of constant 524288 with expression of type 'unsigned char' is always false}}
operator|(
operator|(
name|long
operator|)
literal|0x80000
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
literal|0x80000
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
literal|0x80000
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
literal|0x80000
operator|<
name|b
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
literal|0x80000
operator|<
operator|(
name|unsigned
name|long
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
literal|0x80000
operator|<
operator|(
name|unsigned
name|int
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
literal|0x80000
operator|<
operator|(
name|unsigned
name|short
operator|)
name|b
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
literal|0x80000
operator|<
operator|(
name|unsigned
name|char
operator|)
name|b
operator|)
operator|+
comment|// (a,0x80000)
operator|(
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|==
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of constant 524288 with expression of type 'short' is always false}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of constant 524288 with expression of type 'signed char' is always false}}
operator|(
operator|(
name|long
operator|)
name|a
operator|==
operator|(
name|unsigned
name|long
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|==
operator|(
name|unsigned
name|int
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|==
operator|(
name|unsigned
name|short
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|==
operator|(
name|unsigned
name|char
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
literal|0x80000
operator|)
operator|+
operator|(
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|long
operator|)
name|a
operator|<
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|int
operator|)
name|a
operator|<
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|short
operator|)
name|a
operator|<
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of constant 524288 with expression of type 'short' is always true}}
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of constant 524288 with expression of type 'signed char' is always true}}
operator|(
operator|(
name|long
operator|)
name|a
operator|<
operator|(
name|unsigned
name|long
operator|)
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|int
operator|)
name|a
operator|<
operator|(
name|unsigned
name|int
operator|)
literal|0x80000
operator|)
operator|+
comment|// expected-warning {{comparison of integers of different signs}}
operator|(
operator|(
name|short
operator|)
name|a
operator|<
operator|(
name|unsigned
name|short
operator|)
literal|0x80000
operator|)
operator|+
operator|(
operator|(
name|signed
name|char
operator|)
name|a
operator|<
operator|(
name|unsigned
name|char
operator|)
literal|0x80000
operator|)
operator|+
comment|// We should be able to avoid warning about this.
operator|(
name|b
operator|!=
operator|(
name|a
operator|<
literal|4
condition|?
literal|1
else|:
literal|2
operator|)
operator|)
operator|+
literal|10
return|;
block|}
end_function

begin_function
name|int
name|equal
parameter_list|(
name|char
modifier|*
name|a
parameter_list|,
specifier|const
name|char
modifier|*
name|b
parameter_list|)
block|{
return|return
name|a
operator|==
name|b
return|;
block|}
end_function

begin_decl_stmt
name|int
name|arrays
argument_list|(
name|char
argument_list|(
operator|*
name|a
argument_list|)
index|[
literal|5
index|]
argument_list|,
name|char
argument_list|(
operator|*
name|b
argument_list|)
index|[
literal|10
index|]
argument_list|,
name|char
argument_list|(
operator|*
name|c
argument_list|)
index|[
literal|5
index|]
argument_list|)
block|{
name|int
name|d
init|=
operator|(
name|a
operator|==
name|c
operator|)
decl_stmt|;
return|return
name|a
operator|==
name|b
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
block|}
end_decl_stmt

begin_function
name|int
name|pointers
parameter_list|(
name|int
modifier|*
name|a
parameter_list|)
block|{
return|return
name|a
operator|>
literal|0
return|;
comment|// expected-warning {{ordered comparison between pointer and zero ('int *' and 'int') is an extension}}
return|return
name|a
operator|>
literal|42
return|;
comment|// expected-warning {{ordered comparison between pointer and integer ('int *' and 'int')}}
return|return
name|a
operator|>
operator|(
name|void
operator|*
operator|)
literal|0
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
block|}
end_function

begin_function
name|int
name|function_pointers
parameter_list|(
name|int
function_decl|(
modifier|*
name|a
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|,
name|int
function_decl|(
modifier|*
name|b
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|,
name|void
function_decl|(
modifier|*
name|c
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|)
block|{
return|return
name|a
operator|>
name|b
return|;
comment|// expected-warning {{ordered comparison of function pointers}}
return|return
name|function_pointers
operator|>
name|function_pointers
return|;
comment|// expected-warning {{self-comparison always evaluates to false}} expected-warning{{ordered comparison of function pointers}}
return|return
name|a
operator|>
name|c
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
return|return
name|a
operator|==
operator|(
name|void
operator|*
operator|)
literal|0
return|;
return|return
name|a
operator|==
operator|(
name|void
operator|*
operator|)
literal|1
return|;
comment|// expected-warning {{equality comparison between function pointer and void pointer}}
block|}
end_function

begin_function
name|int
name|void_pointers
parameter_list|(
name|void
modifier|*
name|foo
parameter_list|)
block|{
return|return
name|foo
operator|==
operator|(
name|void
operator|*
operator|)
literal|0
return|;
return|return
name|foo
operator|==
operator|(
name|void
operator|*
operator|)
literal|1
return|;
block|}
end_function

begin_function
name|int
name|test1
parameter_list|(
name|int
name|i
parameter_list|)
block|{
enum|enum
name|en
block|{
name|zero
block|}
enum|;
return|return
name|i
operator|>
name|zero
return|;
block|}
end_function

begin_comment
comment|// PR5937
end_comment

begin_function
name|int
name|test2
parameter_list|(
name|int
name|i32
parameter_list|)
block|{
struct|struct
name|foo
block|{
name|unsigned
name|int
name|u8
range|:
literal|8
decl_stmt|;
name|unsigned
name|long
name|long
name|u31
range|:
literal|31
decl_stmt|;
name|unsigned
name|long
name|long
name|u32
range|:
literal|32
decl_stmt|;
name|unsigned
name|long
name|long
name|u63
range|:
literal|63
decl_stmt|;
name|unsigned
name|long
name|long
name|u64
range|:
literal|64
decl_stmt|;
block|}
modifier|*
name|x
struct|;
if|if
condition|(
name|x
operator|->
name|u8
operator|==
name|i32
condition|)
block|{
comment|// comparison in int32, exact
return|return
literal|0
return|;
block|}
elseif|else
if|if
condition|(
name|x
operator|->
name|u31
operator|==
name|i32
condition|)
block|{
comment|// comparison in int32, exact
return|return
literal|1
return|;
block|}
elseif|else
if|if
condition|(
name|x
operator|->
name|u32
operator|==
name|i32
condition|)
block|{
comment|// expected-warning {{comparison of integers of different signs}}
return|return
literal|2
return|;
block|}
elseif|else
if|if
condition|(
name|x
operator|->
name|u63
operator|==
name|i32
condition|)
block|{
comment|// comparison in uint64, exact because ==
return|return
literal|3
return|;
block|}
elseif|else
if|if
condition|(
name|x
operator|->
name|u64
operator|==
name|i32
condition|)
block|{
comment|// expected-warning {{comparison of integers of different signs}}
return|return
literal|4
return|;
block|}
else|else
block|{
return|return
literal|5
return|;
block|}
block|}
end_function

begin_comment
comment|// PR5887
end_comment

begin_function
name|void
name|test3
parameter_list|()
block|{
name|unsigned
name|short
name|x
decl_stmt|,
name|y
decl_stmt|;
name|unsigned
name|int
name|z
decl_stmt|;
if|if
condition|(
operator|(
name|x
operator|>
name|y
condition|?
name|x
else|:
name|y
operator|)
operator|>
name|z
condition|)
operator|(
name|void
operator|)
literal|0
expr_stmt|;
block|}
end_function

begin_comment
comment|// PR5961
end_comment

begin_decl_stmt
specifier|extern
name|char
modifier|*
name|ptr4
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|test4
parameter_list|()
block|{
name|long
name|value
decl_stmt|;
if|if
condition|(
name|value
operator|<
operator|(
name|unsigned
name|long
operator|)
operator|&
name|ptr4
condition|)
comment|// expected-warning {{comparison of integers of different signs}}
return|return;
block|}
end_function

begin_comment
comment|// PR4807
end_comment

begin_function
name|int
name|test5
parameter_list|(
name|unsigned
name|int
name|x
parameter_list|)
block|{
return|return
operator|(
name|x
operator|<
literal|0
operator|)
comment|// expected-warning {{comparison of unsigned expression< 0 is always false}}
operator|&&
operator|(
literal|0
operator|>
name|x
operator|)
comment|// expected-warning {{comparison of 0> unsigned expression is always false}}
operator|&&
operator|(
name|x
operator|>=
literal|0
operator|)
comment|// expected-warning {{comparison of unsigned expression>= 0 is always true}}
operator|&&
operator|(
literal|0
operator|<=
name|x
operator|)
return|;
comment|// expected-warning {{comparison of 0<= unsigned expression is always true}}
block|}
end_function

begin_function
name|int
name|test6
parameter_list|(
name|unsigned
name|i
parameter_list|,
name|unsigned
name|power
parameter_list|)
block|{
name|unsigned
name|x
init|=
operator|(
name|i
operator|<
operator|(
literal|1
operator|<<
name|power
operator|)
condition|?
name|i
else|:
literal|0
operator|)
decl_stmt|;
return|return
name|x
operator|!=
literal|3
condition|?
literal|1
operator|<<
name|power
else|:
name|i
return|;
block|}
end_function

begin_comment
comment|//<rdar://problem/8414119> enum>= (enum)0 comparison should not generate any warnings
end_comment

begin_enum
enum|enum
name|rdar8414119_Vals
block|{
name|X
block|,
name|Y
block|,
name|Z
block|}
enum|;
end_enum

begin_define
define|#
directive|define
name|ZERO
value|0
end_define

begin_define
define|#
directive|define
name|CHECK
parameter_list|(
name|x
parameter_list|)
value|(x>= X)
end_define

begin_function
name|void
name|rdar8414119_foo
parameter_list|(
name|enum
name|rdar8414119_Vals
name|v
parameter_list|)
block|{
if|if
condition|(
name|CHECK
argument_list|(
name|v
argument_list|)
condition|)
comment|// no-warning
return|return;
if|if
condition|(
name|v
operator|>=
name|X
condition|)
comment|// no-warning
return|return;
block|}
end_function

begin_function
name|int
name|rdar8414119_bar
parameter_list|(
name|unsigned
name|x
parameter_list|)
block|{
return|return
name|x
operator|>=
name|ZERO
return|;
comment|// no-warning
block|}
end_function

begin_undef
undef|#
directive|undef
name|ZERO
end_undef

begin_undef
undef|#
directive|undef
name|CHECK
end_undef

begin_function
name|int
name|rdar8511238
parameter_list|()
block|{
enum|enum
name|A
block|{
name|A_foo
block|,
name|A_bar
block|}
enum|;
name|enum
name|A
name|a
decl_stmt|;
if|if
condition|(
name|a
operator|<
literal|0
condition|)
comment|// expected-warning {{comparison of unsigned enum expression< 0 is always false}}
return|return
literal|0
return|;
return|return
literal|20
return|;
block|}
end_function

begin_comment
comment|// PR10336
end_comment

begin_function
name|int
name|test9
parameter_list|(
name|int
name|sv
parameter_list|,
name|unsigned
name|uv
parameter_list|,
name|long
name|slv
parameter_list|)
block|{
return|return
name|sv
operator|==
operator|(
name|uv
operator|^=
name|slv
operator|)
return|;
comment|// expected-warning {{comparison of integers of different signs: 'int' and 'unsigned int'}}
block|}
end_function

begin_function
name|void
name|test10
parameter_list|(
name|void
parameter_list|)
block|{
name|int
name|si
decl_stmt|;
name|unsigned
name|int
name|ui
decl_stmt|;
name|long
name|sl
decl_stmt|;
name|_Bool
name|b
decl_stmt|;
name|b
operator|=
operator|(
name|si
operator|==
operator|(
name|ui
operator|=
name|sl
operator|)
operator|)
expr_stmt|;
comment|// expected-warning {{comparison of integers of different signs: 'int' and 'unsigned int'}}
name|b
operator|=
operator|(
name|si
operator|==
operator|(
name|ui
operator|=
name|sl
operator|&
literal|15
operator|)
operator|)
expr_stmt|;
block|}
end_function

begin_comment
comment|// PR11572
end_comment

begin_struct
struct|struct
name|test11S
block|{
name|unsigned
name|x
range|:
literal|30
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
name|test11
parameter_list|(
name|unsigned
name|y
parameter_list|,
name|struct
name|test11S
modifier|*
name|p
parameter_list|)
block|{
return|return
name|y
operator|>
operator|(
name|p
operator|->
name|x
operator|>>
literal|24
operator|)
return|;
comment|// no-warning
block|}
end_function

begin_typedef
typedef|typedef
name|char
name|one_char
index|[
literal|1
index|]
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|char
name|two_chars
index|[
literal|2
index|]
typedef|;
end_typedef

begin_function
name|void
name|test12
parameter_list|(
name|unsigned
name|a
parameter_list|)
block|{
if|if
condition|(
literal|0
operator|&&
operator|-
literal|1
operator|>
name|a
condition|)
block|{ }
block|}
end_function

end_unit

