begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -fsyntax-only -pedantic -verify %s
end_comment

begin_function
name|int
name|test
parameter_list|(
name|char
modifier|*
name|C
parameter_list|)
block|{
comment|// nothing here should warn.
return|return
name|C
operator|!=
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
return|;
return|return
name|C
operator|!=
operator|(
name|void
operator|*
operator|)
literal|0
return|;
return|return
name|C
operator|!=
literal|0
return|;
block|}
end_function

begin_function
name|int
name|equal
parameter_list|(
name|char
modifier|*
name|a
parameter_list|,
specifier|const
name|char
modifier|*
name|b
parameter_list|)
block|{
return|return
name|a
operator|==
name|b
return|;
block|}
end_function

begin_decl_stmt
name|int
name|arrays
argument_list|(
name|char
argument_list|(
operator|*
name|a
argument_list|)
index|[
literal|5
index|]
argument_list|,
name|char
argument_list|(
operator|*
name|b
argument_list|)
index|[
literal|10
index|]
argument_list|,
name|char
argument_list|(
operator|*
name|c
argument_list|)
index|[
literal|5
index|]
argument_list|)
block|{
name|int
name|d
init|=
operator|(
name|a
operator|==
name|c
operator|)
decl_stmt|;
return|return
name|a
operator|==
name|b
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
block|}
end_decl_stmt

begin_function
name|int
name|pointers
parameter_list|(
name|int
modifier|*
name|a
parameter_list|)
block|{
return|return
name|a
operator|>
literal|0
return|;
comment|// expected-warning {{ordered comparison between pointer and integer}}
return|return
name|a
operator|>
operator|(
name|void
operator|*
operator|)
literal|0
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
block|}
end_function

begin_function
name|int
name|function_pointers
parameter_list|(
name|int
function_decl|(
modifier|*
name|a
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|,
name|int
function_decl|(
modifier|*
name|b
function_decl|)
parameter_list|(
name|int
parameter_list|)
parameter_list|)
block|{
return|return
name|a
operator|>
name|b
return|;
comment|// expected-warning {{ordered comparison of function pointers}}
return|return
name|function_pointers
operator|>
name|function_pointers
return|;
comment|// expected-warning {{ordered comparison of function pointers}}
return|return
name|a
operator|==
operator|(
name|void
operator|*
operator|)
literal|0
return|;
return|return
name|a
operator|==
operator|(
name|void
operator|*
operator|)
literal|1
return|;
comment|// expected-warning {{comparison of distinct pointer types}}
block|}
end_function

end_unit

