begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -pedantic -Wsign-compare %s
end_comment

begin_function
name|void
name|foo
parameter_list|()
block|{
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// FIXME: GCC doesn't consider the the following two statements to be errors.
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
operator|(
name|int
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-error {{incomplete type 'void' is not assignable}}
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
operator|(
name|double
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-error {{incomplete type 'void' is not assignable}}
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|int
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-error {{incomplete type 'void' is not assignable}} expected-warning {{pointer type mismatch ('double *' and 'int *')}}
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|double
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
operator|*
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-error {{incomplete type 'void' is not assignable}}
name|double
modifier|*
name|dp
decl_stmt|;
name|int
modifier|*
name|ip
decl_stmt|;
name|void
modifier|*
name|vp
decl_stmt|;
name|dp
operator|=
name|vp
expr_stmt|;
name|vp
operator|=
name|dp
expr_stmt|;
name|ip
operator|=
name|dp
expr_stmt|;
comment|// expected-warning {{incompatible pointer types assigning 'double *', expected 'int *'}}
name|dp
operator|=
name|ip
expr_stmt|;
comment|// expected-warning {{incompatible pointer types assigning 'int *', expected 'double *'}}
name|dp
operator|=
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
literal|0
expr_stmt|;
name|vp
operator|=
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
literal|0
expr_stmt|;
name|ip
operator|=
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
literal|0
expr_stmt|;
comment|// expected-warning {{incompatible pointer types assigning 'double *', expected 'int *'}}
specifier|const
name|int
modifier|*
name|cip
decl_stmt|;
name|vp
operator|=
operator|(
literal|0
condition|?
name|vp
else|:
name|cip
operator|)
expr_stmt|;
comment|// expected-warning {{discards qualifiers}}
name|vp
operator|=
operator|(
literal|0
condition|?
name|cip
else|:
name|vp
operator|)
expr_stmt|;
comment|// expected-warning {{discards qualifiers}}
name|int
name|i
init|=
literal|2
decl_stmt|;
name|int
argument_list|(
operator|*
name|pf
argument_list|)
index|[
literal|2
index|]
expr_stmt|;
name|int
argument_list|(
operator|*
name|pv
argument_list|)
index|[
name|i
index|]
expr_stmt|;
name|pf
operator|=
operator|(
name|i
condition|?
name|pf
else|:
name|pv
operator|)
expr_stmt|;
enum|enum
block|{
name|xxx
block|,
name|yyy
block|,
name|zzz
block|}
name|e
enum|,
operator|*
name|ee
expr_stmt|;
name|short
name|x
decl_stmt|;
name|ee
operator|=
name|ee
condition|?
operator|&
name|x
else|:
name|ee
condition|?
operator|&
name|i
else|:
operator|&
name|e
expr_stmt|;
comment|// expected-warning {{pointer type mismatch}}
typedef|typedef
name|void
modifier|*
name|asdf
typedef|;
operator|*
operator|(
literal|0
condition|?
operator|(
name|asdf
operator|)
literal|0
else|:
operator|&
name|x
operator|)
operator|=
literal|10
expr_stmt|;
name|unsigned
name|long
name|test0
init|=
literal|5
decl_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|long
operator|)
name|test0
else|:
name|test0
expr_stmt|;
comment|// expected-warning {{operands of ? are integers of different signs}}
name|test0
operator|=
name|test0
condition|?
operator|(
name|int
operator|)
name|test0
else|:
name|test0
expr_stmt|;
comment|// expected-warning {{operands of ? are integers of different signs}}
name|test0
operator|=
name|test0
condition|?
operator|(
name|short
operator|)
name|test0
else|:
name|test0
expr_stmt|;
comment|// expected-warning {{operands of ? are integers of different signs}}
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|long
operator|)
name|test0
expr_stmt|;
comment|// expected-warning {{operands of ? are integers of different signs}}
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|int
operator|)
name|test0
expr_stmt|;
comment|// expected-warning {{operands of ? are integers of different signs}}
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|short
operator|)
name|test0
expr_stmt|;
comment|// expected-warning {{operands of ? are integers of different signs}}
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|long
operator|)
literal|10
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|int
operator|)
literal|10
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|short
operator|)
literal|10
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|long
operator|)
literal|10
else|:
name|test0
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|int
operator|)
literal|10
else|:
name|test0
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|short
operator|)
literal|10
else|:
name|test0
expr_stmt|;
enum|enum
name|Enum
block|{
name|EVal
block|}
enum|;
name|test0
operator|=
name|test0
condition|?
name|EVal
else|:
name|test0
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
name|EVal
else|:
operator|(
name|int
operator|)
name|test0
expr_stmt|;
comment|// okay: EVal is an int
name|test0
operator|=
name|test0
condition|?
operator|(
name|unsigned
operator|)
name|EVal
else|:
operator|(
name|int
operator|)
name|test0
expr_stmt|;
comment|// expected-warning {{operands of ? are integers of different signs}}
block|}
end_function

begin_function
name|int
name|Postgresql
parameter_list|()
block|{
name|char
name|x
decl_stmt|;
return|return
operator|(
operator|(
operator|(
operator|(
operator|&
name|x
operator|)
operator|!=
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|)
condition|?
operator|(
operator|*
operator|(
operator|&
name|x
operator|)
operator|=
operator|(
operator|(
name|char
operator|)
literal|1
operator|)
operator|)
else|:
call|(
name|void
call|)
argument_list|(
operator|(
name|void
operator|*
operator|)
literal|0
argument_list|)
operator|)
operator|,
call|(
name|unsigned
name|long
call|)
argument_list|(
operator|(
name|void
operator|*
operator|)
literal|0
argument_list|)
operator|)
return|;
comment|// expected-warning {{C99 forbids conditional expressions with only one void side}}
block|}
end_function

begin_define
define|#
directive|define
name|nil
value|((void*) 0)
end_define

begin_function_decl
specifier|extern
name|int
name|f1
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|int
name|f0
parameter_list|(
name|int
name|a
parameter_list|)
block|{
comment|// GCC considers this a warning.
return|return
name|a
condition|?
name|f1
argument_list|()
else|:
name|nil
return|;
comment|// expected-warning {{pointer/integer type mismatch in conditional expression ('int' and 'void *')}} expected-warning {{incompatible pointer to integer conversion returning 'void *', expected 'int'}}
block|}
end_function

end_unit

