begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -pedantic -Wsign-conversion %s
end_comment

begin_function
name|void
name|foo
parameter_list|()
block|{
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// FIXME: GCC doesn't consider the following two statements to be errors.
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
operator|(
name|int
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-error {{incomplete type 'void' is not assignable}}
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
operator|(
name|double
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-error {{incomplete type 'void' is not assignable}}
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|int
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-error {{incomplete type 'void' is not assignable}} expected-warning {{pointer type mismatch ('double *' and 'int *')}}
operator|*
operator|(
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|double
operator|*
operator|)
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
operator|*
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-error {{incomplete type 'void' is not assignable}}
name|double
modifier|*
name|dp
decl_stmt|;
name|int
modifier|*
name|ip
decl_stmt|;
name|void
modifier|*
name|vp
decl_stmt|;
name|dp
operator|=
name|vp
expr_stmt|;
name|vp
operator|=
name|dp
expr_stmt|;
name|ip
operator|=
name|dp
expr_stmt|;
comment|// expected-warning {{incompatible pointer types assigning to 'int *' from 'double *'}}
name|dp
operator|=
name|ip
expr_stmt|;
comment|// expected-warning {{incompatible pointer types assigning to 'double *' from 'int *'}}
name|dp
operator|=
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
literal|0
expr_stmt|;
name|vp
operator|=
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
literal|0
expr_stmt|;
name|ip
operator|=
literal|0
condition|?
operator|(
name|double
operator|*
operator|)
literal|0
else|:
operator|(
name|void
operator|*
operator|)
literal|0
expr_stmt|;
comment|// expected-warning {{incompatible pointer types assigning to 'int *' from 'double *'}}
specifier|const
name|int
modifier|*
name|cip
decl_stmt|;
name|vp
operator|=
operator|(
literal|0
condition|?
name|vp
else|:
name|cip
operator|)
expr_stmt|;
comment|// expected-warning {{discards qualifiers}}
name|vp
operator|=
operator|(
literal|0
condition|?
name|cip
else|:
name|vp
operator|)
expr_stmt|;
comment|// expected-warning {{discards qualifiers}}
name|int
name|i
init|=
literal|2
decl_stmt|;
name|int
argument_list|(
operator|*
name|pf
argument_list|)
index|[
literal|2
index|]
expr_stmt|;
name|int
argument_list|(
operator|*
name|pv
argument_list|)
index|[
name|i
index|]
expr_stmt|;
name|pf
operator|=
operator|(
name|i
condition|?
name|pf
else|:
name|pv
operator|)
expr_stmt|;
enum|enum
block|{
name|xxx
block|,
name|yyy
block|,
name|zzz
block|}
name|e
enum|,
operator|*
name|ee
expr_stmt|;
name|short
name|x
decl_stmt|;
name|ee
operator|=
name|ee
condition|?
operator|&
name|x
else|:
name|ee
condition|?
operator|&
name|i
else|:
operator|&
name|e
expr_stmt|;
comment|// expected-warning {{pointer type mismatch}}
typedef|typedef
name|void
modifier|*
name|asdf
typedef|;
operator|*
operator|(
literal|0
condition|?
operator|(
name|asdf
operator|)
literal|0
else|:
operator|&
name|x
operator|)
operator|=
literal|10
expr_stmt|;
name|unsigned
name|long
name|test0
init|=
literal|5
decl_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|long
operator|)
name|test0
else|:
name|test0
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'long' to 'unsigned long'}}
name|test0
operator|=
name|test0
condition|?
operator|(
name|int
operator|)
name|test0
else|:
name|test0
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'int' to 'unsigned long'}}
name|test0
operator|=
name|test0
condition|?
operator|(
name|short
operator|)
name|test0
else|:
name|test0
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'short' to 'unsigned long'}}
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|long
operator|)
name|test0
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'long' to 'unsigned long'}}
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|int
operator|)
name|test0
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'int' to 'unsigned long'}}
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|short
operator|)
name|test0
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'short' to 'unsigned long'}}
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|long
operator|)
literal|10
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|int
operator|)
literal|10
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
name|test0
else|:
operator|(
name|short
operator|)
literal|10
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|long
operator|)
literal|10
else|:
name|test0
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|int
operator|)
literal|10
else|:
name|test0
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|short
operator|)
literal|10
else|:
name|test0
expr_stmt|;
name|int
name|test1
decl_stmt|;
enum|enum
name|Enum
block|{
name|EVal
block|}
enum|;
name|test0
operator|=
name|test0
condition|?
name|EVal
else|:
name|test0
expr_stmt|;
name|test1
operator|=
name|test0
condition|?
name|EVal
else|:
operator|(
name|int
operator|)
name|test0
expr_stmt|;
name|test0
operator|=
name|test0
condition|?
operator|(
name|unsigned
operator|)
name|EVal
else|:
operator|(
name|int
operator|)
name|test0
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'int' to 'unsigned long'}}
name|test0
operator|=
name|test0
condition|?
name|EVal
else|:
name|test1
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'int' to 'unsigned long'}}
name|test0
operator|=
name|test0
condition|?
name|test1
else|:
name|EVal
expr_stmt|;
comment|// expected-warning {{operand of ? changes signedness: 'int' to 'unsigned long'}}
specifier|const
name|int
modifier|*
name|const_int
decl_stmt|;
name|int
modifier|*
name|nonconst_int
decl_stmt|;
operator|*
operator|(
name|test0
condition|?
name|const_int
else|:
name|nonconst_int
operator|)
operator|=
literal|42
expr_stmt|;
comment|// expected-error {{read-only variable is not assignable}}
operator|*
operator|(
name|test0
condition|?
name|nonconst_int
else|:
name|const_int
operator|)
operator|=
literal|42
expr_stmt|;
comment|// expected-error {{read-only variable is not assignable}}
comment|// The composite type here should be "int (*)[12]", fine for the sizeof
name|int
argument_list|(
operator|*
name|incomplete
argument_list|)
index|[]
expr_stmt|;
name|int
argument_list|(
operator|*
name|complete
argument_list|)
index|[
literal|12
index|]
expr_stmt|;
sizeof|sizeof
argument_list|(
operator|*
operator|(
name|test0
condition|?
name|incomplete
else|:
name|complete
operator|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{expression result unused}}
sizeof|sizeof
argument_list|(
operator|*
operator|(
name|test0
condition|?
name|complete
else|:
name|incomplete
operator|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{expression result unused}}
name|int
name|__attribute__
argument_list|(
operator|(
name|address_space
argument_list|(
literal|2
argument_list|)
operator|)
argument_list|)
modifier|*
name|adr2
decl_stmt|;
name|int
name|__attribute__
argument_list|(
operator|(
name|address_space
argument_list|(
literal|3
argument_list|)
operator|)
argument_list|)
modifier|*
name|adr3
decl_stmt|;
name|test0
condition|?
name|adr2
else|:
name|adr3
expr_stmt|;
comment|// expected-warning {{pointer type mismatch}} expected-warning {{expression result unused}}
comment|// Make sure address-space mask ends up in the result type
operator|(
name|test0
condition|?
operator|(
name|test0
condition|?
name|adr2
else|:
name|adr2
operator|)
else|:
name|nonconst_int
operator|)
expr_stmt|;
comment|// expected-warning {{pointer type mismatch}} expected-warning {{expression result unused}}
block|}
end_function

begin_function
name|int
name|Postgresql
parameter_list|()
block|{
name|char
name|x
decl_stmt|;
return|return
operator|(
operator|(
operator|(
operator|(
operator|&
name|x
operator|)
operator|!=
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
operator|)
condition|?
operator|(
operator|*
operator|(
operator|&
name|x
operator|)
operator|=
operator|(
operator|(
name|char
operator|)
literal|1
operator|)
operator|)
else|:
call|(
name|void
call|)
argument_list|(
operator|(
name|void
operator|*
operator|)
literal|0
argument_list|)
operator|)
operator|,
call|(
name|unsigned
name|long
call|)
argument_list|(
operator|(
name|void
operator|*
operator|)
literal|0
argument_list|)
operator|)
return|;
comment|// expected-warning {{C99 forbids conditional expressions with only one void side}}
block|}
end_function

begin_define
define|#
directive|define
name|nil
value|((void*) 0)
end_define

begin_function_decl
specifier|extern
name|int
name|f1
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|int
name|f0
parameter_list|(
name|int
name|a
parameter_list|)
block|{
comment|// GCC considers this a warning.
return|return
name|a
condition|?
name|f1
argument_list|()
else|:
name|nil
return|;
comment|// expected-warning {{pointer/integer type mismatch in conditional expression ('int' and 'void *')}} expected-warning {{incompatible pointer to integer conversion returning 'void *' from a function with result type 'int'}}
block|}
end_function

begin_function
name|int
name|f2
parameter_list|(
name|int
name|x
parameter_list|)
block|{
comment|// We can suppress this because the immediate context wants an int.
return|return
operator|(
name|x
operator|!=
literal|0
operator|)
condition|?
literal|0U
else|:
name|x
return|;
block|}
end_function

begin_define
define|#
directive|define
name|NULL
value|(void*)0
end_define

begin_function
name|void
name|PR9236
parameter_list|()
block|{
struct|struct
name|A
block|{
name|int
name|i
decl_stmt|;
block|}
name|A1
struct|;
call|(
name|void
call|)
argument_list|(
literal|1
condition|?
name|A1
else|:
name|NULL
argument_list|)
expr_stmt|;
comment|// expected-error{{non-pointer operand type 'struct A' incompatible with NULL}}
call|(
name|void
call|)
argument_list|(
literal|1
condition|?
name|NULL
else|:
name|A1
argument_list|)
expr_stmt|;
comment|// expected-error{{non-pointer operand type 'struct A' incompatible with NULL}}
call|(
name|void
call|)
argument_list|(
literal|1
condition|?
literal|0
else|:
name|A1
argument_list|)
expr_stmt|;
comment|// expected-error{{incompatible operand types}}
call|(
name|void
call|)
argument_list|(
literal|1
condition|?
operator|(
name|void
operator|*
operator|)
literal|0
else|:
name|A1
argument_list|)
expr_stmt|;
comment|// expected-error{{incompatible operand types}}
call|(
name|void
call|)
argument_list|(
literal|1
condition|?
name|A1
else|:
operator|(
name|void
operator|*
operator|)
literal|0
argument_list|)
expr_stmt|;
comment|// expected-error{{incompatible operand types}}
call|(
name|void
call|)
argument_list|(
literal|1
condition|?
name|A1
else|:
operator|(
name|NULL
operator|)
argument_list|)
expr_stmt|;
comment|// expected-error{{non-pointer operand type 'struct A' incompatible with NULL}}
block|}
end_function

end_unit

