begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -triple x86_64-apple-darwin %s
end_comment

begin_comment
comment|// This file tests -Wconstant-conversion, a subcategory of -Wconversion
end_comment

begin_comment
comment|// which is on by default.
end_comment

begin_comment
comment|// rdar://problem/6792488
end_comment

begin_function
name|void
name|test_6792488
parameter_list|(
name|void
parameter_list|)
block|{
name|int
name|x
init|=
literal|0x3ff0000000000000U
decl_stmt|;
comment|// expected-warning {{implicit conversion from 'unsigned long' to 'int' changes value from 4607182418800017408 to 0}}
block|}
end_function

begin_function
name|void
name|test_7809123
parameter_list|(
name|void
parameter_list|)
block|{
struct|struct
block|{
name|int
name|i5
range|:
literal|5
decl_stmt|;
block|}
name|a
struct|;
name|a
operator|.
name|i5
operator|=
literal|36
expr_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from 36 to 4}}
block|}
end_function

begin_function
name|void
name|test
parameter_list|()
block|{
struct|struct
block|{
name|int
name|bit
range|:
literal|1
decl_stmt|;
block|}
name|a
struct|;
name|a
operator|.
name|bit
operator|=
literal|1
expr_stmt|;
comment|// shouldn't warn
block|}
end_function

begin_enum
enum|enum
name|Test2
block|{
name|K_zero
block|,
name|K_one
block|}
enum|;
end_enum

begin_function
name|enum
name|Test2
name|test2
parameter_list|(
name|enum
name|Test2
modifier|*
name|t
parameter_list|)
block|{
operator|*
name|t
operator|=
literal|20
expr_stmt|;
return|return
literal|10
return|;
comment|// shouldn't warn
block|}
end_function

begin_function
name|void
name|test3
parameter_list|()
block|{
struct|struct
name|A
block|{
name|unsigned
name|int
name|foo
range|:
literal|2
decl_stmt|;
name|int
name|bar
range|:
literal|2
decl_stmt|;
block|}
struct|;
name|struct
name|A
name|a
init|=
block|{
literal|0
block|,
literal|10
block|}
decl_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from 10 to -2}}
name|struct
name|A
name|b
index|[]
init|=
block|{
literal|0
block|,
literal|10
block|,
literal|0
block|,
literal|0
block|}
decl_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from 10 to -2}}
name|struct
name|A
name|c
index|[]
init|=
block|{
block|{
literal|10
block|,
literal|0
block|}
block|}
decl_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from 10 to 2}}
name|struct
name|A
name|d
init|=
operator|(
expr|struct
name|A
operator|)
block|{
literal|10
block|,
literal|0
block|}
decl_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from 10 to 2}}
name|struct
name|A
name|e
init|=
block|{
operator|.
name|foo
operator|=
literal|10
block|}
decl_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from 10 to 2}}
block|}
end_function

begin_function
name|void
name|test4
parameter_list|()
block|{
struct|struct
name|A
block|{
name|char
name|c
range|:
literal|2
decl_stmt|;
block|}
name|a
struct|;
name|a
operator|.
name|c
operator|=
literal|0x101
expr_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from 257 to 1}}
block|}
end_function

begin_function
name|void
name|test5
parameter_list|()
block|{
struct|struct
name|A
block|{
name|_Bool
name|b
range|:
literal|1
decl_stmt|;
block|}
name|a
struct|;
comment|// Don't warn about this implicit conversion to bool, or at least
comment|// don't warn about it just because it's a bit-field.
name|a
operator|.
name|b
operator|=
literal|100
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test6
parameter_list|()
block|{
comment|// Test that unreachable code doesn't trigger the truncation warning.
name|unsigned
name|char
name|x
init|=
literal|0
condition|?
literal|65535
else|:
literal|1
decl_stmt|;
comment|// no-warning
name|unsigned
name|char
name|y
init|=
literal|1
condition|?
literal|65535
else|:
literal|1
decl_stmt|;
comment|// expected-warning {{changes value}}
block|}
end_function

begin_function
name|void
name|test7
parameter_list|()
block|{
struct|struct
block|{
name|unsigned
name|int
name|twoBits1
range|:
literal|2
decl_stmt|;
name|unsigned
name|int
name|twoBits2
range|:
literal|2
decl_stmt|;
name|unsigned
name|int
name|reserved
range|:
literal|28
decl_stmt|;
block|}
name|f
struct|;
name|f
operator|.
name|twoBits1
operator|=
operator|~
literal|0
expr_stmt|;
comment|// no-warning
name|f
operator|.
name|twoBits1
operator|=
operator|~
literal|1
expr_stmt|;
comment|// no-warning
name|f
operator|.
name|twoBits2
operator|=
operator|~
literal|2
expr_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from -3 to 1}}
name|f
operator|.
name|twoBits1
operator|&=
operator|~
literal|1
expr_stmt|;
comment|// no-warning
name|f
operator|.
name|twoBits2
operator|&=
operator|~
literal|2
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_function
name|void
name|test8
parameter_list|()
block|{
enum|enum
name|E
block|{
name|A
block|,
name|B
block|,
name|C
block|}
enum|;
struct|struct
block|{
name|enum
name|E
name|x
range|:
literal|1
decl_stmt|;
block|}
name|f
struct|;
name|f
operator|.
name|x
operator|=
name|C
expr_stmt|;
comment|// expected-warning {{implicit truncation from 'int' to bit-field changes value from 2 to 0}}
block|}
end_function

begin_function
name|void
name|test9
parameter_list|()
block|{
specifier|const
name|char
name|max_char
init|=
literal|0x7F
decl_stmt|;
specifier|const
name|short
name|max_short
init|=
literal|0x7FFF
decl_stmt|;
specifier|const
name|int
name|max_int
init|=
literal|0x7FFFFFFF
decl_stmt|;
specifier|const
name|short
name|max_char_plus_one
init|=
operator|(
name|short
operator|)
name|max_char
operator|+
literal|1
decl_stmt|;
specifier|const
name|int
name|max_short_plus_one
init|=
operator|(
name|int
operator|)
name|max_short
operator|+
literal|1
decl_stmt|;
specifier|const
name|long
name|max_int_plus_one
init|=
operator|(
name|long
operator|)
name|max_int
operator|+
literal|1
decl_stmt|;
name|char
name|new_char
init|=
name|max_char_plus_one
decl_stmt|;
comment|// expected-warning {{implicit conversion from 'const short' to 'char' changes value from 128 to -128}}
name|short
name|new_short
init|=
name|max_short_plus_one
decl_stmt|;
comment|// expected-warning {{implicit conversion from 'const int' to 'short' changes value from 32768 to -32768}}
name|int
name|new_int
init|=
name|max_int_plus_one
decl_stmt|;
comment|// expected-warning {{implicit conversion from 'const long' to 'int' changes value from 2147483648 to -2147483648}}
name|char
name|hex_char
init|=
literal|0x80
decl_stmt|;
name|short
name|hex_short
init|=
literal|0x8000
decl_stmt|;
name|int
name|hex_int
init|=
literal|0x80000000
decl_stmt|;
name|char
name|oct_char
init|=
literal|0200
decl_stmt|;
name|short
name|oct_short
init|=
literal|0100000
decl_stmt|;
name|int
name|oct_int
init|=
literal|020000000000
decl_stmt|;
name|char
name|bin_char
init|=
literal|0b10000000
decl_stmt|;
name|short
name|bin_short
init|=
literal|0b1000000000000000
decl_stmt|;
name|int
name|bin_int
init|=
literal|0b10000000000000000000000000000000
decl_stmt|;
define|#
directive|define
name|CHAR_MACRO_HEX
value|0xff
name|char
name|macro_char_hex
init|=
name|CHAR_MACRO_HEX
decl_stmt|;
define|#
directive|define
name|CHAR_MACRO_DEC
value|255
name|char
name|macro_char_dec
init|=
name|CHAR_MACRO_DEC
decl_stmt|;
comment|// expected-warning {{implicit conversion from 'int' to 'char' changes value from 255 to -1}}
name|char
name|array_init
index|[]
init|=
block|{
literal|255
block|,
literal|127
block|,
literal|128
block|,
literal|129
block|,
literal|0
block|}
decl_stmt|;
block|}
end_function

begin_define
define|#
directive|define
name|A
value|1
end_define

begin_function
name|void
name|test10
parameter_list|()
block|{
struct|struct
name|S
block|{
name|unsigned
name|a
range|:
literal|4
decl_stmt|;
block|}
name|s
struct|;
name|s
operator|.
name|a
operator|=
operator|-
literal|1
expr_stmt|;
name|s
operator|.
name|a
operator|=
literal|15
expr_stmt|;
name|s
operator|.
name|a
operator|=
operator|-
literal|8
expr_stmt|;
name|s
operator|.
name|a
operator|=
operator|~
literal|0
expr_stmt|;
name|s
operator|.
name|a
operator|=
operator|~
literal|0U
expr_stmt|;
name|s
operator|.
name|a
operator|=
operator|~
operator|(
literal|1
operator|<<
name|A
operator|)
expr_stmt|;
name|s
operator|.
name|a
operator|=
operator|-
literal|9
expr_stmt|;
comment|// expected-warning{{implicit truncation from 'int' to bit-field changes value from -9 to 7}}
name|s
operator|.
name|a
operator|=
literal|16
expr_stmt|;
comment|// expected-warning{{implicit truncation from 'int' to bit-field changes value from 16 to 0}}
block|}
end_function

end_unit

