begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -Wconversion -nostdinc -isystem %S/Inputs -triple x86_64-apple-darwin %s -Wno-unreachable-code
end_comment

begin_include
include|#
directive|include
file|<conversion.h>
end_include

begin_define
define|#
directive|define
name|BIG
value|0x7f7f7f7f7f7f7f7fL
end_define

begin_function
name|void
name|test0
parameter_list|(
name|char
name|c
parameter_list|,
name|short
name|s
parameter_list|,
name|int
name|i
parameter_list|,
name|long
name|l
parameter_list|,
name|long
name|long
name|ll
parameter_list|)
block|{
name|c
operator|=
name|c
expr_stmt|;
name|c
operator|=
name|s
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|c
operator|=
name|i
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|c
operator|=
name|l
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|s
operator|=
name|c
expr_stmt|;
name|s
operator|=
name|s
expr_stmt|;
name|s
operator|=
name|i
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|s
operator|=
name|l
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|i
operator|=
name|c
expr_stmt|;
name|i
operator|=
name|s
expr_stmt|;
name|i
operator|=
name|i
expr_stmt|;
name|i
operator|=
name|l
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|l
operator|=
name|c
expr_stmt|;
name|l
operator|=
name|s
expr_stmt|;
name|l
operator|=
name|i
expr_stmt|;
name|l
operator|=
name|l
expr_stmt|;
name|c
operator|=
operator|(
name|char
operator|)
literal|0
expr_stmt|;
name|c
operator|=
operator|(
name|short
operator|)
literal|0
expr_stmt|;
name|c
operator|=
operator|(
name|int
operator|)
literal|0
expr_stmt|;
name|c
operator|=
operator|(
name|long
operator|)
literal|0
expr_stmt|;
name|s
operator|=
operator|(
name|char
operator|)
literal|0
expr_stmt|;
name|s
operator|=
operator|(
name|short
operator|)
literal|0
expr_stmt|;
name|s
operator|=
operator|(
name|int
operator|)
literal|0
expr_stmt|;
name|s
operator|=
operator|(
name|long
operator|)
literal|0
expr_stmt|;
name|i
operator|=
operator|(
name|char
operator|)
literal|0
expr_stmt|;
name|i
operator|=
operator|(
name|short
operator|)
literal|0
expr_stmt|;
name|i
operator|=
operator|(
name|int
operator|)
literal|0
expr_stmt|;
name|i
operator|=
operator|(
name|long
operator|)
literal|0
expr_stmt|;
name|l
operator|=
operator|(
name|char
operator|)
literal|0
expr_stmt|;
name|l
operator|=
operator|(
name|short
operator|)
literal|0
expr_stmt|;
name|l
operator|=
operator|(
name|int
operator|)
literal|0
expr_stmt|;
name|l
operator|=
operator|(
name|long
operator|)
literal|0
expr_stmt|;
name|c
operator|=
operator|(
name|char
operator|)
name|BIG
expr_stmt|;
name|c
operator|=
operator|(
name|short
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'short' to 'char' changes value}}
name|c
operator|=
operator|(
name|int
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'int' to 'char' changes value}}
name|c
operator|=
operator|(
name|long
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'long' to 'char' changes value}}
name|s
operator|=
operator|(
name|char
operator|)
name|BIG
expr_stmt|;
name|s
operator|=
operator|(
name|short
operator|)
name|BIG
expr_stmt|;
name|s
operator|=
operator|(
name|int
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'int' to 'short' changes value}}
name|s
operator|=
operator|(
name|long
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'long' to 'short' changes value}}
name|i
operator|=
operator|(
name|char
operator|)
name|BIG
expr_stmt|;
name|i
operator|=
operator|(
name|short
operator|)
name|BIG
expr_stmt|;
name|i
operator|=
operator|(
name|int
operator|)
name|BIG
expr_stmt|;
name|i
operator|=
operator|(
name|long
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'long' to 'int' changes value}}
name|l
operator|=
operator|(
name|char
operator|)
name|BIG
expr_stmt|;
name|l
operator|=
operator|(
name|short
operator|)
name|BIG
expr_stmt|;
name|l
operator|=
operator|(
name|int
operator|)
name|BIG
expr_stmt|;
name|l
operator|=
operator|(
name|long
operator|)
name|BIG
expr_stmt|;
block|}
end_function

begin_function
name|char
name|test1
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|int
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|short
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'long long' to 'char' changes value}}
return|return
operator|(
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'long' to 'char' changes value}}
return|return
operator|(
name|int
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'int' to 'char' changes value}}
return|return
operator|(
name|short
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'short' to 'char' changes value}}
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function
name|short
name|test2
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|int
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|short
operator|)
name|ll
return|;
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'long long' to 'short' changes value}}
return|return
operator|(
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'long' to 'short' changes value}}
return|return
operator|(
name|int
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'int' to 'short' changes value}}
return|return
operator|(
name|short
operator|)
name|BIG
return|;
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function
name|int
name|test3
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit conversion loses integer precision}}
return|return
operator|(
name|int
operator|)
name|ll
return|;
return|return
operator|(
name|short
operator|)
name|ll
return|;
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'long long' to 'int' changes value}}
return|return
operator|(
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit conversion from 'long' to 'int' changes value}}
return|return
operator|(
name|int
operator|)
name|BIG
return|;
return|return
operator|(
name|short
operator|)
name|BIG
return|;
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function
name|long
name|test4
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
return|return
operator|(
name|long
operator|)
name|ll
return|;
return|return
operator|(
name|int
operator|)
name|ll
return|;
return|return
operator|(
name|short
operator|)
name|ll
return|;
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
return|return
operator|(
name|long
operator|)
name|BIG
return|;
return|return
operator|(
name|int
operator|)
name|BIG
return|;
return|return
operator|(
name|short
operator|)
name|BIG
return|;
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function
name|long
name|long
name|test5
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
return|return
operator|(
name|long
operator|)
name|ll
return|;
return|return
operator|(
name|int
operator|)
name|ll
return|;
return|return
operator|(
name|short
operator|)
name|ll
return|;
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
return|return
operator|(
name|long
operator|)
name|BIG
return|;
return|return
operator|(
name|int
operator|)
name|BIG
return|;
return|return
operator|(
name|short
operator|)
name|BIG
return|;
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function_decl
name|void
name|takes_char
parameter_list|(
name|char
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_short
parameter_list|(
name|short
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_int
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_long
parameter_list|(
name|long
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_longlong
parameter_list|(
name|long
name|long
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_float
parameter_list|(
name|float
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_double
parameter_list|(
name|double
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_longdouble
parameter_list|(
name|long
name|double
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|test6
parameter_list|(
name|char
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test7
parameter_list|(
name|short
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test8
parameter_list|(
name|int
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test9
parameter_list|(
name|long
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test10
parameter_list|(
name|long
name|long
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test11
parameter_list|(
name|float
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test12
parameter_list|(
name|double
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses floating-point precision}}
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test13
parameter_list|(
name|long
name|double
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion turns floating-point number into integer}}
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses floating-point precision}}
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit conversion loses floating-point precision}}
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test14
parameter_list|(
name|long
name|l
parameter_list|)
block|{
comment|// Fine because of the boolean whitelist.
name|char
name|c
decl_stmt|;
name|c
operator|=
operator|(
name|l
operator|==
literal|4
operator|)
expr_stmt|;
name|c
operator|=
operator|(
operator|(
name|l
operator|<=
literal|4
operator|)
operator|&&
operator|(
name|l
operator|>=
literal|0
operator|)
operator|)
expr_stmt|;
name|c
operator|=
operator|(
operator|(
name|l
operator|<=
literal|4
operator|)
operator|&&
operator|(
name|l
operator|>=
literal|0
operator|)
operator|)
operator|||
operator|(
name|l
operator|>
literal|20
operator|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test15
parameter_list|(
name|char
name|c
parameter_list|)
block|{
name|c
operator|=
name|c
operator|+
literal|1
operator|+
name|c
operator|*
literal|2
expr_stmt|;
name|c
operator|=
operator|(
name|short
operator|)
name|c
operator|+
literal|1
operator|+
name|c
operator|*
literal|2
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
block|}
end_function

begin_comment
comment|// PR 5422
end_comment

begin_decl_stmt
specifier|extern
name|void
modifier|*
name|test16_external
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|test16
parameter_list|(
name|void
parameter_list|)
block|{
name|int
name|a
init|=
operator|(
name|unsigned
name|long
operator|)
operator|&
name|test16_external
decl_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
block|}
end_function

begin_comment
comment|// PR 5938
end_comment

begin_function
name|void
name|test17
parameter_list|()
block|{
union|union
block|{
name|unsigned
name|long
name|long
name|a
range|:
literal|8
decl_stmt|;
name|unsigned
name|long
name|long
name|b
range|:
literal|32
decl_stmt|;
name|unsigned
name|long
name|long
name|c
decl_stmt|;
block|}
name|U
union|;
name|unsigned
name|int
name|x
decl_stmt|;
name|x
operator|=
name|U
operator|.
name|a
expr_stmt|;
name|x
operator|=
name|U
operator|.
name|b
expr_stmt|;
name|x
operator|=
name|U
operator|.
name|c
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision}}
block|}
end_function

begin_comment
comment|// PR 5939
end_comment

begin_function
name|void
name|test18
parameter_list|()
block|{
union|union
block|{
name|unsigned
name|long
name|long
name|a
range|:
literal|1
decl_stmt|;
name|unsigned
name|long
name|long
name|b
decl_stmt|;
block|}
name|U
union|;
name|int
name|x
decl_stmt|;
name|x
operator|=
operator|(
name|U
operator|.
name|a
condition|?
literal|0
else|:
literal|1
operator|)
expr_stmt|;
name|x
operator|=
operator|(
name|U
operator|.
name|b
condition|?
literal|0
else|:
literal|1
operator|)
expr_stmt|;
block|}
end_function

begin_comment
comment|// None of these should warn.
end_comment

begin_function
name|unsigned
name|char
name|test19
parameter_list|(
name|unsigned
name|long
name|u64
parameter_list|)
block|{
name|unsigned
name|char
name|x1
init|=
name|u64
operator|&
literal|0xff
decl_stmt|;
name|unsigned
name|char
name|x2
init|=
name|u64
operator|>>
literal|56
decl_stmt|;
name|unsigned
name|char
name|mask
init|=
literal|0xee
decl_stmt|;
name|unsigned
name|char
name|x3
init|=
name|u64
operator|&
name|mask
decl_stmt|;
return|return
name|x1
operator|+
name|x2
operator|+
name|x3
return|;
block|}
end_function

begin_comment
comment|//<rdar://problem/7631400>
end_comment

begin_function
name|void
name|test_7631400
parameter_list|(
name|void
parameter_list|)
block|{
comment|// This should show up despite the caret being inside a macro substitution
name|char
name|s
init|=
name|LONG_MAX
decl_stmt|;
comment|// expected-warning {{implicit conversion from 'long' to 'char' changes value}}
block|}
end_function

begin_comment
comment|//<rdar://problem/7676608>: assertion for compound operators with non-integral RHS
end_comment

begin_function_decl
name|void
name|f7676608
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|test_7676608
parameter_list|(
name|void
parameter_list|)
block|{
name|float
name|q
init|=
literal|0.7f
decl_stmt|;
name|char
name|c
init|=
literal|5
decl_stmt|;
name|f7676608
argument_list|(
name|c
operator|*=
name|q
argument_list|)
expr_stmt|;
block|}
end_function

begin_comment
comment|//<rdar://problem/7904686>
end_comment

begin_function
name|void
name|test_7904686
parameter_list|(
name|void
parameter_list|)
block|{
specifier|const
name|int
name|i
init|=
operator|-
literal|1
decl_stmt|;
name|unsigned
name|u1
init|=
name|i
decl_stmt|;
comment|// expected-warning {{implicit conversion changes signedness}}
name|u1
operator|=
name|i
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness}}
name|unsigned
name|u2
init|=
operator|-
literal|1
decl_stmt|;
comment|// expected-warning {{implicit conversion changes signedness}}
name|u2
operator|=
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness}}
block|}
end_function

begin_comment
comment|//<rdar://problem/8232669>: don't warn about conversions required by
end_comment

begin_comment
comment|// contexts in system headers
end_comment

begin_function
name|void
name|test_8232669
parameter_list|(
name|void
parameter_list|)
block|{
name|unsigned
name|bitset
index|[
literal|20
index|]
decl_stmt|;
name|SETBIT
argument_list|(
name|bitset
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|unsigned
name|y
init|=
literal|50
decl_stmt|;
name|SETBIT
argument_list|(
name|bitset
argument_list|,
name|y
argument_list|)
expr_stmt|;
define|#
directive|define
name|USER_SETBIT
parameter_list|(
name|set
parameter_list|,
name|bit
parameter_list|)
value|do { int i = bit; set[i/(8*sizeof(set[0]))] |= (1<< (i%(8*sizeof(set)))); } while(0)
name|USER_SETBIT
argument_list|(
name|bitset
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning 2 {{implicit conversion changes signedness}}
block|}
end_function

end_unit

