begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -fsyntax-only -verify -Wconversion -triple x86_64-apple-darwin %s
end_comment

begin_define
define|#
directive|define
name|BIG
value|0x7f7f7f7f7f7f7f7fL
end_define

begin_function
name|void
name|test0
parameter_list|(
name|char
name|c
parameter_list|,
name|short
name|s
parameter_list|,
name|int
name|i
parameter_list|,
name|long
name|l
parameter_list|,
name|long
name|long
name|ll
parameter_list|)
block|{
name|c
operator|=
name|c
expr_stmt|;
name|c
operator|=
name|s
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|c
operator|=
name|i
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|c
operator|=
name|l
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|s
operator|=
name|c
expr_stmt|;
name|s
operator|=
name|s
expr_stmt|;
name|s
operator|=
name|i
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|s
operator|=
name|l
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|i
operator|=
name|c
expr_stmt|;
name|i
operator|=
name|s
expr_stmt|;
name|i
operator|=
name|i
expr_stmt|;
name|i
operator|=
name|l
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|l
operator|=
name|c
expr_stmt|;
name|l
operator|=
name|s
expr_stmt|;
name|l
operator|=
name|i
expr_stmt|;
name|l
operator|=
name|l
expr_stmt|;
name|c
operator|=
operator|(
name|char
operator|)
literal|0
expr_stmt|;
name|c
operator|=
operator|(
name|short
operator|)
literal|0
expr_stmt|;
name|c
operator|=
operator|(
name|int
operator|)
literal|0
expr_stmt|;
name|c
operator|=
operator|(
name|long
operator|)
literal|0
expr_stmt|;
name|s
operator|=
operator|(
name|char
operator|)
literal|0
expr_stmt|;
name|s
operator|=
operator|(
name|short
operator|)
literal|0
expr_stmt|;
name|s
operator|=
operator|(
name|int
operator|)
literal|0
expr_stmt|;
name|s
operator|=
operator|(
name|long
operator|)
literal|0
expr_stmt|;
name|i
operator|=
operator|(
name|char
operator|)
literal|0
expr_stmt|;
name|i
operator|=
operator|(
name|short
operator|)
literal|0
expr_stmt|;
name|i
operator|=
operator|(
name|int
operator|)
literal|0
expr_stmt|;
name|i
operator|=
operator|(
name|long
operator|)
literal|0
expr_stmt|;
name|l
operator|=
operator|(
name|char
operator|)
literal|0
expr_stmt|;
name|l
operator|=
operator|(
name|short
operator|)
literal|0
expr_stmt|;
name|l
operator|=
operator|(
name|int
operator|)
literal|0
expr_stmt|;
name|l
operator|=
operator|(
name|long
operator|)
literal|0
expr_stmt|;
name|c
operator|=
operator|(
name|char
operator|)
name|BIG
expr_stmt|;
name|c
operator|=
operator|(
name|short
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|c
operator|=
operator|(
name|int
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|c
operator|=
operator|(
name|long
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|s
operator|=
operator|(
name|char
operator|)
name|BIG
expr_stmt|;
name|s
operator|=
operator|(
name|short
operator|)
name|BIG
expr_stmt|;
name|s
operator|=
operator|(
name|int
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|s
operator|=
operator|(
name|long
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|i
operator|=
operator|(
name|char
operator|)
name|BIG
expr_stmt|;
name|i
operator|=
operator|(
name|short
operator|)
name|BIG
expr_stmt|;
name|i
operator|=
operator|(
name|int
operator|)
name|BIG
expr_stmt|;
name|i
operator|=
operator|(
name|long
operator|)
name|BIG
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|l
operator|=
operator|(
name|char
operator|)
name|BIG
expr_stmt|;
name|l
operator|=
operator|(
name|short
operator|)
name|BIG
expr_stmt|;
name|l
operator|=
operator|(
name|int
operator|)
name|BIG
expr_stmt|;
name|l
operator|=
operator|(
name|long
operator|)
name|BIG
expr_stmt|;
block|}
end_function

begin_function
name|char
name|test1
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|int
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|short
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|int
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|short
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function
name|short
name|test2
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|int
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|short
operator|)
name|ll
return|;
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|int
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|short
operator|)
name|BIG
return|;
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function
name|int
name|test3
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|long
operator|)
name|ll
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|int
operator|)
name|ll
return|;
return|return
operator|(
name|short
operator|)
name|ll
return|;
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|long
operator|)
name|BIG
return|;
comment|// expected-warning {{implicit cast loses integer precision}}
return|return
operator|(
name|int
operator|)
name|BIG
return|;
return|return
operator|(
name|short
operator|)
name|BIG
return|;
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function
name|long
name|test4
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
return|return
operator|(
name|long
operator|)
name|ll
return|;
return|return
operator|(
name|int
operator|)
name|ll
return|;
return|return
operator|(
name|short
operator|)
name|ll
return|;
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
return|return
operator|(
name|long
operator|)
name|BIG
return|;
return|return
operator|(
name|int
operator|)
name|BIG
return|;
return|return
operator|(
name|short
operator|)
name|BIG
return|;
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function
name|long
name|long
name|test5
parameter_list|(
name|long
name|long
name|ll
parameter_list|)
block|{
return|return
operator|(
name|long
name|long
operator|)
name|ll
return|;
return|return
operator|(
name|long
operator|)
name|ll
return|;
return|return
operator|(
name|int
operator|)
name|ll
return|;
return|return
operator|(
name|short
operator|)
name|ll
return|;
return|return
operator|(
name|char
operator|)
name|ll
return|;
return|return
operator|(
name|long
name|long
operator|)
name|BIG
return|;
return|return
operator|(
name|long
operator|)
name|BIG
return|;
return|return
operator|(
name|int
operator|)
name|BIG
return|;
return|return
operator|(
name|short
operator|)
name|BIG
return|;
return|return
operator|(
name|char
operator|)
name|BIG
return|;
block|}
end_function

begin_function_decl
name|void
name|takes_char
parameter_list|(
name|char
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_short
parameter_list|(
name|short
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_int
parameter_list|(
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_long
parameter_list|(
name|long
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_longlong
parameter_list|(
name|long
name|long
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_float
parameter_list|(
name|float
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_double
parameter_list|(
name|double
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|takes_longdouble
parameter_list|(
name|long
name|double
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|test6
parameter_list|(
name|char
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test7
parameter_list|(
name|short
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test8
parameter_list|(
name|int
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test9
parameter_list|(
name|long
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test10
parameter_list|(
name|long
name|long
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test11
parameter_list|(
name|float
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test12
parameter_list|(
name|double
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses floating-point precision}}
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test13
parameter_list|(
name|long
name|double
name|v
parameter_list|)
block|{
name|takes_char
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_short
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_int
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_long
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_longlong
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast turns floating-point number into integer}}
name|takes_float
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses floating-point precision}}
name|takes_double
argument_list|(
name|v
argument_list|)
expr_stmt|;
comment|// expected-warning {{implicit cast loses floating-point precision}}
name|takes_longdouble
argument_list|(
name|v
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test14
parameter_list|(
name|long
name|l
parameter_list|)
block|{
comment|// Fine because of the boolean whitelist.
name|char
name|c
decl_stmt|;
name|c
operator|=
operator|(
name|l
operator|==
literal|4
operator|)
expr_stmt|;
name|c
operator|=
operator|(
operator|(
name|l
operator|<=
literal|4
operator|)
operator|&&
operator|(
name|l
operator|>=
literal|0
operator|)
operator|)
expr_stmt|;
name|c
operator|=
operator|(
operator|(
name|l
operator|<=
literal|4
operator|)
operator|&&
operator|(
name|l
operator|>=
literal|0
operator|)
operator|)
operator|||
operator|(
name|l
operator|>
literal|20
operator|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test15
parameter_list|(
name|char
name|c
parameter_list|)
block|{
name|c
operator|=
name|c
operator|+
literal|1
operator|+
name|c
operator|*
literal|2
expr_stmt|;
name|c
operator|=
operator|(
name|short
operator|)
name|c
operator|+
literal|1
operator|+
name|c
operator|*
literal|2
expr_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
block|}
end_function

begin_comment
comment|// PR 5422
end_comment

begin_decl_stmt
specifier|extern
name|void
modifier|*
name|test16_external
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|test16
parameter_list|(
name|void
parameter_list|)
block|{
name|int
name|a
init|=
operator|(
name|unsigned
name|long
operator|)
operator|&
name|test16_external
decl_stmt|;
comment|// expected-warning {{implicit cast loses integer precision}}
block|}
end_function

end_unit

