begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -fsyntax-only -std=c99 -verify -pedantic %s
end_comment

begin_decl_stmt
name|int
name|x
index|[
literal|10
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|x
index|[]
init|=
block|{
literal|1
block|,
literal|2
block|,
literal|3
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|testx
index|[
operator|(
sizeof|sizeof
argument_list|(
name|x
argument_list|)
operator|==
sizeof|sizeof
argument_list|(
name|int
argument_list|)
operator|*
literal|10
operator|)
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_expr_stmt
name|int
argument_list|(
operator|*
name|a
argument_list|)
argument_list|(
name|int
argument_list|(
operator|*
name|x
argument_list|)
index|[
literal|10
index|]
argument_list|,
name|int
argument_list|(
operator|*
name|y
argument_list|)
index|[]
argument_list|)
expr_stmt|;
end_expr_stmt

begin_expr_stmt
name|int
argument_list|(
operator|*
name|a
argument_list|)
argument_list|(
name|int
argument_list|(
operator|*
name|x
argument_list|)
index|[]
argument_list|,
name|int
argument_list|(
operator|*
name|y
argument_list|)
index|[
literal|5
index|]
argument_list|)
expr_stmt|;
end_expr_stmt

begin_function
name|void
name|b
parameter_list|()
block|{
name|int
name|x
index|[
literal|10
index|]
decl_stmt|,
name|y
index|[
literal|5
index|]
decl_stmt|;
name|a
argument_list|(
operator|&
name|x
argument_list|,
operator|&
name|y
argument_list|)
expr_stmt|;
name|a
argument_list|(
operator|&
name|y
argument_list|,
operator|&
name|y
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible pointer}}
name|a
argument_list|(
operator|&
name|x
argument_list|,
operator|&
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible pointer}}
block|}
end_function

end_unit

