begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|/* RUN: %clang_cc1 -fsyntax-only -verify -std=c90 -pedantic %s  */
end_comment

begin_function
name|void
name|foo
parameter_list|(
name|void
parameter_list|)
block|{
struct_decl|struct
name|b
struct_decl|;
name|struct
name|b
modifier|*
name|x
init|=
literal|0
decl_stmt|;
name|struct
name|b
modifier|*
name|y
init|=
operator|&
operator|*
name|x
decl_stmt|;
block|}
end_function

begin_function
name|void
name|foo2
parameter_list|(
name|void
parameter_list|)
block|{
typedef|typedef
name|int
argument_list|(
operator|*
name|arrayptr
argument_list|)
index|[]
expr_stmt|;
name|arrayptr
name|x
init|=
literal|0
decl_stmt|;
name|arrayptr
name|y
init|=
operator|&
operator|*
name|x
decl_stmt|;
block|}
end_function

begin_function
name|void
name|foo3
parameter_list|(
name|void
parameter_list|)
block|{
name|void
modifier|*
name|x
init|=
literal|0
decl_stmt|;
name|void
modifier|*
name|y
init|=
operator|&
operator|*
name|x
decl_stmt|;
comment|/* expected-warning{{address of an expression of type 'void'}} */
block|}
end_function

begin_decl_stmt
specifier|extern
specifier|const
name|void
name|cv1
decl_stmt|;
end_decl_stmt

begin_function
specifier|const
name|void
modifier|*
name|foo4
parameter_list|(
name|void
parameter_list|)
block|{
return|return
operator|&
name|cv1
return|;
block|}
end_function

begin_decl_stmt
specifier|extern
name|void
name|cv2
decl_stmt|;
end_decl_stmt

begin_function
name|void
modifier|*
name|foo5
parameter_list|(
name|void
parameter_list|)
block|{
return|return
operator|&
name|cv2
return|;
comment|/* expected-warning{{address of an expression of type 'void'}} */
block|}
end_function

begin_typedef
typedef|typedef
specifier|const
name|void
name|CVT
typedef|;
end_typedef

begin_decl_stmt
specifier|extern
name|CVT
name|cv3
decl_stmt|;
end_decl_stmt

begin_function
specifier|const
name|void
modifier|*
name|foo6
parameter_list|(
name|void
parameter_list|)
block|{
return|return
operator|&
name|cv3
return|;
block|}
end_function

end_unit

