begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -triple x86_64-unknown-unknown %s
end_comment

begin_decl_stmt
name|int
name|complete_array_from_init
index|[]
init|=
block|{
literal|1
block|,
literal|2
block|,
index|[
literal|10
index|]
operator|=
literal|5
block|,
literal|1
block|,
literal|2
block|,
index|[
literal|5
index|]
operator|=
literal|2
block|,
literal|6
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|complete_array_from_init_check
index|[
operator|(
operator|(
sizeof|sizeof
argument_list|(
name|complete_array_from_init
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
name|int
argument_list|)
operator|)
operator|==
literal|13
operator|)
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|iarray
index|[
literal|10
index|]
init|=
block|{
index|[
literal|0
index|]
operator|=
literal|1
block|,
index|[
literal|1
operator|...
literal|5
index|]
operator|=
literal|2
block|,
index|[
literal|6
operator|...
literal|6
index|]
operator|=
literal|3
block|,
index|[
literal|8
operator|...
literal|7
index|]
operator|=
literal|4
block|,
comment|// expected-error{{array designator range [8, 7] is empty}}
index|[
literal|10
index|]
operator|=
literal|5
block|,
index|[
operator|-
literal|1
index|]
operator|=
literal|6
comment|// expected-error{{array designator value '-1' is negative}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|iarray2
index|[
literal|10
index|]
init|=
block|{
index|[
literal|10
index|]
operator|=
literal|1
block|,
comment|// expected-error{{array designator index (10) exceeds array bounds (10)}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|iarray3
index|[
literal|10
index|]
init|=
block|{
index|[
literal|3
index|]
literal|2
block|,
comment|// expected-warning{{use of GNU 'missing =' extension in designator}}
index|[
literal|5
operator|...
literal|12
index|]
operator|=
literal|2
comment|// expected-error{{array designator index (12) exceeds array bounds (10)}}
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|point
block|{
name|double
name|x
decl_stmt|;
name|double
name|y
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|point
name|p1
init|=
block|{
operator|.
name|y
operator|=
literal|1.0
block|,
name|x
operator|:
literal|2.0
block|,
comment|// expected-warning{{}}
operator|.
name|a
operator|=
literal|4.0
block|,
comment|// expected-error{{field designator 'a' does not refer to any field in type 'struct point'}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|point
name|p2
init|=
block|{
index|[
literal|1
index|]
operator|=
literal|1.0
comment|// expected-error{{array designator cannot initialize non-array type}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|point
name|array
index|[
literal|10
index|]
init|=
block|{
index|[
literal|0
index|]
operator|.
name|x
operator|=
literal|1.0
block|,
index|[
literal|1
index|]
operator|.
name|y
operator|=
literal|2.0
block|,
index|[
literal|2
index|]
operator|.
name|z
operator|=
literal|3.0
block|,
comment|// expected-error{{field designator 'z' does not refer to any field in type 'struct point'}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|point
name|array2
index|[
literal|10
index|]
init|=
block|{
index|[
literal|10
index|]
operator|.
name|x
operator|=
literal|2.0
block|,
comment|// expected-error{{array designator index (10) exceeds array bounds (10)}}
index|[
literal|4
operator|...
literal|5
index|]
operator|.
name|y
operator|=
literal|2.0
block|,
comment|// expected-note 2 {{previous initialization is here}}
index|[
literal|4
operator|...
literal|6
index|]
operator|=
block|{
operator|.
name|x
operator|=
literal|3
block|,
operator|.
name|y
operator|=
literal|4.0
block|}
comment|// expected-warning 2 {{subobject initialization overrides initialization of other fields within its enclosing subobject}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|point
name|array3
index|[
literal|10
index|]
init|=
block|{
operator|.
name|x
operator|=
literal|5
comment|// expected-error{{field designator cannot initialize a non-struct, non-union type}}
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|rect
block|{
name|struct
name|point
name|top_left
decl_stmt|;
name|struct
name|point
name|bottom_right
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|rect
name|window
init|=
block|{
operator|.
name|top_left
operator|.
name|x
operator|=
literal|1.0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|rect
name|windows
index|[]
init|=
block|{
index|[
literal|2
index|]
operator|.
name|top_left
operator|=
block|{
literal|1.0
block|,
literal|2.0
block|}
block|,
index|[
literal|4
index|]
operator|.
name|bottom_right
operator|=
block|{
operator|.
name|y
operator|=
literal|1.0
block|}
block|,
block|{
block|{
operator|.
name|y
operator|=
literal|7.0
block|,
operator|.
name|x
operator|=
literal|8.0
block|}
block|,
block|{
operator|.
name|x
operator|=
literal|5.0
block|}
block|}
block|,
index|[
literal|3
index|]
operator|=
block|{
operator|.
name|top_left
operator|=
block|{
literal|1.1
block|,
literal|2.2
block|}
block|,
operator|.
name|bottom_right
operator|=
block|{
operator|.
name|y
operator|=
literal|1.1
block|}
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|windows_size
index|[
operator|(
operator|(
sizeof|sizeof
argument_list|(
name|windows
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
expr|struct
name|rect
argument_list|)
operator|)
operator|==
literal|6
operator|)
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|rect
name|windows_bad
index|[
literal|3
index|]
init|=
block|{
index|[
literal|2
index|]
operator|.
name|top_left
operator|=
block|{
block|{
operator|.
name|x
operator|=
literal|1.1
block|}
block|}
block|,
comment|// expected-error{{designator in initializer for scalar type}}
index|[
literal|1
index|]
operator|.
name|top_left
operator|=
block|{
operator|.
name|x
operator|=
literal|1.1
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|gui
block|{
name|struct
name|rect
name|windows
index|[
literal|10
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|gui
name|gui
index|[]
init|=
block|{
index|[
literal|5
index|]
operator|.
name|windows
index|[
literal|3
index|]
operator|.
name|top_left
operator|.
name|x
operator|=
block|{
literal|7.0
block|}
comment|// expected-warning{{braces around scalar initializer}}
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|translator
block|{
struct|struct
name|wonky
block|{
name|int
modifier|*
name|ptr
decl_stmt|;
block|}
name|wonky
struct|;
name|struct
name|rect
name|window
decl_stmt|;
name|struct
name|point
name|offset
decl_stmt|;
block|}
name|tran
init|=
block|{
operator|.
name|window
operator|=
block|{
operator|.
name|top_left
operator|=
block|{
literal|1.0
block|,
literal|2.0
block|}
block|}
block|,
block|{
operator|.
name|x
operator|=
literal|5.0
block|,
operator|.
name|y
operator|=
literal|6.0
block|}
block|,
operator|.
name|wonky
operator|=
block|{
literal|0
block|}
block|}
struct|;
end_struct

begin_decl_stmt
name|int
name|anint
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
block|{
name|int
name|x
decl_stmt|,
modifier|*
name|y
decl_stmt|;
block|}
name|z
index|[]
init|=
block|{
index|[
literal|0
index|]
operator|.
name|x
operator|=
literal|2
block|,
operator|&
name|z
index|[
literal|0
index|]
operator|.
name|x
block|}
struct|;
end_struct

begin_struct
struct|struct
name|outer
block|{
struct|struct
name|inner
block|{
name|int
name|x
decl_stmt|,
modifier|*
name|y
decl_stmt|;
block|}
name|in
struct|,
modifier|*
name|inp
struct|;
block|}
name|zz
index|[]
init|=
block|{
index|[
literal|0
index|]
operator|.
name|in
operator|.
name|x
operator|=
literal|2
block|,
operator|&
name|zz
index|[
literal|0
index|]
operator|.
name|in
operator|.
name|x
block|,
operator|&
name|zz
index|[
literal|0
index|]
operator|.
name|in
block|,
literal|0
block|,
operator|&
name|anint
block|,
operator|&
name|zz
index|[
literal|1
index|]
operator|.
name|in
block|,
index|[
literal|3
index|]
operator|.
name|in
operator|=
block|{
operator|.
name|y
operator|=
operator|&
name|anint
block|,
operator|.
name|x
operator|=
literal|17
block|}
block|,
index|[
literal|7
index|]
operator|.
name|in
operator|.
name|y
operator|=
operator|&
name|anint
block|,
operator|&
name|zz
index|[
literal|0
index|]
operator|.
name|in
block|,
index|[
literal|4
index|]
operator|.
name|in
operator|.
name|y
operator|=
operator|&
name|anint
block|,
index|[
literal|5
index|]
operator|.
name|in
operator|.
name|x
operator|=
literal|12
block|}
struct|;
end_struct

begin_decl_stmt
name|int
name|zz_sizecheck
index|[
sizeof|sizeof
argument_list|(
name|zz
argument_list|)
operator|/
sizeof|sizeof
argument_list|(
expr|struct
name|outer
argument_list|)
operator|==
literal|8
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|disklabel_ops
block|{
struct|struct
block|{}
name|type
struct|;
name|int
name|labelsize
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|disklabel_ops
name|disklabel64_ops
init|=
block|{
operator|.
name|labelsize
operator|=
expr|sizeof
operator|(
expr|struct
name|disklabel_ops
operator|)
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// PR clang/3378
end_comment

begin_decl_stmt
name|int
name|bitwidth
index|[]
init|=
block|{
index|[
operator|(
name|long
name|long
name|int
operator|)
literal|1
index|]
operator|=
literal|5
block|,
index|[
operator|(
name|short
name|int
operator|)
literal|2
index|]
operator|=
literal|2
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|a
index|[]
init|=
block|{
index|[
sizeof|sizeof
argument_list|(
name|int
argument_list|)
index|]
operator|=
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|a2
index|[]
init|=
block|{
index|[
literal|0
operator|...
sizeof|sizeof
argument_list|(
name|int
argument_list|)
index|]
operator|=
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// Test warnings about initializers overriding previous initializers
end_comment

begin_struct
struct|struct
name|X
block|{
name|int
name|a
decl_stmt|,
name|b
decl_stmt|,
name|c
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|int
name|counter
init|=
literal|0
decl_stmt|;
end_decl_stmt

begin_function
name|int
name|get8
parameter_list|()
block|{
operator|++
name|counter
expr_stmt|;
return|return
literal|8
return|;
block|}
end_function

begin_function
name|void
name|test
parameter_list|()
block|{
name|struct
name|X
name|xs
index|[]
init|=
block|{
index|[
literal|0
index|]
operator|=
operator|(
expr|struct
name|X
operator|)
block|{
literal|1
block|,
literal|2
block|}
block|,
comment|// expected-note 2 {{previous initialization is here}}
index|[
literal|0
index|]
operator|.
name|c
operator|=
literal|3
block|,
comment|// expected-warning{{subobject initialization overrides initialization of other fields within its enclosing subobject}}
operator|(
expr|struct
name|X
operator|)
block|{
literal|4
block|,
literal|5
block|,
literal|6
block|}
block|,
comment|// expected-note{{previous initialization is here}}
index|[
literal|1
index|]
operator|.
name|b
operator|=
name|get8
argument_list|()
block|,
comment|// expected-warning{{subobject initialization overrides initialization of other fields within its enclosing subobject}}
index|[
literal|0
index|]
operator|.
name|b
operator|=
literal|8
comment|// expected-warning{{subobject initialization overrides initialization of other fields within its enclosing subobject}}
block|}
decl_stmt|;
block|}
end_function

begin_union
union|union
block|{
name|char
name|c
decl_stmt|;
name|long
name|l
decl_stmt|;
block|}
name|u1
init|=
block|{
operator|.
name|l
operator|=
literal|0xFFFF
block|}
union|;
end_union

begin_decl_stmt
specifier|extern
name|float
name|global_float
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|XX
block|{
name|int
name|a
decl_stmt|,
modifier|*
name|b
decl_stmt|;
block|}
struct|;
end_struct

begin_struct
struct|struct
name|XY
block|{
name|int
name|before
decl_stmt|;
name|struct
name|XX
name|xx
decl_stmt|,
modifier|*
name|xp
decl_stmt|;
name|float
modifier|*
name|after
decl_stmt|;
block|}
name|xy
index|[]
init|=
block|{
literal|0
block|,
literal|0
block|,
operator|&
name|xy
index|[
literal|0
index|]
operator|.
name|xx
operator|.
name|a
block|,
operator|&
name|xy
index|[
literal|0
index|]
operator|.
name|xx
block|,
operator|&
name|global_float
block|,
index|[
literal|1
index|]
operator|.
name|xx
operator|=
literal|0
block|,
operator|&
name|xy
index|[
literal|1
index|]
operator|.
name|xx
operator|.
name|a
block|,
operator|&
name|xy
index|[
literal|1
index|]
operator|.
name|xx
block|,
operator|&
name|global_float
block|,
literal|0
block|,
comment|// expected-note{{previous initialization is here}}
literal|0
block|,
comment|// expected-note{{previous initialization is here}}
index|[
literal|2
index|]
operator|.
name|before
operator|=
literal|0
block|,
comment|// expected-warning{{initializer overrides prior initialization of this subobject}}
literal|0
block|,
comment|// expected-warning{{initializer overrides prior initialization of this subobject}}
operator|&
name|xy
index|[
literal|2
index|]
operator|.
name|xx
operator|.
name|a
block|,
operator|&
name|xy
index|[
literal|2
index|]
operator|.
name|xx
block|,
operator|&
name|global_float
block|}
struct|;
end_struct

begin_comment
comment|// PR3519
end_comment

begin_struct
struct|struct
name|foo
block|{
name|int
name|arr
index|[
literal|10
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|foo
name|Y
index|[
literal|10
index|]
init|=
block|{
index|[
literal|1
index|]
operator|.
name|arr
index|[
literal|1
index|]
operator|=
literal|2
block|,
index|[
literal|4
index|]
operator|.
name|arr
index|[
literal|2
index|]
operator|=
literal|4
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|bar
block|{
name|struct
name|foo
name|f
decl_stmt|;
name|float
modifier|*
name|arr
index|[
literal|10
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
specifier|extern
name|float
name|f
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|bar
name|saloon
init|=
block|{
operator|.
name|f
operator|.
name|arr
index|[
literal|3
index|]
operator|=
literal|1
block|,
operator|.
name|arr
operator|=
block|{
operator|&
name|f
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_typedef
typedef|typedef
name|unsigned
name|char
name|u_char
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|short
name|u_short
typedef|;
end_typedef

begin_union
union|union
name|wibble
block|{
name|u_char
name|arr1
index|[
literal|6
index|]
decl_stmt|;
name|u_short
name|arr2
index|[
literal|3
index|]
decl_stmt|;
block|}
union|;
end_union

begin_decl_stmt
specifier|const
name|union
name|wibble
name|wobble
init|=
block|{
operator|.
name|arr2
index|[
literal|0
index|]
operator|=
literal|0xffff
block|,
operator|.
name|arr2
index|[
literal|1
index|]
operator|=
literal|0xffff
block|,
operator|.
name|arr2
index|[
literal|2
index|]
operator|=
literal|0xffff
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
specifier|const
name|union
name|wibble
name|wobble2
init|=
block|{
operator|.
name|arr2
operator|=
block|{
literal|4
block|,
literal|5
block|,
literal|6
block|}
block|,
literal|7
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning{{excess elements in union initializer}}
end_comment

begin_comment
comment|// PR3778
end_comment

begin_struct
struct|struct
name|s
block|{
union|union
block|{
name|int
name|i
decl_stmt|;
block|}
union|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|s
name|si
init|=
block|{
block|{
operator|.
name|i
operator|=
literal|1
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|double
name|d0
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|c0
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|float
name|f0
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|i0
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|Enigma
block|{
union|union
block|{
struct|struct
block|{
struct|struct
block|{
name|double
modifier|*
name|double_ptr
decl_stmt|;
name|char
modifier|*
name|string
decl_stmt|;
block|}
struct|;
name|float
modifier|*
name|float_ptr
decl_stmt|;
block|}
struct|;
name|int
modifier|*
name|int_ptr
decl_stmt|;
block|}
union|;
name|char
modifier|*
name|string2
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|Enigma
name|enigma
init|=
block|{
operator|.
name|double_ptr
operator|=
operator|&
name|d0
block|,
operator|&
name|c0
block|,
operator|&
name|f0
block|,
comment|// expected-note{{previous}}
operator|&
name|c0
block|,
operator|.
name|float_ptr
operator|=
operator|&
name|f0
comment|// expected-warning{{overrides}}
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|/// PR16644
end_comment

begin_typedef
typedef|typedef
union|union
block|{
struct|struct
block|{
name|int
name|zero
decl_stmt|;
name|int
name|one
decl_stmt|;
name|int
name|two
decl_stmt|;
name|int
name|three
decl_stmt|;
block|}
name|a
struct|;
name|int
name|b
index|[
literal|4
index|]
decl_stmt|;
block|}
name|union_16644_t
typedef|;
end_typedef

begin_decl_stmt
name|union_16644_t
name|union_16644_instance_0
init|=
block|{
operator|.
name|b
index|[
literal|0
index|]
operator|=
literal|0
block|,
comment|//                               expected-note{{previous}}
operator|.
name|a
operator|.
name|one
operator|=
literal|1
block|,
comment|// expected-warning{{overrides}} expected-note{{previous}}
operator|.
name|b
index|[
literal|2
index|]
operator|=
literal|2
block|,
comment|// expected-warning{{overrides}} expected-note{{previous}}
operator|.
name|a
operator|.
name|three
operator|=
literal|3
block|,
comment|// expected-warning{{overrides}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|union_16644_t
name|union_16644_instance_1
init|=
block|{
operator|.
name|a
operator|.
name|three
operator|=
literal|13
block|,
comment|//                               expected-note{{previous}}
operator|.
name|b
index|[
literal|2
index|]
operator|=
literal|12
block|,
comment|// expected-warning{{overrides}} expected-note{{previous}}
operator|.
name|a
operator|.
name|one
operator|=
literal|11
block|,
comment|// expected-warning{{overrides}} expected-note{{previous}}
operator|.
name|b
index|[
literal|0
index|]
operator|=
literal|10
block|,
comment|// expected-warning{{overrides}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|union_16644_t
name|union_16644_instance_2
init|=
block|{
operator|.
name|a
operator|.
name|one
operator|=
literal|21
block|,
comment|//                               expected-note{{previous}}
operator|.
name|b
index|[
literal|1
index|]
operator|=
literal|20
block|,
comment|// expected-warning{{overrides}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|union_16644_t
name|union_16644_instance_3
init|=
block|{
operator|.
name|b
index|[
literal|1
index|]
operator|=
literal|30
block|,
comment|//                               expected-note{{previous}}
operator|.
name|a
operator|=
block|{
comment|// expected-warning{{overrides}}
operator|.
name|one
operator|=
literal|31
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|union_16644_t
name|union_16644_instance_4
index|[
literal|2
index|]
init|=
block|{
index|[
literal|0
index|]
operator|.
name|a
operator|.
name|one
operator|=
literal|2
block|,
index|[
literal|1
index|]
operator|.
name|a
operator|.
name|zero
operator|=
literal|3
block|,
comment|//                               expected-note{{previous}}
index|[
literal|0
index|]
operator|.
name|a
operator|.
name|zero
operator|=
literal|5
block|,
index|[
literal|1
index|]
operator|.
name|b
index|[
literal|1
index|]
operator|=
literal|4
comment|// expected-warning{{overrides}}
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|/// PR4073
end_comment

begin_comment
comment|/// Should use evaluate to fold aggressively and emit a warning if not an ice.
end_comment

begin_decl_stmt
specifier|extern
name|int
name|crazy_x
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|crazy_Y
index|[]
init|=
block|{
index|[
literal|0
operator|?
name|crazy_x
operator|:
literal|4
index|]
operator|=
literal|1
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// PR5843
end_comment

begin_struct
struct|struct
name|expr
block|{
name|int
name|nargs
decl_stmt|;
union|union
block|{
name|unsigned
name|long
name|int
name|num
decl_stmt|;
name|struct
name|expr
modifier|*
name|args
index|[
literal|3
index|]
decl_stmt|;
block|}
name|val
union|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|expr
name|expr0
init|=
block|{
operator|.
name|nargs
operator|=
literal|2
block|,
operator|.
name|val
operator|=
block|{
operator|.
name|args
operator|=
block|{
index|[
literal|0
index|]
operator|=
operator|(
expr|struct
name|expr
operator|*
operator|)
literal|0
block|,
index|[
literal|1
index|]
operator|=
operator|(
expr|struct
name|expr
operator|*
operator|)
literal|0
block|}
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// PR6955
end_comment

begin_struct
struct|struct
name|ds
block|{
struct|struct
block|{
struct|struct
block|{
name|unsigned
name|int
name|a
decl_stmt|;
block|}
struct|;
name|unsigned
name|int
name|b
decl_stmt|;
struct|struct
block|{
name|unsigned
name|int
name|c
decl_stmt|;
block|}
struct|;
block|}
struct|;
block|}
struct|;
end_struct

begin_comment
comment|// C1X lookup-based anonymous member init cases
end_comment

begin_decl_stmt
name|struct
name|ds
name|ds0
init|=
block|{
block|{
block|{
operator|.
name|a
operator|=
literal|1
comment|// expected-note{{previous initialization is here}}
block|}
block|}
block|,
operator|.
name|a
operator|=
literal|2
block|,
comment|// expected-warning{{initializer overrides prior initialization of this subobject}}
operator|.
name|b
operator|=
literal|3
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|ds
name|ds1
init|=
block|{
operator|.
name|c
operator|=
literal|0
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|ds
name|ds2
init|=
block|{
block|{
block|{
operator|.
name|a
operator|=
literal|0
block|,
operator|.
name|b
operator|=
literal|1
comment|// expected-error{{field designator 'b' does not refer to any field}}
block|}
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// Check initializer override warnings overriding a character in a string
end_comment

begin_struct
struct|struct
name|overwrite_string_struct
block|{
name|char
name|L
index|[
literal|6
index|]
decl_stmt|;
name|int
name|M
decl_stmt|;
block|}
name|overwrite_string
index|[]
init|=
block|{
block|{
block|{
literal|"foo"
block|}
block|,
literal|1
block|}
block|,
comment|// expected-note {{previous initialization is here}}
index|[
literal|0
index|]
operator|.
name|L
index|[
literal|2
index|]
operator|=
literal|'x'
comment|// expected-warning{{subobject initialization overrides initialization of other fields}}
block|}
struct|;
end_struct

begin_struct
struct|struct
name|overwrite_string_struct2
block|{
name|char
name|L
index|[
literal|6
index|]
decl_stmt|;
name|int
name|M
decl_stmt|;
block|}
name|overwrite_string2
index|[]
init|=
block|{
block|{
block|{
literal|"foo"
block|}
block|,
literal|1
block|}
block|,
comment|// expected-note{{previous initialization is here}}
index|[
literal|0
index|]
operator|.
name|L
index|[
literal|4
index|]
operator|=
literal|'x'
comment|// expected-warning{{subobject initialization overrides initialization of other fields}}
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|overwrite_string_struct
name|overwrite_string3
index|[]
init|=
block|{
literal|"foo"
block|,
literal|1
block|,
comment|// expected-note{{previous initialization is here}}
index|[
literal|0
index|]
operator|.
name|L
index|[
literal|4
index|]
operator|=
literal|'x'
comment|// expected-warning{{subobject initialization overrides initialization of other fields}}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|overwrite_string_struct
name|overwrite_string4
index|[]
init|=
block|{
block|{
block|{
literal|'f'
block|,
literal|'o'
block|,
literal|'o'
block|}
block|,
literal|1
block|}
block|,
index|[
literal|0
index|]
operator|.
name|L
index|[
literal|4
index|]
operator|=
literal|'x'
comment|// no-warning
block|}
decl_stmt|;
end_decl_stmt

end_unit

