begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -Wassign-enum -Wswitch-enum -Wcovered-switch-default %s
end_comment

begin_enum
enum|enum
name|Enum
block|{
name|A0
init|=
literal|1
block|,
name|A1
init|=
literal|10
block|}
enum|;
end_enum

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|enum_extensibility
argument_list|(
name|closed
argument_list|)
operator|)
argument_list|)
name|EnumClosed
block|{
name|B0
operator|=
literal|1
operator|,
name|B1
operator|=
literal|10
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|enum_extensibility
argument_list|(
name|open
argument_list|)
operator|)
argument_list|)
name|EnumOpen
block|{
name|C0
operator|=
literal|1
operator|,
name|C1
operator|=
literal|10
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|flag_enum
operator|)
argument_list|)
name|EnumFlag
block|{
name|D0
operator|=
literal|1
operator|,
name|D1
operator|=
literal|8
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|flag_enum
operator|,
name|enum_extensibility
argument_list|(
name|closed
argument_list|)
operator|)
argument_list|)
name|EnumFlagClosed
block|{
name|E0
operator|=
literal|1
operator|,
name|E1
operator|=
literal|8
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|flag_enum
operator|,
name|enum_extensibility
argument_list|(
name|open
argument_list|)
operator|)
argument_list|)
name|EnumFlagOpen
block|{
name|F0
operator|=
literal|1
operator|,
name|F1
operator|=
literal|8
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|enum_extensibility
argument_list|(
name|arg1
argument_list|)
operator|)
argument_list|)
name|EnumInvalidArg
block|{
comment|// expected-warning{{'enum_extensibility' attribute argument not supported: 'arg1'}}
name|X0
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_comment
comment|// FIXME: The warning should mention that enum_extensibility takes only one argument.
end_comment

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|enum_extensibility
argument_list|(
name|closed
argument_list|,
name|open
argument_list|)
operator|)
argument_list|)
name|EnumTooManyArgs
block|{
comment|// expected-error{{use of undeclared identifier 'open'}}
name|X1
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_decl_stmt
name|enum
name|__attribute__
argument_list|(
operator|(
name|enum_extensibility
argument_list|()
operator|)
argument_list|)
name|EnumTooFewArgs
block|{
comment|// expected-error{{'enum_extensibility' attribute takes one argument}}
name|X2
block|}
end_decl_stmt

begin_empty_stmt
empty_stmt|;
end_empty_stmt

begin_struct
struct|struct
name|__attribute__
argument_list|(
argument|(enum_extensibility(open))
argument_list|)
name|S
block|{
comment|// expected-warning{{'enum_extensibility' attribute only applies to enums}}{
block|}
struct|;
end_struct

begin_function
name|void
name|test
parameter_list|()
block|{
name|enum
name|Enum
name|t0
init|=
literal|100
decl_stmt|;
comment|// expected-warning{{integer constant not in range of enumerated type}}
name|t0
operator|=
literal|1
expr_stmt|;
switch|switch
condition|(
name|t0
condition|)
block|{
comment|// expected-warning{{enumeration value 'A1' not handled in switch}}
case|case
name|A0
case|:
break|break;
case|case
literal|16
case|:
break|break;
comment|// expected-warning{{case value not in enumerated type}}
block|}
switch|switch
condition|(
name|t0
condition|)
block|{
case|case
name|A0
case|:
break|break;
case|case
name|A1
case|:
break|break;
default|default:
break|break;
comment|// expected-warning{{default label in switch which covers all enumeration}}
block|}
name|enum
name|EnumClosed
name|t1
init|=
literal|100
decl_stmt|;
comment|// expected-warning{{integer constant not in range of enumerated type}}
name|t1
operator|=
literal|1
expr_stmt|;
switch|switch
condition|(
name|t1
condition|)
block|{
comment|// expected-warning{{enumeration value 'B1' not handled in switch}}
case|case
name|B0
case|:
break|break;
case|case
literal|16
case|:
break|break;
comment|// expected-warning{{case value not in enumerated type}}
block|}
switch|switch
condition|(
name|t1
condition|)
block|{
case|case
name|B0
case|:
break|break;
case|case
name|B1
case|:
break|break;
default|default:
break|break;
comment|// expected-warning{{default label in switch which covers all enumeration}}
block|}
name|enum
name|EnumOpen
name|t2
init|=
literal|100
decl_stmt|;
name|t2
operator|=
literal|1
expr_stmt|;
switch|switch
condition|(
name|t2
condition|)
block|{
comment|// expected-warning{{enumeration value 'C1' not handled in switch}}
case|case
name|C0
case|:
break|break;
case|case
literal|16
case|:
break|break;
block|}
switch|switch
condition|(
name|t2
condition|)
block|{
case|case
name|C0
case|:
break|break;
case|case
name|C1
case|:
break|break;
default|default:
break|break;
block|}
name|enum
name|EnumFlag
name|t3
init|=
literal|5
decl_stmt|;
comment|// expected-warning{{integer constant not in range of enumerated type}}
name|t3
operator|=
literal|9
expr_stmt|;
switch|switch
condition|(
name|t3
condition|)
block|{
comment|// expected-warning{{enumeration value 'D1' not handled in switch}}
case|case
name|D0
case|:
break|break;
case|case
literal|9
case|:
break|break;
case|case
literal|16
case|:
break|break;
comment|// expected-warning{{case value not in enumerated type}}
block|}
switch|switch
condition|(
name|t3
condition|)
block|{
case|case
name|D0
case|:
break|break;
case|case
name|D1
case|:
break|break;
default|default:
break|break;
block|}
name|enum
name|EnumFlagClosed
name|t4
init|=
literal|5
decl_stmt|;
comment|// expected-warning{{integer constant not in range of enumerated type}}
name|t4
operator|=
literal|9
expr_stmt|;
switch|switch
condition|(
name|t4
condition|)
block|{
comment|// expected-warning{{enumeration value 'E1' not handled in switch}}
case|case
name|E0
case|:
break|break;
case|case
literal|9
case|:
break|break;
case|case
literal|16
case|:
break|break;
comment|// expected-warning{{case value not in enumerated type}}
block|}
switch|switch
condition|(
name|t4
condition|)
block|{
case|case
name|E0
case|:
break|break;
case|case
name|E1
case|:
break|break;
default|default:
break|break;
block|}
name|enum
name|EnumFlagOpen
name|t5
init|=
literal|5
decl_stmt|;
name|t5
operator|=
literal|9
expr_stmt|;
switch|switch
condition|(
name|t5
condition|)
block|{
comment|// expected-warning{{enumeration value 'F1' not handled in switch}}
case|case
name|F0
case|:
break|break;
case|case
literal|9
case|:
break|break;
case|case
literal|16
case|:
break|break;
block|}
switch|switch
condition|(
name|t5
condition|)
block|{
case|case
name|F0
case|:
break|break;
case|case
name|F1
case|:
break|break;
default|default:
break|break;
block|}
block|}
end_function

end_unit

