begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -verify -fsyntax-only
end_comment

begin_struct
struct|struct
name|xx
block|{
name|int
name|bitf
range|:
literal|1
decl_stmt|;
block|}
struct|;
end_struct

begin_struct
struct|struct
name|entry
block|{
name|struct
name|xx
modifier|*
name|whatever
decl_stmt|;
name|int
name|value
decl_stmt|;
name|int
name|bitf
range|:
literal|1
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|void
name|add_one
parameter_list|(
name|int
modifier|*
name|p
parameter_list|)
block|{
operator|(
operator|*
name|p
operator|)
operator|++
expr_stmt|;
block|}
end_function

begin_function
name|void
name|test
parameter_list|()
block|{
specifier|register
name|struct
name|entry
modifier|*
name|p
decl_stmt|;
name|add_one
argument_list|(
operator|&
name|p
operator|->
name|value
argument_list|)
expr_stmt|;
name|struct
name|entry
name|pvalue
decl_stmt|;
name|add_one
argument_list|(
operator|&
name|p
operator|->
name|bitf
argument_list|)
expr_stmt|;
comment|// expected-error {{address of bit-field requested}}
name|add_one
argument_list|(
operator|&
name|pvalue
operator|.
name|bitf
argument_list|)
expr_stmt|;
comment|// expected-error {{address of bit-field requested}}
name|add_one
argument_list|(
operator|&
name|p
operator|->
name|whatever
operator|->
name|bitf
argument_list|)
expr_stmt|;
comment|// expected-error {{address of bit-field requested}}
block|}
end_function

begin_function
name|void
name|foo
parameter_list|()
block|{
specifier|register
name|int
name|x
index|[
literal|10
index|]
decl_stmt|;
operator|&
name|x
index|[
literal|10
index|]
expr_stmt|;
comment|// expected-error {{address of register variable requested}}
specifier|register
name|int
modifier|*
name|y
decl_stmt|;
name|int
modifier|*
name|x2
init|=
operator|&
name|y
decl_stmt|;
comment|// expected-error {{address of register variable requested}}
name|int
modifier|*
name|x3
init|=
operator|&
name|y
index|[
literal|10
index|]
decl_stmt|;
block|}
end_function

begin_function
name|void
name|testVectorComponentAccess
parameter_list|()
block|{
typedef|typedef
name|float
name|v4sf
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
specifier|static
name|v4sf
name|q
decl_stmt|;
name|float
modifier|*
name|r
init|=
operator|&
name|q
index|[
literal|0
index|]
decl_stmt|;
comment|// expected-error {{address of vector element requested}}
block|}
end_function

begin_typedef
typedef|typedef
name|__attribute__
argument_list|(
argument|( ext_vector_type(
literal|4
argument|) )
argument_list|)
name|float
name|float4
typedef|;
end_typedef

begin_function
name|float
modifier|*
name|testExtVectorComponentAccess
parameter_list|(
name|float4
name|x
parameter_list|)
block|{
return|return
operator|&
name|x
operator|.
name|w
return|;
comment|// expected-error {{address of vector element requested}}
block|}
end_function

begin_function
name|void
name|f0
parameter_list|()
block|{
specifier|register
name|int
modifier|*
name|x0
decl_stmt|;
name|int
modifier|*
name|_dummy0
init|=
operator|&
operator|(
operator|*
name|x0
operator|)
decl_stmt|;
specifier|register
name|int
modifier|*
name|x1
decl_stmt|;
name|int
modifier|*
name|_dummy1
init|=
operator|&
operator|(
operator|*
operator|(
name|x1
operator|+
literal|1
operator|)
operator|)
decl_stmt|;
block|}
end_function

begin_comment
comment|// FIXME: The checks for this function are broken; we should error
end_comment

begin_comment
comment|// on promoting a register array to a pointer! (C99 6.3.2.1p3)
end_comment

begin_function
name|void
name|f1
parameter_list|()
block|{
specifier|register
name|int
name|x0
index|[
literal|10
index|]
decl_stmt|;
name|int
modifier|*
name|_dummy00
init|=
name|x0
decl_stmt|;
comment|// fixme-error {{address of register variable requested}}
name|int
modifier|*
name|_dummy01
init|=
operator|&
operator|(
operator|*
name|x0
operator|)
decl_stmt|;
comment|// fixme-error {{address of register variable requested}}
specifier|register
name|int
name|x1
index|[
literal|10
index|]
decl_stmt|;
name|int
modifier|*
name|_dummy1
init|=
operator|&
operator|(
operator|*
operator|(
name|x1
operator|+
literal|1
operator|)
operator|)
decl_stmt|;
comment|// fixme-error {{address of register variable requested}}
specifier|register
name|int
modifier|*
name|x2
decl_stmt|;
name|int
modifier|*
name|_dummy2
init|=
operator|&
operator|(
operator|*
operator|(
name|x2
operator|+
literal|1
operator|)
operator|)
decl_stmt|;
specifier|register
name|int
name|x3
index|[
literal|10
index|]
index|[
literal|10
index|]
index|[
literal|10
index|]
decl_stmt|;
name|int
argument_list|(
operator|*
name|_dummy3
argument_list|)
index|[
literal|10
index|]
operator|=
operator|&
name|x3
index|[
literal|0
index|]
index|[
literal|0
index|]
expr_stmt|;
comment|// expected-error {{address of register variable requested}}
specifier|register
struct|struct
block|{
name|int
name|f0
index|[
literal|10
index|]
decl_stmt|;
block|}
name|x4
struct|;
name|int
modifier|*
name|_dummy4
init|=
operator|&
name|x4
operator|.
name|f0
index|[
literal|2
index|]
decl_stmt|;
comment|// expected-error {{address of register variable requested}}
block|}
end_function

begin_function
name|void
name|f2
parameter_list|()
block|{
specifier|register
name|int
modifier|*
name|y
decl_stmt|;
name|int
modifier|*
name|_dummy0
init|=
operator|&
name|y
decl_stmt|;
comment|// expected-error {{address of register variable requested}}
name|int
modifier|*
name|_dummy1
init|=
operator|&
name|y
index|[
literal|10
index|]
decl_stmt|;
block|}
end_function

begin_function
name|void
name|f3
parameter_list|()
block|{
specifier|extern
name|void
name|f4
parameter_list|()
function_decl|;
name|void
function_decl|(
modifier|*
name|_dummy0
function_decl|)
parameter_list|()
init|=
operator|&
operator|*
operator|*
operator|*
operator|*
name|f4
function_decl|;
block|}
end_function

begin_function
name|void
name|f4
parameter_list|()
block|{
specifier|register
specifier|_Complex
name|int
name|x
decl_stmt|;
name|int
modifier|*
name|_dummy0
init|=
operator|&
name|__real__
name|x
decl_stmt|;
comment|// expected-error {{address of register variable requested}}
block|}
end_function

begin_function
name|void
name|f5
parameter_list|()
block|{
specifier|register
name|int
name|arr
index|[
literal|2
index|]
decl_stmt|;
comment|/* This is just here because if we happened to support this as an      lvalue we would need to give a warning. Note that gcc warns about      this as a register before it warns about it as an invalid      lvalue. */
name|int
modifier|*
name|_dummy0
init|=
operator|&
operator|(
name|int
operator|*
operator|)
name|arr
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue}}
name|int
modifier|*
name|_dummy1
init|=
operator|&
operator|(
name|arr
operator|+
literal|1
operator|)
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue}}
block|}
end_function

begin_function
name|void
name|f6
parameter_list|(
specifier|register
name|int
name|x
parameter_list|)
block|{
name|int
modifier|*
name|dummy0
init|=
operator|&
name|x
decl_stmt|;
comment|// expected-error {{address of register variable requested}}
block|}
end_function

begin_function
name|char
modifier|*
name|f7
parameter_list|()
block|{
specifier|register
struct|struct
block|{
name|char
modifier|*
name|x
decl_stmt|;
block|}
name|t1
init|=
block|{
literal|"Hello"
block|}
struct|;
name|char
modifier|*
name|dummy1
init|=
operator|&
operator|(
name|t1
operator|.
name|x
index|[
literal|0
index|]
operator|)
decl_stmt|;
struct|struct
block|{
name|int
name|a
range|:
literal|10
decl_stmt|;
struct|struct
block|{
name|int
name|b
range|:
literal|10
decl_stmt|;
block|}
struct|;
block|}
name|t2
struct|;
name|int
modifier|*
name|dummy2
init|=
operator|&
operator|(
name|t2
operator|.
name|a
operator|)
decl_stmt|;
comment|// expected-error {{address of bit-field requested}}
name|int
modifier|*
name|dummy3
init|=
operator|&
operator|(
name|t2
operator|.
name|b
operator|)
decl_stmt|;
comment|// expected-error {{address of bit-field requested}}
name|void
modifier|*
name|t3
init|=
operator|&
operator|(
operator|*
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
decl_stmt|;
block|}
end_function

begin_function
name|void
name|f8
parameter_list|()
block|{
name|void
modifier|*
name|dummy0
init|=
operator|&
name|f8
argument_list|()
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue of type 'void'}}
specifier|extern
name|void
name|v
decl_stmt|;
name|void
modifier|*
name|dummy1
init|=
operator|&
operator|(
literal|1
condition|?
name|v
else|:
name|f8
argument_list|()
operator|)
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue of type 'void'}}
name|void
modifier|*
name|dummy2
init|=
operator|&
operator|(
name|f8
argument_list|()
expr|,
name|v
operator|)
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue of type 'void'}}
name|void
modifier|*
name|dummy3
init|=
operator|&
operator|(
block|{ ; }
operator|)
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue of type 'void'}}
block|}
end_function

end_unit

