begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -fsyntax-only -verify -Wformat-nonliteral %s
end_comment

begin_include
include|#
directive|include
file|<stdarg.h>
end_include

begin_typedef
typedef|typedef
name|__typeof
argument_list|(
argument|sizeof(int)
argument_list|)
name|size_t
expr_stmt|;
end_typedef

begin_typedef
typedef|typedef
name|struct
name|_FILE
name|FILE
typedef|;
end_typedef

begin_function_decl
name|int
name|fprintf
parameter_list|(
name|FILE
modifier|*
parameter_list|,
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
modifier|...
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|printf
parameter_list|(
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
modifier|...
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|snprintf
parameter_list|(
name|char
modifier|*
specifier|restrict
parameter_list|,
name|size_t
parameter_list|,
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
modifier|...
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|sprintf
parameter_list|(
name|char
modifier|*
specifier|restrict
parameter_list|,
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
modifier|...
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|vasprintf
parameter_list|(
name|char
modifier|*
modifier|*
parameter_list|,
specifier|const
name|char
modifier|*
parameter_list|,
name|va_list
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|vfprintf
parameter_list|(
name|FILE
modifier|*
parameter_list|,
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
name|va_list
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|vprintf
parameter_list|(
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
name|va_list
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|vsnprintf
parameter_list|(
name|char
modifier|*
parameter_list|,
name|size_t
parameter_list|,
specifier|const
name|char
modifier|*
parameter_list|,
name|va_list
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|vsprintf
parameter_list|(
name|char
modifier|*
specifier|restrict
parameter_list|,
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
name|va_list
parameter_list|)
function_decl|;
end_function_decl

begin_decl_stmt
name|char
modifier|*
name|global_fmt
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|check_string_literal
parameter_list|(
name|FILE
modifier|*
name|fp
parameter_list|,
specifier|const
name|char
modifier|*
name|s
parameter_list|,
name|char
modifier|*
name|buf
parameter_list|,
modifier|...
parameter_list|)
block|{
name|char
modifier|*
name|b
decl_stmt|;
name|va_list
name|ap
decl_stmt|;
name|va_start
argument_list|(
name|ap
argument_list|,
name|buf
argument_list|)
expr_stmt|;
name|printf
argument_list|(
name|s
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string literal}}
name|vprintf
argument_list|(
name|s
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// // no-warning
name|fprintf
argument_list|(
name|fp
argument_list|,
name|s
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string literal}}
name|vfprintf
argument_list|(
name|fp
argument_list|,
name|s
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// no-warning
name|asprintf
argument_list|(
operator|&
name|b
argument_list|,
name|s
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string lit}}
name|vasprintf
argument_list|(
operator|&
name|b
argument_list|,
name|s
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// no-warning
name|sprintf
argument_list|(
name|buf
argument_list|,
name|s
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string literal}}
name|snprintf
argument_list|(
name|buf
argument_list|,
literal|2
argument_list|,
name|s
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string lit}}
name|__builtin___sprintf_chk
argument_list|(
name|buf
argument_list|,
literal|0
argument_list|,
operator|-
literal|1
argument_list|,
name|s
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string literal}}
name|__builtin___snprintf_chk
argument_list|(
name|buf
argument_list|,
literal|2
argument_list|,
literal|0
argument_list|,
operator|-
literal|1
argument_list|,
name|s
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string lit}}
name|vsprintf
argument_list|(
name|buf
argument_list|,
name|s
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// no-warning
name|vsnprintf
argument_list|(
name|buf
argument_list|,
literal|2
argument_list|,
name|s
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// no-warning
name|vsnprintf
argument_list|(
name|buf
argument_list|,
literal|2
argument_list|,
name|global_fmt
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string literal}}
name|__builtin___vsnprintf_chk
argument_list|(
name|buf
argument_list|,
literal|2
argument_list|,
literal|0
argument_list|,
operator|-
literal|1
argument_list|,
name|s
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// no-warning
name|__builtin___vsnprintf_chk
argument_list|(
name|buf
argument_list|,
literal|2
argument_list|,
literal|0
argument_list|,
operator|-
literal|1
argument_list|,
name|global_fmt
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string literal}}
comment|// rdar://6079877
name|printf
argument_list|(
literal|"abc"
literal|"%*d"
argument_list|,
operator|(
name|unsigned
operator|)
literal|1
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-warning {{field width should have type 'int'}}
name|printf
argument_list|(
literal|"abc\ def"
literal|"%*d"
argument_list|,
operator|(
name|unsigned
operator|)
literal|1
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-warning {{field width should have type 'int'}}
block|}
end_function

begin_function
name|void
name|check_conditional_literal
parameter_list|(
specifier|const
name|char
modifier|*
name|s
parameter_list|,
name|int
name|i
parameter_list|)
block|{
name|printf
argument_list|(
name|i
operator|==
literal|1
condition|?
literal|"yes"
else|:
literal|"no"
argument_list|)
expr_stmt|;
comment|// no-warning
name|printf
argument_list|(
name|i
operator|==
literal|0
condition|?
operator|(
name|i
operator|==
literal|1
condition|?
literal|"yes"
else|:
literal|"no"
operator|)
else|:
literal|"dont know"
argument_list|)
expr_stmt|;
comment|// no-warning
name|printf
argument_list|(
name|i
operator|==
literal|0
condition|?
operator|(
name|i
operator|==
literal|1
condition|?
name|s
else|:
literal|"no"
operator|)
else|:
literal|"dont know"
argument_list|)
expr_stmt|;
comment|// expected-warning{{format string is not a string literal}}
block|}
end_function

begin_function
name|void
name|check_writeback_specifier
parameter_list|()
block|{
name|int
name|x
decl_stmt|;
name|char
modifier|*
name|b
decl_stmt|;
name|printf
argument_list|(
literal|"%n"
argument_list|,
operator|&
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{'%n' in format string discouraged}}
name|sprintf
argument_list|(
name|b
argument_list|,
literal|"%d%%%n"
argument_list|,
literal|1
argument_list|,
operator|&
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{'%n' in format string dis}}
block|}
end_function

begin_function
name|void
name|check_invalid_specifier
parameter_list|(
name|FILE
modifier|*
name|fp
parameter_list|,
name|char
modifier|*
name|buf
parameter_list|)
block|{
name|printf
argument_list|(
literal|"%s%lb%d"
argument_list|,
literal|"unix"
argument_list|,
literal|10
argument_list|,
literal|20
argument_list|)
expr_stmt|;
comment|// expected-warning {{lid conversion '%lb'}}
name|fprintf
argument_list|(
name|fp
argument_list|,
literal|"%%%l"
argument_list|)
expr_stmt|;
comment|// expected-warning {{lid conversion '%l'}}
name|sprintf
argument_list|(
name|buf
argument_list|,
literal|"%%%%%ld%d%d"
argument_list|,
literal|1
argument_list|,
literal|2
argument_list|,
literal|3
argument_list|)
expr_stmt|;
comment|// no-warning
name|snprintf
argument_list|(
name|buf
argument_list|,
literal|2
argument_list|,
literal|"%%%%%ld%;%d"
argument_list|,
literal|1
argument_list|,
literal|2
argument_list|,
literal|3
argument_list|)
expr_stmt|;
comment|// expected-warning {{sion '%;'}}
block|}
end_function

begin_function
name|void
name|check_null_char_string
parameter_list|(
name|char
modifier|*
name|b
parameter_list|)
block|{
name|printf
argument_list|(
literal|"\0this is bogus%d"
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-warning {{string contains '\0'}}
name|snprintf
argument_list|(
name|b
argument_list|,
literal|10
argument_list|,
literal|"%%%%%d\0%d"
argument_list|,
literal|1
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-warning {{string contains '\0'}}
name|printf
argument_list|(
literal|"%\0d"
argument_list|,
literal|1
argument_list|)
expr_stmt|;
comment|// expected-warning {{string contains '\0'}}
block|}
end_function

begin_function
name|void
name|check_empty_format_string
parameter_list|(
name|char
modifier|*
name|buf
parameter_list|,
modifier|...
parameter_list|)
block|{
name|va_list
name|ap
decl_stmt|;
name|va_start
argument_list|(
name|ap
argument_list|,
name|buf
argument_list|)
expr_stmt|;
name|vprintf
argument_list|(
literal|""
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is empty}}
name|sprintf
argument_list|(
name|buf
argument_list|,
literal|""
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is empty}}
block|}
end_function

begin_function
name|void
name|check_wide_string
parameter_list|(
name|char
modifier|*
name|b
parameter_list|,
modifier|...
parameter_list|)
block|{
name|va_list
name|ap
decl_stmt|;
name|va_start
argument_list|(
name|ap
argument_list|,
name|b
argument_list|)
expr_stmt|;
name|printf
argument_list|(
literal|L"foo %d"
argument_list|,
literal|2
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible pointer types}}, expected-warning {{should not be a wide string}}
name|vsprintf
argument_list|(
name|b
argument_list|,
literal|L"bar %d"
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// expected-warning {{incompatible pointer types}}, expected-warning {{should not be a wide string}}
block|}
end_function

begin_function
name|void
name|check_asterisk_precision_width
parameter_list|(
name|int
name|x
parameter_list|)
block|{
name|printf
argument_list|(
literal|"%*d"
argument_list|)
expr_stmt|;
comment|// expected-warning {{'*' specified field width is missing a matching 'int' argument}}
name|printf
argument_list|(
literal|"%.*d"
argument_list|)
expr_stmt|;
comment|// expected-warning {{'.*' specified field precision is missing a matching 'int' argument}}
name|printf
argument_list|(
literal|"%*d"
argument_list|,
literal|12
argument_list|,
name|x
argument_list|)
expr_stmt|;
comment|// no-warning
name|printf
argument_list|(
literal|"%*d"
argument_list|,
literal|"foo"
argument_list|,
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{field width should have type 'int', but argument has type 'char *'}}
name|printf
argument_list|(
literal|"%.*d"
argument_list|,
literal|"foo"
argument_list|,
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{field precision should have type 'int', but argument has type 'char *'}}
block|}
end_function

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|format
argument_list|(
name|printf
argument_list|,
literal|1
argument_list|,
literal|3
argument_list|)
operator|)
argument_list|)
name|myprintf
argument_list|(
specifier|const
name|char
operator|*
argument_list|,
name|int
name|blah
argument_list|,
operator|...
argument_list|)
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|test_myprintf
parameter_list|()
block|{
name|myprintf
argument_list|(
literal|"%d"
argument_list|,
literal|17
argument_list|,
literal|18
argument_list|)
expr_stmt|;
comment|// okay
block|}
end_function

begin_function
name|void
name|test_constant_bindings
parameter_list|(
name|void
parameter_list|)
block|{
specifier|const
name|char
modifier|*
specifier|const
name|s1
init|=
literal|"hello"
decl_stmt|;
specifier|const
name|char
name|s2
index|[]
init|=
literal|"hello"
decl_stmt|;
specifier|const
name|char
modifier|*
name|s3
init|=
literal|"hello"
decl_stmt|;
name|char
modifier|*
specifier|const
name|s4
init|=
literal|"hello"
decl_stmt|;
specifier|extern
specifier|const
name|char
name|s5
index|[]
decl_stmt|;
name|printf
argument_list|(
name|s1
argument_list|)
expr_stmt|;
comment|// no-warning
name|printf
argument_list|(
name|s2
argument_list|)
expr_stmt|;
comment|// no-warning
name|printf
argument_list|(
name|s3
argument_list|)
expr_stmt|;
comment|// expected-warning{{not a string literal}}
name|printf
argument_list|(
name|s4
argument_list|)
expr_stmt|;
comment|// expected-warning{{not a string literal}}
name|printf
argument_list|(
name|s5
argument_list|)
expr_stmt|;
comment|// expected-warning{{not a string literal}}
block|}
end_function

begin_comment
comment|// Test what happens when -Wformat-security only.
end_comment

begin_pragma
pragma|#
directive|pragma
name|GCC
name|diagnostic
name|ignored
literal|"-Wformat-nonliteral"
end_pragma

begin_pragma
pragma|#
directive|pragma
name|GCC
name|diagnostic
name|warning
literal|"-Wformat-security"
end_pragma

begin_function
name|void
name|test9
parameter_list|(
name|char
modifier|*
name|P
parameter_list|)
block|{
name|int
name|x
decl_stmt|;
name|printf
argument_list|(
name|P
argument_list|)
expr_stmt|;
comment|// expected-warning {{format string is not a string literal (potentially insecure)}}
name|printf
argument_list|(
name|P
argument_list|,
literal|42
argument_list|)
expr_stmt|;
name|printf
argument_list|(
literal|"%n"
argument_list|,
operator|&
name|x
argument_list|)
expr_stmt|;
comment|// expected-warning {{use of '%n' in format string discouraged }}
block|}
end_function

begin_function
name|void
name|torture
parameter_list|(
name|va_list
name|v8
parameter_list|)
block|{
name|vprintf
argument_list|(
literal|"%*.*d"
argument_list|,
name|v8
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
end_function

end_unit

