begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|/* RUN: %clang_cc1 -fsyntax-only -verify -triple i386-apple-darwin9 -pedantic -std=c89 %s  */
end_comment

begin_function_decl
name|int
name|scanf
parameter_list|(
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
modifier|...
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|int
name|printf
parameter_list|(
specifier|const
name|char
modifier|*
specifier|restrict
parameter_list|,
modifier|...
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|foo
parameter_list|(
name|char
modifier|*
modifier|*
name|sp
parameter_list|,
name|float
modifier|*
name|fp
parameter_list|,
name|int
modifier|*
name|ip
parameter_list|)
block|{
name|scanf
argument_list|(
literal|"%as"
argument_list|,
name|sp
argument_list|)
expr_stmt|;
comment|/* expected-warning{{'a' length modifier is not supported by ISO C}} */
name|scanf
argument_list|(
literal|"%a[abc]"
argument_list|,
name|sp
argument_list|)
expr_stmt|;
comment|/* expected-warning{{'a' length modifier is not supported by ISO C}} */
comment|/* TODO: Warn that the 'a' conversion specifier is a C99 feature. */
name|scanf
argument_list|(
literal|"%a"
argument_list|,
name|fp
argument_list|)
expr_stmt|;
name|scanf
argument_list|(
literal|"%afoobar"
argument_list|,
name|fp
argument_list|)
expr_stmt|;
name|printf
argument_list|(
literal|"%a"
argument_list|,
literal|1.0
argument_list|)
expr_stmt|;
name|printf
argument_list|(
literal|"%as"
argument_list|,
literal|1.0
argument_list|)
expr_stmt|;
name|printf
argument_list|(
literal|"%aS"
argument_list|,
literal|1.0
argument_list|)
expr_stmt|;
name|printf
argument_list|(
literal|"%a["
argument_list|,
literal|1.0
argument_list|)
expr_stmt|;
name|printf
argument_list|(
literal|"%afoo"
argument_list|,
literal|1.0
argument_list|)
expr_stmt|;
name|scanf
argument_list|(
literal|"%da"
argument_list|,
name|ip
argument_list|)
expr_stmt|;
comment|/* Test argument type check for the 'a' length modifier. */
name|scanf
argument_list|(
literal|"%as"
argument_list|,
name|fp
argument_list|)
expr_stmt|;
comment|/* expected-warning{{format specifies type 'char **' but the argument has type 'float *'}}                        expected-warning{{'a' length modifier is not supported by ISO C}} */
name|scanf
argument_list|(
literal|"%aS"
argument_list|,
name|fp
argument_list|)
expr_stmt|;
comment|/* expected-warning{{format specifies type 'wchar_t **' (aka 'int **') but the argument has type 'float *'}}                        expected-warning{{'a' length modifier is not supported by ISO C}}                        expected-warning{{'S' conversion specifier is not supported by ISO C}} */
name|scanf
argument_list|(
literal|"%a[abc]"
argument_list|,
name|fp
argument_list|)
expr_stmt|;
comment|/* expected-warning{{format specifies type 'char **' but the argument has type 'float *'}}                            expected-warning{{'a' length modifier is not supported by ISO C}} */
block|}
end_function

end_unit

