begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -std=c1x -fsyntax-only -verify %s
end_comment

begin_function
name|void
name|foo
parameter_list|(
name|int
name|n
parameter_list|)
block|{
operator|(
name|void
operator|)
generic_selection|_Generic(
literal|0
generic_selection|,
type|struct
name|A
association|:
literal|0
association_list|,
comment|// expected-error {{type 'struct A' in generic association incomplete}}
name|void
expr|(
block|)
function|: 0
operator|,
comment|// expected-error {{type 'void ()' in generic association not an object type}}
function|int[n]: 0
end_function

begin_empty_stmt
unit|)
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-error {{type 'int [n]' in generic association is a variably modified type}}
end_comment

begin_expr_stmt
operator|(
name|void
operator|)
generic_selection|_Generic(
literal|0
generic_selection|,
name|void
expr|(
operator|*
end_expr_stmt

begin_expr_stmt
unit|)
operator|(
operator|)
operator|:
literal|0
operator|,
comment|// expected-note {{compatible type 'void (*)()' specified here}}
name|void
argument_list|(
operator|*
argument_list|)
argument_list|(
name|void
argument_list|)
operator|:
literal|0
end_expr_stmt

begin_empty_stmt
unit|)
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-error {{type 'void (*)(void)' in generic association compatible with previously specified type 'void (*)()'}}
end_comment

begin_expr_stmt
operator|(
name|void
operator|)
generic_selection|_Generic(
operator|(
name|void
argument_list|(
operator|*
argument_list|)
operator|(
operator|)
operator|)
literal|0
generic_selection|,
comment|// expected-error {{controlling expression type 'void (*)()' compatible with 2 generic association types}}
name|void
expr|(
operator|*
end_expr_stmt

begin_expr_stmt
unit|)
operator|(
name|int
operator|)
operator|:
literal|0
operator|,
comment|// expected-note {{compatible type 'void (*)(int)' specified here}}
name|void
argument_list|(
operator|*
argument_list|)
argument_list|(
name|void
argument_list|)
operator|:
literal|0
end_expr_stmt

begin_empty_stmt
unit|)
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note {{compatible type 'void (*)(void)' specified here}}
end_comment

begin_expr_stmt
operator|(
name|void
operator|)
generic_selection|_Generic(
literal|0
generic_selection|,
comment|// expected-error {{controlling expression type 'int' not compatible with any generic association type}}
name|char
association|:
literal|0
association_list|,
name|short
association|:
literal|0
association_list|,
name|long
association|:
literal|0
generic_selection|)
expr_stmt|;
end_expr_stmt

begin_decl_stmt
name|int
name|a1
index|[
generic_selection|_Generic(
literal|0
generic_selection|,
name|int
association|:
literal|1
association_list|,
name|short
association|:
literal|2
association_list|,
name|float
association|:
literal|3
association_list|,
type|default
association|:
literal|4
generic_selection|)
operator|==
literal|1
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|a2
index|[
generic_selection|_Generic(
literal|0
generic_selection|,
type|default
association|:
literal|1
association_list|,
name|short
association|:
literal|2
association_list|,
name|float
association|:
literal|3
association_list|,
name|int
association|:
literal|4
generic_selection|)
operator|==
literal|4
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|a3
index|[
generic_selection|_Generic(
literal|0L
generic_selection|,
name|int
association|:
literal|1
association_list|,
name|short
association|:
literal|2
association_list|,
name|float
association|:
literal|3
association_list|,
type|default
association|:
literal|4
generic_selection|)
operator|==
literal|4
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|a4
index|[
generic_selection|_Generic(
literal|0L
generic_selection|,
type|default
association|:
literal|1
association_list|,
name|short
association|:
literal|2
association_list|,
name|float
association|:
literal|3
association_list|,
name|int
association|:
literal|4
generic_selection|)
operator|==
literal|1
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|a5
index|[
generic_selection|_Generic(
literal|0
generic_selection|,
name|int
association|:
literal|1
association_list|,
name|short
association|:
literal|2
association_list|,
name|float
association|:
literal|3
generic_selection|)
operator|==
literal|1
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|int
name|a6
index|[
generic_selection|_Generic(
literal|0
generic_selection|,
name|short
association|:
literal|1
association_list|,
name|float
association|:
literal|2
association_list|,
name|int
association|:
literal|3
generic_selection|)
operator|==
literal|3
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
end_decl_stmt

unit|}
end_unit

