begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -verify -fsyntax-only
end_comment

begin_typedef
typedef|typedef
name|int
name|int32_t
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|char
name|Boolean
typedef|;
end_typedef

begin_function
name|void
name|func
parameter_list|()
block|{
name|int32_t
modifier|*
name|vector
index|[
literal|16
index|]
decl_stmt|;
specifier|const
name|char
name|compDesc
index|[
literal|16
operator|+
literal|1
index|]
decl_stmt|;
name|int32_t
name|compCount
init|=
literal|0
decl_stmt|;
if|if
condition|(
name|_CFCalendarDecomposeAbsoluteTimeV
argument_list|(
name|compDesc
argument_list|,
name|vector
argument_list|,
name|compCount
argument_list|)
condition|)
block|{
comment|// expected-note {{previous implicit declaration is here}} \
name|expected
operator|-
name|warning
block|{
block|{
name|implicit
name|declaration
name|of
name|function
literal|'_CFCalendarDecomposeAbsoluteTimeV'
name|is
name|invalid
name|in
name|C99
block|}
block|}
block|}
return|return
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
return|;
comment|// expected-warning {{void function 'func' should not return a value}}
block|}
end_function

begin_function
name|Boolean
name|_CFCalendarDecomposeAbsoluteTimeV
parameter_list|(
specifier|const
name|char
modifier|*
name|componentDesc
parameter_list|,
name|int32_t
modifier|*
modifier|*
name|vector
parameter_list|,
name|int32_t
name|count
parameter_list|)
block|{
comment|// expected-error{{conflicting types for '_CFCalendarDecomposeAbsoluteTimeV'}}
return|return
literal|0
return|;
block|}
end_function

end_unit

