begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -verify -fsyntax-only
end_comment

begin_typedef
typedef|typedef
name|unsigned
name|long
name|long
name|uint64_t
typedef|;
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|long
name|long
name|uint32_t
typedef|;
end_typedef

begin_function_decl
name|uint64_t
name|f0
parameter_list|(
name|uint64_t
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|uint64_t
name|f1
parameter_list|(
name|uint64_t
parameter_list|,
name|uint32_t
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|uint64_t
name|f2
parameter_list|(
name|uint64_t
parameter_list|,
modifier|...
parameter_list|)
function_decl|;
end_function_decl

begin_decl_stmt
specifier|static
specifier|const
name|uint64_t
name|overflow
init|=
literal|1
operator|*
literal|4608
operator|*
literal|1024
operator|*
literal|1024
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning {{overflow in expression; result is 536870912 with type 'int'}}
end_comment

begin_function
name|uint64_t
name|check_integer_overflows
parameter_list|(
name|int
name|i
parameter_list|)
block|{
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
name|uint64_t
name|overflow
init|=
literal|4608
operator|*
literal|1024
operator|*
literal|1024
decl_stmt|,
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
name|overflow2
init|=
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
decl_stmt|,
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
name|overflow3
init|=
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
operator|*
name|i
argument_list|)
decl_stmt|,
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
name|overflow4
init|=
operator|(
literal|1ULL
operator|*
operator|(
operator|(
literal|4608
operator|)
operator|*
operator|(
operator|(
literal|1024
operator|)
operator|*
operator|(
literal|1024
operator|)
operator|)
operator|)
operator|+
literal|2ULL
operator|)
decl_stmt|,
comment|// expected-warning@+1 2{{overflow in expression; result is 536870912 with type 'int'}}
name|multi_overflow
init|=
call|(
name|uint64_t
call|)
argument_list|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
operator|*
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
argument_list|)
decl_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
name|overflow
operator|+=
name|overflow2
operator|=
name|overflow3
operator|=
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
expr_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
name|overflow
operator|+=
name|overflow2
operator|=
name|overflow3
operator|=
literal|4608
operator|*
literal|1024
operator|*
literal|1024
expr_stmt|;
name|uint64_t
name|not_overflow
init|=
literal|4608
operator|*
literal|1024
operator|*
literal|1024ULL
decl_stmt|;
name|uint64_t
name|not_overflow2
init|=
operator|(
literal|1ULL
operator|*
operator|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
argument_list|)
operator|*
operator|(
literal|1024
operator|*
literal|1024
operator|)
operator|)
operator|+
literal|2ULL
operator|)
decl_stmt|;
comment|// expected-warning@+1 2{{overflow in expression; result is 536870912 with type 'int'}}
name|overflow
operator|=
literal|4608
operator|*
literal|1024
operator|*
literal|1024
condition|?
literal|4608
operator|*
literal|1024
operator|*
literal|1024
else|:
literal|0
expr_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
name|overflow
operator|=
literal|0
condition|?
literal|0
else|:
literal|4608
operator|*
literal|1024
operator|*
literal|1024
expr_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
if|if
condition|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
condition|)
return|return
literal|0
return|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
if|if
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
condition|)
return|return
literal|1
return|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
if|if
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
condition|)
return|return
literal|2
return|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
if|if
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
operator|*
name|i
argument_list|)
condition|)
return|return
literal|3
return|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
if|if
condition|(
operator|(
literal|1ULL
operator|*
operator|(
operator|(
literal|4608
operator|)
operator|*
operator|(
operator|(
literal|1024
operator|)
operator|*
operator|(
literal|1024
operator|)
operator|)
operator|)
operator|+
literal|2ULL
operator|)
condition|)
return|return
literal|4
return|;
comment|// expected-warning@+1 2{{overflow in expression; result is 536870912 with type 'int'}}
if|if
condition|(
call|(
name|uint64_t
call|)
argument_list|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
operator|*
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
argument_list|)
condition|)
return|return
literal|5
return|;
switch|switch
condition|(
name|i
condition|)
block|{
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
case|case
literal|4608
operator|*
literal|1024
operator|*
literal|1024
case|:
return|return
literal|6
return|;
comment|// expected-warning@+1 {{overflow in expression; result is 537919488 with type 'int'}}
case|case
call|(
name|uint64_t
call|)
argument_list|(
literal|4609
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
case|:
return|return
literal|7
return|;
comment|// expected-error@+1 {{expression is not an integer constant expression}}
case|case
operator|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
operator|*
name|i
argument_list|)
operator|)
case|:
return|return
literal|8
return|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
case|case
operator|(
operator|(
literal|1ULL
operator|*
operator|(
operator|(
literal|4608
operator|)
operator|*
operator|(
operator|(
literal|1024
operator|)
operator|*
operator|(
literal|1024
operator|)
operator|)
operator|)
operator|+
literal|2ULL
operator|)
operator|)
case|:
return|return
literal|9
return|;
comment|// expected-warning@+2 2{{overflow in expression; result is 536870912 with type 'int'}}
comment|// expected-warning@+1 {{overflow converting case value to switch condition type (288230376151711744 to 0)}}
case|case
operator|(
call|(
name|uint64_t
call|)
argument_list|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
operator|*
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
argument_list|)
operator|)
case|:
return|return
literal|10
return|;
block|}
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
while|while
condition|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
condition|)
empty_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
while|while
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
condition|)
empty_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
while|while
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
condition|)
empty_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
while|while
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
operator|*
name|i
argument_list|)
condition|)
empty_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
while|while
condition|(
operator|(
literal|1ULL
operator|*
operator|(
operator|(
literal|4608
operator|)
operator|*
operator|(
operator|(
literal|1024
operator|)
operator|*
operator|(
literal|1024
operator|)
operator|)
operator|)
operator|+
literal|2ULL
operator|)
condition|)
empty_stmt|;
comment|// expected-warning@+1 2{{overflow in expression; result is 536870912 with type 'int'}}
while|while
condition|(
call|(
name|uint64_t
call|)
argument_list|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
operator|*
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
argument_list|)
condition|)
empty_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
do|do
block|{ }
do|while
condition|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
condition|)
do|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
do|do
block|{ }
do|while
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
condition|)
do|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
do|do
block|{ }
do|while
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
condition|)
do|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
do|do
block|{ }
do|while
condition|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
operator|*
name|i
argument_list|)
condition|)
do|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
do|do
block|{ }
do|while
condition|(
operator|(
literal|1ULL
operator|*
operator|(
operator|(
literal|4608
operator|)
operator|*
operator|(
operator|(
literal|1024
operator|)
operator|*
operator|(
literal|1024
operator|)
operator|)
operator|)
operator|+
literal|2ULL
operator|)
condition|)
do|;
comment|// expected-warning@+1 2{{overflow in expression; result is 536870912 with type 'int'}}
do|do
block|{ }
do|while
condition|(
call|(
name|uint64_t
call|)
argument_list|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
operator|*
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
argument_list|)
condition|)
do|;
comment|// expected-warning@+3 {{overflow in expression; result is 536870912 with type 'int'}}
comment|// expected-warning@+3 {{overflow in expression; result is 536870912 with type 'int'}}
comment|// expected-warning@+3 {{overflow in expression; result is 536870912 with type 'int'}}
for|for
control|(
name|uint64_t
name|i
init|=
literal|4608
operator|*
literal|1024
operator|*
literal|1024
init|;
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
condition|;
name|i
operator|+=
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
operator|*
name|i
argument_list|)
control|)
empty_stmt|;
comment|// expected-warning@+3 {{overflow in expression; result is 536870912 with type 'int'}}
comment|// expected-warning@+3 2{{overflow in expression; result is 536870912 with type 'int'}}
comment|// expected-warning@+3 2{{overflow in expression; result is 536870912 with type 'int'}}
for|for
control|(
name|uint64_t
name|i
init|=
operator|(
literal|1ULL
operator|*
operator|(
operator|(
literal|4608
operator|)
operator|*
operator|(
operator|(
literal|1024
operator|)
operator|*
operator|(
literal|1024
operator|)
operator|)
operator|)
operator|+
literal|2ULL
operator|)
init|;
operator|(
call|(
name|uint64_t
call|)
argument_list|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
operator|*
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
argument_list|)
operator|)
condition|;
name|i
operator|=
operator|(
operator|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
operator|)
operator|+
operator|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
operator|)
operator|)
control|)
empty_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
specifier|_Complex
name|long
name|long
name|x
init|=
literal|4608
operator|*
literal|1024
operator|*
literal|1024
decl_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
operator|(
name|__real__
name|x
operator|)
operator|=
literal|4608
operator|*
literal|1024
operator|*
literal|1024
expr_stmt|;
comment|// expected-warning@+1 {{overflow in expression; result is 536870912 with type 'int'}}
operator|(
name|__imag__
name|x
operator|)
operator|=
literal|4608
operator|*
literal|1024
operator|*
literal|1024
expr_stmt|;
comment|// expected-warning@+4 {{overflow in expression; result is 536870912 with type 'int'}}
comment|// expected-warning@+3 {{array index 536870912 is past the end of the array (which contains 10 elements)}}
comment|// expected-note@+1 {{array 'a' declared here}}
name|uint64_t
name|a
index|[
literal|10
index|]
decl_stmt|;
name|a
index|[
literal|4608
operator|*
literal|1024
operator|*
literal|1024
index|]
operator|=
literal|1i
expr_stmt|;
comment|// expected-warning@+1 2{{overflow in expression; result is 536870912 with type 'int'}}
return|return
operator|(
operator|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
operator|)
operator|+
operator|(
call|(
name|uint64_t
call|)
argument_list|(
literal|4608
operator|*
literal|1024
operator|*
literal|1024
argument_list|)
operator|)
operator|)
return|;
block|}
end_function

begin_struct
struct|struct
name|s
block|{
name|unsigned
name|x
decl_stmt|;
name|unsigned
name|y
decl_stmt|;
block|}
name|s
init|=
block|{
operator|.
name|y
operator|=
literal|5
block|,
operator|.
name|x
operator|=
literal|4
operator|*
literal|1024
operator|*
literal|1024
operator|*
literal|1024
comment|// expected-warning {{overflow in expression; result is 0 with type 'int'}}
block|}
struct|;
end_struct

end_unit

