begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -verify -fsyntax-only
end_comment

begin_typedef
typedef|typedef
name|struct
name|_zend_module_entry
name|zend_module_entry
typedef|;
end_typedef

begin_struct
struct|struct
name|_zend_module_entry
block|{
name|_efree
argument_list|(
operator|(
name|p
operator|)
argument_list|)
expr_stmt|;
comment|// expected-error{{type name requires a specifier or qualifier}} \
name|expected
operator|-
name|error
block|{
block|{
name|field
literal|'_efree'
name|declared
name|as
name|a
name|function
block|}
block|}
expr|\
name|expected
operator|-
name|warning
block|{
block|{
name|type
name|specifier
name|missing
block|,
name|defaults
name|to
literal|'int'
block|}
block|}
expr|\
name|expected
operator|-
name|warning
block|{
block|{
name|type
name|specifier
name|missing
block|,
name|defaults
name|to
literal|'int'
block|}
block|}
block|}
struct|;
end_struct

begin_typedef
typedef|typedef
struct|struct
name|_zend_function_entry
block|{ }
name|zend_function_entry
typedef|;
end_typedef

begin_typedef
typedef|typedef
struct|struct
name|_zend_pcre_globals
block|{ }
name|zend_pcre_globals
typedef|;
end_typedef

begin_decl_stmt
name|zend_pcre_globals
name|pcre_globals
decl_stmt|;
end_decl_stmt

begin_function
specifier|static
name|void
name|zm_globals_ctor_pcre
parameter_list|(
name|zend_pcre_globals
modifier|*
name|pcre_globals
parameter_list|)
block|{ }
end_function

begin_function
specifier|static
name|void
name|zm_globals_dtor_pcre
parameter_list|(
name|zend_pcre_globals
modifier|*
name|pcre_globals
parameter_list|)
block|{ }
end_function

begin_function
specifier|static
name|void
name|zm_info_pcre
parameter_list|(
name|zend_module_entry
modifier|*
name|zend_module
parameter_list|)
block|{ }
end_function

begin_function
specifier|static
name|int
name|zm_startup_pcre
parameter_list|(
name|int
name|type
parameter_list|,
name|int
name|module_number
parameter_list|)
block|{ }
end_function

begin_function
specifier|static
name|int
name|zm_shutdown_pcre
parameter_list|(
name|int
name|type
parameter_list|,
name|int
name|module_number
parameter_list|)
block|{
name|zend_function_entry
name|pcre_functions
index|[]
init|=
block|{
block|{ }
block|;
comment|// expected-error{{expected '}'}} expected-note {{to match this '{'}}
name|zend_module_entry
name|pcre_module_entry
operator|=
block|{
sizeof|sizeof
argument_list|(
name|zend_module_entry
argument_list|)
block|,
literal|20071006
block|,
literal|0
block|,
literal|0
block|,
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
block|,
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
block|,
literal|"pcre"
block|,
name|pcre_functions
block|,
name|zm_startup_pcre
block|,
name|zm_shutdown_pcre
block|,
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
block|,
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
block|,
name|zm_info_pcre
block|,
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
block|,
sizeof|sizeof
argument_list|(
name|zend_pcre_globals
argument_list|)
block|,
operator|&
name|pcre_globals
block|,
operator|(
operator|(
name|void
argument_list|(
operator|*
argument_list|)
argument_list|(
name|void
operator|*
argument_list|)
operator|)
operator|(
name|zm_globals_ctor_pcre
operator|)
operator|)
block|,
operator|(
operator|(
name|void
argument_list|(
operator|*
argument_list|)
argument_list|(
name|void
operator|*
argument_list|)
operator|)
operator|(
name|zm_globals_dtor_pcre
operator|)
operator|)
block|,
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
block|,
literal|0
block|,
literal|0
block|,
operator|(
operator|(
name|void
operator|*
operator|)
literal|0
operator|)
block|,
literal|0
block|}
block|; }
end_function

end_unit

