begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -Wmissing-field-initializers %s
end_comment

begin_comment
comment|// This was PR4808.
end_comment

begin_struct
struct|struct
name|Foo
block|{
name|int
name|a
decl_stmt|,
name|b
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|Foo
name|foo0
init|=
block|{
literal|1
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning {{missing field 'b' initializer}}
end_comment

begin_decl_stmt
name|struct
name|Foo
name|foo1
init|=
block|{
operator|.
name|a
operator|=
literal|1
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// designator avoids MFI warning
end_comment

begin_decl_stmt
name|struct
name|Foo
name|foo2
init|=
block|{
operator|.
name|b
operator|=
literal|1
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// designator avoids MFI warning
end_comment

begin_decl_stmt
name|struct
name|Foo
name|bar0
index|[]
init|=
block|{
block|{
literal|1
block|,
literal|2
block|}
block|,
block|{
literal|1
block|}
block|,
comment|// expected-warning {{missing field 'b' initializer}}
block|{
literal|1
block|,
literal|2
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|Foo
name|bar1
index|[]
init|=
block|{
literal|1
block|,
literal|2
block|,
literal|1
block|,
literal|2
block|,
literal|1
block|}
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-warning {{missing field 'b' initializer}}
end_comment

begin_struct
struct|struct
name|One
block|{
name|int
name|a
decl_stmt|;
name|int
name|b
decl_stmt|;
block|}
struct|;
end_struct

begin_struct
struct|struct
name|Two
block|{
name|float
name|c
decl_stmt|;
name|float
name|d
decl_stmt|;
name|float
name|e
decl_stmt|;
block|}
struct|;
end_struct

begin_struct
struct|struct
name|Three
block|{
union|union
block|{
name|struct
name|One
name|one
decl_stmt|;
name|struct
name|Two
name|two
decl_stmt|;
block|}
name|both
union|;
block|}
struct|;
end_struct

begin_decl_stmt
name|struct
name|Three
name|t0
init|=
block|{
block|{
operator|.
name|one
operator|=
block|{
literal|1
block|,
literal|2
block|}
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|Three
name|t1
init|=
block|{
block|{
operator|.
name|two
operator|=
block|{
literal|1.0f
block|,
literal|2.0f
block|,
literal|3.0f
block|}
block|}
block|}
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|struct
name|Three
name|data
index|[]
init|=
block|{
block|{
block|{
operator|.
name|one
operator|=
block|{
literal|1
block|,
literal|2
block|}
block|}
block|}
block|,
block|{
block|{
operator|.
name|one
operator|=
block|{
literal|1
block|}
block|}
block|}
block|,
comment|// expected-warning {{missing field 'b' initializer}}
block|{
block|{
operator|.
name|two
operator|=
block|{
literal|1.0f
block|,
literal|2.0f
block|,
literal|3.0f
block|}
block|}
block|}
block|,
block|{
block|{
operator|.
name|two
operator|=
block|{
literal|1.0f
block|,
literal|2.0f
block|}
block|}
block|}
comment|// expected-warning {{missing field 'e' initializer}}
block|}
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
block|{
name|int
label|:
literal|5
expr_stmt|;
name|int
name|a
decl_stmt|;
name|int
label|:
literal|5
expr_stmt|;
name|int
name|b
decl_stmt|;
name|int
label|:
literal|5
block|}
name|noNamedImplicit
index|[]
init|=
block|{
block|{
literal|1
block|,
literal|2
block|}
block|,
block|{
literal|1
block|}
comment|// expected-warning {{missing field 'b' initializer}}
block|}
struct|;
end_struct

end_unit

