begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -fblocks -fsyntax-only -verify %s
end_comment

begin_function_decl
name|int
name|f1
parameter_list|(
name|int
name|x
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-warning{{'nonnull' attribute applied to function with no pointer arguments}}
end_comment

begin_function_decl
name|int
name|f2
parameter_list|(
name|int
modifier|*
name|x
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
parameter_list|(
function_decl|1
end_function_decl

begin_empty_stmt
unit|)))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|int
name|f3
parameter_list|(
name|int
modifier|*
name|x
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
parameter_list|(
function_decl|0
end_function_decl

begin_empty_stmt
unit|)))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-error {{'nonnull' attribute parameter 1 is out of bounds}}
end_comment

begin_function_decl
name|int
name|f4
parameter_list|(
name|int
modifier|*
name|x
parameter_list|,
name|int
modifier|*
name|y
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
parameter_list|(
function_decl|1
operator|,
function_decl|2
end_function_decl

begin_empty_stmt
unit|)))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|int
name|f5
parameter_list|(
name|int
modifier|*
name|x
parameter_list|,
name|int
modifier|*
name|y
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
parameter_list|(
function_decl|2
operator|,
function_decl|1
end_function_decl

begin_empty_stmt
unit|)))
empty_stmt|;
end_empty_stmt

begin_function_decl
specifier|extern
name|void
name|func1
parameter_list|(
name|void
function_decl|(
modifier|^
name|block1
function_decl|)
parameter_list|()
parameter_list|,
name|void
function_decl|(
modifier|^
name|block2
function_decl|)
parameter_list|()
parameter_list|,
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
specifier|extern
name|void
name|func3
parameter_list|(
name|void
function_decl|(
modifier|^
name|block1
function_decl|)
parameter_list|()
parameter_list|,
name|int
parameter_list|,
name|void
function_decl|(
modifier|^
name|block2
function_decl|)
parameter_list|()
parameter_list|,
name|int
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
parameter_list|(
function_decl|1
operator|,
function_decl|3
end_function_decl

begin_empty_stmt
unit|)))
empty_stmt|;
end_empty_stmt

begin_function_decl
specifier|extern
name|void
name|func4
parameter_list|(
name|void
function_decl|(
modifier|^
name|block1
function_decl|)
parameter_list|()
parameter_list|,
name|void
function_decl|(
modifier|^
name|block2
function_decl|)
parameter_list|()
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(nonnull
parameter_list|(
function_decl|1
end_function_decl

begin_expr_stmt
unit|)))
name|__attribute__
argument_list|(
operator|(
name|nonnull
argument_list|(
literal|2
argument_list|)
operator|)
argument_list|)
expr_stmt|;
end_expr_stmt

begin_function
name|void
name|foo
parameter_list|(
name|int
name|i1
parameter_list|,
name|int
name|i2
parameter_list|,
name|int
name|i3
parameter_list|,
name|void
function_decl|(
modifier|^
name|cp1
function_decl|)
parameter_list|()
parameter_list|,
name|void
function_decl|(
modifier|^
name|cp2
function_decl|)
parameter_list|()
parameter_list|,
name|void
function_decl|(
modifier|^
name|cp3
function_decl|)
parameter_list|()
parameter_list|)
block|{
name|func1
argument_list|(
name|cp1
argument_list|,
name|cp2
argument_list|,
name|i1
argument_list|)
expr_stmt|;
name|func1
argument_list|(
literal|0
argument_list|,
name|cp2
argument_list|,
name|i1
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee which requires a non-null argument}}
name|func1
argument_list|(
name|cp1
argument_list|,
literal|0
argument_list|,
name|i1
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee which requires a non-null argument}}
name|func1
argument_list|(
name|cp1
argument_list|,
name|cp2
argument_list|,
literal|0
argument_list|)
expr_stmt|;
name|func3
argument_list|(
literal|0
argument_list|,
name|i2
argument_list|,
name|cp3
argument_list|,
name|i3
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee which requires a non-null argument}}
name|func3
argument_list|(
name|cp3
argument_list|,
name|i2
argument_list|,
literal|0
argument_list|,
name|i3
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee which requires a non-null argument}}
name|func4
argument_list|(
literal|0
argument_list|,
name|cp1
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee which requires a non-null argument}}
name|func4
argument_list|(
name|cp1
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning {{null passed to a callee which requires a non-null argument}}
block|}
end_function

end_unit

