begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -fsyntax-only -verify %s
end_comment

begin_define
define|#
directive|define
name|offsetof
parameter_list|(
name|TYPE
parameter_list|,
name|MEMBER
parameter_list|)
value|__builtin_offsetof (TYPE, MEMBER)
end_define

begin_typedef
typedef|typedef
struct|struct
name|P
block|{
name|int
name|i
decl_stmt|;
name|float
name|f
decl_stmt|;
block|}
name|PT
typedef|;
end_typedef

begin_struct
struct|struct
name|external_sun3_core
block|{
name|unsigned
name|c_regs
decl_stmt|;
name|PT
name|X
index|[
literal|100
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|void
name|swap
parameter_list|()
block|{
name|int
name|x
decl_stmt|;
name|x
operator|=
name|offsetof
argument_list|(
expr|struct
name|external_sun3_core
argument_list|,
name|c_regs
argument_list|)
expr_stmt|;
name|x
operator|=
name|__builtin_offsetof
argument_list|(
expr|struct
name|external_sun3_core
argument_list|,
name|X
index|[
literal|42
index|]
operator|.
name|f
argument_list|)
expr_stmt|;
name|x
operator|=
name|__builtin_offsetof
argument_list|(
expr|struct
name|external_sun3_core
argument_list|,
name|X
index|[
literal|42
index|]
operator|.
name|f2
argument_list|)
expr_stmt|;
comment|// expected-error {{no member named 'f2'}}
name|x
operator|=
name|__builtin_offsetof
argument_list|(
name|int
argument_list|,
name|X
index|[
literal|42
index|]
operator|.
name|f2
argument_list|)
expr_stmt|;
comment|// expected-error {{offsetof requires struct}}
name|int
name|a
index|[
name|__builtin_offsetof
argument_list|(
expr|struct
name|external_sun3_core
argument_list|,
name|X
argument_list|)
operator|==
literal|4
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
name|int
name|b
index|[
name|__builtin_offsetof
argument_list|(
expr|struct
name|external_sun3_core
argument_list|,
name|X
index|[
literal|42
index|]
argument_list|)
operator|==
literal|340
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
name|int
name|c
index|[
name|__builtin_offsetof
argument_list|(
expr|struct
name|external_sun3_core
argument_list|,
name|X
index|[
literal|42
index|]
operator|.
name|f2
argument_list|)
operator|==
literal|344
condition|?
literal|1
else|:
operator|-
literal|1
index|]
decl_stmt|;
comment|// expected-error {{no member named 'f2'}}
block|}
end_function

begin_function_decl
specifier|extern
name|int
name|f
parameter_list|()
function_decl|;
end_function_decl

begin_struct
struct|struct
name|s1
block|{
name|int
name|a
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|int
name|v1
init|=
name|offsetof
argument_list|(
expr|struct
name|s1
argument_list|,
name|a
argument_list|)
operator|==
literal|0
condition|?
literal|0
else|:
name|f
argument_list|()
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|s2
block|{
name|int
name|a
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|int
name|v2
init|=
operator|(
name|int
operator|)
operator|(
operator|&
operator|(
operator|(
expr|struct
name|s2
operator|*
operator|)
literal|0
operator|)
operator|->
name|a
operator|)
operator|==
literal|0
condition|?
literal|0
else|:
name|f
argument_list|()
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
name|s3
block|{
name|int
name|a
decl_stmt|;
block|}
struct|;
end_struct

begin_decl_stmt
name|int
name|v3
init|=
name|__builtin_offsetof
argument_list|(
expr|struct
name|s3
argument_list|,
name|a
argument_list|)
operator|==
literal|0
condition|?
literal|0
else|:
name|f
argument_list|()
decl_stmt|;
end_decl_stmt

begin_comment
comment|// PR3396
end_comment

begin_struct
struct|struct
name|sockaddr_un
block|{
name|unsigned
name|char
name|sun_len
decl_stmt|;
name|char
name|sun_path
index|[
literal|104
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|int
name|a
parameter_list|(
name|int
name|len
parameter_list|)
block|{
name|int
name|a
index|[
name|__builtin_offsetof
argument_list|(
expr|struct
name|sockaddr_un
argument_list|,
name|sun_path
index|[
name|len
operator|+
literal|1
index|]
argument_list|)
index|]
decl_stmt|;
block|}
end_function

begin_comment
comment|// PR4079
end_comment

begin_union
union|union
name|x
block|{
struct|struct
block|{
name|int
name|x
decl_stmt|;
block|}
struct|;
block|}
union|;
end_union

begin_macro
name|int
end_macro

begin_expr_stmt
name|x
index|[
name|__builtin_offsetof
argument_list|(
expr|union
name|x
argument_list|,
name|x
argument_list|)
index|]
expr_stmt|;
end_expr_stmt

begin_comment
comment|// rdar://problem/7222956
end_comment

begin_struct_decl
struct_decl|struct
name|incomplete
struct_decl|;
end_struct_decl

begin_comment
comment|// expected-note 2 {{forward declaration of 'struct incomplete'}}
end_comment

begin_decl_stmt
name|int
name|test1
index|[
name|__builtin_offsetof
argument_list|(
expr|struct
name|incomplete
argument_list|,
name|foo
argument_list|)
index|]
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error {{offsetof of incomplete type 'struct incomplete'}}
end_comment

begin_decl_stmt
name|int
name|test1
index|[
name|__builtin_offsetof
argument_list|(
expr|struct
name|incomplete
index|[
literal|10
index|]
argument_list|,
index|[
literal|4
index|]
operator|.
name|foo
argument_list|)
index|]
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error {{array has incomplete element type 'struct incomplete'}}
end_comment

end_unit

