begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-apple-darwin -fsyntax-only -Wtautological-constant-out-of-range-compare -verify %s
end_comment

begin_comment
comment|// rdar://12202422
end_comment

begin_function_decl
name|int
name|value
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|int
name|main
parameter_list|()
block|{
name|int
name|a
init|=
name|value
argument_list|()
decl_stmt|;
if|if
condition|(
name|a
operator|==
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|!=
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|<
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|<=
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|>
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|>=
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|==
name|a
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|!=
name|a
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always true}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|<
name|a
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|<=
name|a
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|>
name|a
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always true}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|>=
name|a
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|==
literal|0x1234567812345678LL
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|==
operator|-
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant -1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|<
operator|-
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant -1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|>
operator|-
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant -1311768465173141112 with expression of type 'int' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|<=
operator|-
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant -1311768465173141112 with expression of type 'int' is always false}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|>=
operator|-
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant -1311768465173141112 with expression of type 'int' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|a
operator|==
literal|0x12345678L
condition|)
comment|// no warning
return|return
literal|1
return|;
name|short
name|s
init|=
name|value
argument_list|()
decl_stmt|;
if|if
condition|(
name|s
operator|==
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always false}}
return|return
literal|0
return|;
if|if
condition|(
name|s
operator|!=
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|s
operator|<
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|s
operator|<=
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always true}}
return|return
literal|0
return|;
if|if
condition|(
name|s
operator|>
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always false}}
return|return
literal|0
return|;
if|if
condition|(
name|s
operator|>=
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always false}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|==
name|s
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always false}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|!=
name|s
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always true}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|<
name|s
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always false}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|<=
name|s
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always false}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|>
name|s
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always true}}
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|>=
name|s
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'short' is always true}}
return|return
literal|0
return|;
name|long
name|l
init|=
name|value
argument_list|()
decl_stmt|;
if|if
condition|(
name|l
operator|==
literal|0x1234567812345678L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|l
operator|!=
literal|0x1234567812345678L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|l
operator|<
literal|0x1234567812345678L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|l
operator|<=
literal|0x1234567812345678L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|l
operator|>
literal|0x1234567812345678L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|l
operator|>=
literal|0x1234567812345678L
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|==
name|l
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|!=
name|l
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|<
name|l
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|<=
name|l
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|>
name|l
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x1234567812345678L
operator|>=
name|l
condition|)
return|return
literal|0
return|;
name|unsigned
name|un
init|=
literal|0
decl_stmt|;
if|if
condition|(
name|un
operator|==
literal|0x0000000000000000L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|un
operator|!=
literal|0x0000000000000000L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|un
operator|<
literal|0x0000000000000000L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|un
operator|<=
literal|0x0000000000000000L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|un
operator|>
literal|0x0000000000000000L
condition|)
return|return
literal|0
return|;
if|if
condition|(
name|un
operator|>=
literal|0x0000000000000000L
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x0000000000000000L
operator|==
name|un
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x0000000000000000L
operator|!=
name|un
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x0000000000000000L
operator|<
name|un
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x0000000000000000L
operator|<=
name|un
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x0000000000000000L
operator|>
name|un
condition|)
return|return
literal|0
return|;
if|if
condition|(
literal|0x0000000000000000L
operator|>=
name|un
condition|)
return|return
literal|0
return|;
name|float
name|fl
init|=
literal|0
decl_stmt|;
if|if
condition|(
name|fl
operator|==
literal|0x0000000000000000L
condition|)
comment|// no warning
return|return
literal|0
return|;
name|float
name|dl
init|=
literal|0
decl_stmt|;
if|if
condition|(
name|dl
operator|==
literal|0x0000000000000000L
condition|)
comment|// no warning
return|return
literal|0
return|;
enum|enum
name|E
block|{
name|yes
block|,
name|no
block|,
name|maybe
block|}
enum|;
name|enum
name|E
name|e
decl_stmt|;
if|if
condition|(
name|e
operator|==
literal|0x1234567812345678L
condition|)
comment|// expected-warning {{comparison of constant 1311768465173141112 with expression of type 'enum E' is always false}}
return|return
literal|0
return|;
return|return
literal|1
return|;
block|}
end_function

end_unit

