begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s
end_comment

begin_function_decl
name|char
modifier|*
name|foo
parameter_list|(
name|float
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note 3 {{candidate function}}
end_comment

begin_function
name|void
name|test_foo_1
parameter_list|(
name|float
name|fv
parameter_list|,
name|double
name|dv
parameter_list|,
name|float
specifier|_Complex
name|fc
parameter_list|,
name|double
specifier|_Complex
name|dc
parameter_list|)
block|{
name|char
modifier|*
name|cp1
init|=
name|foo
argument_list|(
name|fv
argument_list|)
decl_stmt|;
name|char
modifier|*
name|cp2
init|=
name|foo
argument_list|(
name|dv
argument_list|)
decl_stmt|;
comment|// Note: GCC and EDG reject these two, but they are valid C99 conversions
name|char
modifier|*
name|cp3
init|=
name|foo
argument_list|(
name|fc
argument_list|)
decl_stmt|;
name|char
modifier|*
name|cp4
init|=
name|foo
argument_list|(
name|dc
argument_list|)
decl_stmt|;
block|}
end_function

begin_function_decl
name|int
modifier|*
name|foo
parameter_list|(
name|float
specifier|_Complex
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note 3 {{candidate function}}
end_comment

begin_function
name|void
name|test_foo_2
parameter_list|(
name|float
name|fv
parameter_list|,
name|double
name|dv
parameter_list|,
name|float
specifier|_Complex
name|fc
parameter_list|,
name|double
specifier|_Complex
name|dc
parameter_list|)
block|{
name|char
modifier|*
name|cp1
init|=
name|foo
argument_list|(
name|fv
argument_list|)
decl_stmt|;
name|char
modifier|*
name|cp2
init|=
name|foo
argument_list|(
name|dv
argument_list|)
decl_stmt|;
comment|// expected-error{{call to 'foo' is ambiguous; candidates are:}}
name|int
modifier|*
name|ip
init|=
name|foo
argument_list|(
name|fc
argument_list|)
decl_stmt|;
name|int
modifier|*
name|lp
init|=
name|foo
argument_list|(
name|dc
argument_list|)
decl_stmt|;
comment|// expected-error{{call to 'foo' is ambiguous; candidates are:}}
block|}
end_function

begin_function_decl
name|long
modifier|*
name|foo
parameter_list|(
name|double
specifier|_Complex
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note {{candidate function}}
end_comment

begin_function
name|void
name|test_foo_3
parameter_list|(
name|float
name|fv
parameter_list|,
name|double
name|dv
parameter_list|,
name|float
specifier|_Complex
name|fc
parameter_list|,
name|double
specifier|_Complex
name|dc
parameter_list|)
block|{
name|char
modifier|*
name|cp1
init|=
name|foo
argument_list|(
name|fv
argument_list|)
decl_stmt|;
name|char
modifier|*
name|cp2
init|=
name|foo
argument_list|(
name|dv
argument_list|)
decl_stmt|;
comment|// expected-error{{call to 'foo' is ambiguous; candidates are:}}
name|int
modifier|*
name|ip
init|=
name|foo
argument_list|(
name|fc
argument_list|)
decl_stmt|;
name|long
modifier|*
name|lp
init|=
name|foo
argument_list|(
name|dc
argument_list|)
decl_stmt|;
block|}
end_function

begin_function_decl
name|char
modifier|*
name|promote_or_convert
parameter_list|(
name|double
specifier|_Complex
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note 2 {{candidate function}}
end_comment

begin_function_decl
name|int
modifier|*
name|promote_or_convert
parameter_list|(
name|long
name|double
specifier|_Complex
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_comment
comment|// expected-note 2 {{candidate function}}
end_comment

begin_function
name|void
name|test_promote_or_convert
parameter_list|(
name|float
name|f
parameter_list|,
name|float
specifier|_Complex
name|fc
parameter_list|)
block|{
name|char
modifier|*
name|cp
init|=
name|promote_or_convert
argument_list|(
name|fc
argument_list|)
decl_stmt|;
comment|// expected-error{{call to 'promote_or_convert' is ambiguous; candidates are:}}
name|int
modifier|*
name|ip2
init|=
name|promote_or_convert
argument_list|(
name|f
argument_list|)
decl_stmt|;
comment|// expected-error{{call to 'promote_or_convert' is ambiguous; candidates are:}}
block|}
end_function

begin_function_decl
name|char
modifier|*
name|promote_or_convert2
parameter_list|(
name|float
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|int
modifier|*
name|promote_or_convert2
parameter_list|(
name|double
specifier|_Complex
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function
name|void
name|test_promote_or_convert2
parameter_list|(
name|float
specifier|_Complex
name|fc
parameter_list|)
block|{
name|int
modifier|*
name|cp
init|=
name|promote_or_convert2
argument_list|(
name|fc
argument_list|)
decl_stmt|;
block|}
end_function

begin_function_decl
name|char
modifier|*
name|promote_or_convert3
parameter_list|(
name|int
specifier|_Complex
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|int
modifier|*
name|promote_or_convert3
parameter_list|(
name|long
specifier|_Complex
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__overloadable__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function
name|void
name|test_promote_or_convert3
parameter_list|(
name|short
specifier|_Complex
name|sc
parameter_list|)
block|{
name|char
modifier|*
name|cp
init|=
name|promote_or_convert3
argument_list|(
name|sc
argument_list|)
decl_stmt|;
block|}
end_function

end_unit

