begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -Wparentheses -fsyntax-only -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -Wparentheses -fixit %s -o - | %clang_cc1 -Wparentheses -Werror -
end_comment

begin_comment
comment|// Test the various warnings under -Wparentheses
end_comment

begin_function
name|void
name|if_assign
parameter_list|(
name|void
parameter_list|)
block|{
name|int
name|i
decl_stmt|;
if|if
condition|(
name|i
operator|=
literal|4
condition|)
block|{}
comment|// expected-warning {{assignment as a condition}}
if|if
condition|(
operator|(
name|i
operator|=
literal|4
operator|)
condition|)
block|{}
block|}
end_function

begin_function
name|void
name|bitwise_rel
parameter_list|(
name|unsigned
name|i
parameter_list|)
block|{
call|(
name|void
call|)
argument_list|(
name|i
operator|&
literal|0x2
operator|==
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning {{& has lower precedence than ==}}
call|(
name|void
call|)
argument_list|(
literal|0
operator|==
name|i
operator|&
literal|0x2
argument_list|)
expr_stmt|;
comment|// expected-warning {{& has lower precedence than ==}}
call|(
name|void
call|)
argument_list|(
name|i
operator|&
literal|0xff
operator|<
literal|30
argument_list|)
expr_stmt|;
comment|// expected-warning {{& has lower precedence than<}}
call|(
name|void
call|)
argument_list|(
operator|(
name|i
operator|&
literal|0x2
operator|)
operator|==
literal|0
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|i
operator|&
operator|(
literal|0x2
operator|==
literal|0
operator|)
argument_list|)
expr_stmt|;
comment|// Eager logical op
call|(
name|void
call|)
argument_list|(
name|i
operator|==
literal|1
operator||
name|i
operator|==
literal|2
operator||
name|i
operator|==
literal|3
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|i
operator|!=
literal|1
operator|&
name|i
operator|!=
literal|2
operator|&
name|i
operator|!=
literal|3
argument_list|)
expr_stmt|;
block|}
end_function

end_unit

