begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -Wno-pragmas -verify %s
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|annotate
name|(
literal|"test"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|function
name|,
name|variable
name|)
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_comment
comment|// Check for contradictions in rules for attribute without a strict subject set:
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|annotate
name|(
literal|"subRuleContradictions"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|variable
name|,
name|variable
name|(
name|is_parameter
name|)
name|,
name|function
name|(
name|is_member
name|)
name|,
name|variable
name|(
name|is_global
name|)
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{redundant attribute subject matcher sub-rule 'variable(is_parameter)'; 'variable' already matches those declarations}}
end_comment

begin_comment
comment|// expected-error@-2 {{redundant attribute subject matcher sub-rule 'variable(is_global)'; 'variable' already matches those declarations}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|annotate
name|(
literal|"subRuleContradictions2"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|function
name|(
name|is_member
name|)
name|,
name|function
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{redundant attribute subject matcher sub-rule 'function(is_member)'; 'function' already matches those declarations}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|annotate
name|(
literal|"subRuleContradictions3"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|variable
name|,
name|variable
name|(
name|unless
name|(
name|is_parameter
name|)
name|)
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{redundant attribute subject matcher sub-rule 'variable(unless(is_parameter))'; 'variable' already matches those declarations}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|annotate
name|(
literal|"negatedSubRuleContradictions1"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|variable
name|(
name|is_parameter
name|)
name|,
name|variable
name|(
name|unless
name|(
name|is_parameter
name|)
name|)
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{negated attribute subject matcher sub-rule 'variable(unless(is_parameter))' contradicts sub-rule 'variable(is_parameter)'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|annotate
name|(
literal|"negatedSubRuleContradictions2"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|variable
name|(
name|unless
name|(
name|is_parameter
name|)
name|)
name|,
name|variable
name|(
name|is_thread_local
name|)
name|,
name|function
name|,
name|variable
name|(
name|is_global
name|)
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{negated attribute subject matcher sub-rule 'variable(unless(is_parameter))' contradicts sub-rule 'variable(is_global)'}}
end_comment

begin_comment
comment|// We have just one error, don't error on 'variable(is_global)'
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_comment
comment|// Verify the strict subject set verification.
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|function
name|)
name|)
end_pragma

begin_comment
comment|// No error
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|function
name|,
name|variable
name|)
name|)
end_pragma

begin_comment
comment|// No error
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|function
name|,
name|variable
name|,
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|)
name|)
end_pragma

begin_comment
comment|// No error
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|variable
name|,
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|function
name|)
name|)
end_pragma

begin_comment
comment|// No error
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|function
name|,
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|variable
name|,
name|enum
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'abi_tag' can't be applied to 'enum'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|enum_constant
name|,
name|function
name|,
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|variable
name|,
name|variable
name|(
name|is_parameter
name|)
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'abi_tag' can't be applied to 'variable(is_parameter)', and 'enum_constant'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|function
name|,
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|enum
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'abi_tag' can't be applied to 'enum'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_comment
comment|// Verify the non-strict subject set verification.
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|function
name|)
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|variable
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|)
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|function
name|,
name|variable
name|)
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|variable
name|,
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|)
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|function
name|)
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|function
name|,
name|variable
name|)
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|function
name|,
name|variable
name|,
name|enum
name|,
name|enum_constant
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'abi_tag' can't be applied to 'enum_constant', and 'enum'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"a"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|enum
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'abi_tag' can't be applied to 'enum'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_comment
comment|// Handle attributes whose subjects are supported only in other language modes:
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"b"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|namespace
name|,
name|record
name|(
name|unless
name|(
name|is_union
name|)
name|)
name|,
name|variable
name|,
name|function
name|)
name|)
end_pragma

begin_comment
comment|// 'namespace' is accepted!
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|abi_tag
name|(
literal|"b"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|namespace
name|)
name|)
end_pragma

begin_comment
comment|// 'namespace' is accepted!
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|objc_subclassing_restricted
name|)
name|)
name|,
name|apply_to
name|=
name|objc_interface
name|)
end_pragma

begin_comment
comment|// No error!
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|objc_subclassing_restricted
name|)
name|)
name|,
name|apply_to
name|=
name|objc_interface
name|)
end_pragma

begin_comment
comment|// No error!
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|objc_subclassing_restricted
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|objc_interface
name|,
name|objc_protocol
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'objc_subclassing_restricted' can't be applied to 'objc_protocol'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|objc_subclassing_restricted
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|objc_protocol
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'objc_subclassing_restricted' can't be applied to 'objc_protocol'}}
end_comment

begin_comment
comment|// Don't report an error about missing 'objc_interface' as we aren't parsing
end_comment

begin_comment
comment|// Objective-C.
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|objc_subclassing_restricted
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|objc_interface
name|,
name|objc_protocol
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'objc_subclassing_restricted' can't be applied to 'objc_protocol'}}
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|objc_subclassing_restricted
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|objc_protocol
name|)
name|)
end_pragma

begin_comment
comment|// expected-error@-1 {{attribute 'objc_subclassing_restricted' can't be applied to 'objc_protocol'}}
end_comment

begin_comment
comment|// Don't report an error about missing 'objc_interface' as we aren't parsing
end_comment

begin_comment
comment|// Objective-C.
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_comment
comment|// Use of matchers from other language modes should not cause for attributes
end_comment

begin_comment
comment|// without subject list:
end_comment

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|annotate
name|(
literal|"test"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|objc_method
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|push
name|(
name|__attribute__
name|(
name|(
name|annotate
name|(
literal|"test"
name|)
name|)
name|)
name|,
name|apply_to
name|=
name|any
name|(
name|objc_interface
name|,
name|objc_protocol
name|)
name|)
end_pragma

begin_pragma
pragma|#
directive|pragma
name|clang
name|attribute
name|pop
end_pragma

end_unit

