begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang -Wall -ffreestanding -fsyntax-only -Xclang -verify %s
end_comment

begin_include
include|#
directive|include
file|<limits.h>
end_include

begin_enum
enum|enum
block|{
name|X
init|=
literal|1
operator|<<
literal|0
block|,
name|Y
init|=
literal|1
operator|<<
literal|1
block|,
name|Z
init|=
literal|1
operator|<<
literal|2
block|}
enum|;
end_enum

begin_function
name|void
name|test
parameter_list|()
block|{
name|char
name|c
decl_stmt|;
name|c
operator|=
literal|0
operator|<<
literal|0
expr_stmt|;
name|c
operator|=
literal|0
operator|<<
literal|1
expr_stmt|;
name|c
operator|=
literal|1
operator|<<
literal|0
expr_stmt|;
name|c
operator|=
literal|1
operator|<<
operator|-
literal|0
expr_stmt|;
name|c
operator|=
literal|1
operator|>>
operator|-
literal|0
expr_stmt|;
name|c
operator|=
literal|1
operator|<<
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{shift count is negative}}
name|c
operator|=
literal|1
operator|>>
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{shift count is negative}}
name|c
operator|=
literal|1
operator|<<
name|c
expr_stmt|;
name|c
operator|<<=
literal|0
expr_stmt|;
name|c
operator|>>=
literal|0
expr_stmt|;
name|c
operator|<<=
literal|1
expr_stmt|;
name|c
operator|>>=
literal|1
expr_stmt|;
name|c
operator|<<=
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{shift count is negative}}
name|c
operator|>>=
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{shift count is negative}}
name|c
operator|<<=
literal|999999
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|>>=
literal|999999
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|<<=
name|CHAR_BIT
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|>>=
name|CHAR_BIT
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|<<=
name|CHAR_BIT
operator|+
literal|1
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|>>=
name|CHAR_BIT
operator|+
literal|1
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
call|(
name|void
call|)
argument_list|(
operator|(
name|long
operator|)
name|c
operator|<<
name|CHAR_BIT
argument_list|)
expr_stmt|;
block|}
end_function

begin_define
define|#
directive|define
name|a
value|0
end_define

begin_define
define|#
directive|define
name|ashift
value|8
end_define

begin_enum
enum|enum
block|{
name|b
init|=
operator|(
name|a
operator|<<
name|ashift
operator|)
block|}
enum|;
end_enum

end_unit

