begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -Wall -Wshift-sign-overflow -ffreestanding -fsyntax-only -verify %s
end_comment

begin_include
include|#
directive|include
file|<limits.h>
end_include

begin_define
define|#
directive|define
name|WORD_BIT
value|(sizeof(int) * CHAR_BIT)
end_define

begin_enum
enum|enum
block|{
name|X
init|=
literal|1
operator|<<
literal|0
block|,
name|Y
init|=
literal|1
operator|<<
literal|1
block|,
name|Z
init|=
literal|1
operator|<<
literal|2
block|}
enum|;
end_enum

begin_function
name|void
name|test
parameter_list|()
block|{
name|char
name|c
decl_stmt|;
name|c
operator|=
literal|0
operator|<<
literal|0
expr_stmt|;
name|c
operator|=
literal|0
operator|<<
literal|1
expr_stmt|;
name|c
operator|=
literal|1
operator|<<
literal|0
expr_stmt|;
name|c
operator|=
literal|1
operator|<<
operator|-
literal|0
expr_stmt|;
name|c
operator|=
literal|1
operator|>>
operator|-
literal|0
expr_stmt|;
name|c
operator|=
literal|1
operator|<<
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{shift count is negative}}
name|c
operator|=
literal|1
operator|>>
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{shift count is negative}}
name|c
operator|=
literal|1
operator|<<
name|c
expr_stmt|;
name|c
operator|<<=
literal|0
expr_stmt|;
name|c
operator|>>=
literal|0
expr_stmt|;
name|c
operator|<<=
literal|1
expr_stmt|;
name|c
operator|>>=
literal|1
expr_stmt|;
name|c
operator|<<=
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{shift count is negative}}
name|c
operator|>>=
operator|-
literal|1
expr_stmt|;
comment|// expected-warning {{shift count is negative}}
name|c
operator|<<=
literal|999999
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|>>=
literal|999999
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|<<=
name|CHAR_BIT
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|>>=
name|CHAR_BIT
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|<<=
name|CHAR_BIT
operator|+
literal|1
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
name|c
operator|>>=
name|CHAR_BIT
operator|+
literal|1
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
call|(
name|void
call|)
argument_list|(
operator|(
name|long
operator|)
name|c
operator|<<
name|CHAR_BIT
argument_list|)
expr_stmt|;
name|int
name|i
decl_stmt|;
name|i
operator|=
literal|1
operator|<<
operator|(
name|WORD_BIT
operator|-
literal|2
operator|)
expr_stmt|;
name|i
operator|=
literal|2
operator|<<
operator|(
name|WORD_BIT
operator|-
literal|1
operator|)
expr_stmt|;
comment|// expected-warning {{bits to represent, but 'int' only has}}
name|i
operator|=
literal|1
operator|<<
operator|(
name|WORD_BIT
operator|-
literal|1
operator|)
expr_stmt|;
comment|// expected-warning {{sets the sign bit of the shift expression}}
name|i
operator|=
operator|-
literal|1
operator|<<
operator|(
name|WORD_BIT
operator|-
literal|1
operator|)
expr_stmt|;
name|i
operator|=
literal|0
operator|<<
operator|(
name|WORD_BIT
operator|-
literal|1
operator|)
expr_stmt|;
name|i
operator|=
operator|(
name|char
operator|)
literal|1
operator|<<
operator|(
name|WORD_BIT
operator|-
literal|2
operator|)
expr_stmt|;
name|unsigned
name|u
decl_stmt|;
name|u
operator|=
literal|1U
operator|<<
operator|(
name|WORD_BIT
operator|-
literal|1
operator|)
expr_stmt|;
name|u
operator|=
literal|5U
operator|<<
operator|(
name|WORD_BIT
operator|-
literal|1
operator|)
expr_stmt|;
name|long
name|long
name|int
name|lli
decl_stmt|;
name|lli
operator|=
name|INT_MIN
operator|<<
literal|2
expr_stmt|;
comment|// expected-warning {{bits to represent, but 'int' only has}}
name|lli
operator|=
literal|1LL
operator|<<
operator|(
expr|sizeof
operator|(
name|long
name|long
operator|)
operator|*
name|CHAR_BIT
operator|-
literal|2
operator|)
expr_stmt|;
block|}
end_function

begin_define
define|#
directive|define
name|a
value|0
end_define

begin_define
define|#
directive|define
name|ashift
value|8
end_define

begin_enum
enum|enum
block|{
name|b
init|=
operator|(
name|a
operator|<<
name|ashift
operator|)
block|}
enum|;
end_enum

begin_comment
comment|// Don't warn for negative shifts in code that is unreachable.
end_comment

begin_function
name|void
name|test_pr5544
parameter_list|()
block|{
call|(
name|void
call|)
argument_list|(
operator|(
operator|(
literal|1
operator|)
operator|>
literal|63
operator|&&
operator|(
literal|1
operator|)
operator|<
literal|128
condition|?
operator|(
operator|(
operator|(
name|unsigned
name|long
name|long
operator|)
literal|1
operator|)
operator|<<
operator|(
operator|(
literal|1
operator|)
operator|-
literal|64
operator|)
operator|)
else|:
operator|(
name|unsigned
name|long
name|long
operator|)
literal|0
operator|)
argument_list|)
expr_stmt|;
comment|// no-warning
block|}
end_function

begin_function
name|void
name|test_shift_too_much
parameter_list|(
name|char
name|x
parameter_list|)
block|{
if|if
condition|(
literal|0
condition|)
call|(
name|void
call|)
argument_list|(
name|x
operator|>>
literal|80
argument_list|)
expr_stmt|;
comment|// no-warning
call|(
name|void
call|)
argument_list|(
name|x
operator|>>
literal|80
argument_list|)
expr_stmt|;
comment|// expected-warning {{shift count>= width of type}}
block|}
end_function

end_unit

