begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s -triple x86_64-apple-darwin9
end_comment

begin_function
name|void
name|f1
parameter_list|(
name|int
name|a
parameter_list|)
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to function}}
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-error {{'va_start' used in function with fixed args}}
block|}
end_function

begin_function
name|void
name|f2
parameter_list|(
name|int
name|a
parameter_list|,
name|int
name|b
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
literal|10
argument_list|)
expr_stmt|;
comment|// expected-warning {{second parameter of 'va_start' not last named argument}}
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-warning {{second parameter of 'va_start' not last named argument}}
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|b
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|f3
parameter_list|(
name|float
name|a
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
operator|(
name|a
operator|)
argument_list|)
expr_stmt|;
block|}
end_function

begin_comment
comment|// stdarg: PR3075
end_comment

begin_function
name|void
name|f4
parameter_list|(
specifier|const
name|char
modifier|*
name|msg
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_stdarg_start
argument_list|(
operator|(
name|ap
operator|)
argument_list|,
operator|(
name|msg
operator|)
argument_list|)
expr_stmt|;
name|__builtin_va_end
argument_list|(
name|ap
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|f5
parameter_list|()
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// expected-error {{'va_start' used in function with fixed args}}
block|}
end_function

begin_function
name|void
name|f6
parameter_list|(
name|int
name|a
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|)
expr_stmt|;
comment|// expected-error {{too few arguments to function}}
block|}
end_function

begin_comment
comment|// PR3350
end_comment

begin_function
name|void
name|foo
parameter_list|(
name|__builtin_va_list
name|authors
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_va_start
argument_list|(
name|authors
argument_list|,
name|authors
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|authors
argument_list|,
name|int
argument_list|)
expr_stmt|;
name|__builtin_va_end
argument_list|(
name|authors
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|f7
parameter_list|(
name|int
name|a
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// FIXME: This error message is sub-par.
name|__builtin_va_arg
argument_list|(
name|ap
argument_list|,
name|int
argument_list|)
operator|=
literal|1
expr_stmt|;
comment|// expected-error {{expression is not assignable}}
name|int
modifier|*
name|x
init|=
operator|&
name|__builtin_va_arg
argument_list|(
name|ap
argument_list|,
name|int
argument_list|)
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue}}
name|__builtin_va_end
argument_list|(
name|ap
argument_list|)
expr_stmt|;
block|}
end_function

begin_function
name|void
name|f8
parameter_list|(
name|int
name|a
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|ap
argument_list|,
name|void
argument_list|)
expr_stmt|;
comment|// expected-error {{second argument to 'va_arg' is of incomplete type 'void'}}
name|__builtin_va_end
argument_list|(
name|ap
argument_list|)
expr_stmt|;
block|}
end_function

begin_enum
enum|enum
name|E
block|{
name|x
init|=
operator|-
literal|1
block|,
name|y
init|=
literal|2
block|,
name|z
init|=
literal|10000
block|}
enum|;
end_enum

begin_function
name|void
name|f9
parameter_list|(
name|__builtin_va_list
name|args
parameter_list|)
block|{
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args
argument_list|,
name|float
argument_list|)
expr_stmt|;
comment|// expected-warning {{second argument to 'va_arg' is of promotable type 'float'}}
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args
argument_list|,
expr|enum
name|E
argument_list|)
expr_stmt|;
comment|// Don't warn here in C
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args
argument_list|,
name|short
argument_list|)
expr_stmt|;
comment|// expected-warning {{second argument to 'va_arg' is of promotable type 'short'}}
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args
argument_list|,
name|char
argument_list|)
expr_stmt|;
comment|// expected-warning {{second argument to 'va_arg' is of promotable type 'char'}}
block|}
end_function

begin_function
name|void
name|f10
parameter_list|(
name|int
name|a
parameter_list|,
modifier|...
parameter_list|)
block|{
name|int
name|i
decl_stmt|;
name|__builtin_va_list
name|ap
decl_stmt|;
name|i
operator|=
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-error {{assigning to 'int' from incompatible type 'void'}}
name|__builtin_va_end
argument_list|(
name|ap
argument_list|)
expr_stmt|;
block|}
end_function

end_unit

