begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify %s -triple x86_64-apple-darwin9
end_comment

begin_comment
comment|// rdar://6726818
end_comment

begin_function
name|void
name|f1
parameter_list|()
block|{
specifier|const
name|__builtin_va_list
name|args2
decl_stmt|;
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args2
argument_list|,
name|int
argument_list|)
expr_stmt|;
comment|// expected-error {{first argument to 'va_arg' is of type 'const __builtin_va_list' and not 'va_list'}}
block|}
end_function

begin_function
name|void
name|f2
parameter_list|(
name|int
name|a
parameter_list|,
modifier|...
parameter_list|)
block|{
name|__builtin_ms_va_list
name|ap
decl_stmt|;
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-error {{'__builtin_ms_va_start' used in System V ABI function}}
block|}
end_function

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g1
argument_list|(
name|int
name|a
argument_list|)
block|{
name|__builtin_ms_va_list
name|ap
decl_stmt|;
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-error {{too many arguments to function}}
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-error {{'va_start' used in function with fixed args}}
block|}
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g2
argument_list|(
name|int
name|a
argument_list|,
name|int
name|b
argument_list|,
operator|...
argument_list|)
block|{
name|__builtin_ms_va_list
name|ap
decl_stmt|;
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
literal|10
argument_list|)
expr_stmt|;
comment|// expected-warning {{second argument to 'va_start' is not the last named parameter}}
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-warning {{second argument to 'va_start' is not the last named parameter}}
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|b
argument_list|)
expr_stmt|;
block|}
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g3
argument_list|(
name|float
name|a
argument_list|,
operator|...
argument_list|)
block|{
comment|// expected-note 2{{parameter of type 'float' is declared here}}
name|__builtin_ms_va_list
name|ap
decl_stmt|;
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing an object that undergoes default argument promotion to 'va_start' has undefined behavior}}
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
operator|(
name|a
operator|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{passing an object that undergoes default argument promotion to 'va_start' has undefined behavior}}
block|}
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g5
argument_list|()
block|{
name|__builtin_ms_va_list
name|ap
decl_stmt|;
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|ap
argument_list|)
expr_stmt|;
comment|// expected-error {{'va_start' used in function with fixed args}}
block|}
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g6
argument_list|(
name|int
name|a
argument_list|,
operator|...
argument_list|)
block|{
name|__builtin_ms_va_list
name|ap
decl_stmt|;
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|)
expr_stmt|;
comment|// expected-error {{too few arguments to function}}
block|}
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|bar
argument_list|(
name|__builtin_ms_va_list
name|authors
argument_list|,
operator|...
argument_list|)
block|{
name|__builtin_ms_va_start
argument_list|(
name|authors
argument_list|,
name|authors
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|authors
argument_list|,
name|int
argument_list|)
expr_stmt|;
name|__builtin_ms_va_end
argument_list|(
name|authors
argument_list|)
expr_stmt|;
block|}
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g7
argument_list|(
name|int
name|a
argument_list|,
operator|...
argument_list|)
block|{
name|__builtin_ms_va_list
name|ap
decl_stmt|;
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// FIXME: This error message is sub-par.
name|__builtin_va_arg
argument_list|(
name|ap
argument_list|,
name|int
argument_list|)
operator|=
literal|1
expr_stmt|;
comment|// expected-error {{expression is not assignable}}
name|int
modifier|*
name|x
init|=
operator|&
name|__builtin_va_arg
argument_list|(
name|ap
argument_list|,
name|int
argument_list|)
decl_stmt|;
comment|// expected-error {{cannot take the address of an rvalue}}
name|__builtin_ms_va_end
argument_list|(
name|ap
argument_list|)
expr_stmt|;
block|}
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g8
argument_list|(
name|int
name|a
argument_list|,
operator|...
argument_list|)
block|{
name|__builtin_ms_va_list
name|ap
decl_stmt|;
name|__builtin_ms_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|ap
argument_list|,
name|void
argument_list|)
expr_stmt|;
comment|// expected-error {{second argument to 'va_arg' is of incomplete type 'void'}}
name|__builtin_ms_va_end
argument_list|(
name|ap
argument_list|)
expr_stmt|;
block|}
end_decl_stmt

begin_enum
enum|enum
name|E
block|{
name|x
init|=
operator|-
literal|1
block|,
name|y
init|=
literal|2
block|,
name|z
init|=
literal|10000
block|}
enum|;
end_enum

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g9
argument_list|(
name|__builtin_ms_va_list
name|args
argument_list|)
block|{
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args
argument_list|,
name|float
argument_list|)
expr_stmt|;
comment|// expected-warning {{second argument to 'va_arg' is of promotable type 'float'}}
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args
argument_list|,
expr|enum
name|E
argument_list|)
expr_stmt|;
comment|// no-warning
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args
argument_list|,
name|short
argument_list|)
expr_stmt|;
comment|// expected-warning {{second argument to 'va_arg' is of promotable type 'short'}}
operator|(
name|void
operator|)
name|__builtin_va_arg
argument_list|(
name|args
argument_list|,
name|char
argument_list|)
expr_stmt|;
comment|// expected-warning {{second argument to 'va_arg' is of promotable type 'char'}}
block|}
end_decl_stmt

begin_decl_stmt
name|void
name|__attribute__
argument_list|(
operator|(
name|ms_abi
operator|)
argument_list|)
name|g10
argument_list|(
name|int
name|a
argument_list|,
operator|...
argument_list|)
block|{
name|__builtin_va_list
name|ap
decl_stmt|;
name|__builtin_va_start
argument_list|(
name|ap
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-error {{'va_start' used in Win64 ABI function}}
block|}
end_decl_stmt

end_unit

