begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 %s -verify -fsyntax-only -Weverything -triple x86_64-apple-darwin10
end_comment

begin_comment
comment|// Test the compatibility of clang's vector extensions with gcc's vector
end_comment

begin_comment
comment|// extensions for C. Notably&&, ||, ?: and ! are not available.
end_comment

begin_typedef
typedef|typedef
name|long
name|long
name|v2i64
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|int
name|v2i32
name|__attribute__
typedef|((
name|vector_size
typedef|(8)));
end_typedef

begin_typedef
typedef|typedef
name|short
name|v2i16
name|__attribute__
typedef|((
name|vector_size
typedef|(4)));
end_typedef

begin_typedef
typedef|typedef
name|char
name|v2i8
name|__attribute__
typedef|((
name|vector_size
typedef|(2)));
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|long
name|long
name|v2u64
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|int
name|v2u32
name|__attribute__
typedef|((
name|vector_size
typedef|(8)));
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|short
name|v2u16
name|__attribute__
typedef|((
name|vector_size
typedef|(4)));
end_typedef

begin_typedef
typedef|typedef
name|unsigned
name|char
name|v2u8
name|__attribute__
typedef|((
name|vector_size
typedef|(2)));
end_typedef

begin_typedef
typedef|typedef
name|float
name|v4f32
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|double
name|v2f64
name|__attribute__
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_typedef
typedef|typedef
name|double
name|v4f64
name|__attribute__
typedef|((
name|vector_size
typedef|(32)));
end_typedef

begin_typedef
typedef|typedef
name|int
name|v4i32
name|__attribute
typedef|((
name|vector_size
typedef|(16)));
end_typedef

begin_function_decl
name|void
name|arithmeticTest
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|logicTest
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|comparisonTest
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|floatTestSignedType
parameter_list|(
name|char
name|a
parameter_list|,
name|short
name|b
parameter_list|,
name|int
name|c
parameter_list|,
name|long
name|long
name|d
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|floatTestUnsignedType
parameter_list|(
name|unsigned
name|char
name|a
parameter_list|,
name|unsigned
name|short
name|b
parameter_list|,
name|unsigned
name|int
name|c
parameter_list|,
name|unsigned
name|long
name|long
name|d
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|floatTestConstant
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|intTestType
parameter_list|(
name|char
name|a
parameter_list|,
name|short
name|b
parameter_list|,
name|int
name|c
parameter_list|,
name|long
name|long
name|d
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|intTestTypeUnsigned
parameter_list|(
name|unsigned
name|char
name|a
parameter_list|,
name|unsigned
name|short
name|b
parameter_list|,
name|unsigned
name|int
name|c
parameter_list|,
name|unsigned
name|long
name|long
name|d
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|uintTestType
parameter_list|(
name|char
name|a
parameter_list|,
name|short
name|b
parameter_list|,
name|int
name|c
parameter_list|,
name|long
name|long
name|d
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|uintTestTypeUnsigned
parameter_list|(
name|unsigned
name|char
name|a
parameter_list|,
name|unsigned
name|short
name|b
parameter_list|,
name|unsigned
name|int
name|c
parameter_list|,
name|unsigned
name|long
name|long
name|d
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|uintTestConstant
parameter_list|(
name|v2u64
name|v2u64_a
parameter_list|,
name|v2u32
name|v2u32_a
parameter_list|,
name|v2u16
name|v2u16_a
parameter_list|,
name|v2u8
name|v2u8_a
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|intTestConstant
parameter_list|(
name|v2i64
name|v2i64_a
parameter_list|,
name|v2i32
name|v2i32_a
parameter_list|,
name|v2i16
name|v2i16_a
parameter_list|,
name|v2i8
name|v2i8_a
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|arithmeticTest
parameter_list|(
name|void
parameter_list|)
block|{
name|v2i64
name|v2i64_a
init|=
operator|(
name|v2i64
operator|)
block|{
literal|0
block|,
literal|1
block|}
decl_stmt|;
name|v2i64
name|v2i64_r
decl_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|+
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|-
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|*
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|/
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|%
literal|1
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|+
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|-
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|*
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|/
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|%
name|v2i64_a
expr_stmt|;
name|v2i64_a
operator|+=
literal|1
expr_stmt|;
name|v2i64_a
operator|-=
literal|1
expr_stmt|;
name|v2i64_a
operator|*=
literal|1
expr_stmt|;
name|v2i64_a
operator|/=
literal|1
expr_stmt|;
name|v2i64_a
operator|%=
literal|1
expr_stmt|;
block|}
end_function

begin_function
name|void
name|comparisonTest
parameter_list|(
name|void
parameter_list|)
block|{
name|v2i64
name|v2i64_a
init|=
operator|(
name|v2i64
operator|)
block|{
literal|0
block|,
literal|1
block|}
decl_stmt|;
name|v2i64
name|v2i64_r
decl_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|==
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|!=
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|<
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|>
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|<=
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|>=
literal|1
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|==
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|!=
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|<
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|>
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|<=
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|>=
name|v2i64_a
expr_stmt|;
block|}
end_function

begin_function
name|void
name|logicTest
parameter_list|(
name|void
parameter_list|)
block|{
name|v2i64
name|v2i64_a
init|=
operator|(
name|v2i64
operator|)
block|{
literal|0
block|,
literal|1
block|}
decl_stmt|;
name|v2i64
name|v2i64_b
init|=
operator|(
name|v2i64
operator|)
block|{
literal|2
block|,
literal|1
block|}
decl_stmt|;
name|v2i64
name|v2i64_c
init|=
operator|(
name|v2i64
operator|)
block|{
literal|3
block|,
literal|1
block|}
decl_stmt|;
name|v2i64
name|v2i64_r
decl_stmt|;
name|v2i64_r
operator|=
operator|!
name|v2i64_a
expr_stmt|;
comment|// expected-error {{invalid argument type 'v2i64' (vector of 2 'long long' values) to unary expression}}
name|v2i64_r
operator|=
operator|~
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
condition|?
name|v2i64_b
else|:
name|v2i64_c
expr_stmt|;
comment|// expected-error {{used type 'v2i64' (vector of 2 'long long' values) where arithmetic or pointer type is required}}
name|v2i64_r
operator|=
name|v2i64_a
operator|&
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator||
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|^
literal|1
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|&
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator||
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|^
name|v2i64_a
expr_stmt|;
name|v2i64_a
operator|&=
literal|1
expr_stmt|;
name|v2i64_a
operator||=
literal|1
expr_stmt|;
name|v2i64_a
operator|^=
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|&&
literal|1
expr_stmt|;
comment|// expected-error {{logical expression with vector type 'v2i64' (vector of 2 'long long' values) and non-vector type 'int' is only supported in C++}}
name|v2i64_r
operator|=
name|v2i64_a
operator|||
literal|1
expr_stmt|;
comment|// expected-error {{logical expression with vector type 'v2i64' (vector of 2 'long long' values) and non-vector type 'int' is only supported in C++}}
name|v2i64_r
operator|=
name|v2i64_a
operator|&&
name|v2i64_a
expr_stmt|;
comment|// expected-error {{logical expression with vector types 'v2i64' (vector of 2 'long long' values) and 'v2i64' is only supported in C++}}
name|v2i64_r
operator|=
name|v2i64_a
operator|||
name|v2i64_a
expr_stmt|;
comment|// expected-error {{logical expression with vector types 'v2i64' (vector of 2 'long long' values) and 'v2i64' is only supported in C++}}
name|v2i64_r
operator|=
name|v2i64_a
operator|<<
literal|1
expr_stmt|;
name|v2i64_r
operator|=
name|v2i64_a
operator|>>
literal|1
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|<<
name|v2i64_a
expr_stmt|;
name|v2i64_r
operator|=
literal|1
operator|>>
name|v2i64_a
expr_stmt|;
name|v2i64_a
operator|<<=
literal|1
expr_stmt|;
name|v2i64_a
operator|>>=
literal|1
expr_stmt|;
block|}
end_function

begin_comment
comment|// For operations with floating point types, we check that interger constants
end_comment

begin_comment
comment|// can be respresented, or failing that checking based on the integer types.
end_comment

begin_function
name|void
name|floatTestConstant
parameter_list|(
name|void
parameter_list|)
block|{
comment|// Test that constants added to floats must be expressible as floating point
comment|// numbers.
name|v4f32
name|v4f32_a
init|=
block|{
literal|0.4f
block|,
literal|0.4f
block|,
literal|0.4f
block|,
literal|0.4f
block|}
decl_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
literal|1
expr_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
literal|0xFFFFFF
expr_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
operator|(
operator|-
literal|1567563LL
operator|)
expr_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
operator|(
literal|16777208
operator|)
expr_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
operator|(
literal|16777219
operator|)
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'int' and vector type 'v4f32' (vector of 4 'float' values) as implicit conversion would cause truncation}}
block|}
end_function

begin_function_decl
name|void
name|floatTestConstantComparison
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|void
name|doubleTestConstantComparison
parameter_list|(
name|void
parameter_list|)
function_decl|;
end_function_decl

begin_function
name|void
name|floatTestConstantComparison
parameter_list|(
name|void
parameter_list|)
block|{
name|v4f32
name|v4f32_a
init|=
block|{
literal|0.4f
block|,
literal|0.4f
block|,
literal|0.4f
block|,
literal|0.4f
block|}
decl_stmt|;
name|v4i32
name|v4i32_r
decl_stmt|;
name|v4i32_r
operator|=
name|v4f32_a
operator|>
literal|0.4f
expr_stmt|;
name|v4i32_r
operator|=
name|v4f32_a
operator|>=
literal|0.4f
expr_stmt|;
name|v4i32_r
operator|=
name|v4f32_a
operator|<
literal|0.4f
expr_stmt|;
name|v4i32_r
operator|=
name|v4f32_a
operator|<=
literal|0.4f
expr_stmt|;
name|v4i32_r
operator|=
name|v4f32_a
operator|==
literal|0.4f
expr_stmt|;
comment|// expected-warning {{comparing floating point with == or != is unsafe}}
name|v4i32_r
operator|=
name|v4f32_a
operator|!=
literal|0.4f
expr_stmt|;
comment|// expected-warning {{comparing floating point with == or != is unsafe}}
block|}
end_function

begin_function
name|void
name|doubleTestConstantComparison
parameter_list|(
name|void
parameter_list|)
block|{
name|v2f64
name|v2f64_a
init|=
block|{
literal|0.4
block|,
literal|0.4
block|}
decl_stmt|;
name|v2i64
name|v2i64_r
decl_stmt|;
name|v2i64_r
operator|=
name|v2f64_a
operator|>
literal|0.4
expr_stmt|;
name|v2i64_r
operator|=
name|v2f64_a
operator|>=
literal|0.4
expr_stmt|;
name|v2i64_r
operator|=
name|v2f64_a
operator|<
literal|0.4
expr_stmt|;
name|v2i64_r
operator|=
name|v2f64_a
operator|<=
literal|0.4
expr_stmt|;
name|v2i64_r
operator|=
name|v2f64_a
operator|==
literal|0.4
expr_stmt|;
comment|// expected-warning {{comparing floating point with == or != is unsafe}}
name|v2i64_r
operator|=
name|v2f64_a
operator|!=
literal|0.4
expr_stmt|;
comment|// expected-warning {{comparing floating point with == or != is unsafe}}
block|}
end_function

begin_function
name|void
name|floatTestUnsignedType
parameter_list|(
name|unsigned
name|char
name|a
parameter_list|,
name|unsigned
name|short
name|b
parameter_list|,
name|unsigned
name|int
name|c
parameter_list|,
name|unsigned
name|long
name|long
name|d
parameter_list|)
block|{
name|v4f32
name|v4f32_a
init|=
block|{
literal|0.4f
block|,
literal|0.4f
block|,
literal|0.4f
block|,
literal|0.4f
block|}
decl_stmt|;
name|v4f64
name|v4f64_b
init|=
block|{
literal|0.4
block|,
literal|0.4
block|,
literal|0.4
block|,
literal|0.4
block|}
decl_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
name|a
expr_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
name|b
expr_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
name|c
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned int' and vector type 'v4f32' (vector of 4 'float' values) as implicit conversion would cause truncation}}
name|v4f32_a
operator|=
name|v4f32_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned long long' and vector type 'v4f32' (vector of 4 'float' values) as implicit conversion would cause truncation}}
name|v4f64_b
operator|=
name|v4f64_b
operator|+
name|a
expr_stmt|;
name|v4f64_b
operator|=
name|v4f64_b
operator|+
name|b
expr_stmt|;
name|v4f64_b
operator|=
name|v4f64_b
operator|+
name|c
expr_stmt|;
name|v4f64_b
operator|=
name|v4f64_b
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned long long' and vector type 'v4f64' (vector of 4 'double' values) as implicit conversion would cause truncation}}
block|}
end_function

begin_function
name|void
name|floatTestSignedType
parameter_list|(
name|char
name|a
parameter_list|,
name|short
name|b
parameter_list|,
name|int
name|c
parameter_list|,
name|long
name|long
name|d
parameter_list|)
block|{
name|v4f32
name|v4f32_a
init|=
block|{
literal|0.4f
block|,
literal|0.4f
block|,
literal|0.4f
block|,
literal|0.4f
block|}
decl_stmt|;
name|v4f64
name|v4f64_b
init|=
block|{
literal|0.4
block|,
literal|0.4
block|,
literal|0.4
block|,
literal|0.4
block|}
decl_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
name|a
expr_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
name|b
expr_stmt|;
name|v4f32_a
operator|=
name|v4f32_a
operator|+
name|c
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'int' and vector type 'v4f32' (vector of 4 'float' values) as implicit conversion would cause truncation}}
name|v4f32_a
operator|=
name|v4f32_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'long long' and vector type 'v4f32' (vector of 4 'float' values) as implicit conversion would cause truncation}}
name|v4f64_b
operator|=
name|v4f64_b
operator|+
name|a
expr_stmt|;
name|v4f64_b
operator|=
name|v4f64_b
operator|+
name|b
expr_stmt|;
name|v4f64_b
operator|=
name|v4f64_b
operator|+
name|c
expr_stmt|;
name|v4f64_b
operator|=
name|v4f64_b
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'long long' and vector type 'v4f64' (vector of 4 'double' values) as implicit conversion would cause truncation}}
block|}
end_function

begin_function
name|void
name|intTestType
parameter_list|(
name|char
name|a
parameter_list|,
name|short
name|b
parameter_list|,
name|int
name|c
parameter_list|,
name|long
name|long
name|d
parameter_list|)
block|{
name|v2i64
name|v2i64_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2i32
name|v2i32_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2i16
name|v2i16_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2i8
name|v2i8_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2i64_a
operator|=
name|v2i64_a
operator|+
name|d
expr_stmt|;
name|v2i64_a
operator|=
name|v2i64_a
operator|+
name|c
expr_stmt|;
name|v2i64_a
operator|=
name|v2i64_a
operator|+
name|b
expr_stmt|;
name|v2i64_a
operator|=
name|v2i64_a
operator|+
name|a
expr_stmt|;
name|v2i32_a
operator|=
name|v2i32_a
operator|+
name|d
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'long long' to 'v2i32' (vector of 2 'int' values)}}
name|v2i32_a
operator|=
name|v2i32_a
operator|+
name|c
expr_stmt|;
name|v2i32_a
operator|=
name|v2i32_a
operator|+
name|b
expr_stmt|;
name|v2i32_a
operator|=
name|v2i32_a
operator|+
name|a
expr_stmt|;
name|v2i16_a
operator|=
name|v2i16_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'long long' and vector type 'v2i16' (vector of 2 'short' values) as implicit conversion would cause truncation}}
name|v2i16_a
operator|=
name|v2i16_a
operator|+
name|c
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'int' to 'v2i16' (vector of 2 'short' values)}}
name|v2i16_a
operator|=
name|v2i16_a
operator|+
name|b
expr_stmt|;
name|v2i16_a
operator|=
name|v2i16_a
operator|+
name|a
expr_stmt|;
name|v2i8_a
operator|=
name|v2i8_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'long long' and vector type 'v2i8' (vector of 2 'char' values) as implicit conversion would cause truncation}}
name|v2i8_a
operator|=
name|v2i8_a
operator|+
name|c
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'int' and vector type 'v2i8' (vector of 2 'char' values) as implicit conversion would cause truncation}}
name|v2i8_a
operator|=
name|v2i8_a
operator|+
name|b
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'short' to 'v2i8' (vector of 2 'char' values)}}
name|v2i8_a
operator|=
name|v2i8_a
operator|+
name|a
expr_stmt|;
block|}
end_function

begin_function
name|void
name|intTestTypeUnsigned
parameter_list|(
name|unsigned
name|char
name|a
parameter_list|,
name|unsigned
name|short
name|b
parameter_list|,
name|unsigned
name|int
name|c
parameter_list|,
name|unsigned
name|long
name|long
name|d
parameter_list|)
block|{
name|v2i64
name|v2i64_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2i32
name|v2i32_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2i16
name|v2i16_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2i8
name|v2i8_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2i64_a
operator|=
name|v2i64_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned long long' and vector type 'v2i64' (vector of 2 'long long' values) as implicit conversion would cause truncation}}
name|v2i64_a
operator|=
name|v2i64_a
operator|+
name|c
expr_stmt|;
name|v2i64_a
operator|=
name|v2i64_a
operator|+
name|b
expr_stmt|;
name|v2i64_a
operator|=
name|v2i64_a
operator|+
name|a
expr_stmt|;
name|v2i32_a
operator|=
name|v2i32_a
operator|+
name|d
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'unsigned long long' to 'v2i32' (vector of 2 'int' values)}}
name|v2i32_a
operator|=
name|v2i32_a
operator|+
name|c
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned int' and vector type 'v2i32' (vector of 2 'int' values) as implicit conversion would cause truncation}}
name|v2i32_a
operator|=
name|v2i32_a
operator|+
name|b
expr_stmt|;
name|v2i32_a
operator|=
name|v2i32_a
operator|+
name|a
expr_stmt|;
name|v2i16_a
operator|=
name|v2i16_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned long long' and vector type 'v2i16' (vector of 2 'short' values) as implicit conversion would cause truncation}}
name|v2i16_a
operator|=
name|v2i16_a
operator|+
name|c
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'unsigned int' to 'v2i16' (vector of 2 'short' values)}}
name|v2i16_a
operator|=
name|v2i16_a
operator|+
name|b
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned short' and vector type 'v2i16' (vector of 2 'short' values) as implicit conversion would cause truncation}}
name|v2i16_a
operator|=
name|v2i16_a
operator|+
name|a
expr_stmt|;
name|v2i8_a
operator|=
name|v2i8_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned long long' and vector type 'v2i8' (vector of 2 'char' values) as implicit conversion would cause truncation}}
name|v2i8_a
operator|=
name|v2i8_a
operator|+
name|c
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned int' and vector type 'v2i8' (vector of 2 'char' values) as implicit conversion would cause truncation}}
name|v2i8_a
operator|=
name|v2i8_a
operator|+
name|b
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'unsigned short' to 'v2i8' (vector of 2 'char' values)}}
name|v2i8_a
operator|=
name|v2i8_a
operator|+
name|a
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned char' and vector type 'v2i8' (vector of 2 'char' values) as implicit conversion would cause truncation}}
block|}
end_function

begin_function
name|void
name|uintTestType
parameter_list|(
name|char
name|a
parameter_list|,
name|short
name|b
parameter_list|,
name|int
name|c
parameter_list|,
name|long
name|long
name|d
parameter_list|)
block|{
name|v2u64
name|v2u64_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2u32
name|v2u32_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2u16
name|v2u16_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2u8
name|v2u8_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2u64_a
operator|=
name|v2u64_a
operator|+
name|d
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'long long' to 'v2u64' (vector of 2 'unsigned long long' values)}}
name|v2u64_a
operator|=
name|v2u64_a
operator|+
name|c
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'int' to 'v2u64' (vector of 2 'unsigned long long' values)}}
name|v2u64_a
operator|=
name|v2u64_a
operator|+
name|b
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'short' to 'v2u64' (vector of 2 'unsigned long long' values)}}
name|v2u64_a
operator|=
name|v2u64_a
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'char' to 'v2u64' (vector of 2 'unsigned long long' values)}}
name|v2u32_a
operator|=
name|v2u32_a
operator|+
name|d
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'long long' to 'v2u32' (vector of 2 'unsigned int' values)}}
name|v2u32_a
operator|=
name|v2u32_a
operator|+
name|c
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'int' to 'v2u32' (vector of 2 'unsigned int' values)}}
name|v2u32_a
operator|=
name|v2u32_a
operator|+
name|b
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'short' to 'v2u32' (vector of 2 'unsigned int' values)}}
name|v2u32_a
operator|=
name|v2u32_a
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'char' to 'v2u32' (vector of 2 'unsigned int' values)}}
name|v2u16_a
operator|=
name|v2u16_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'long long' and vector type 'v2u16' (vector of 2 'unsigned short' values) as implicit conversion would cause truncation}}
name|v2u16_a
operator|=
name|v2u16_a
operator|+
name|c
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'int' to 'v2u16' (vector of 2 'unsigned short' values)}}
name|v2u16_a
operator|=
name|v2u16_a
operator|+
name|b
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'short' to 'v2u16' (vector of 2 'unsigned short' values)}}
name|v2u16_a
operator|=
name|v2u16_a
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'char' to 'v2u16' (vector of 2 'unsigned short' values)}}
name|v2u8_a
operator|=
name|v2u8_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'long long' and vector type 'v2u8' (vector of 2 'unsigned char' values) as implicit conversion would cause truncation}}
name|v2u8_a
operator|=
name|v2u8_a
operator|+
name|c
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'int' and vector type 'v2u8' (vector of 2 'unsigned char' values) as implicit conversion would cause truncation}}
name|v2u8_a
operator|=
name|v2u8_a
operator|+
name|b
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'short' to 'v2u8' (vector of 2 'unsigned char' values)}}
name|v2u8_a
operator|=
name|v2u8_a
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{implicit conversion changes signedness: 'char' to 'v2u8' (vector of 2 'unsigned char' values)}}
block|}
end_function

begin_function
name|void
name|uintTestTypeUnsigned
parameter_list|(
name|unsigned
name|char
name|a
parameter_list|,
name|unsigned
name|short
name|b
parameter_list|,
name|unsigned
name|int
name|c
parameter_list|,
name|unsigned
name|long
name|long
name|d
parameter_list|)
block|{
name|v2u64
name|v2u64_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2u32
name|v2u32_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2u16
name|v2u16_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2u8
name|v2u8_a
init|=
block|{
literal|1
block|,
literal|2
block|}
decl_stmt|;
name|v2u64_a
operator|=
name|v2u64_a
operator|+
name|d
expr_stmt|;
name|v2u64_a
operator|=
name|v2u64_a
operator|+
name|c
expr_stmt|;
name|v2u64_a
operator|=
name|v2u64_a
operator|+
name|b
expr_stmt|;
name|v2u64_a
operator|=
name|v2u64_a
operator|+
name|a
expr_stmt|;
name|v2u32_a
operator|=
name|v2u32_a
operator|+
name|d
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'unsigned long long' to 'v2u32' (vector of 2 'unsigned int' values)}}
name|v2u32_a
operator|=
name|v2u32_a
operator|+
name|c
expr_stmt|;
name|v2u32_a
operator|=
name|v2u32_a
operator|+
name|b
expr_stmt|;
name|v2u32_a
operator|=
name|v2u32_a
operator|+
name|a
expr_stmt|;
name|v2u16_a
operator|=
name|v2u16_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned long long' and vector type 'v2u16' (vector of 2 'unsigned short' values) as implicit conversion would cause truncation}}
name|v2u16_a
operator|=
name|v2u16_a
operator|+
name|c
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'unsigned int' to 'v2u16' (vector of 2 'unsigned short' values)}}
name|v2u16_a
operator|=
name|v2u16_a
operator|+
name|b
expr_stmt|;
name|v2u16_a
operator|=
name|v2u16_a
operator|+
name|a
expr_stmt|;
name|v2u8_a
operator|=
name|v2u8_a
operator|+
name|d
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned long long' and vector type 'v2u8' (vector of 2 'unsigned char' values) as implicit conversion would cause truncation}}
name|v2u8_a
operator|=
name|v2u8_a
operator|+
name|c
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned int' and vector type 'v2u8' (vector of 2 'unsigned char' values) as implicit conversion would cause truncation}}
name|v2u8_a
operator|=
name|v2u8_a
operator|+
name|b
expr_stmt|;
comment|// expected-warning {{implicit conversion loses integer precision: 'unsigned short' to 'v2u8' (vector of 2 'unsigned char' values)}}
name|v2u8_a
operator|=
name|v2u8_a
operator|+
name|a
expr_stmt|;
block|}
end_function

begin_function
name|void
name|uintTestConstant
parameter_list|(
name|v2u64
name|v2u64_a
parameter_list|,
name|v2u32
name|v2u32_a
parameter_list|,
name|v2u16
name|v2u16_a
parameter_list|,
name|v2u8
name|v2u8_a
parameter_list|)
block|{
name|v2u64_a
operator|=
name|v2u64_a
operator|+
literal|0xFFFFFFFFFFFFFFFF
expr_stmt|;
name|v2u32_a
operator|=
name|v2u32_a
operator|+
literal|0xFFFFFFFF
expr_stmt|;
name|v2u16_a
operator|=
name|v2u16_a
operator|+
literal|0xFFFF
expr_stmt|;
name|v2u8_a
operator|=
name|v2u8_a
operator|+
literal|0xFF
expr_stmt|;
name|v2u32_a
operator|=
name|v2u32_a
operator|+
literal|0x1FFFFFFFF
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'long' to 'v2u32' (vector of 2 'unsigned int' values) changes value from 8589934591 to 4294967295}}
name|v2u16_a
operator|=
name|v2u16_a
operator|+
literal|0x1FFFF
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'int' to 'v2u16' (vector of 2 'unsigned short' values) changes value from 131071 to 65535}}
name|v2u8_a
operator|=
name|v2u8_a
operator|+
literal|0x1FF
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'int' and vector type 'v2u8' (vector of 2 'unsigned char' values) as implicit conversion would cause truncation}}
block|}
end_function

begin_function
name|void
name|intTestConstant
parameter_list|(
name|v2i64
name|v2i64_a
parameter_list|,
name|v2i32
name|v2i32_a
parameter_list|,
name|v2i16
name|v2i16_a
parameter_list|,
name|v2i8
name|v2i8_a
parameter_list|)
block|{
comment|// Legal upper bounds.
name|v2i64_a
operator|=
name|v2i64_a
operator|+
operator|(
name|long
name|long
operator|)
literal|0x7FFFFFFFFFFFFFFF
expr_stmt|;
name|v2i32_a
operator|=
name|v2i32_a
operator|+
operator|(
name|int
operator|)
literal|0x7FFFFFFF
expr_stmt|;
name|v2i16_a
operator|=
name|v2i16_a
operator|+
operator|(
name|short
operator|)
literal|0x7FFF
expr_stmt|;
name|v2i8_a
operator|=
name|v2i8_a
operator|+
operator|(
name|char
operator|)
literal|0x7F
expr_stmt|;
comment|// Legal lower bounds.
name|v2i64_a
operator|=
name|v2i64_a
operator|+
operator|(
operator|-
literal|9223372036854775807
operator|)
expr_stmt|;
name|v2i32_a
operator|=
name|v2i32_a
operator|+
operator|(
operator|-
literal|2147483648
operator|)
expr_stmt|;
name|v2i16_a
operator|=
name|v2i16_a
operator|+
operator|(
operator|-
literal|32768
operator|)
expr_stmt|;
name|v2i8_a
operator|=
name|v2i8_a
operator|+
operator|(
operator|-
literal|128
operator|)
expr_stmt|;
comment|// One increment/decrement more than the type can hold
name|v2i32_a
operator|=
name|v2i32_a
operator|+
literal|2147483648
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'long' to 'v2i32' (vector of 2 'int' values) changes value from 2147483648 to -2147483648}}
name|v2i16_a
operator|=
name|v2i16_a
operator|+
literal|32768
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'int' to 'v2i16' (vector of 2 'short' values) changes value from 32768 to -32768}}
name|v2i8_a
operator|=
name|v2i8_a
operator|+
literal|128
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'int' to 'v2i8' (vector of 2 'char' values) changes value from 128 to -128}}
name|v2i32_a
operator|=
name|v2i32_a
operator|+
operator|(
operator|-
literal|2147483649
operator|)
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'long' to 'v2i32' (vector of 2 'int' values) changes value from -2147483649 to 2147483647}}
name|v2i16_a
operator|=
name|v2i16_a
operator|+
operator|(
operator|-
literal|32769
operator|)
expr_stmt|;
comment|// expected-warning {{implicit conversion from 'int' to 'v2i16' (vector of 2 'short' values) changes value from -32769 to 32767}}
name|v2i8_a
operator|=
name|v2i8_a
operator|+
operator|(
operator|-
literal|129
operator|)
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'int' and vector type 'v2i8' (vector of 2 'char' values) as implicit conversion would cause truncation}}
block|}
end_function

end_unit

