begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang %s -fsyntax-only -Xclang -verify -fblocks -Wunreachable-code -Wno-unused-value -Wno-covered-switch-default
end_comment

begin_function_decl
name|int
name|halt
parameter_list|()
function_decl|__attribute__
parameter_list|(
function_decl|(noreturn
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|int
name|live
parameter_list|()
function_decl|;
end_function_decl

begin_function_decl
name|int
name|dead
parameter_list|()
function_decl|;
end_function_decl

begin_function
name|void
name|test1
parameter_list|()
block|{
goto|goto
name|c
goto|;
name|d
label|:
goto|goto
name|e
goto|;
comment|// expected-warning {{will never be executed}}
name|c
label|:
empty_stmt|;
name|int
name|i
decl_stmt|;
return|return;
goto|goto
name|b
goto|;
comment|// expected-warning {{will never be executed}}
goto|goto
name|a
goto|;
comment|// expected-warning {{will never be executed}}
name|b
label|:
name|i
operator|=
literal|1
expr_stmt|;
name|a
label|:
name|i
operator|=
literal|2
expr_stmt|;
goto|goto
name|f
goto|;
name|e
label|:
goto|goto
name|d
goto|;
name|f
label|:
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test2
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
switch|switch
condition|(
name|live
argument_list|()
condition|)
block|{
case|case
literal|1
case|:
name|halt
argument_list|()
operator|,
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
case|case
literal|2
case|:
name|live
argument_list|()
operator|,
name|halt
argument_list|()
operator|,
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
case|case
literal|3
case|:
name|live
argument_list|()
operator|+
comment|// expected-warning {{will never be executed}}
name|halt
argument_list|()
expr_stmt|;
name|dead
argument_list|()
expr_stmt|;
case|case
literal|4
case|:
name|a4
label|:
name|live
argument_list|()
operator|,
name|halt
argument_list|()
expr_stmt|;
goto|goto
name|a4
goto|;
comment|// expected-warning {{will never be executed}}
case|case
literal|5
case|:
goto|goto
name|a5
goto|;
name|c5
label|:
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
goto|goto
name|b5
goto|;
name|a5
label|:
name|live
argument_list|()
operator|,
name|halt
argument_list|()
expr_stmt|;
name|b5
label|:
goto|goto
name|c5
goto|;
case|case
literal|6
case|:
if|if
condition|(
name|live
argument_list|()
condition|)
goto|goto
name|e6
goto|;
name|live
argument_list|()
operator|,
name|halt
argument_list|()
expr_stmt|;
name|d6
label|:
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
goto|goto
name|b6
goto|;
name|c6
label|:
name|dead
argument_list|()
expr_stmt|;
goto|goto
name|b6
goto|;
name|e6
label|:
name|live
argument_list|()
operator|,
name|halt
argument_list|()
expr_stmt|;
name|b6
label|:
goto|goto
name|c6
goto|;
case|case
literal|7
case|:
name|halt
argument_list|()
operator|+
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
operator|-
comment|// expected-warning {{will never be executed}}
name|halt
argument_list|()
expr_stmt|;
case|case
literal|8
case|:
name|i
comment|// expected-warning {{will never be executed}}
operator|+=
name|halt
argument_list|()
expr_stmt|;
case|case
literal|9
case|:
name|halt
argument_list|()
condition|?
comment|// expected-warning {{will never be executed}}
name|dead
argument_list|()
else|:
name|dead
argument_list|()
expr_stmt|;
case|case
literal|10
case|:
operator|(
comment|// expected-warning {{will never be executed}}
name|float
operator|)
name|halt
argument_list|()
expr_stmt|;
case|case
literal|11
case|:
block|{
name|int
name|a
index|[
literal|5
index|]
decl_stmt|;
name|live
argument_list|()
operator|,
name|a
index|[
name|halt
argument_list|()
comment|// expected-warning {{will never be executed}}
index|]
expr_stmt|;
block|}
block|}
block|}
end_function

begin_enum
enum|enum
name|Cases
block|{
name|C1
block|,
name|C2
block|,
name|C3
block|}
enum|;
end_enum

begin_function
name|int
name|test_enum_cases
parameter_list|(
name|enum
name|Cases
name|C
parameter_list|)
block|{
switch|switch
condition|(
name|C
condition|)
block|{
case|case
name|C1
case|:
case|case
name|C2
case|:
case|case
name|C3
case|:
return|return
literal|1
return|;
default|default:
block|{
name|int
name|i
init|=
literal|0
decl_stmt|;
comment|// expected-warning{{will never be executed}}
operator|++
name|i
expr_stmt|;
return|return
name|i
return|;
block|}
block|}
block|}
end_function

begin_comment
comment|// Handle unreachable code triggered by macro expansions.
end_comment

begin_function_decl
name|void
name|__myassert_rtn
parameter_list|(
specifier|const
name|char
modifier|*
parameter_list|,
specifier|const
name|char
modifier|*
parameter_list|,
name|int
parameter_list|,
specifier|const
name|char
modifier|*
parameter_list|)
function_decl|__attribute__
parameter_list|(
function_decl|(__noreturn__
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_define
define|#
directive|define
name|myassert
parameter_list|(
name|e
parameter_list|)
define|\
value|(__builtin_expect(!(e), 0) ? __myassert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
end_define

begin_function
name|void
name|test_assert
parameter_list|()
block|{
name|myassert
argument_list|(
literal|0
operator|&&
literal|"unreachable"
argument_list|)
expr_stmt|;
return|return;
comment|// no-warning
block|}
end_function

begin_comment
comment|// Test case for PR 9774.  Tests that dead code in macros aren't warned about.
end_comment

begin_define
define|#
directive|define
name|MY_MAX
parameter_list|(
name|a
parameter_list|,
name|b
parameter_list|)
value|((a)>= (b) ? (a) : (b))
end_define

begin_function
name|void
name|PR9774
parameter_list|(
name|int
modifier|*
name|s
parameter_list|)
block|{
for|for
control|(
name|int
name|i
init|=
literal|0
init|;
name|i
operator|<
name|MY_MAX
argument_list|(
literal|2
argument_list|,
literal|3
argument_list|)
condition|;
name|i
operator|++
control|)
comment|// no-warning
name|s
index|[
name|i
index|]
operator|=
literal|0
expr_stmt|;
block|}
end_function

end_unit

