begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-apple-darwin10 -fsyntax-only -Wcast-align -verify %s
end_comment

begin_comment
comment|// Simple casts.
end_comment

begin_function
name|void
name|test0
parameter_list|(
name|char
modifier|*
name|P
parameter_list|)
block|{
name|char
modifier|*
name|a
init|=
operator|(
name|char
operator|*
operator|)
name|P
decl_stmt|;
name|short
modifier|*
name|b
init|=
operator|(
name|short
operator|*
operator|)
name|P
decl_stmt|;
comment|// expected-warning {{cast from 'char *' to 'short *' increases required alignment from 1 to 2}}
name|int
modifier|*
name|c
init|=
operator|(
name|int
operator|*
operator|)
name|P
decl_stmt|;
comment|// expected-warning {{cast from 'char *' to 'int *' increases required alignment from 1 to 4}}
block|}
end_function

begin_comment
comment|// Casts from void* are a special case.
end_comment

begin_function
name|void
name|test1
parameter_list|(
name|void
modifier|*
name|P
parameter_list|)
block|{
name|char
modifier|*
name|a
init|=
operator|(
name|char
operator|*
operator|)
name|P
decl_stmt|;
name|short
modifier|*
name|b
init|=
operator|(
name|short
operator|*
operator|)
name|P
decl_stmt|;
name|int
modifier|*
name|c
init|=
operator|(
name|int
operator|*
operator|)
name|P
decl_stmt|;
specifier|const
specifier|volatile
name|void
modifier|*
name|P2
init|=
name|P
decl_stmt|;
name|char
modifier|*
name|d
init|=
operator|(
name|char
operator|*
operator|)
name|P2
decl_stmt|;
name|short
modifier|*
name|e
init|=
operator|(
name|short
operator|*
operator|)
name|P2
decl_stmt|;
name|int
modifier|*
name|f
init|=
operator|(
name|int
operator|*
operator|)
name|P2
decl_stmt|;
specifier|const
name|char
modifier|*
name|g
init|=
operator|(
specifier|const
name|char
operator|*
operator|)
name|P2
decl_stmt|;
specifier|const
name|short
modifier|*
name|h
init|=
operator|(
specifier|const
name|short
operator|*
operator|)
name|P2
decl_stmt|;
specifier|const
name|int
modifier|*
name|i
init|=
operator|(
specifier|const
name|int
operator|*
operator|)
name|P2
decl_stmt|;
specifier|const
specifier|volatile
name|char
modifier|*
name|j
init|=
operator|(
specifier|const
specifier|volatile
name|char
operator|*
operator|)
name|P2
decl_stmt|;
specifier|const
specifier|volatile
name|short
modifier|*
name|k
init|=
operator|(
specifier|const
specifier|volatile
name|short
operator|*
operator|)
name|P2
decl_stmt|;
specifier|const
specifier|volatile
name|int
modifier|*
name|l
init|=
operator|(
specifier|const
specifier|volatile
name|int
operator|*
operator|)
name|P2
decl_stmt|;
block|}
end_function

begin_comment
comment|// Aligned struct.
end_comment

begin_macro
name|__attribute__
argument_list|(
argument|(aligned(
literal|16
argument|))
argument_list|)
end_macro

begin_struct
struct|struct
name|A
block|{
name|char
name|buffer
index|[
literal|16
index|]
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|void
name|test2
parameter_list|(
name|char
modifier|*
name|P
parameter_list|)
block|{
name|struct
name|A
modifier|*
name|a
init|=
operator|(
expr|struct
name|A
operator|*
operator|)
name|P
decl_stmt|;
comment|// expected-warning {{cast from 'char *' to 'struct A *' increases required alignment from 1 to 16}}
block|}
end_function

begin_comment
comment|// Incomplete type.
end_comment

begin_function
name|void
name|test3
parameter_list|(
name|char
modifier|*
name|P
parameter_list|)
block|{
name|struct
name|B
modifier|*
name|b
init|=
operator|(
expr|struct
name|B
operator|*
operator|)
name|P
decl_stmt|;
block|}
end_function

end_unit

