begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-unknown-unknown -fsyntax-only -Wcast-qual -verify %s
end_comment

begin_comment
comment|// RUN: %clang_cc1 -triple x86_64-unknown-unknown -x c++ -fsyntax-only -Wcast-qual -verify %s
end_comment

begin_include
include|#
directive|include
file|<stdint.h>
end_include

begin_function
name|void
name|foo
parameter_list|()
block|{
specifier|const
name|char
modifier|*
specifier|const
name|ptr
init|=
literal|0
decl_stmt|;
specifier|const
name|char
modifier|*
specifier|const
modifier|*
name|ptrptr
init|=
literal|0
decl_stmt|;
name|char
modifier|*
name|y
init|=
operator|(
name|char
operator|*
operator|)
name|ptr
decl_stmt|;
comment|// expected-warning {{cast from 'const char *' to 'char *' drops const qualifier}}
name|char
modifier|*
modifier|*
name|y1
init|=
operator|(
name|char
operator|*
operator|*
operator|)
name|ptrptr
decl_stmt|;
comment|// expected-warning {{cast from 'const char *const' to 'char *' drops const qualifier}}
specifier|const
name|char
modifier|*
modifier|*
name|y2
init|=
operator|(
specifier|const
name|char
operator|*
operator|*
operator|)
name|ptrptr
decl_stmt|;
comment|// expected-warning {{cast from 'const char *const *' to 'const char **' drops const qualifier}}
name|char
modifier|*
name|z
init|=
operator|(
name|char
operator|*
operator|)
operator|(
name|uintptr_t
operator|)
operator|(
specifier|const
name|void
operator|*
operator|)
name|ptr
decl_stmt|;
comment|// no warning
name|char
modifier|*
name|z1
init|=
operator|(
name|char
operator|*
operator|)
operator|(
specifier|const
name|void
operator|*
operator|)
name|ptr
decl_stmt|;
comment|// expected-warning {{cast from 'const void *' to 'char *' drops const qualifier}}
specifier|volatile
name|char
modifier|*
name|vol
init|=
literal|0
decl_stmt|;
name|char
modifier|*
name|vol2
init|=
operator|(
name|char
operator|*
operator|)
name|vol
decl_stmt|;
comment|// expected-warning {{cast from 'volatile char *' to 'char *' drops volatile qualifier}}
specifier|const
specifier|volatile
name|char
modifier|*
name|volc
init|=
literal|0
decl_stmt|;
name|char
modifier|*
name|volc2
init|=
operator|(
name|char
operator|*
operator|)
name|volc
decl_stmt|;
comment|// expected-warning {{cast from 'const volatile char *' to 'char *' drops const and volatile qualifiers}}
name|int
modifier|*
modifier|*
name|intptrptr
decl_stmt|;
specifier|const
name|int
modifier|*
modifier|*
name|intptrptrc
init|=
operator|(
specifier|const
name|int
operator|*
operator|*
operator|)
name|intptrptr
decl_stmt|;
comment|// expected-warning {{cast from 'int **' to 'const int **' must have all intermediate pointers const qualified}}
specifier|volatile
name|int
modifier|*
modifier|*
name|intptrptrv
init|=
operator|(
specifier|volatile
name|int
operator|*
operator|*
operator|)
name|intptrptr
decl_stmt|;
comment|// expected-warning {{cast from 'int **' to 'volatile int **' must have all intermediate pointers const qualified}}
name|int
modifier|*
name|intptr
decl_stmt|;
specifier|const
name|int
modifier|*
name|intptrc
init|=
operator|(
specifier|const
name|int
operator|*
operator|)
name|intptr
decl_stmt|;
comment|// no warning
specifier|const
name|char
modifier|*
modifier|*
name|charptrptrc
decl_stmt|;
name|char
modifier|*
modifier|*
name|charptrptr
init|=
operator|(
name|char
operator|*
operator|*
operator|)
name|charptrptrc
decl_stmt|;
comment|// expected-warning {{cast from 'const char *' to 'char *' drops const qualifier}}
specifier|const
name|char
modifier|*
name|constcharptr
decl_stmt|;
name|char
modifier|*
name|charptr
init|=
operator|(
name|char
operator|*
operator|)
name|constcharptr
decl_stmt|;
comment|// expected-warning {{cast from 'const char *' to 'char *' drops const qualifier}}
specifier|const
name|char
modifier|*
name|constcharptr2
init|=
operator|(
name|char
operator|*
operator|)
name|constcharptr
decl_stmt|;
comment|// expected-warning {{cast from 'const char *' to 'char *' drops const qualifier}}
specifier|const
name|char
modifier|*
name|charptr2
init|=
operator|(
name|char
operator|*
operator|)
name|charptr
decl_stmt|;
comment|// no warning
block|}
end_function

begin_function
name|void
name|bar_0
parameter_list|()
block|{
struct|struct
name|C
block|{
specifier|const
name|int
name|a
decl_stmt|;
name|int
name|b
decl_stmt|;
block|}
struct|;
specifier|const
name|struct
name|C
name|S
init|=
block|{
literal|0
block|,
literal|0
block|}
decl_stmt|;
operator|*
operator|(
name|int
operator|*
operator|)
operator|(
operator|&
name|S
operator|.
name|a
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-warning {{cast from 'const int *' to 'int *' drops const qualifier}}
operator|*
operator|(
name|int
operator|*
operator|)
operator|(
operator|&
name|S
operator|.
name|b
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-warning {{cast from 'const int *' to 'int *' drops const qualifier}}
block|}
end_function

begin_function
name|void
name|bar_1
parameter_list|()
block|{
struct|struct
name|C
block|{
specifier|const
name|int
name|a
decl_stmt|;
name|int
name|b
decl_stmt|;
block|}
struct|;
name|struct
name|C
name|S
init|=
block|{
literal|0
block|,
literal|0
block|}
decl_stmt|;
name|S
operator|.
name|b
operator|=
literal|0
expr_stmt|;
comment|// no warning
operator|*
operator|(
name|int
operator|*
operator|)
operator|(
operator|&
name|S
operator|.
name|a
operator|)
operator|=
literal|0
expr_stmt|;
comment|// expected-warning {{cast from 'const int *' to 'int *' drops const qualifier}}
operator|*
operator|(
name|int
operator|*
operator|)
operator|(
operator|&
name|S
operator|.
name|b
operator|)
operator|=
literal|0
expr_stmt|;
comment|// no warning
block|}
end_function

end_unit

