begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang %s -fsyntax-only -Xclang -verify -fblocks -Wunreachable-code -Wno-unused-value
end_comment

begin_function_decl
name|int
name|halt
parameter_list|()
function_decl|__attribute__
parameter_list|(
function_decl|(noreturn
end_function_decl

begin_empty_stmt
unit|))
empty_stmt|;
end_empty_stmt

begin_function_decl
name|int
name|live
parameter_list|()
function_decl|;
end_function_decl

begin_function_decl
name|int
name|dead
parameter_list|()
function_decl|;
end_function_decl

begin_function
name|void
name|test1
parameter_list|()
block|{
goto|goto
name|c
goto|;
name|d
label|:
goto|goto
name|e
goto|;
comment|// expected-warning {{will never be executed}}
name|c
label|:
empty_stmt|;
name|int
name|i
decl_stmt|;
return|return;
goto|goto
name|b
goto|;
comment|// expected-warning {{will never be executed}}
goto|goto
name|a
goto|;
comment|// expected-warning {{will never be executed}}
name|b
label|:
name|i
operator|=
literal|1
expr_stmt|;
name|a
label|:
name|i
operator|=
literal|2
expr_stmt|;
goto|goto
name|f
goto|;
name|e
label|:
goto|goto
name|d
goto|;
name|f
label|:
empty_stmt|;
block|}
end_function

begin_function
name|void
name|test2
parameter_list|()
block|{
name|int
name|i
decl_stmt|;
switch|switch
condition|(
name|live
argument_list|()
condition|)
block|{
case|case
literal|1
case|:
name|halt
argument_list|()
operator|,
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
case|case
literal|2
case|:
name|live
argument_list|()
operator|,
name|halt
argument_list|()
operator|,
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
case|case
literal|3
case|:
name|live
argument_list|()
operator|+
comment|// expected-warning {{will never be executed}}
name|halt
argument_list|()
expr_stmt|;
name|dead
argument_list|()
expr_stmt|;
case|case
literal|4
case|:
name|a4
label|:
name|live
argument_list|()
operator|,
name|halt
argument_list|()
expr_stmt|;
goto|goto
name|a4
goto|;
comment|// expected-warning {{will never be executed}}
case|case
literal|5
case|:
goto|goto
name|a5
goto|;
name|c5
label|:
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
goto|goto
name|b5
goto|;
name|a5
label|:
name|live
argument_list|()
operator|,
name|halt
argument_list|()
expr_stmt|;
name|b5
label|:
goto|goto
name|c5
goto|;
case|case
literal|6
case|:
if|if
condition|(
name|live
argument_list|()
condition|)
goto|goto
name|e6
goto|;
name|live
argument_list|()
operator|,
name|halt
argument_list|()
expr_stmt|;
name|d6
label|:
name|dead
argument_list|()
expr_stmt|;
comment|// expected-warning {{will never be executed}}
goto|goto
name|b6
goto|;
name|c6
label|:
name|dead
argument_list|()
expr_stmt|;
goto|goto
name|b6
goto|;
name|e6
label|:
name|live
argument_list|()
operator|,
name|halt
argument_list|()
expr_stmt|;
name|b6
label|:
goto|goto
name|c6
goto|;
case|case
literal|7
case|:
name|halt
argument_list|()
operator|+
comment|// expected-warning {{will never be executed}}
name|dead
argument_list|()
expr_stmt|;
operator|-
comment|// expected-warning {{will never be executed}}
name|halt
argument_list|()
expr_stmt|;
case|case
literal|8
case|:
name|i
operator|+=
comment|// expected-warning {{will never be executed}}
name|halt
argument_list|()
expr_stmt|;
case|case
literal|9
case|:
name|halt
argument_list|()
condition|?
comment|// expected-warning {{will never be executed}}
name|dead
argument_list|()
else|:
name|dead
argument_list|()
expr_stmt|;
case|case
literal|10
case|:
operator|(
comment|// expected-warning {{will never be executed}}
name|float
operator|)
name|halt
argument_list|()
expr_stmt|;
case|case
literal|11
case|:
block|{
name|int
name|a
index|[
literal|5
index|]
decl_stmt|;
name|live
argument_list|()
operator|,
name|a
index|[
name|halt
argument_list|()
index|]
expr_stmt|;
comment|// expected-warning {{will never be executed}}
block|}
block|}
block|}
end_function

end_unit

