begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -Wstrlcpy-strlcat-size -verify -fsyntax-only %s
end_comment

begin_typedef
typedef|typedef
name|__SIZE_TYPE__
name|size_t
typedef|;
end_typedef

begin_function_decl
name|size_t
name|strlcpy
parameter_list|(
name|char
modifier|*
specifier|restrict
name|dst
parameter_list|,
specifier|const
name|char
modifier|*
specifier|restrict
name|src
parameter_list|,
name|size_t
name|size
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|size_t
name|strlcat
parameter_list|(
name|char
modifier|*
specifier|restrict
name|dst
parameter_list|,
specifier|const
name|char
modifier|*
specifier|restrict
name|src
parameter_list|,
name|size_t
name|size
parameter_list|)
function_decl|;
end_function_decl

begin_function_decl
name|size_t
name|strlen
parameter_list|(
specifier|const
name|char
modifier|*
name|s
parameter_list|)
function_decl|;
end_function_decl

begin_decl_stmt
name|char
name|s1
index|[
literal|100
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
name|s2
index|[
literal|200
index|]
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|char
modifier|*
name|s3
decl_stmt|;
end_decl_stmt

begin_struct
struct|struct
block|{
name|char
name|f1
index|[
literal|100
index|]
decl_stmt|;
name|char
name|f2
index|[
literal|100
index|]
index|[
literal|3
index|]
decl_stmt|;
block|}
name|s4
struct|,
modifier|*
modifier|*
name|s5
struct|;
end_struct

begin_decl_stmt
name|int
name|x
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|f
parameter_list|(
name|void
parameter_list|)
block|{
name|strlcpy
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
sizeof|sizeof
argument_list|(
name|s1
argument_list|)
argument_list|)
expr_stmt|;
comment|// no warning
name|strlcpy
argument_list|(
name|s1
argument_list|,
name|s2
argument_list|,
sizeof|sizeof
argument_list|(
name|s2
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcpy' call appears to be size of the source; expected the size of the destination}} expected-note {{change size argument to be the size of the destination}}
name|strlcpy
argument_list|(
name|s1
argument_list|,
name|s3
argument_list|,
name|strlen
argument_list|(
name|s3
argument_list|)
operator|+
literal|1
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcpy' call appears to be size of the source; expected the size of the destination}} expected-note {{change size argument to be the size of the destination}}
name|strlcat
argument_list|(
name|s2
argument_list|,
name|s3
argument_list|,
sizeof|sizeof
argument_list|(
name|s3
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcat' call appears to be size of the source; expected the size of the destination}} expected-note {{change size argument to be the size of the destination}}
name|strlcpy
argument_list|(
name|s4
operator|.
name|f1
argument_list|,
name|s2
argument_list|,
sizeof|sizeof
argument_list|(
name|s2
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcpy' call appears to be size of the source; expected the size of the destination}} expected-note {{change size argument to be the size of the destination}}
name|strlcpy
argument_list|(
operator|(
operator|*
name|s5
operator|)
operator|->
name|f2
index|[
name|x
index|]
argument_list|,
name|s2
argument_list|,
sizeof|sizeof
argument_list|(
name|s2
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcpy' call appears to be size of the source; expected the size of the destination}} expected-note {{change size argument to be the size of the destination}}
name|strlcpy
argument_list|(
name|s1
operator|+
literal|3
argument_list|,
name|s2
argument_list|,
sizeof|sizeof
argument_list|(
name|s2
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcpy' call appears to be size of the source; expected the size of the destination}}
block|}
end_function

begin_comment
comment|// Don't issue FIXIT for flexible arrays.
end_comment

begin_struct
struct|struct
name|S
block|{
name|int
name|y
decl_stmt|;
name|char
name|x
index|[]
decl_stmt|;
block|}
struct|;
end_struct

begin_function
name|void
name|flexible_arrays
parameter_list|(
name|struct
name|S
modifier|*
name|s
parameter_list|)
block|{
name|char
name|str
index|[]
init|=
literal|"hi"
decl_stmt|;
name|strlcpy
argument_list|(
name|s
operator|->
name|x
argument_list|,
name|str
argument_list|,
sizeof|sizeof
argument_list|(
name|str
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcpy' call appears to be size of the source; expected the size of the destination}}
block|}
end_function

begin_comment
comment|// Don't issue FIXIT for destinations of size 1.
end_comment

begin_function
name|void
name|size_1
parameter_list|()
block|{
name|char
name|z
index|[
literal|1
index|]
decl_stmt|;
name|char
name|str
index|[]
init|=
literal|"hi"
decl_stmt|;
name|strlcpy
argument_list|(
name|z
argument_list|,
name|str
argument_list|,
sizeof|sizeof
argument_list|(
name|str
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcpy' call appears to be size of the source; expected the size of the destination}}
block|}
end_function

begin_comment
comment|// Support VLAs.
end_comment

begin_function
name|void
name|vlas
parameter_list|(
name|int
name|size
parameter_list|)
block|{
name|char
name|z
index|[
name|size
index|]
decl_stmt|;
name|char
name|str
index|[]
init|=
literal|"hi"
decl_stmt|;
name|strlcpy
argument_list|(
name|z
argument_list|,
name|str
argument_list|,
sizeof|sizeof
argument_list|(
name|str
argument_list|)
argument_list|)
expr_stmt|;
comment|// expected-warning {{size argument in 'strlcpy' call appears to be size of the source; expected the size of the destination}} expected-note {{change size argument to be the size of the destination}}
block|}
end_function

end_unit

