begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -Wtautological-compare %s
end_comment

begin_define
define|#
directive|define
name|mydefine
value|2
end_define

begin_function
name|void
name|f
parameter_list|(
name|int
name|x
parameter_list|)
block|{
if|if
condition|(
operator|(
literal|8
operator|&
name|x
operator|)
operator|==
literal|3
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to false}}
if|if
condition|(
operator|(
name|x
operator|&
literal|8
operator|)
operator|==
literal|4
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to false}}
if|if
condition|(
operator|(
name|x
operator|&
literal|8
operator|)
operator|!=
literal|4
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to true}}
if|if
condition|(
operator|(
literal|2
operator|&
name|x
operator|)
operator|!=
literal|4
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to true}}
if|if
condition|(
operator|(
name|x
operator||
literal|4
operator|)
operator|==
literal|3
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to false}}
if|if
condition|(
operator|(
name|x
operator||
literal|3
operator|)
operator|!=
literal|4
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to true}}
if|if
condition|(
operator|(
literal|5
operator||
name|x
operator|)
operator|!=
literal|3
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to true}}
if|if
condition|(
operator|(
name|x
operator|&
literal|0x15
operator|)
operator|==
literal|0x13
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to false}}
if|if
condition|(
operator|(
literal|0x23
operator||
name|x
operator|)
operator|==
literal|0x155
condition|)
block|{}
comment|// expected-warning {{bitwise comparison always evaluates to false}}
if|if
condition|(
operator|(
name|x
operator|&
literal|8
operator|)
operator|==
literal|8
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator|&
literal|8
operator|)
operator|!=
literal|8
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator||
literal|4
operator|)
operator|==
literal|4
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator||
literal|4
operator|)
operator|!=
literal|4
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator|&
literal|9
operator|)
operator|==
literal|8
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator|&
literal|9
operator|)
operator|!=
literal|8
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator||
literal|4
operator|)
operator|==
literal|5
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator||
literal|4
operator|)
operator|!=
literal|5
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator|&
name|mydefine
operator|)
operator|==
literal|8
condition|)
block|{}
if|if
condition|(
operator|(
name|x
operator||
name|mydefine
operator|)
operator|==
literal|4
condition|)
block|{}
block|}
end_function

end_unit

