begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -fsyntax-only -verify -std=c11 -Wno-unused %s
end_comment

begin_function_decl
name|int
name|f
parameter_list|(
name|int
parameter_list|,
name|int
parameter_list|)
function_decl|;
end_function_decl

begin_typedef
typedef|typedef
struct|struct
name|A
block|{
name|int
name|x
decl_stmt|,
name|y
decl_stmt|;
block|}
name|A
typedef|;
end_typedef

begin_function
name|void
name|test
parameter_list|()
block|{
name|int
name|a
decl_stmt|;
name|int
name|xs
index|[
literal|10
index|]
decl_stmt|;
name|a
operator|+
operator|++
name|a
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access to 'a'}}
name|a
operator|=
operator|++
name|a
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications to 'a'}}
name|a
operator|+
name|a
operator|++
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access to 'a'}}
name|a
operator|=
name|a
operator|++
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications to 'a'}}
operator|(
name|a
operator|++
operator|,
name|a
operator|++
operator|)
expr_stmt|;
comment|// ok
operator|++
name|a
operator|+
operator|++
name|a
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|++
operator|+
name|a
operator|++
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|=
name|xs
index|[
operator|++
name|a
index|]
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|=
name|xs
index|[
name|a
operator|++
index|]
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|=
operator|(
operator|++
name|a
operator|,
operator|++
name|a
operator|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|=
operator|(
name|a
operator|++
operator|,
operator|++
name|a
operator|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|=
operator|(
name|a
operator|++
operator|,
name|a
operator|++
operator|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|f
argument_list|(
name|a
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// ok
name|f
argument_list|(
name|a
operator|=
literal|0
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
name|f
argument_list|(
name|a
argument_list|,
name|a
operator|+=
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
name|f
argument_list|(
name|a
operator|=
literal|0
argument_list|,
name|a
operator|=
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|=
name|f
argument_list|(
operator|++
name|a
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// ok
name|a
operator|=
name|f
argument_list|(
name|a
operator|++
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// ok
name|a
operator|=
name|f
argument_list|(
operator|++
name|a
argument_list|,
name|a
operator|++
argument_list|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
operator|++
name|a
operator|+
name|f
argument_list|(
operator|++
name|a
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|f
argument_list|(
operator|++
name|a
argument_list|,
literal|0
argument_list|)
operator|+
operator|++
name|a
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|++
operator|+
name|f
argument_list|(
name|a
operator|++
argument_list|,
literal|0
argument_list|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|f
argument_list|(
name|a
operator|++
argument_list|,
literal|0
argument_list|)
operator|+
name|a
operator|++
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|=
operator|++
name|a
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|a
operator|+=
operator|++
name|a
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
name|A
name|agg1
init|=
block|{
name|a
operator|++
block|,
name|a
operator|++
block|}
decl_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|A
name|agg2
init|=
block|{
name|a
operator|++
operator|+
name|a
block|,
name|a
operator|++
block|}
decl_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
operator|(
name|xs
index|[
literal|2
index|]
operator|&&
operator|(
name|a
operator|=
literal|0
operator|)
operator|)
operator|+
name|a
expr_stmt|;
comment|// ok
operator|(
literal|0
operator|&&
operator|(
name|a
operator|=
literal|0
operator|)
operator|)
operator|+
name|a
expr_stmt|;
comment|// ok
operator|(
literal|1
operator|&&
operator|(
name|a
operator|=
literal|0
operator|)
operator|)
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
operator|(
name|xs
index|[
literal|3
index|]
operator|||
operator|(
name|a
operator|=
literal|0
operator|)
operator|)
operator|+
name|a
expr_stmt|;
comment|// ok
operator|(
literal|0
operator|||
operator|(
name|a
operator|=
literal|0
operator|)
operator|)
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
operator|(
literal|1
operator|||
operator|(
name|a
operator|=
literal|0
operator|)
operator|)
operator|+
name|a
expr_stmt|;
comment|// ok
operator|(
name|xs
index|[
literal|4
index|]
condition|?
name|a
else|:
operator|++
name|a
operator|)
operator|+
name|a
expr_stmt|;
comment|// ok
operator|(
literal|0
condition|?
name|a
else|:
operator|++
name|a
operator|)
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
operator|(
literal|1
condition|?
name|a
else|:
operator|++
name|a
operator|)
operator|+
name|a
expr_stmt|;
comment|// ok
operator|(
name|xs
index|[
literal|5
index|]
condition|?
operator|++
name|a
else|:
operator|++
name|a
operator|)
operator|+
name|a
expr_stmt|;
comment|// FIXME: warn here
operator|(
operator|++
name|a
operator|,
name|xs
index|[
literal|6
index|]
condition|?
operator|++
name|a
else|:
literal|0
operator|)
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
comment|// Here, the read of the fourth 'a' might happen before or after the write to
comment|// the second 'a'.
name|a
operator|+=
operator|(
name|a
operator|++
operator|,
name|a
operator|)
operator|+
name|a
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access}}
name|int
modifier|*
name|p
init|=
name|xs
decl_stmt|;
name|a
operator|=
operator|*
operator|(
name|a
operator|++
operator|,
name|p
operator|)
expr_stmt|;
comment|// ok
name|a
operator|=
name|a
operator|++
operator|&&
name|a
expr_stmt|;
comment|// ok
name|A
modifier|*
name|q
init|=
operator|&
name|agg1
decl_stmt|;
operator|(
name|q
operator|=
operator|&
name|agg2
operator|)
operator|->
name|y
operator|=
name|q
operator|->
name|x
expr_stmt|;
comment|// expected-warning {{unsequenced modification and access to 'q'}}
comment|// This has undefined behavior if a == 0; otherwise, the side-effect of the
comment|// increment is sequenced before the value computation of 'f(a, a)', which is
comment|// sequenced before the value computation of the '&&', which is sequenced
comment|// before the assignment. We treat the sequencing in '&&' as being
comment|// unconditional.
name|a
operator|=
name|a
operator|++
operator|&&
name|f
argument_list|(
name|a
argument_list|,
name|a
argument_list|)
expr_stmt|;
comment|// This has undefined behavior if a != 0. FIXME: We should diagnose this.
operator|(
name|a
operator|&&
name|a
operator|++
operator|)
operator|+
name|a
expr_stmt|;
operator|(
name|xs
index|[
literal|7
index|]
operator|&&
operator|++
name|a
operator|)
operator|*
operator|(
operator|!
name|xs
index|[
literal|7
index|]
operator|&&
operator|++
name|a
operator|)
expr_stmt|;
comment|// ok
name|xs
index|[
literal|0
index|]
operator|=
operator|(
name|a
operator|=
literal|1
operator|,
name|a
operator|)
expr_stmt|;
comment|// ok
name|xs
index|[
literal|8
index|]
condition|?
operator|++
name|a
operator|+
name|a
operator|++
else|:
literal|0
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|xs
index|[
literal|8
index|]
condition|?
literal|0
else|:
operator|++
name|a
operator|+
name|a
operator|++
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|xs
index|[
literal|8
index|]
condition|?
operator|++
name|a
else|:
name|a
operator|++
expr_stmt|;
comment|// ok
name|xs
index|[
literal|8
index|]
operator|&&
operator|(
operator|++
name|a
operator|+
name|a
operator|++
operator|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
name|xs
index|[
literal|8
index|]
operator|||
operator|(
operator|++
name|a
operator|+
name|a
operator|++
operator|)
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
operator|(
name|__builtin_classify_type
argument_list|(
operator|++
name|a
argument_list|)
condition|?
literal|1
else|:
literal|0
operator|)
operator|+
operator|++
name|a
expr_stmt|;
comment|// ok
operator|(
name|__builtin_constant_p
argument_list|(
operator|++
name|a
argument_list|)
condition|?
literal|1
else|:
literal|0
operator|)
operator|+
operator|++
name|a
expr_stmt|;
comment|// ok
operator|(
name|__builtin_expect
argument_list|(
operator|++
name|a
argument_list|,
literal|0
argument_list|)
condition|?
literal|1
else|:
literal|0
operator|)
operator|+
operator|++
name|a
expr_stmt|;
comment|// expected-warning {{multiple unsequenced modifications}}
generic_selection|_Generic(
operator|++
name|a
generic_selection|,
type|default
association|:
literal|0
generic_selection|)
operator|+
operator|++
name|a
expr_stmt|;
comment|// ok
sizeof|sizeof
argument_list|(
operator|++
name|a
argument_list|)
operator|+
operator|++
name|a
expr_stmt|;
comment|// ok
alignof|_Alignof
argument_list|(
operator|++
name|a
argument_list|)
operator|+
operator|++
name|a
expr_stmt|;
comment|// expected-warning {{extension}}
block|}
end_function

end_unit

