begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: clang-cc -Wchar-subscripts -fsyntax-only -verify %s
end_comment

begin_function
name|void
name|t1
parameter_list|()
block|{
name|int
name|array
index|[
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
name|char
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|array
index|[
name|subscript
index|]
decl_stmt|;
comment|// expected-warning{{array subscript is of type 'char'}}
block|}
end_function

begin_function
name|void
name|t2
parameter_list|()
block|{
name|int
name|array
index|[
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
name|char
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|subscript
index|[
name|array
index|]
decl_stmt|;
comment|// expected-warning{{array subscript is of type 'char'}}
block|}
end_function

begin_function
name|void
name|t3
parameter_list|()
block|{
name|int
modifier|*
name|array
init|=
literal|0
decl_stmt|;
name|char
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|array
index|[
name|subscript
index|]
decl_stmt|;
comment|// expected-warning{{array subscript is of type 'char'}}
block|}
end_function

begin_function
name|void
name|t4
parameter_list|()
block|{
name|int
modifier|*
name|array
init|=
literal|0
decl_stmt|;
name|char
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|subscript
index|[
name|array
index|]
decl_stmt|;
comment|// expected-warning{{array subscript is of type 'char'}}
block|}
end_function

begin_function_decl
name|char
name|returnsChar
parameter_list|()
function_decl|;
end_function_decl

begin_function
name|void
name|t5
parameter_list|()
block|{
name|int
modifier|*
name|array
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|array
index|[
name|returnsChar
argument_list|()
index|]
decl_stmt|;
comment|// expected-warning{{array subscript is of type 'char'}}
block|}
end_function

begin_function
name|void
name|t6
parameter_list|()
block|{
name|int
name|array
index|[
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
name|signed
name|char
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|array
index|[
name|subscript
index|]
decl_stmt|;
comment|// no warning for explicit signed char
block|}
end_function

begin_function
name|void
name|t7
parameter_list|()
block|{
name|int
name|array
index|[
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
name|unsigned
name|char
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|array
index|[
name|subscript
index|]
decl_stmt|;
comment|// no warning for unsigned char
block|}
end_function

begin_typedef
typedef|typedef
name|char
name|CharTy
typedef|;
end_typedef

begin_function
name|void
name|t8
parameter_list|()
block|{
name|int
name|array
index|[
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
name|CharTy
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|array
index|[
name|subscript
index|]
decl_stmt|;
comment|// expected-warning{{array subscript is of type 'char'}}
block|}
end_function

begin_typedef
typedef|typedef
name|signed
name|char
name|SignedCharTy
typedef|;
end_typedef

begin_function
name|void
name|t9
parameter_list|()
block|{
name|int
name|array
index|[
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
name|SignedCharTy
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|array
index|[
name|subscript
index|]
decl_stmt|;
comment|// no warning for explicit signed char
block|}
end_function

begin_typedef
typedef|typedef
name|unsigned
name|char
name|UnsignedCharTy
typedef|;
end_typedef

begin_function
name|void
name|t10
parameter_list|()
block|{
name|int
name|array
index|[
literal|1
index|]
init|=
block|{
literal|0
block|}
decl_stmt|;
name|UnsignedCharTy
name|subscript
init|=
literal|0
decl_stmt|;
name|int
name|val
init|=
name|array
index|[
name|subscript
index|]
decl_stmt|;
comment|// no warning for unsigned char
block|}
end_function

end_unit

