begin_unit|revision:0.9.5;language:C;cregit-version:0.0.1
begin_comment
comment|// RUN: %clang_cc1 -triple s390x-linux-gnu -fzvector \
end_comment

begin_comment
comment|// RUN:  -fno-lax-vector-conversions -W -Wall -Wconversion \
end_comment

begin_comment
comment|// RUN:  -Werror -fsyntax-only -verify %s
end_comment

begin_decl_stmt
name|vector
name|signed
name|char
name|sc
decl_stmt|,
name|sc2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|unsigned
name|char
name|uc
decl_stmt|,
name|uc2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|bool
name|char
name|bc
decl_stmt|,
name|bc2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|signed
name|short
name|ss
decl_stmt|,
name|ss2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|unsigned
name|short
name|us
decl_stmt|,
name|us2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|bool
name|short
name|bs
decl_stmt|,
name|bs2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|signed
name|int
name|si
decl_stmt|,
name|si2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|unsigned
name|int
name|ui
decl_stmt|,
name|ui2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|bool
name|int
name|bi
decl_stmt|,
name|bi2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|signed
name|long
name|long
name|sl
decl_stmt|,
name|sl2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|unsigned
name|long
name|long
name|ul
decl_stmt|,
name|ul2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|bool
name|long
name|long
name|bl
decl_stmt|,
name|bl2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|double
name|fd
decl_stmt|,
name|fd2
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|vector
name|long
name|ll
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error {{cannot use 'long' with '__vector'}}
end_comment

begin_decl_stmt
name|vector
name|float
name|ff
decl_stmt|;
end_decl_stmt

begin_comment
comment|// expected-error {{cannot use 'float' with '__vector'}}
end_comment

begin_decl_stmt
name|signed
name|char
name|sc_scalar
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|unsigned
name|char
name|uc_scalar
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|signed
name|short
name|ss_scalar
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|unsigned
name|short
name|us_scalar
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|signed
name|int
name|si_scalar
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|unsigned
name|int
name|ui_scalar
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|signed
name|long
name|sl_scalar
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|unsigned
name|long
name|ul_scalar
decl_stmt|;
end_decl_stmt

begin_decl_stmt
name|double
name|fd_scalar
decl_stmt|;
end_decl_stmt

begin_function
name|void
name|foo
parameter_list|(
name|void
parameter_list|)
block|{
comment|// -------------------------------------------------------------------------
comment|// Test assignment.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc2
expr_stmt|;
name|uc
operator|=
name|uc2
expr_stmt|;
name|bc
operator|=
name|bc2
expr_stmt|;
name|ss
operator|=
name|ss2
expr_stmt|;
name|us
operator|=
name|us2
expr_stmt|;
name|bs
operator|=
name|bs2
expr_stmt|;
name|si
operator|=
name|si2
expr_stmt|;
name|ui
operator|=
name|ui2
expr_stmt|;
name|bi
operator|=
name|bi2
expr_stmt|;
name|sl
operator|=
name|sl2
expr_stmt|;
name|ul
operator|=
name|ul2
expr_stmt|;
name|bl
operator|=
name|bl2
expr_stmt|;
name|fd
operator|=
name|fd2
expr_stmt|;
name|sc
operator|=
name|uc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|uc
operator|=
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|uc
operator|=
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|uc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|sc_scalar
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|uc_scalar
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|uc
operator|=
name|sc_scalar
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|uc
operator|=
name|uc_scalar
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|sc_scalar
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|uc_scalar
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|ss2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|si2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|sl2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|fd2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|ss
operator|=
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|si
operator|=
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sl
operator|=
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|fd
operator|=
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|uc
operator|=
name|us2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|uc
operator|=
name|ui2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|uc
operator|=
name|ul2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|uc
operator|=
name|fd2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|us
operator|=
name|uc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|ui
operator|=
name|uc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|ul
operator|=
name|uc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|fd
operator|=
name|uc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|us2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|ui2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|ul2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|fd2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bs
operator|=
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bi
operator|=
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bl
operator|=
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|fd
operator|=
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
comment|// -------------------------------------------------------------------------
comment|// Test casts to same element width.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
operator|(
name|vector
name|signed
name|char
operator|)
name|bc2
expr_stmt|;
name|bc
operator|=
operator|(
name|vector
name|bool
name|char
operator|)
name|uc2
expr_stmt|;
name|uc
operator|=
operator|(
name|vector
name|unsigned
name|char
operator|)
name|sc2
expr_stmt|;
name|ss
operator|=
operator|(
name|vector
name|signed
name|short
operator|)
name|bs2
expr_stmt|;
name|bs
operator|=
operator|(
name|vector
name|bool
name|short
operator|)
name|us2
expr_stmt|;
name|us
operator|=
operator|(
name|vector
name|unsigned
name|short
operator|)
name|ss2
expr_stmt|;
name|si
operator|=
operator|(
name|vector
name|signed
name|int
operator|)
name|bi2
expr_stmt|;
name|bi
operator|=
operator|(
name|vector
name|bool
name|int
operator|)
name|ui2
expr_stmt|;
name|ui
operator|=
operator|(
name|vector
name|unsigned
name|int
operator|)
name|si2
expr_stmt|;
name|sl
operator|=
operator|(
name|vector
name|signed
name|long
name|long
operator|)
name|bl2
expr_stmt|;
name|bl
operator|=
operator|(
name|vector
name|bool
name|long
name|long
operator|)
name|ul2
expr_stmt|;
name|ul
operator|=
operator|(
name|vector
name|unsigned
name|long
name|long
operator|)
name|fd2
expr_stmt|;
name|fd
operator|=
operator|(
name|vector
name|double
operator|)
name|sl2
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test casts to different element width.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
operator|(
name|vector
name|signed
name|char
operator|)
name|bs2
expr_stmt|;
name|bc
operator|=
operator|(
name|vector
name|bool
name|char
operator|)
name|us2
expr_stmt|;
name|uc
operator|=
operator|(
name|vector
name|unsigned
name|char
operator|)
name|fd2
expr_stmt|;
name|ss
operator|=
operator|(
name|vector
name|signed
name|short
operator|)
name|bi2
expr_stmt|;
name|bs
operator|=
operator|(
name|vector
name|bool
name|short
operator|)
name|ui2
expr_stmt|;
name|us
operator|=
operator|(
name|vector
name|unsigned
name|short
operator|)
name|fd2
expr_stmt|;
name|si
operator|=
operator|(
name|vector
name|signed
name|int
operator|)
name|bl2
expr_stmt|;
name|bi
operator|=
operator|(
name|vector
name|bool
name|int
operator|)
name|ul2
expr_stmt|;
name|ui
operator|=
operator|(
name|vector
name|unsigned
name|int
operator|)
name|fd2
expr_stmt|;
name|sl
operator|=
operator|(
name|vector
name|signed
name|long
name|long
operator|)
name|bc2
expr_stmt|;
name|bl
operator|=
operator|(
name|vector
name|bool
name|long
name|long
operator|)
name|uc2
expr_stmt|;
name|ul
operator|=
operator|(
name|vector
name|unsigned
name|long
name|long
operator|)
name|sc2
expr_stmt|;
name|fd
operator|=
operator|(
name|vector
name|double
operator|)
name|sc2
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test ++.
comment|// -------------------------------------------------------------------------
operator|++
name|sc2
expr_stmt|;
operator|++
name|uc2
expr_stmt|;
operator|++
name|bc2
expr_stmt|;
comment|// expected-error {{cannot increment}}
operator|++
name|ss2
expr_stmt|;
operator|++
name|us2
expr_stmt|;
operator|++
name|bs2
expr_stmt|;
comment|// expected-error {{cannot increment}}
operator|++
name|si2
expr_stmt|;
operator|++
name|ui2
expr_stmt|;
operator|++
name|bi2
expr_stmt|;
comment|// expected-error {{cannot increment}}
operator|++
name|sl2
expr_stmt|;
operator|++
name|ul2
expr_stmt|;
operator|++
name|bl2
expr_stmt|;
comment|// expected-error {{cannot increment}}
operator|++
name|fd2
expr_stmt|;
name|sc
operator|++
expr_stmt|;
name|uc
operator|++
expr_stmt|;
name|bc
operator|++
expr_stmt|;
comment|// expected-error {{cannot increment}}
name|ss
operator|++
expr_stmt|;
name|us
operator|++
expr_stmt|;
name|bs
operator|++
expr_stmt|;
comment|// expected-error {{cannot increment}}
name|si
operator|++
expr_stmt|;
name|ui
operator|++
expr_stmt|;
name|bi
operator|++
expr_stmt|;
comment|// expected-error {{cannot increment}}
name|sl
operator|++
expr_stmt|;
name|ul
operator|++
expr_stmt|;
name|bl
operator|++
expr_stmt|;
comment|// expected-error {{cannot increment}}
name|fd
operator|++
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test --.
comment|// -------------------------------------------------------------------------
operator|--
name|sc2
expr_stmt|;
operator|--
name|uc2
expr_stmt|;
operator|--
name|bc2
expr_stmt|;
comment|// expected-error {{cannot decrement}}
operator|--
name|ss2
expr_stmt|;
operator|--
name|us2
expr_stmt|;
operator|--
name|bs2
expr_stmt|;
comment|// expected-error {{cannot decrement}}
operator|--
name|si2
expr_stmt|;
operator|--
name|ui2
expr_stmt|;
operator|--
name|bi2
expr_stmt|;
comment|// expected-error {{cannot decrement}}
operator|--
name|sl2
expr_stmt|;
operator|--
name|ul2
expr_stmt|;
operator|--
name|bl2
expr_stmt|;
comment|// expected-error {{cannot decrement}}
operator|--
name|fd2
expr_stmt|;
name|sc
operator|--
expr_stmt|;
name|uc
operator|--
expr_stmt|;
name|bc
operator|--
expr_stmt|;
comment|// expected-error {{cannot decrement}}
name|ss
operator|--
expr_stmt|;
name|us
operator|--
expr_stmt|;
name|bs
operator|--
expr_stmt|;
comment|// expected-error {{cannot decrement}}
name|si
operator|--
expr_stmt|;
name|ui
operator|--
expr_stmt|;
name|bi
operator|--
expr_stmt|;
comment|// expected-error {{cannot decrement}}
name|sl
operator|--
expr_stmt|;
name|ul
operator|--
expr_stmt|;
name|bl
operator|--
expr_stmt|;
comment|// expected-error {{cannot decrement}}
name|fd
operator|--
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test unary +.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
operator|+
name|sc2
expr_stmt|;
name|uc
operator|=
operator|+
name|uc2
expr_stmt|;
name|bc
operator|=
operator|+
name|bc2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
name|ss
operator|=
operator|+
name|ss2
expr_stmt|;
name|us
operator|=
operator|+
name|us2
expr_stmt|;
name|bs
operator|=
operator|+
name|bs2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
name|si
operator|=
operator|+
name|si2
expr_stmt|;
name|ui
operator|=
operator|+
name|ui2
expr_stmt|;
name|bi
operator|=
operator|+
name|bi2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
name|sl
operator|=
operator|+
name|sl2
expr_stmt|;
name|ul
operator|=
operator|+
name|ul2
expr_stmt|;
name|bl
operator|=
operator|+
name|bl2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
name|fd
operator|=
operator|+
name|fd2
expr_stmt|;
name|sc
operator|=
operator|+
name|si2
expr_stmt|;
comment|// expected-error {{assigning to}}
name|ui
operator|=
operator|+
name|si2
expr_stmt|;
comment|// expected-error {{assigning to}}
name|ui
operator|=
operator|+
name|bi2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
comment|// -------------------------------------------------------------------------
comment|// Test unary -.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
operator|-
name|sc2
expr_stmt|;
name|uc
operator|=
operator|-
name|uc2
expr_stmt|;
name|bc
operator|=
operator|-
name|bc2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
name|ss
operator|=
operator|-
name|ss2
expr_stmt|;
name|us
operator|=
operator|-
name|us2
expr_stmt|;
name|bs
operator|=
operator|-
name|bs2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
name|si
operator|=
operator|-
name|si2
expr_stmt|;
name|ui
operator|=
operator|-
name|ui2
expr_stmt|;
name|bi
operator|=
operator|-
name|bi2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
name|sl
operator|=
operator|-
name|sl2
expr_stmt|;
name|ul
operator|=
operator|-
name|ul2
expr_stmt|;
name|bl
operator|=
operator|-
name|bl2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
name|fd
operator|=
operator|-
name|fd2
expr_stmt|;
name|sc
operator|=
operator|-
name|si2
expr_stmt|;
comment|// expected-error {{assigning to}}
name|ui
operator|=
operator|-
name|si2
expr_stmt|;
comment|// expected-error {{assigning to}}
name|ui
operator|=
operator|-
name|bi2
expr_stmt|;
comment|// expected-error {{invalid argument type}}
comment|// -------------------------------------------------------------------------
comment|// Test ~.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
operator|~
name|sc2
expr_stmt|;
name|uc
operator|=
operator|~
name|uc2
expr_stmt|;
name|bc
operator|=
operator|~
name|bc2
expr_stmt|;
name|ss
operator|=
operator|~
name|ss2
expr_stmt|;
name|us
operator|=
operator|~
name|us2
expr_stmt|;
name|bs
operator|=
operator|~
name|bs2
expr_stmt|;
name|si
operator|=
operator|~
name|si2
expr_stmt|;
name|ui
operator|=
operator|~
name|ui2
expr_stmt|;
name|bi
operator|=
operator|~
name|bi2
expr_stmt|;
name|sl
operator|=
operator|~
name|sl2
expr_stmt|;
name|ul
operator|=
operator|~
name|ul2
expr_stmt|;
name|bl
operator|=
operator|~
name|bl2
expr_stmt|;
name|fd
operator|=
operator|~
name|fd2
expr_stmt|;
comment|// expected-error {{invalid argument}}
name|sc
operator|=
operator|~
name|si2
expr_stmt|;
comment|// expected-error {{assigning to}}
name|ui
operator|=
operator|~
name|si2
expr_stmt|;
comment|// expected-error {{assigning to}}
name|ui
operator|=
operator|~
name|bi2
expr_stmt|;
comment|// expected-error {{assigning to}}
comment|// -------------------------------------------------------------------------
comment|// Test binary +.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|+
name|sc2
expr_stmt|;
name|sc
operator|=
name|sc
operator|+
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|uc
operator|+
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|sc
operator|+
name|bc2
expr_stmt|;
name|sc
operator|=
name|bc
operator|+
name|sc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|+
name|uc2
expr_stmt|;
name|uc
operator|=
name|sc
operator|+
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|uc
operator|+
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|bc
operator|+
name|uc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|+
name|bc2
expr_stmt|;
name|bc
operator|=
name|bc
operator|+
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|bc
operator|+
name|uc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|uc
operator|+
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|bc
operator|+
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|sc
operator|+
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|sc
operator|+
name|sc_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|+
name|uc_scalar
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned char' and vector type '__vector signed char' (vector of 16 'signed char' values) as implicit conversion would cause truncation}}
name|sc
operator|=
name|sc_scalar
operator|+
name|sc
expr_stmt|;
name|sc
operator|=
name|uc_scalar
operator|+
name|sc
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned char' and vector type '__vector signed char' (vector of 16 'signed char' values) as implicit conversion would cause truncation}}
name|uc
operator|=
name|uc
operator|+
name|sc_scalar
expr_stmt|;
comment|// expected-error {{implicit conversion changes signedness: 'signed char' to '__vector unsigned char' (vector of 16 'unsigned char' values)}}
name|uc
operator|=
name|uc
operator|+
name|uc_scalar
expr_stmt|;
name|uc
operator|=
name|sc_scalar
operator|+
name|uc
expr_stmt|;
comment|// expected-error {{implicit conversion changes signedness: 'signed char' to '__vector unsigned char' (vector of 16 'unsigned char' values)}}
name|uc
operator|=
name|uc_scalar
operator|+
name|uc
expr_stmt|;
name|ss
operator|=
name|ss
operator|+
name|ss2
expr_stmt|;
name|us
operator|=
name|us
operator|+
name|us2
expr_stmt|;
name|bs
operator|=
name|bs
operator|+
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|=
name|si
operator|+
name|si2
expr_stmt|;
name|ui
operator|=
name|ui
operator|+
name|ui2
expr_stmt|;
name|bi
operator|=
name|bi
operator|+
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|=
name|sl
operator|+
name|sl2
expr_stmt|;
name|ul
operator|=
name|ul
operator|+
name|ul2
expr_stmt|;
name|bl
operator|=
name|bl
operator|+
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|fd
operator|+
name|fd2
expr_stmt|;
name|fd
operator|=
name|fd
operator|+
name|ul2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|fd
operator|=
name|sl
operator|+
name|fd2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|+=
name|sc2
expr_stmt|;
name|sc
operator|+=
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|+=
name|bc2
expr_stmt|;
name|uc
operator|+=
name|uc2
expr_stmt|;
name|uc
operator|+=
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|+=
name|bc2
expr_stmt|;
name|bc
operator|+=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|+=
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|bc
operator|+=
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|+=
name|ss2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|+=
name|si2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|+=
name|sl2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|+=
name|fd2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|+=
name|sc_scalar
expr_stmt|;
name|sc
operator|+=
name|uc_scalar
expr_stmt|;
comment|// expected-error {{cannot convert between scalar type 'unsigned char' and vector type '__vector signed char' (vector of 16 'signed char' values) as implicit conversion would cause truncation}}
name|uc
operator|+=
name|sc_scalar
expr_stmt|;
comment|// expected-error {{implicit conversion changes signedness: 'signed char' to '__vector unsigned char' (vector of 16 'unsigned char' values)}}
name|uc
operator|+=
name|uc_scalar
expr_stmt|;
name|ss
operator|+=
name|ss2
expr_stmt|;
name|us
operator|+=
name|us2
expr_stmt|;
name|bs
operator|+=
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|+=
name|si2
expr_stmt|;
name|ui
operator|+=
name|ui2
expr_stmt|;
name|bi
operator|+=
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|+=
name|sl2
expr_stmt|;
name|ul
operator|+=
name|ul2
expr_stmt|;
name|bl
operator|+=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|+=
name|fd2
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that binary + rules apply to binary - too.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|-
name|sc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|-
name|uc2
expr_stmt|;
name|bc
operator|=
name|bc
operator|-
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|=
name|uc
operator|-
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|sc
operator|-
name|bc2
expr_stmt|;
name|uc
operator|=
name|bc
operator|-
name|uc2
expr_stmt|;
name|sc
operator|-=
name|sc2
expr_stmt|;
name|uc
operator|-=
name|uc2
expr_stmt|;
name|bc
operator|-=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|-=
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|-=
name|bc2
expr_stmt|;
name|bc
operator|-=
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|ss
operator|-=
name|ss2
expr_stmt|;
name|us
operator|-=
name|us2
expr_stmt|;
name|bs
operator|-=
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|-=
name|si2
expr_stmt|;
name|ui
operator|-=
name|ui2
expr_stmt|;
name|bi
operator|-=
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|-=
name|sl2
expr_stmt|;
name|ul
operator|-=
name|ul2
expr_stmt|;
name|bl
operator|-=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|-=
name|fd2
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that binary + rules apply to * too.  64-bit integer multiplication
comment|// is not required by the spec and so isn't tested here.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|*
name|sc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|*
name|uc2
expr_stmt|;
name|bc
operator|=
name|bc
operator|*
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|=
name|uc
operator|*
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|sc
operator|*
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|bc
operator|*
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|*=
name|sc2
expr_stmt|;
name|uc
operator|*=
name|uc2
expr_stmt|;
name|bc
operator|*=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|*=
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|*=
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|bc
operator|*=
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|ss
operator|*=
name|ss2
expr_stmt|;
name|us
operator|*=
name|us2
expr_stmt|;
name|bs
operator|*=
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|*=
name|si2
expr_stmt|;
name|ui
operator|*=
name|ui2
expr_stmt|;
name|bi
operator|*=
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|*=
name|sl2
expr_stmt|;
name|ul
operator|*=
name|ul2
expr_stmt|;
name|bl
operator|*=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|*=
name|fd2
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that * rules apply to / too.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|/
name|sc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|/
name|uc2
expr_stmt|;
name|bc
operator|=
name|bc
operator|/
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|=
name|uc
operator|/
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|sc
operator|/
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|bc
operator|/
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|/=
name|sc2
expr_stmt|;
name|uc
operator|/=
name|uc2
expr_stmt|;
name|bc
operator|/=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|/=
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|/=
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|bc
operator|/=
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|ss
operator|/=
name|ss2
expr_stmt|;
name|us
operator|/=
name|us2
expr_stmt|;
name|bs
operator|/=
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|/=
name|si2
expr_stmt|;
name|ui
operator|/=
name|ui2
expr_stmt|;
name|bi
operator|/=
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|/=
name|sl2
expr_stmt|;
name|ul
operator|/=
name|ul2
expr_stmt|;
name|bl
operator|/=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|/=
name|fd2
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that / rules apply to % too, except that doubles are not allowed.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|%
name|sc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|%
name|uc2
expr_stmt|;
name|bc
operator|=
name|bc
operator|%
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|=
name|uc
operator|%
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|sc
operator|%
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|bc
operator|%
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|%=
name|sc2
expr_stmt|;
name|uc
operator|%=
name|uc2
expr_stmt|;
name|bc
operator|%=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|%=
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|%=
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|bc
operator|%=
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|ss
operator|%=
name|ss2
expr_stmt|;
name|us
operator|%=
name|us2
expr_stmt|;
name|bs
operator|%=
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|%=
name|si2
expr_stmt|;
name|ui
operator|%=
name|ui2
expr_stmt|;
name|bi
operator|%=
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|%=
name|sl2
expr_stmt|;
name|ul
operator|%=
name|ul2
expr_stmt|;
name|bl
operator|%=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|%=
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
comment|// -------------------------------------------------------------------------
comment|// Test&.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|&
name|sc2
expr_stmt|;
name|sc
operator|=
name|sc
operator|&
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|uc
operator|&
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|sc
operator|&
name|bc2
expr_stmt|;
name|sc
operator|=
name|bc
operator|&
name|sc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|&
name|uc2
expr_stmt|;
name|uc
operator|=
name|sc
operator|&
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|uc
operator|&
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|bc
operator|&
name|uc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|&
name|bc2
expr_stmt|;
name|bc
operator|=
name|bc
operator|&
name|bc2
expr_stmt|;
name|bc
operator|=
name|bc
operator|&
name|uc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|uc
operator|&
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|bc
operator|&
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|sc
operator|&
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|fd
operator|=
name|fd
operator|&
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|bl
operator|&
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|fd
operator|&
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|fd
operator|&
name|sl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|fd
operator|&
name|ul2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|&=
name|sc2
expr_stmt|;
name|sc
operator|&=
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|&=
name|bc2
expr_stmt|;
name|uc
operator|&=
name|uc2
expr_stmt|;
name|uc
operator|&=
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|&=
name|bc2
expr_stmt|;
name|bc
operator|&=
name|bc2
expr_stmt|;
name|bc
operator|&=
name|sc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|bc
operator|&=
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|&=
name|ss2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|&=
name|si2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|&=
name|sl2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|&=
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|us
operator|&=
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|ui
operator|&=
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|ul
operator|&=
name|bc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|fd
operator|&=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|ss
operator|&=
name|ss2
expr_stmt|;
name|us
operator|&=
name|us2
expr_stmt|;
name|bs
operator|&=
name|bs2
expr_stmt|;
name|si
operator|&=
name|si2
expr_stmt|;
name|ui
operator|&=
name|ui2
expr_stmt|;
name|bi
operator|&=
name|bi2
expr_stmt|;
name|sl
operator|&=
name|sl2
expr_stmt|;
name|ul
operator|&=
name|ul2
expr_stmt|;
name|bl
operator|&=
name|bl2
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that& rules apply to | too.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator||
name|sc2
expr_stmt|;
name|sc
operator|=
name|sc
operator||
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|sc
operator||
name|bc2
expr_stmt|;
name|uc
operator|=
name|uc
operator||
name|uc2
expr_stmt|;
name|uc
operator|=
name|sc
operator||
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|bc
operator||
name|uc2
expr_stmt|;
name|bc
operator|=
name|bc
operator||
name|bc2
expr_stmt|;
name|bc
operator|=
name|uc
operator||
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|bc
operator||
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|fd
operator|=
name|fd
operator||
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|bl
operator||
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|ss
operator||=
name|ss2
expr_stmt|;
name|us
operator||=
name|us2
expr_stmt|;
name|bs
operator||=
name|bs2
expr_stmt|;
name|si
operator||=
name|si2
expr_stmt|;
name|ui
operator||=
name|ui2
expr_stmt|;
name|bi
operator||=
name|bi2
expr_stmt|;
name|sl
operator||=
name|sl2
expr_stmt|;
name|ul
operator||=
name|ul2
expr_stmt|;
name|bl
operator||=
name|bl2
expr_stmt|;
name|fd
operator||=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator||=
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
comment|// -------------------------------------------------------------------------
comment|// Test that& rules apply to ^ too.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|^
name|sc2
expr_stmt|;
name|sc
operator|=
name|sc
operator|^
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|sc
operator|=
name|sc
operator|^
name|bc2
expr_stmt|;
name|uc
operator|=
name|uc
operator|^
name|uc2
expr_stmt|;
name|uc
operator|=
name|sc
operator|^
name|uc2
expr_stmt|;
comment|// expected-error {{cannot convert}}
name|uc
operator|=
name|bc
operator|^
name|uc2
expr_stmt|;
name|bc
operator|=
name|bc
operator|^
name|bc2
expr_stmt|;
name|bc
operator|=
name|uc
operator|^
name|bc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|bc
operator|=
name|bc
operator|^
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|fd
operator|=
name|fd
operator|^
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|bl
operator|^
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|ss
operator|^=
name|ss2
expr_stmt|;
name|us
operator|^=
name|us2
expr_stmt|;
name|bs
operator|^=
name|bs2
expr_stmt|;
name|si
operator|^=
name|si2
expr_stmt|;
name|ui
operator|^=
name|ui2
expr_stmt|;
name|bi
operator|^=
name|bi2
expr_stmt|;
name|sl
operator|^=
name|sl2
expr_stmt|;
name|ul
operator|^=
name|ul2
expr_stmt|;
name|bl
operator|^=
name|bl2
expr_stmt|;
name|fd
operator|^=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|^=
name|fd2
expr_stmt|;
comment|// expected-error {{invalid operands}}
comment|// -------------------------------------------------------------------------
comment|// Test<<.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|<<
name|sc2
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
name|uc2
expr_stmt|;
name|sc
operator|=
name|uc
operator|<<
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|sc
operator|<<
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|=
name|bc
operator|<<
name|sc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|uc
operator|=
name|uc
operator|<<
name|uc2
expr_stmt|;
name|uc
operator|=
name|sc
operator|<<
name|uc2
expr_stmt|;
comment|// expected-error {{assigning to}}
name|uc
operator|=
name|uc
operator|<<
name|sc2
expr_stmt|;
name|uc
operator|=
name|bc
operator|<<
name|uc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|uc
operator|=
name|uc
operator|<<
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|bc
operator|<<
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|bc
operator|<<
name|uc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|uc
operator|<<
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|bc
operator|<<
name|sc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|sc
operator|<<
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|=
name|sc
operator|<<
literal|1
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
literal|1.0f
expr_stmt|;
comment|// expected-error {{integer is required}}
name|sc
operator|=
name|sc
operator|<<
name|sc_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
name|uc_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
name|ss_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
name|us_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
name|si_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
name|ui_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
name|sl_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|<<
name|ul_scalar
expr_stmt|;
name|sc
operator|=
name|sc_scalar
operator|<<
name|sc
expr_stmt|;
comment|// expected-error {{first operand is not a vector}}
name|sc
operator|=
name|uc_scalar
operator|<<
name|sc
expr_stmt|;
comment|// expected-error {{first operand is not a vector}}
name|uc
operator|=
name|uc
operator|<<
name|sc_scalar
expr_stmt|;
name|uc
operator|=
name|uc
operator|<<
name|uc_scalar
expr_stmt|;
name|uc
operator|=
name|sc_scalar
operator|<<
name|uc
expr_stmt|;
comment|// expected-error {{first operand is not a vector}}
name|uc
operator|=
name|uc_scalar
operator|<<
name|uc
expr_stmt|;
comment|// expected-error {{first operand is not a vector}}
name|ss
operator|=
name|ss
operator|<<
name|ss2
expr_stmt|;
name|ss
operator|=
name|ss
operator|<<
name|ss_scalar
expr_stmt|;
name|us
operator|=
name|us
operator|<<
name|us2
expr_stmt|;
name|us
operator|=
name|us
operator|<<
name|us_scalar
expr_stmt|;
name|bs
operator|=
name|bs
operator|<<
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|=
name|si
operator|<<
name|si2
expr_stmt|;
name|si
operator|=
name|si
operator|<<
name|si_scalar
expr_stmt|;
name|ui
operator|=
name|ui
operator|<<
name|ui2
expr_stmt|;
name|ui
operator|=
name|ui
operator|<<
name|ui_scalar
expr_stmt|;
name|bi
operator|=
name|bi
operator|<<
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|=
name|sl
operator|<<
name|sl2
expr_stmt|;
name|sl
operator|=
name|sl
operator|<<
name|sl_scalar
expr_stmt|;
name|ul
operator|=
name|ul
operator|<<
name|ul2
expr_stmt|;
name|ul
operator|=
name|ul
operator|<<
name|ul_scalar
expr_stmt|;
name|bl
operator|=
name|bl
operator|<<
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|fd
operator|<<
name|fd2
expr_stmt|;
comment|// expected-error {{integer is required}}
name|fd
operator|=
name|fd
operator|<<
name|ul2
expr_stmt|;
comment|// expected-error {{integer is required}}
name|fd
operator|=
name|sl
operator|<<
name|fd2
expr_stmt|;
comment|// expected-error {{integer is required}}
name|sc
operator|<<=
name|sc2
expr_stmt|;
name|sc
operator|<<=
name|uc2
expr_stmt|;
name|sc
operator|<<=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|<<=
name|sc_scalar
expr_stmt|;
name|uc
operator|<<=
name|uc2
expr_stmt|;
name|uc
operator|<<=
name|sc2
expr_stmt|;
name|uc
operator|<<=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|uc
operator|<<=
name|uc_scalar
expr_stmt|;
name|bc
operator|<<=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|<<=
name|sc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|<<=
name|uc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|<<=
name|ss2
expr_stmt|;
comment|// expected-error {{vector operands do not have the same number of elements}}
name|sc
operator|<<=
name|si2
expr_stmt|;
comment|// expected-error {{vector operands do not have the same number of elements}}
name|sc
operator|<<=
name|sl2
expr_stmt|;
comment|// expected-error {{vector operands do not have the same number of elements}}
name|sc
operator|<<=
name|fd2
expr_stmt|;
comment|// expected-error {{integer is required}}
name|ss
operator|<<=
name|ss2
expr_stmt|;
name|ss
operator|<<=
name|ss_scalar
expr_stmt|;
name|us
operator|<<=
name|us2
expr_stmt|;
name|us
operator|<<=
name|us_scalar
expr_stmt|;
name|bs
operator|<<=
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|<<=
name|si2
expr_stmt|;
name|si
operator|<<=
name|si_scalar
expr_stmt|;
name|ui
operator|<<=
name|ui2
expr_stmt|;
name|ui
operator|<<=
name|ui_scalar
expr_stmt|;
name|bi
operator|<<=
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|<<=
name|sl2
expr_stmt|;
name|sl
operator|<<=
name|sl_scalar
expr_stmt|;
name|ul
operator|<<=
name|ul2
expr_stmt|;
name|ul
operator|<<=
name|ul_scalar
expr_stmt|;
name|bl
operator|<<=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|<<=
name|fd2
expr_stmt|;
comment|// expected-error {{integer is required}}
comment|// -------------------------------------------------------------------------
comment|// Test>>.
comment|// -------------------------------------------------------------------------
name|sc
operator|=
name|sc
operator|>>
name|sc2
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
name|uc2
expr_stmt|;
name|sc
operator|=
name|uc
operator|>>
name|sc2
expr_stmt|;
comment|// expected-error {{incompatible type}}
name|sc
operator|=
name|sc
operator|>>
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|=
name|bc
operator|>>
name|sc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|uc
operator|=
name|uc
operator|>>
name|uc2
expr_stmt|;
name|uc
operator|=
name|sc
operator|>>
name|uc2
expr_stmt|;
comment|// expected-error {{assigning to}}
name|uc
operator|=
name|uc
operator|>>
name|sc2
expr_stmt|;
name|uc
operator|=
name|bc
operator|>>
name|uc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|uc
operator|=
name|uc
operator|>>
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|bc
operator|>>
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|bc
operator|>>
name|uc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|uc
operator|>>
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|bc
operator|>>
name|sc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|=
name|sc
operator|>>
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|=
name|sc
operator|>>
literal|1
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
literal|1.0f
expr_stmt|;
comment|// expected-error {{integer is required}}
name|sc
operator|=
name|sc
operator|>>
name|sc_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
name|uc_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
name|ss_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
name|us_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
name|si_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
name|ui_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
name|sl_scalar
expr_stmt|;
name|sc
operator|=
name|sc
operator|>>
name|ul_scalar
expr_stmt|;
name|sc
operator|=
name|sc_scalar
operator|>>
name|sc
expr_stmt|;
comment|// expected-error {{first operand is not a vector}}
name|sc
operator|=
name|uc_scalar
operator|>>
name|sc
expr_stmt|;
comment|// expected-error {{first operand is not a vector}}
name|uc
operator|=
name|uc
operator|>>
name|sc_scalar
expr_stmt|;
name|uc
operator|=
name|uc
operator|>>
name|uc_scalar
expr_stmt|;
name|uc
operator|=
name|sc_scalar
operator|>>
name|uc
expr_stmt|;
comment|// expected-error {{first operand is not a vector}}
name|uc
operator|=
name|uc_scalar
operator|>>
name|uc
expr_stmt|;
comment|// expected-error {{first operand is not a vector}}
name|ss
operator|=
name|ss
operator|>>
name|ss2
expr_stmt|;
name|ss
operator|=
name|ss
operator|>>
name|ss_scalar
expr_stmt|;
name|us
operator|=
name|us
operator|>>
name|us2
expr_stmt|;
name|us
operator|=
name|us
operator|>>
name|us_scalar
expr_stmt|;
name|bs
operator|=
name|bs
operator|>>
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|=
name|si
operator|>>
name|si2
expr_stmt|;
name|si
operator|=
name|si
operator|>>
name|si_scalar
expr_stmt|;
name|ui
operator|=
name|ui
operator|>>
name|ui2
expr_stmt|;
name|ui
operator|=
name|ui
operator|>>
name|ui_scalar
expr_stmt|;
name|bi
operator|=
name|bi
operator|>>
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|=
name|sl
operator|>>
name|sl2
expr_stmt|;
name|sl
operator|=
name|sl
operator|>>
name|sl_scalar
expr_stmt|;
name|ul
operator|=
name|ul
operator|>>
name|ul2
expr_stmt|;
name|ul
operator|=
name|ul
operator|>>
name|ul_scalar
expr_stmt|;
name|bl
operator|=
name|bl
operator|>>
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|=
name|fd
operator|>>
name|fd2
expr_stmt|;
comment|// expected-error {{integer is required}}
name|fd
operator|=
name|fd
operator|>>
name|ul2
expr_stmt|;
comment|// expected-error {{integer is required}}
name|fd
operator|=
name|sl
operator|>>
name|fd2
expr_stmt|;
comment|// expected-error {{integer is required}}
name|sc
operator|>>=
name|sc2
expr_stmt|;
name|sc
operator|>>=
name|uc2
expr_stmt|;
name|sc
operator|>>=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|>>=
name|sc_scalar
expr_stmt|;
name|uc
operator|>>=
name|uc2
expr_stmt|;
name|uc
operator|>>=
name|sc2
expr_stmt|;
name|uc
operator|>>=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|uc
operator|>>=
name|uc_scalar
expr_stmt|;
name|bc
operator|>>=
name|bc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|>>=
name|sc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|bc
operator|>>=
name|uc2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sc
operator|>>=
name|ss2
expr_stmt|;
comment|// expected-error {{vector operands do not have the same number of elements}}
name|sc
operator|>>=
name|si2
expr_stmt|;
comment|// expected-error {{vector operands do not have the same number of elements}}
name|sc
operator|>>=
name|sl2
expr_stmt|;
comment|// expected-error {{vector operands do not have the same number of elements}}
name|sc
operator|>>=
name|fd2
expr_stmt|;
comment|// expected-error {{integer is required}}
name|ss
operator|>>=
name|ss2
expr_stmt|;
name|ss
operator|>>=
name|ss_scalar
expr_stmt|;
name|us
operator|>>=
name|us2
expr_stmt|;
name|us
operator|>>=
name|us_scalar
expr_stmt|;
name|bs
operator|>>=
name|bs2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|si
operator|>>=
name|si2
expr_stmt|;
name|si
operator|>>=
name|si_scalar
expr_stmt|;
name|ui
operator|>>=
name|ui2
expr_stmt|;
name|ui
operator|>>=
name|ui_scalar
expr_stmt|;
name|bi
operator|>>=
name|bi2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|sl
operator|>>=
name|sl2
expr_stmt|;
name|sl
operator|>>=
name|sl_scalar
expr_stmt|;
name|ul
operator|>>=
name|ul2
expr_stmt|;
name|ul
operator|>>=
name|ul_scalar
expr_stmt|;
name|bl
operator|>>=
name|bl2
expr_stmt|;
comment|// expected-error {{invalid operands}}
name|fd
operator|>>=
name|fd2
expr_stmt|;
comment|// expected-error {{integer is required}}
comment|// -------------------------------------------------------------------------
comment|// Test ==.
comment|// -------------------------------------------------------------------------
call|(
name|void
call|)
argument_list|(
name|sc
operator|==
name|sc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|uc
operator|==
name|uc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bc
operator|==
name|bc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sc
operator|==
name|uc
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
call|(
name|void
call|)
argument_list|(
name|sc
operator|==
name|bc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|uc
operator|==
name|sc
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
call|(
name|void
call|)
argument_list|(
name|uc
operator|==
name|bc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bc
operator|==
name|sc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bc
operator|==
name|uc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ss
operator|==
name|ss2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|us
operator|==
name|us2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bs
operator|==
name|bs2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|si
operator|==
name|si2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ui
operator|==
name|ui2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bi
operator|==
name|bi2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sl
operator|==
name|sl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ul
operator|==
name|ul2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bl
operator|==
name|bl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|fd
operator|==
name|fd2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|fd
operator|==
name|ul
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
call|(
name|void
call|)
argument_list|(
name|ul
operator|==
name|fd
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
comment|// -------------------------------------------------------------------------
comment|// Test that == rules apply to != too.
comment|// -------------------------------------------------------------------------
call|(
name|void
call|)
argument_list|(
name|sc
operator|!=
name|sc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|uc
operator|!=
name|uc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bc
operator|!=
name|bc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sc
operator|!=
name|uc
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
call|(
name|void
call|)
argument_list|(
name|sc
operator|!=
name|bc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ss
operator|!=
name|ss2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|us
operator|!=
name|us2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bs
operator|!=
name|bs2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|si
operator|!=
name|si2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ui
operator|!=
name|ui2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bi
operator|!=
name|bi2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sl
operator|!=
name|sl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ul
operator|!=
name|ul2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bl
operator|!=
name|bl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|fd
operator|!=
name|fd2
argument_list|)
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that == rules apply to<= too.
comment|// -------------------------------------------------------------------------
call|(
name|void
call|)
argument_list|(
name|sc
operator|<=
name|sc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|uc
operator|<=
name|uc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bc
operator|<=
name|bc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sc
operator|<=
name|uc
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
call|(
name|void
call|)
argument_list|(
name|sc
operator|<=
name|bc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ss
operator|<=
name|ss2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|us
operator|<=
name|us2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bs
operator|<=
name|bs2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|si
operator|<=
name|si2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ui
operator|<=
name|ui2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bi
operator|<=
name|bi2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sl
operator|<=
name|sl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ul
operator|<=
name|ul2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bl
operator|<=
name|bl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|fd
operator|<=
name|fd2
argument_list|)
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that == rules apply to>= too.
comment|// -------------------------------------------------------------------------
call|(
name|void
call|)
argument_list|(
name|sc
operator|>=
name|sc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|uc
operator|>=
name|uc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bc
operator|>=
name|bc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sc
operator|>=
name|uc
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
call|(
name|void
call|)
argument_list|(
name|sc
operator|>=
name|bc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ss
operator|>=
name|ss2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|us
operator|>=
name|us2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bs
operator|>=
name|bs2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|si
operator|>=
name|si2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ui
operator|>=
name|ui2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bi
operator|>=
name|bi2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sl
operator|>=
name|sl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ul
operator|>=
name|ul2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bl
operator|>=
name|bl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|fd
operator|>=
name|fd2
argument_list|)
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that == rules apply to< too.
comment|// -------------------------------------------------------------------------
call|(
name|void
call|)
argument_list|(
name|sc
operator|<
name|sc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|uc
operator|<
name|uc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bc
operator|<
name|bc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sc
operator|<
name|uc
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
call|(
name|void
call|)
argument_list|(
name|sc
operator|<
name|bc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ss
operator|<
name|ss2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|us
operator|<
name|us2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bs
operator|<
name|bs2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|si
operator|<
name|si2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ui
operator|<
name|ui2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bi
operator|<
name|bi2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sl
operator|<
name|sl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ul
operator|<
name|ul2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bl
operator|<
name|bl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|fd
operator|<
name|fd2
argument_list|)
expr_stmt|;
comment|// -------------------------------------------------------------------------
comment|// Test that == rules apply to> too.
comment|// -------------------------------------------------------------------------
call|(
name|void
call|)
argument_list|(
name|sc
operator|>
name|sc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|uc
operator|>
name|uc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bc
operator|>
name|bc2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sc
operator|>
name|uc
argument_list|)
expr_stmt|;
comment|// expected-error {{cannot convert}}
call|(
name|void
call|)
argument_list|(
name|sc
operator|>
name|bc
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ss
operator|>
name|ss2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|us
operator|>
name|us2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bs
operator|>
name|bs2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|si
operator|>
name|si2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ui
operator|>
name|ui2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bi
operator|>
name|bi2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|sl
operator|>
name|sl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|ul
operator|>
name|ul2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|bl
operator|>
name|bl2
argument_list|)
expr_stmt|;
call|(
name|void
call|)
argument_list|(
name|fd
operator|>
name|fd2
argument_list|)
expr_stmt|;
block|}
end_function

end_unit

